

<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">

    <title>My Blog</title>
    
    <meta name="author" content="Shirin Glander">

    <!-- Enable responsive viewport -->
    <meta name="viewport" content="width=device-width, initial-scale=1.0">

    <!-- Bootstrap styles -->
    <link href="/assets/themes/bootstrap-3/bootstrap/css/bootstrap.min.css" rel="stylesheet">
    <!-- Optional theme -->
    <link href="/assets/themes/bootstrap-3/bootstrap/css/bootstrap-theme.min.css" rel="stylesheet">
    <!-- Sticky Footer -->
    <link href="/assets/themes/bootstrap-3/bootstrap/css/bs-sticky-footer.css" rel="stylesheet">

    <!-- Custom styles -->
    <link href="/assets/themes/bootstrap-3/css/style.css?body=1" rel="stylesheet" type="text/css" media="all">

    <!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
      <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
    <![endif]-->
	  
    <!-- Fav and touch icons -->
    <!-- Update these with your own images      -->
      <link rel="shortcut icon" type="image/x-icon" href="http://localhost:4000/assets/images/iconified/favicon.ico">
      <!-- <link rel="shortcut icon" type="image/png" href="http://localhost:4000/assets/images/iconified/apple-touch-icon.png"> -->
      <link rel="apple-touch-icon" type="image/png" href="http://localhost:4000/assets/images/iconified/apple-touch-icon.png">
      <link rel="apple-touch-icon" type="image/png" sizes="76x76" href="http://localhost:4000/images/iconified/apple-touch-icon-76x76.png">
      <link rel="apple-touch-icon" type="image/png" sizes="114x114" href="http://localhost:4000/images/iconified/apple-touch-icon-114x114.png">

    <!-- atom & rss feed -->
    <link href="/atom.xml" type="application/atom+xml" rel="alternate" title="Sitewide ATOM Feed">
    <link href="/rss.xml" type="application/rss+xml" rel="alternate" title="Sitewide RSS Feed">
  </head>
  
 <div class="header">
      <div id="wrap">
      <nav class="navbar navbar-default" role="navigation">
        <!-- Brand and toggle get grouped for better mobile display -->
        <div class="navbar-header">
          <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#jb-navbar-collapse">
            <span class="sr-only">Toggle navigation</span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
          </button>
          <!-- <a class="navbar-brand" href="/">Shirin's playgRound</a> -->
		  <a class="navbar-brand" href="/"><img src="https://raw.githubusercontent.com/ShirinG/ShirinG.github.io/master/assets/images/logo.png" alt="logo" />Shirin's playgRound</a>
        </div>

        <!-- Collect the nav links, forms, and other content for toggling -->
        <div class="collapse navbar-collapse" id="jb-navbar-collapse">
          <ul class="nav navbar-nav">
            
            
            


  
    
      
    
  
    
      
    
  
    
      
      	
      	<li><a href="/about">About me</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/archive">Archive</a></li>
      	
      
    
  
    
      
    
  
    
      
      	
      	<li><a href="/categories">Categories</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/feeds">Feeds</a></li>
      	
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
      	
      	<li><a href="/tags">Tags</a></li>
      	
      
    
  
    
  
    
  
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  



          </ul>
          <form class="navbar-form navbar-right" role="search">
            <div class="form-group">
              <input type="text" class="form-control" placeholder="Search">
            </div>
            <button type="submit" class="btn btn-default">Submit</button>
          </form>
        </div><!-- /.navbar-collapse -->
      </nav>

    </div>
</div>

  <body>
    <div id="wrap">
      <div class="container">
        <!-- This loops through the paginated posts -->

  <h1><a href="/machine_learning/2016/12/02/flu_outcome_ML_2_post">Extreme Gradient Boosting and Preprocessing in Machine Learning - Addendum to predicting flu outcome with R</a></h1>
  <p class="author">
    <span class="date">2016-12-02 00:00:00 +0100</span>
  </p>
  <div class="content">
    <p>In <a href="https://shiring.github.io/machine_learning/2016/11/27/flu_outcome_ML_post">last week’s post</a> I explored whether machine learning models can be applied to predict flu deaths from the 2013 outbreak of influenza A H7N9 in China. There, I compared random forests, elastic-net regularized generalized linear models, k-nearest neighbors, penalized discriminant analysis, stabilized linear discriminant analysis, nearest shrunken centroids, single C5.0 tree and partial least squares.</p>

<p><br /></p>

<h1 id="extreme-gradient-boosting">Extreme gradient boosting</h1>

<p><a href="http://xgboost.readthedocs.io/en/latest/model.html">Extreme gradient boosting (XGBoost)</a> is a faster and improved implementation of <a href="https://en.wikipedia.org/wiki/Gradient_boosting">gradient boosting</a> for supervised learning and has recently been very successfully applied in Kaggle competitions. Because I’ve heard XGBoost’s praise being sung everywhere lately, I wanted to get my feet wet with it too. So this week I want to compare the prediction success of gradient boosting with the same dataset. Additionally, I want to test the influence of different preprocessing methods on the outcome.</p>

<p><br /></p>

<blockquote>
  <p>“XGBoost uses a more regularized model formalization to control over-fitting, which gives it better performance.” Tianqi Chen, developer of xgboost</p>
</blockquote>

<p>XGBoost is a tree ensemble model, which means the sum of predictions from a set of classification and regression trees (CART). In that, XGBoost is similar to Random Forests but it uses a different approach to model training.</p>

<p><br /></p>

<p>Starting with the same test and training data (partitioned into validation test and validation train subsets) from <a href="https://shiring.github.io/machine_learning/2016/11/27/flu_outcome_ML_post">last week’s post</a>, I am training extreme gradient boosting models as implemented in the <a href="ftp://cran.r-project.org/pub/R/web/packages/xgboost/vignettes/xgboost.pdf">xgboost</a> and <a href="http://topepo.github.io/caret/index.html">caret</a> packages with different preprocessing settings.</p>

<p>Out of the different implementations and variations of gradient boosting algorithms, caret performed best on PCA-preprocessed data in the validation set. These paramteres were then used to predict the outcome in the test set and compare it to last week’s predictions.</p>

<p>Compared to last week, there is much less uncertainty in the predictions from XGBoost. Overall, I would say that this algorithm is superior to the others I have used before.</p>

<p><br /></p>

<h2 id="xgboost">xgboost</h2>

<p>Extreme gradient boosting is implemented in the xgboost package.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># install the stable/pre-compiled version from CRAN
</span><span class="w">
</span><span class="n">install.packages</span><span class="p">(</span><span class="s1">'xgboost'</span><span class="p">)</span><span class="w">

</span><span class="c1"># or install from weekly updated drat repo
</span><span class="w">
</span><span class="n">install.packages</span><span class="p">(</span><span class="s2">"drat"</span><span class="p">,</span><span class="w"> </span><span class="n">repos</span><span class="o">=</span><span class="s2">"https://cran.rstudio.com"</span><span class="p">)</span><span class="w">
</span><span class="n">drat</span><span class="o">:::</span><span class="n">addRepo</span><span class="p">(</span><span class="s2">"dmlc"</span><span class="p">)</span><span class="w">
</span><span class="n">install.packages</span><span class="p">(</span><span class="s2">"xgboost"</span><span class="p">,</span><span class="w"> </span><span class="n">repos</span><span class="o">=</span><span class="s2">"http://dmlc.ml/drat/"</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"source"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>XGBoost supports only numbers, so the outcome classes have to be converted into integers and both training and test data have to be in numeric matrix format.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">xgboost</span><span class="p">)</span><span class="w">
</span><span class="n">matrix_train</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">val_train_X</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span><span class="n">outcome_death_train</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">val_train_data</span><span class="o">$</span><span class="n">outcome</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Death"</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w">

</span><span class="n">matrix_test</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">val_test_X</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span><span class="n">outcome_death_test</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Death"</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w">

</span><span class="n">xgb_train_matrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">xgb.DMatrix</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.matrix</span><span class="p">(</span><span class="n">matrix_train</span><span class="p">),</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">outcome_death_train</span><span class="p">)</span><span class="w">
</span><span class="n">xgb_test_matrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">xgb.DMatrix</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.matrix</span><span class="p">(</span><span class="n">matrix_test</span><span class="p">),</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">outcome_death_test</span><span class="p">)</span><span class="w">

</span><span class="n">watchlist</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">train</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">xgb_train_matrix</span><span class="p">,</span><span class="w"> </span><span class="n">test</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">xgb_test_matrix</span><span class="p">)</span><span class="w">
</span><span class="n">label</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">getinfo</span><span class="p">(</span><span class="n">xgb_test_matrix</span><span class="p">,</span><span class="w"> </span><span class="s2">"label"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>I am using cross validation to evaluate the error rate.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">param</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="s2">"objective"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"binary:logistic"</span><span class="p">)</span><span class="w">

</span><span class="n">xgb.cv</span><span class="p">(</span><span class="n">param</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">param</span><span class="p">,</span><span class="w"> 
       </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">xgb_train_matrix</span><span class="p">,</span><span class="w"> 
       </span><span class="n">nfold</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w">
       </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">getinfo</span><span class="p">(</span><span class="n">xgb_train_matrix</span><span class="p">,</span><span class="w"> </span><span class="s2">"label"</span><span class="p">),</span><span class="w">
       </span><span class="n">nrounds</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1]  train-error:0.133950+0.022131   test-error:0.341131+0.129919 
## [2]  train-error:0.124941+0.033432   test-error:0.358674+0.105191 
## [3]  train-error:0.115932+0.045501   test-error:0.340156+0.092598 
## [4]  train-error:0.124941+0.033432   test-error:0.340156+0.092598 
## [5]  train-error:0.089616+0.051301   test-error:0.394737+0.098465
</code></pre>
</div>

<p><br /></p>

<h3 id="training-with-gbtree">Training with gbtree</h3>

<p>gbtree is the default booster for xgb.train.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">bst_1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">xgb.train</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">xgb_train_matrix</span><span class="p">,</span><span class="w"> 
                   </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">getinfo</span><span class="p">(</span><span class="n">xgb_train_matrix</span><span class="p">,</span><span class="w"> </span><span class="s2">"label"</span><span class="p">),</span><span class="w">
                   </span><span class="n">max.depth</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> 
                   </span><span class="n">eta</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> 
                   </span><span class="n">nthread</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">4</span><span class="p">,</span><span class="w"> 
                   </span><span class="n">nround</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">50</span><span class="p">,</span><span class="w"> </span><span class="c1"># number of trees used for model building
</span><span class="w">                   </span><span class="n">watchlist</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">watchlist</span><span class="p">,</span><span class="w"> 
                   </span><span class="n">objective</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"binary:logistic"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1]  train-error:0.250000    test-error:0.347826 
## [2]  train-error:0.178571    test-error:0.304348 
## [3]  train-error:0.142857    test-error:0.347826 
## [4]  train-error:0.089286    test-error:0.260870 
## [5]  train-error:0.053571    test-error:0.347826 
## [6]  train-error:0.053571    test-error:0.391304 
## [7]  train-error:0.035714    test-error:0.347826 
## [8]  train-error:0.000000    test-error:0.391304 
## [9]  train-error:0.017857    test-error:0.391304 
## [10] train-error:0.000000    test-error:0.391304 
## [11] train-error:0.000000    test-error:0.391304 
## [12] train-error:0.000000    test-error:0.347826 
## [13] train-error:0.000000    test-error:0.347826 
## [14] train-error:0.000000    test-error:0.347826 
## [15] train-error:0.000000    test-error:0.347826 
## [16] train-error:0.000000    test-error:0.347826 
## [17] train-error:0.000000    test-error:0.347826 
## [18] train-error:0.000000    test-error:0.347826 
## [19] train-error:0.000000    test-error:0.347826 
## [20] train-error:0.000000    test-error:0.347826 
## [21] train-error:0.000000    test-error:0.347826 
## [22] train-error:0.000000    test-error:0.347826 
## [23] train-error:0.000000    test-error:0.347826 
## [24] train-error:0.000000    test-error:0.347826 
## [25] train-error:0.000000    test-error:0.347826 
## [26] train-error:0.000000    test-error:0.347826 
## [27] train-error:0.000000    test-error:0.347826 
## [28] train-error:0.000000    test-error:0.347826 
## [29] train-error:0.000000    test-error:0.347826 
## [30] train-error:0.000000    test-error:0.347826 
## [31] train-error:0.000000    test-error:0.347826 
## [32] train-error:0.000000    test-error:0.347826 
## [33] train-error:0.000000    test-error:0.347826 
## [34] train-error:0.000000    test-error:0.347826 
## [35] train-error:0.000000    test-error:0.347826 
## [36] train-error:0.000000    test-error:0.347826 
## [37] train-error:0.000000    test-error:0.347826 
## [38] train-error:0.000000    test-error:0.347826 
## [39] train-error:0.000000    test-error:0.347826 
## [40] train-error:0.000000    test-error:0.347826 
## [41] train-error:0.000000    test-error:0.347826 
## [42] train-error:0.000000    test-error:0.347826 
## [43] train-error:0.000000    test-error:0.347826 
## [44] train-error:0.000000    test-error:0.347826 
## [45] train-error:0.000000    test-error:0.347826 
## [46] train-error:0.000000    test-error:0.347826 
## [47] train-error:0.000000    test-error:0.347826 
## [48] train-error:0.000000    test-error:0.347826 
## [49] train-error:0.000000    test-error:0.347826 
## [50] train-error:0.000000    test-error:0.347826
</code></pre>
</div>

<p>Each feature is grouped by importance with k-means clustering. Gain is the improvement in accuracy that the addition of a feature brings to the branches it is on.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">matrix_train</span><span class="p">)</span><span class="w">
</span><span class="n">importance_matrix_1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">xgb.importance</span><span class="p">(</span><span class="n">features</span><span class="p">,</span><span class="w"> </span><span class="n">model</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bst_1</span><span class="p">)</span><span class="w">
</span><span class="n">print</span><span class="p">(</span><span class="n">importance_matrix_1</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##                   Feature       Gain      Cover  Frequency
## 1:                    age 0.54677689 0.45726557 0.43333333
## 2:  days_onset_to_outcome 0.14753698 0.11686537 0.11111111
## 3:            early_onset 0.06724531 0.07365416 0.07777778
## 4:          early_outcome 0.06679880 0.08869426 0.07777778
## 5:      province_Shanghai 0.05325295 0.04317030 0.04444444
## 6: days_onset_to_hospital 0.03883631 0.11300882 0.13333333
## 7:         province_other 0.03502971 0.03977657 0.04444444
## 8:               gender_f 0.02610959 0.02484170 0.01111111
## 9:               hospital 0.01841345 0.04272324 0.06666667
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">xgb.ggplot.importance</span><span class="p">(</span><span class="n">importance_matrix_1</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme_minimal</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<p><img src="flu_outcome_ML_2_post_files/figure-markdown_github/unnamed-chunk-6-1.png" style="display: block; margin: auto;" /></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">pred_1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">bst_1</span><span class="p">,</span><span class="w"> </span><span class="n">xgb_test_matrix</span><span class="p">)</span><span class="w">

</span><span class="n">result_1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">case_ID</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">val_test_data</span><span class="p">),</span><span class="w">
                       </span><span class="n">outcome</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="p">,</span><span class="w"> 
                       </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">label</span><span class="p">,</span><span class="w"> 
                       </span><span class="n">prediction_p_death</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">pred_1</span><span class="p">,</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w">
                       </span><span class="n">prediction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.integer</span><span class="p">(</span><span class="n">pred_1</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0.5</span><span class="p">),</span><span class="w">
                       </span><span class="n">prediction_eval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="nf">as.integer</span><span class="p">(</span><span class="n">pred_1</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="n">label</span><span class="p">,</span><span class="w"> </span><span class="s2">"wrong"</span><span class="p">,</span><span class="w"> </span><span class="s2">"correct"</span><span class="p">))</span><span class="w">
</span><span class="n">result_1</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##     case_ID outcome label prediction_p_death prediction prediction_eval
## 1  case_123   Death     1               0.01          0           wrong
## 2  case_127 Recover     0               0.00          0         correct
## 3  case_128 Recover     0               0.96          1           wrong
## 4   case_14 Recover     0               0.76          1           wrong
## 5   case_19   Death     1               0.95          1         correct
## 6    case_2   Death     1               0.85          1         correct
## 7   case_20 Recover     0               0.98          1           wrong
## 8   case_21 Recover     0               0.06          0         correct
## 9   case_34   Death     1               1.00          1         correct
## 10  case_37 Recover     0               0.00          0         correct
## 11   case_5 Recover     0               0.03          0         correct
## 12  case_51 Recover     0               0.27          0         correct
## 13  case_55 Recover     0               0.11          0         correct
## 14   case_6   Death     1               0.26          0           wrong
## 15  case_61   Death     1               1.00          1         correct
## 16  case_65 Recover     0               0.00          0         correct
## 17  case_74 Recover     0               0.92          1           wrong
## 18  case_78   Death     1               0.03          0           wrong
## 19  case_79 Recover     0               0.00          0         correct
## 20   case_8   Death     1               0.06          0           wrong
## 21  case_87   Death     1               0.64          1         correct
## 22  case_91 Recover     0               0.00          0         correct
## 23  case_94 Recover     0               0.01          0         correct
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">err</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="nf">as.integer</span><span class="p">(</span><span class="n">pred_1</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="n">label</span><span class="p">))</span><span class="o">/</span><span class="nf">length</span><span class="p">(</span><span class="n">label</span><span class="p">)</span><span class="w">
</span><span class="n">print</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"test-error ="</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">err</span><span class="p">,</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "test-error = 0.35"
</code></pre>
</div>

<p><br /></p>

<h3 id="training-with-gblinear">Training with gblinear</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">bst_2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">xgb.train</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">xgb_train_matrix</span><span class="p">,</span><span class="w"> 
                   </span><span class="n">booster</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"gblinear"</span><span class="p">,</span><span class="w"> 
                   </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">getinfo</span><span class="p">(</span><span class="n">xgb_train_matrix</span><span class="p">,</span><span class="w"> </span><span class="s2">"label"</span><span class="p">),</span><span class="w">
                   </span><span class="n">max.depth</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> 
                   </span><span class="n">eta</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> 
                   </span><span class="n">nthread</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">4</span><span class="p">,</span><span class="w"> 
                   </span><span class="n">nround</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">50</span><span class="p">,</span><span class="w"> </span><span class="c1"># number of trees used for model building
</span><span class="w">                   </span><span class="n">watchlist</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">watchlist</span><span class="p">,</span><span class="w"> 
                   </span><span class="n">objective</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"binary:logistic"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1]  train-error:0.339286    test-error:0.434783 
## [2]  train-error:0.303571    test-error:0.391304 
## [3]  train-error:0.285714    test-error:0.347826 
## [4]  train-error:0.303571    test-error:0.347826 
## [5]  train-error:0.285714    test-error:0.347826 
## [6]  train-error:0.285714    test-error:0.391304 
## [7]  train-error:0.285714    test-error:0.391304 
## [8]  train-error:0.267857    test-error:0.391304 
## [9]  train-error:0.250000    test-error:0.391304 
## [10] train-error:0.250000    test-error:0.478261 
## [11] train-error:0.250000    test-error:0.434783 
## [12] train-error:0.232143    test-error:0.434783 
## [13] train-error:0.232143    test-error:0.434783 
## [14] train-error:0.232143    test-error:0.434783 
## [15] train-error:0.232143    test-error:0.434783 
## [16] train-error:0.232143    test-error:0.434783 
## [17] train-error:0.232143    test-error:0.434783 
## [18] train-error:0.232143    test-error:0.434783 
## [19] train-error:0.232143    test-error:0.434783 
## [20] train-error:0.232143    test-error:0.434783 
## [21] train-error:0.214286    test-error:0.434783 
## [22] train-error:0.214286    test-error:0.434783 
## [23] train-error:0.214286    test-error:0.434783 
## [24] train-error:0.214286    test-error:0.434783 
## [25] train-error:0.214286    test-error:0.434783 
## [26] train-error:0.214286    test-error:0.434783 
## [27] train-error:0.214286    test-error:0.434783 
## [28] train-error:0.214286    test-error:0.434783 
## [29] train-error:0.214286    test-error:0.434783 
## [30] train-error:0.232143    test-error:0.434783 
## [31] train-error:0.214286    test-error:0.434783 
## [32] train-error:0.214286    test-error:0.434783 
## [33] train-error:0.214286    test-error:0.434783 
## [34] train-error:0.214286    test-error:0.434783 
## [35] train-error:0.214286    test-error:0.434783 
## [36] train-error:0.214286    test-error:0.434783 
## [37] train-error:0.214286    test-error:0.434783 
## [38] train-error:0.214286    test-error:0.434783 
## [39] train-error:0.214286    test-error:0.434783 
## [40] train-error:0.214286    test-error:0.434783 
## [41] train-error:0.214286    test-error:0.434783 
## [42] train-error:0.214286    test-error:0.434783 
## [43] train-error:0.214286    test-error:0.434783 
## [44] train-error:0.214286    test-error:0.434783 
## [45] train-error:0.214286    test-error:0.434783 
## [46] train-error:0.214286    test-error:0.434783 
## [47] train-error:0.214286    test-error:0.434783 
## [48] train-error:0.214286    test-error:0.434783 
## [49] train-error:0.214286    test-error:0.434783 
## [50] train-error:0.214286    test-error:0.434783
</code></pre>
</div>

<p>Each feature is grouped by importance with k-means clustering. Gain is the improvement in accuracy that the addition of a feature brings to the branches it is on.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">matrix_train</span><span class="p">)</span><span class="w">
</span><span class="n">importance_matrix_2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">xgb.importance</span><span class="p">(</span><span class="n">features</span><span class="p">,</span><span class="w"> </span><span class="n">model</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bst_2</span><span class="p">)</span><span class="w">
</span><span class="n">print</span><span class="p">(</span><span class="n">importance_matrix_2</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##                    Feature     Weight
##  1:                    age  0.0439012
##  2:               hospital -0.3131160
##  3:               gender_f  0.8067830
##  4:       province_Jiangsu -0.8807380
##  5:      province_Shanghai -0.4065330
##  6:      province_Zhejiang -0.1284590
##  7:         province_other  0.4510330
##  8:  days_onset_to_outcome  0.0209952
##  9: days_onset_to_hospital -0.0856504
## 10:            early_onset  0.6452190
## 11:          early_outcome  2.5134300
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">xgb.ggplot.importance</span><span class="p">(</span><span class="n">importance_matrix_2</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme_minimal</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<p><img src="flu_outcome_ML_2_post_files/figure-markdown_github/unnamed-chunk-10-1.png" style="display: block; margin: auto;" /></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">pred_2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">bst_2</span><span class="p">,</span><span class="w"> </span><span class="n">xgb_test_matrix</span><span class="p">)</span><span class="w">

</span><span class="n">result_2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">case_ID</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">val_test_data</span><span class="p">),</span><span class="w">
                       </span><span class="n">outcome</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="p">,</span><span class="w"> 
                       </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">label</span><span class="p">,</span><span class="w"> 
                       </span><span class="n">prediction_p_death</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">pred_2</span><span class="p">,</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w">
                       </span><span class="n">prediction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.integer</span><span class="p">(</span><span class="n">pred_2</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0.5</span><span class="p">),</span><span class="w">
                       </span><span class="n">prediction_eval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="nf">as.integer</span><span class="p">(</span><span class="n">pred_2</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="n">label</span><span class="p">,</span><span class="w"> </span><span class="s2">"wrong"</span><span class="p">,</span><span class="w"> </span><span class="s2">"correct"</span><span class="p">))</span><span class="w">
</span><span class="n">result_2</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##     case_ID outcome label prediction_p_death prediction prediction_eval
## 1  case_123   Death     1               0.10          0           wrong
## 2  case_127 Recover     0               0.03          0         correct
## 3  case_128 Recover     0               0.13          0         correct
## 4   case_14 Recover     0               0.15          0         correct
## 5   case_19   Death     1               0.28          0           wrong
## 6    case_2   Death     1               0.28          0           wrong
## 7   case_20 Recover     0               0.80          1           wrong
## 8   case_21 Recover     0               0.79          1           wrong
## 9   case_34   Death     1               0.74          1         correct
## 10  case_37 Recover     0               0.06          0         correct
## 11   case_5 Recover     0               0.14          0         correct
## 12  case_51 Recover     0               0.77          1           wrong
## 13  case_55 Recover     0               0.23          0         correct
## 14   case_6   Death     1               0.47          0           wrong
## 15  case_61   Death     1               0.52          1         correct
## 16  case_65 Recover     0               0.03          0         correct
## 17  case_74 Recover     0               0.70          1           wrong
## 18  case_78   Death     1               0.22          0           wrong
## 19  case_79 Recover     0               0.06          0         correct
## 20   case_8   Death     1               0.36          0           wrong
## 21  case_87   Death     1               0.65          1         correct
## 22  case_91 Recover     0               0.05          0         correct
## 23  case_94 Recover     0               0.07          0         correct
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">err</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="nf">as.integer</span><span class="p">(</span><span class="n">pred_2</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="n">label</span><span class="p">))</span><span class="o">/</span><span class="nf">length</span><span class="p">(</span><span class="n">label</span><span class="p">)</span><span class="w">
</span><span class="n">print</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"test-error ="</span><span class="p">,</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">err</span><span class="p">,</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "test-error = 0.43"
</code></pre>
</div>

<p><br /></p>

<h2 id="caret">caret</h2>

<p>Extreme gradient boosting is also implemented in the caret package. Caret also provides options for preprocessing, of which I will compare a few.</p>

<p><br /></p>

<h3 id="no-preprocessing">No preprocessing</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">caret</span><span class="p">)</span><span class="w">

</span><span class="n">set.seed</span><span class="p">(</span><span class="m">27</span><span class="p">)</span><span class="w">
</span><span class="n">model_xgb_null</span><span class="w"> </span><span class="o">&lt;-</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                 </span><span class="n">data</span><span class="o">=</span><span class="n">val_train_data</span><span class="p">,</span><span class="w">
                 </span><span class="n">method</span><span class="o">=</span><span class="s2">"xgbTree"</span><span class="p">,</span><span class="w">
                 </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w">
                 </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">confusionMatrix</span><span class="p">(</span><span class="n">predict</span><span class="p">(</span><span class="n">model_xgb_null</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]),</span><span class="w"> </span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Confusion Matrix and Statistics
## 
##           Reference
## Prediction Death Recover
##    Death       5       4
##    Recover     4      10
##                                           
##                Accuracy : 0.6522          
##                  95% CI : (0.4273, 0.8362)
##     No Information Rate : 0.6087          
##     P-Value [Acc &gt; NIR] : 0.4216          
##                                           
##                   Kappa : 0.2698          
##  Mcnemar's Test P-Value : 1.0000          
##                                           
##             Sensitivity : 0.5556          
##             Specificity : 0.7143          
##          Pos Pred Value : 0.5556          
##          Neg Pred Value : 0.7143          
##              Prevalence : 0.3913          
##          Detection Rate : 0.2174          
##    Detection Prevalence : 0.3913          
##       Balanced Accuracy : 0.6349          
##                                           
##        'Positive' Class : Death           
## 
</code></pre>
</div>

<p><br /></p>

<h3 id="scaling-and-centering">Scaling and centering</h3>

<p>With this method the column variables are centered (subtracting the column mean from each value in a column) and standardized (dividing by the column standard deviation).</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">set.seed</span><span class="p">(</span><span class="m">27</span><span class="p">)</span><span class="w">
</span><span class="n">model_xgb_sc_cen</span><span class="w"> </span><span class="o">&lt;-</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                 </span><span class="n">data</span><span class="o">=</span><span class="n">val_train_data</span><span class="p">,</span><span class="w">
                 </span><span class="n">method</span><span class="o">=</span><span class="s2">"xgbTree"</span><span class="p">,</span><span class="w">
                 </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"scale"</span><span class="p">,</span><span class="w"> </span><span class="s2">"center"</span><span class="p">),</span><span class="w">
                 </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">confusionMatrix</span><span class="p">(</span><span class="n">predict</span><span class="p">(</span><span class="n">model_xgb_sc_cen</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]),</span><span class="w"> </span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Confusion Matrix and Statistics
## 
##           Reference
## Prediction Death Recover
##    Death       5       4
##    Recover     4      10
##                                           
##                Accuracy : 0.6522          
##                  95% CI : (0.4273, 0.8362)
##     No Information Rate : 0.6087          
##     P-Value [Acc &gt; NIR] : 0.4216          
##                                           
##                   Kappa : 0.2698          
##  Mcnemar's Test P-Value : 1.0000          
##                                           
##             Sensitivity : 0.5556          
##             Specificity : 0.7143          
##          Pos Pred Value : 0.5556          
##          Neg Pred Value : 0.7143          
##              Prevalence : 0.3913          
##          Detection Rate : 0.2174          
##    Detection Prevalence : 0.3913          
##       Balanced Accuracy : 0.6349          
##                                           
##        'Positive' Class : Death           
## 
</code></pre>
</div>

<p>Scaling and centering did not improve the predictions.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">pred_3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_xgb_sc_cen</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">])</span><span class="w">
</span><span class="n">pred_3b</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">predict</span><span class="p">(</span><span class="n">model_xgb_sc_cen</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">

</span><span class="n">result_3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">case_ID</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">val_test_data</span><span class="p">),</span><span class="w">
                       </span><span class="n">outcome</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="p">,</span><span class="w"> 
                       </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">label</span><span class="p">,</span><span class="w"> 
                       </span><span class="n">prediction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pred_3</span><span class="p">,</span><span class="w">
                       </span><span class="n">pred_3b</span><span class="p">)</span><span class="w">
</span><span class="n">result_3</span><span class="o">$</span><span class="n">prediction_eval</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">result_3</span><span class="o">$</span><span class="n">prediction</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="n">result_3</span><span class="o">$</span><span class="n">outcome</span><span class="p">,</span><span class="w"> </span><span class="s2">"wrong"</span><span class="p">,</span><span class="w"> </span><span class="s2">"correct"</span><span class="p">)</span><span class="w">
</span><span class="n">result_3</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##     case_ID outcome label prediction Death Recover prediction_eval
## 1  case_123   Death     1    Recover  0.04    0.96           wrong
## 2  case_127 Recover     0    Recover  0.04    0.96         correct
## 3  case_128 Recover     0      Death  0.81    0.19           wrong
## 4   case_14 Recover     0      Death  0.70    0.30           wrong
## 5   case_19   Death     1      Death  0.67    0.33         correct
## 6    case_2   Death     1      Death  0.66    0.34         correct
## 7   case_20 Recover     0      Death  0.76    0.24           wrong
## 8   case_21 Recover     0    Recover  0.17    0.83         correct
## 9   case_34   Death     1      Death  0.94    0.06         correct
## 10  case_37 Recover     0    Recover  0.02    0.98         correct
## 11   case_5 Recover     0    Recover  0.09    0.91         correct
## 12  case_51 Recover     0    Recover  0.41    0.59         correct
## 13  case_55 Recover     0    Recover  0.12    0.88         correct
## 14   case_6   Death     1    Recover  0.30    0.70           wrong
## 15  case_61   Death     1      Death  0.99    0.01         correct
## 16  case_65 Recover     0    Recover  0.01    0.99         correct
## 17  case_74 Recover     0      Death  0.85    0.15           wrong
## 18  case_78   Death     1    Recover  0.21    0.79           wrong
## 19  case_79 Recover     0    Recover  0.01    0.99         correct
## 20   case_8   Death     1    Recover  0.20    0.80           wrong
## 21  case_87   Death     1      Death  0.65    0.35         correct
## 22  case_91 Recover     0    Recover  0.01    0.99         correct
## 23  case_94 Recover     0    Recover  0.07    0.93         correct
</code></pre>
</div>

<p><br /></p>

<h3 id="box-cox-transformation">Box-Cox transformation</h3>

<p>The Box-Cox power transformation is used to normalize data.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">set.seed</span><span class="p">(</span><span class="m">27</span><span class="p">)</span><span class="w">
</span><span class="n">model_xgb_BoxCox</span><span class="w"> </span><span class="o">&lt;-</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                 </span><span class="n">data</span><span class="o">=</span><span class="n">val_train_data</span><span class="p">,</span><span class="w">
                 </span><span class="n">method</span><span class="o">=</span><span class="s2">"xgbTree"</span><span class="p">,</span><span class="w">
                 </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"BoxCox"</span><span class="p">,</span><span class="w">
                 </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">confusionMatrix</span><span class="p">(</span><span class="n">predict</span><span class="p">(</span><span class="n">model_xgb_BoxCox</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]),</span><span class="w"> </span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Confusion Matrix and Statistics
## 
##           Reference
## Prediction Death Recover
##    Death       5       4
##    Recover     4      10
##                                           
##                Accuracy : 0.6522          
##                  95% CI : (0.4273, 0.8362)
##     No Information Rate : 0.6087          
##     P-Value [Acc &gt; NIR] : 0.4216          
##                                           
##                   Kappa : 0.2698          
##  Mcnemar's Test P-Value : 1.0000          
##                                           
##             Sensitivity : 0.5556          
##             Specificity : 0.7143          
##          Pos Pred Value : 0.5556          
##          Neg Pred Value : 0.7143          
##              Prevalence : 0.3913          
##          Detection Rate : 0.2174          
##    Detection Prevalence : 0.3913          
##       Balanced Accuracy : 0.6349          
##                                           
##        'Positive' Class : Death           
## 
</code></pre>
</div>

<p>Box-Cox transformation did not improve the predictions either.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">pred_4</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_xgb_BoxCox</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">])</span><span class="w">
</span><span class="n">pred_4b</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">predict</span><span class="p">(</span><span class="n">model_xgb_BoxCox</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">

</span><span class="n">result_4</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">case_ID</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">val_test_data</span><span class="p">),</span><span class="w">
                       </span><span class="n">outcome</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="p">,</span><span class="w"> 
                       </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">label</span><span class="p">,</span><span class="w"> 
                       </span><span class="n">prediction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pred_4</span><span class="p">,</span><span class="w">
                       </span><span class="n">pred_4b</span><span class="p">)</span><span class="w">
</span><span class="n">result_4</span><span class="o">$</span><span class="n">prediction_eval</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">result_4</span><span class="o">$</span><span class="n">prediction</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="n">result_4</span><span class="o">$</span><span class="n">outcome</span><span class="p">,</span><span class="w"> </span><span class="s2">"wrong"</span><span class="p">,</span><span class="w"> </span><span class="s2">"correct"</span><span class="p">)</span><span class="w">
</span><span class="n">result_4</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##     case_ID outcome label prediction Death Recover prediction_eval
## 1  case_123   Death     1    Recover  0.01    0.99           wrong
## 2  case_127 Recover     0    Recover  0.01    0.99         correct
## 3  case_128 Recover     0      Death  0.88    0.12           wrong
## 4   case_14 Recover     0      Death  0.81    0.19           wrong
## 5   case_19   Death     1      Death  0.88    0.12         correct
## 6    case_2   Death     1      Death  0.79    0.21         correct
## 7   case_20 Recover     0      Death  0.84    0.16           wrong
## 8   case_21 Recover     0    Recover  0.07    0.93         correct
## 9   case_34   Death     1      Death  0.97    0.03         correct
## 10  case_37 Recover     0    Recover  0.01    0.99         correct
## 11   case_5 Recover     0    Recover  0.06    0.94         correct
## 12  case_51 Recover     0    Recover  0.27    0.73         correct
## 13  case_55 Recover     0    Recover  0.09    0.91         correct
## 14   case_6   Death     1    Recover  0.30    0.70           wrong
## 15  case_61   Death     1      Death  1.00    0.00         correct
## 16  case_65 Recover     0    Recover  0.00    1.00         correct
## 17  case_74 Recover     0      Death  0.90    0.10           wrong
## 18  case_78   Death     1    Recover  0.11    0.89           wrong
## 19  case_79 Recover     0    Recover  0.00    1.00         correct
## 20   case_8   Death     1    Recover  0.13    0.87           wrong
## 21  case_87   Death     1      Death  0.61    0.39         correct
## 22  case_91 Recover     0    Recover  0.00    1.00         correct
## 23  case_94 Recover     0    Recover  0.04    0.96         correct
</code></pre>
</div>

<p><br /></p>

<h4 id="principal-component-analysis-pca">Principal Component Analysis (PCA)</h4>

<p>PCA is used for dimensionality reduction. When applied as a preprocessing method the number of features are reduced by using the eigenvectors of the covariance matrix.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">set.seed</span><span class="p">(</span><span class="m">27</span><span class="p">)</span><span class="w">
</span><span class="n">model_xgb_pca</span><span class="w"> </span><span class="o">&lt;-</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                 </span><span class="n">data</span><span class="o">=</span><span class="n">val_train_data</span><span class="p">,</span><span class="w">
                 </span><span class="n">method</span><span class="o">=</span><span class="s2">"xgbTree"</span><span class="p">,</span><span class="w">
                 </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"pca"</span><span class="p">,</span><span class="w">
                 </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">confusionMatrix</span><span class="p">(</span><span class="n">predict</span><span class="p">(</span><span class="n">model_xgb_pca</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]),</span><span class="w"> </span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Confusion Matrix and Statistics
## 
##           Reference
## Prediction Death Recover
##    Death       4       2
##    Recover     5      12
##                                           
##                Accuracy : 0.6957          
##                  95% CI : (0.4708, 0.8679)
##     No Information Rate : 0.6087          
##     P-Value [Acc &gt; NIR] : 0.2644          
##                                           
##                   Kappa : 0.3207          
##  Mcnemar's Test P-Value : 0.4497          
##                                           
##             Sensitivity : 0.4444          
##             Specificity : 0.8571          
##          Pos Pred Value : 0.6667          
##          Neg Pred Value : 0.7059          
##              Prevalence : 0.3913          
##          Detection Rate : 0.1739          
##    Detection Prevalence : 0.2609          
##       Balanced Accuracy : 0.6508          
##                                           
##        'Positive' Class : Death           
## 
</code></pre>
</div>

<p>PCA did improve the predictions slightly.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">pred_5</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_xgb_pca</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">])</span><span class="w">
</span><span class="n">pred_5b</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">predict</span><span class="p">(</span><span class="n">model_xgb_pca</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">

</span><span class="n">result_5</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">case_ID</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">val_test_data</span><span class="p">),</span><span class="w">
                       </span><span class="n">outcome</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="p">,</span><span class="w"> 
                       </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">label</span><span class="p">,</span><span class="w"> 
                       </span><span class="n">prediction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pred_5</span><span class="p">,</span><span class="w">
                       </span><span class="n">pred_5b</span><span class="p">)</span><span class="w">
</span><span class="n">result_5</span><span class="o">$</span><span class="n">prediction_eval</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">result_5</span><span class="o">$</span><span class="n">prediction</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="n">result_5</span><span class="o">$</span><span class="n">outcome</span><span class="p">,</span><span class="w"> </span><span class="s2">"wrong"</span><span class="p">,</span><span class="w"> </span><span class="s2">"correct"</span><span class="p">)</span><span class="w">
</span><span class="n">result_5</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##     case_ID outcome label prediction Death Recover prediction_eval
## 1  case_123   Death     1    Recover  0.02    0.98           wrong
## 2  case_127 Recover     0    Recover  0.29    0.71         correct
## 3  case_128 Recover     0    Recover  0.01    0.99         correct
## 4   case_14 Recover     0    Recover  0.44    0.56         correct
## 5   case_19   Death     1    Recover  0.07    0.93           wrong
## 6    case_2   Death     1    Recover  0.26    0.74           wrong
## 7   case_20 Recover     0    Recover  0.16    0.84         correct
## 8   case_21 Recover     0      Death  0.57    0.43           wrong
## 9   case_34   Death     1      Death  0.93    0.07         correct
## 10  case_37 Recover     0    Recover  0.03    0.97         correct
## 11   case_5 Recover     0    Recover  0.11    0.89         correct
## 12  case_51 Recover     0    Recover  0.19    0.81         correct
## 13  case_55 Recover     0    Recover  0.16    0.84         correct
## 14   case_6   Death     1      Death  0.55    0.45         correct
## 15  case_61   Death     1    Recover  0.22    0.78           wrong
## 16  case_65 Recover     0    Recover  0.08    0.92         correct
## 17  case_74 Recover     0      Death  0.67    0.33           wrong
## 18  case_78   Death     1      Death  0.87    0.13         correct
## 19  case_79 Recover     0    Recover  0.16    0.84         correct
## 20   case_8   Death     1    Recover  0.22    0.78           wrong
## 21  case_87   Death     1      Death  0.66    0.34         correct
## 22  case_91 Recover     0    Recover  0.01    0.99         correct
## 23  case_94 Recover     0    Recover  0.14    0.86         correct
</code></pre>
</div>

<p><br /></p>

<h3 id="median-imputation">Median imputation</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">set.seed</span><span class="p">(</span><span class="m">27</span><span class="p">)</span><span class="w">
</span><span class="n">model_xgb_medianImpute</span><span class="w"> </span><span class="o">&lt;-</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                 </span><span class="n">data</span><span class="o">=</span><span class="n">val_train_data</span><span class="p">,</span><span class="w">
                 </span><span class="n">method</span><span class="o">=</span><span class="s2">"xgbTree"</span><span class="p">,</span><span class="w">
                 </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"medianImpute"</span><span class="p">,</span><span class="w">
                 </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">confusionMatrix</span><span class="p">(</span><span class="n">predict</span><span class="p">(</span><span class="n">model_xgb_medianImpute</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]),</span><span class="w"> </span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Confusion Matrix and Statistics
## 
##           Reference
## Prediction Death Recover
##    Death       5       4
##    Recover     4      10
##                                           
##                Accuracy : 0.6522          
##                  95% CI : (0.4273, 0.8362)
##     No Information Rate : 0.6087          
##     P-Value [Acc &gt; NIR] : 0.4216          
##                                           
##                   Kappa : 0.2698          
##  Mcnemar's Test P-Value : 1.0000          
##                                           
##             Sensitivity : 0.5556          
##             Specificity : 0.7143          
##          Pos Pred Value : 0.5556          
##          Neg Pred Value : 0.7143          
##              Prevalence : 0.3913          
##          Detection Rate : 0.2174          
##    Detection Prevalence : 0.3913          
##       Balanced Accuracy : 0.6349          
##                                           
##        'Positive' Class : Death           
## 
</code></pre>
</div>

<p>Median imputation did not improve the predictions either.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">pred_6</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_xgb_medianImpute</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">])</span><span class="w">
</span><span class="n">pred_6b</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">predict</span><span class="p">(</span><span class="n">model_xgb_medianImpute</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">

</span><span class="n">result_6</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">case_ID</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">val_test_data</span><span class="p">),</span><span class="w">
                       </span><span class="n">outcome</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="p">,</span><span class="w"> 
                       </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">label</span><span class="p">,</span><span class="w"> 
                       </span><span class="n">prediction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pred_6</span><span class="p">,</span><span class="w">
                       </span><span class="n">pred_6b</span><span class="p">)</span><span class="w">
</span><span class="n">result_6</span><span class="o">$</span><span class="n">prediction_eval</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">result_6</span><span class="o">$</span><span class="n">prediction</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="n">result_6</span><span class="o">$</span><span class="n">outcome</span><span class="p">,</span><span class="w"> </span><span class="s2">"wrong"</span><span class="p">,</span><span class="w"> </span><span class="s2">"correct"</span><span class="p">)</span><span class="w">
</span><span class="n">result_6</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##     case_ID outcome label prediction Death Recover prediction_eval
## 1  case_123   Death     1    Recover  0.04    0.96           wrong
## 2  case_127 Recover     0    Recover  0.04    0.96         correct
## 3  case_128 Recover     0      Death  0.81    0.19           wrong
## 4   case_14 Recover     0      Death  0.70    0.30           wrong
## 5   case_19   Death     1      Death  0.67    0.33         correct
## 6    case_2   Death     1      Death  0.66    0.34         correct
## 7   case_20 Recover     0      Death  0.76    0.24           wrong
## 8   case_21 Recover     0    Recover  0.17    0.83         correct
## 9   case_34   Death     1      Death  0.94    0.06         correct
## 10  case_37 Recover     0    Recover  0.02    0.98         correct
## 11   case_5 Recover     0    Recover  0.09    0.91         correct
## 12  case_51 Recover     0    Recover  0.41    0.59         correct
## 13  case_55 Recover     0    Recover  0.12    0.88         correct
## 14   case_6   Death     1    Recover  0.30    0.70           wrong
## 15  case_61   Death     1      Death  0.99    0.01         correct
## 16  case_65 Recover     0    Recover  0.01    0.99         correct
## 17  case_74 Recover     0      Death  0.85    0.15           wrong
## 18  case_78   Death     1    Recover  0.21    0.79           wrong
## 19  case_79 Recover     0    Recover  0.01    0.99         correct
## 20   case_8   Death     1    Recover  0.20    0.80           wrong
## 21  case_87   Death     1      Death  0.65    0.35         correct
## 22  case_91 Recover     0    Recover  0.01    0.99         correct
## 23  case_94 Recover     0    Recover  0.07    0.93         correct
</code></pre>
</div>

<p><br /></p>

<h1 id="comparison-of-extreme-gradient-boosting-models">Comparison of extreme gradient boosting models</h1>

<h2 id="combining-results">Combining results</h2>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span><span class="w">
</span><span class="n">result</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">left_join</span><span class="p">(</span><span class="n">result_1</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">6</span><span class="p">)],</span><span class="w"> </span><span class="n">result_2</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">6</span><span class="p">)],</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"case_ID"</span><span class="p">)</span><span class="w">
</span><span class="n">result</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">left_join</span><span class="p">(</span><span class="n">result</span><span class="p">,</span><span class="w"> </span><span class="n">result_3</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">7</span><span class="p">)],</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"case_ID"</span><span class="p">)</span><span class="w">
</span><span class="n">result</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">left_join</span><span class="p">(</span><span class="n">result</span><span class="p">,</span><span class="w"> </span><span class="n">result_4</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">7</span><span class="p">)],</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"case_ID"</span><span class="p">)</span><span class="w">
</span><span class="n">result</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">left_join</span><span class="p">(</span><span class="n">result</span><span class="p">,</span><span class="w"> </span><span class="n">result_5</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">7</span><span class="p">)],</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"case_ID"</span><span class="p">)</span><span class="w">
</span><span class="n">result</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">left_join</span><span class="p">(</span><span class="n">result</span><span class="p">,</span><span class="w"> </span><span class="n">result_6</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">7</span><span class="p">)],</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"case_ID"</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">result</span><span class="p">)[</span><span class="o">-</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"pred_xgboost_gbtree"</span><span class="p">,</span><span class="w"> </span><span class="s2">"pred_xgboost_gblinear"</span><span class="p">,</span><span class="w"> </span><span class="s2">"model_xgb_sc_cen"</span><span class="p">,</span><span class="w"> </span><span class="s2">"model_xgb_BoxCox"</span><span class="p">,</span><span class="w"> </span><span class="s2">"pred_xgbTree_pca"</span><span class="p">,</span><span class="w"> </span><span class="s2">"model_xgb_medianImpute"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<h3 id="whats-the-rate-of-correctly-predicted-cases-in-the-validation-data">What’s the rate of correctly predicted cases in the validation data?</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">round</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="n">result</span><span class="o">$</span><span class="n">pred_xgboost_gbtree</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"correct"</span><span class="p">)</span><span class="o">/</span><span class="n">nrow</span><span class="p">(</span><span class="n">result</span><span class="p">),</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 0.65
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">round</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="n">result</span><span class="o">$</span><span class="n">pred_xgboost_gblinear</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"correct"</span><span class="p">)</span><span class="o">/</span><span class="n">nrow</span><span class="p">(</span><span class="n">result</span><span class="p">),</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 0.57
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">round</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="n">result</span><span class="o">$</span><span class="n">model_xgb_sc_cen</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"correct"</span><span class="p">)</span><span class="o">/</span><span class="n">nrow</span><span class="p">(</span><span class="n">result</span><span class="p">),</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 0.65
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">round</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="n">result</span><span class="o">$</span><span class="n">model_xgb_BoxCox</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"correct"</span><span class="p">)</span><span class="o">/</span><span class="n">nrow</span><span class="p">(</span><span class="n">result</span><span class="p">),</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 0.65
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">round</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="n">result</span><span class="o">$</span><span class="n">pred_xgbTree_pca</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"correct"</span><span class="p">)</span><span class="o">/</span><span class="n">nrow</span><span class="p">(</span><span class="n">result</span><span class="p">),</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 0.7
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">round</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="n">result</span><span class="o">$</span><span class="n">model_xgb_medianImpute</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"correct"</span><span class="p">)</span><span class="o">/</span><span class="n">nrow</span><span class="p">(</span><span class="n">result</span><span class="p">),</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 0.65
</code></pre>
</div>

<p>PCA preprocessing combined with caret’s implementation of extreme gradient boosting had the highest number of accurately predicted cases in the validation data set.</p>

<p><br /></p>

<h1 id="predicting-unknown-output">Predicting unknown output</h1>

<p>Because PCA preprocessing and caret’s implementation of extreme gradient boosting had the highest prediction accuracy, this model will be used to predict the unknown cases from the original dataset.</p>

<p>Uncertainty of prediction is accounted for with a cutoff below 0.7.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">set.seed</span><span class="p">(</span><span class="m">27</span><span class="p">)</span><span class="w">
</span><span class="n">model_xgb_pca</span><span class="w"> </span><span class="o">&lt;-</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                 </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">train_data</span><span class="p">,</span><span class="w">
                 </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"xgbTree"</span><span class="p">,</span><span class="w">
                 </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"pca"</span><span class="p">,</span><span class="w">
                 </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">pred</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_xgb_pca</span><span class="p">,</span><span class="w"> </span><span class="n">test_data</span><span class="p">)</span><span class="w">
</span><span class="n">predb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">predict</span><span class="p">(</span><span class="n">model_xgb_pca</span><span class="p">,</span><span class="w"> </span><span class="n">test_data</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">

</span><span class="n">result</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">case_ID</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">test_data</span><span class="p">),</span><span class="w">
                       </span><span class="n">prediction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pred</span><span class="p">,</span><span class="w">
                       </span><span class="n">predb</span><span class="p">)</span><span class="w">
</span><span class="n">result</span><span class="o">$</span><span class="n">predicted_outcome</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">result</span><span class="o">$</span><span class="n">Death</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0.7</span><span class="p">,</span><span class="w"> </span><span class="s2">"Death"</span><span class="p">,</span><span class="w">
                                   </span><span class="n">ifelse</span><span class="p">(</span><span class="n">result</span><span class="o">$</span><span class="n">Recover</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0.7</span><span class="p">,</span><span class="w"> </span><span class="s2">"Recover"</span><span class="p">,</span><span class="w"> </span><span class="s2">"uncertain"</span><span class="p">))</span><span class="w">
</span><span class="n">result</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##     case_ID prediction Death Recover predicted_outcome
## 1  case_100    Recover  0.08    0.92           Recover
## 2  case_101    Recover  0.11    0.89           Recover
## 3  case_102    Recover  0.04    0.96           Recover
## 4  case_103    Recover  0.01    0.99           Recover
## 5  case_104    Recover  0.19    0.81           Recover
## 6  case_105    Recover  0.01    0.99           Recover
## 7  case_108      Death  0.99    0.01             Death
## 8  case_109    Recover  0.30    0.70         uncertain
## 9  case_110    Recover  0.16    0.84           Recover
## 10 case_112      Death  0.54    0.46         uncertain
## 11 case_113    Recover  0.02    0.98           Recover
## 12 case_114    Recover  0.30    0.70         uncertain
## 13 case_115    Recover  0.02    0.98           Recover
## 14 case_118      Death  0.65    0.35         uncertain
## 15 case_120      Death  0.89    0.11             Death
## 16 case_122    Recover  0.00    1.00           Recover
## 17 case_126    Recover  0.04    0.96           Recover
## 18 case_130    Recover  0.42    0.58         uncertain
## 19 case_132    Recover  0.47    0.53         uncertain
## 20 case_136      Death  0.86    0.14             Death
## 21  case_15      Death  0.99    0.01             Death
## 22  case_16      Death  0.91    0.09             Death
## 23  case_22      Death  0.80    0.20             Death
## 24  case_28      Death  0.91    0.09             Death
## 25  case_31      Death  1.00    0.00             Death
## 26  case_32      Death  0.93    0.07             Death
## 27  case_38      Death  0.87    0.13             Death
## 28  case_39      Death  0.88    0.12             Death
## 29   case_4      Death  0.99    0.01             Death
## 30  case_40      Death  1.00    0.00             Death
## 31  case_41      Death  0.95    0.05             Death
## 32  case_42    Recover  0.49    0.51         uncertain
## 33  case_47      Death  0.91    0.09             Death
## 34  case_48      Death  0.95    0.05             Death
## 35  case_52      Death  0.62    0.38         uncertain
## 36  case_54    Recover  0.43    0.57         uncertain
## 37  case_56      Death  0.81    0.19             Death
## 38  case_62      Death  0.58    0.42         uncertain
## 39  case_63    Recover  0.44    0.56         uncertain
## 40  case_66    Recover  0.18    0.82           Recover
## 41  case_67    Recover  0.02    0.98           Recover
## 42  case_68    Recover  0.27    0.73           Recover
## 43  case_69    Recover  0.35    0.65         uncertain
## 44  case_70    Recover  0.02    0.98           Recover
## 45  case_71    Recover  0.17    0.83           Recover
## 46  case_80    Recover  0.05    0.95           Recover
## 47  case_84    Recover  0.02    0.98           Recover
## 48  case_85      Death  0.99    0.01             Death
## 49  case_86    Recover  0.02    0.98           Recover
## 50  case_88    Recover  0.02    0.98           Recover
## 51   case_9      Death  0.98    0.02             Death
## 52  case_90    Recover  0.08    0.92           Recover
## 53  case_92    Recover  0.04    0.96           Recover
## 54  case_93    Recover  0.43    0.57         uncertain
## 55  case_95    Recover  0.11    0.89           Recover
## 56  case_96      Death  0.68    0.32         uncertain
## 57  case_99    Recover  0.04    0.96           Recover
</code></pre>
</div>

<p><br /></p>

<h2 id="comparison-with-predicted-outcome-from-last-weeks-analyses">Comparison with predicted outcome from last week’s analyses</h2>

<p><br /></p>

<h3 id="plotting-predicted-outcome">Plotting predicted outcome</h3>

<p>Results from last week’s analysis come from the following part of the analysis:</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">results</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">randomForest</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_rf</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">test_data</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
</span><span class="n">glmnet</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_glmnet</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">test_data</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
</span><span class="n">kknn</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_kknn</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">test_data</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
</span><span class="n">pda</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_pda</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">test_data</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
</span><span class="n">slda</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_slda</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">test_data</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
</span><span class="n">pam</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_pam</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">test_data</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
</span><span class="n">C</span><span class="m">5.0</span><span class="n">Tree</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_C5.0Tree</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">test_data</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
</span><span class="n">pls</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_pls</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">test_data</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">))</span><span class="w">

</span><span class="n">results</span><span class="o">$</span><span class="n">sum_Death</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rowSums</span><span class="p">(</span><span class="n">results</span><span class="p">[,</span><span class="w"> </span><span class="n">grep</span><span class="p">(</span><span class="s2">"Death"</span><span class="p">,</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">results</span><span class="p">))])</span><span class="w">
</span><span class="n">results</span><span class="o">$</span><span class="n">sum_Recover</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rowSums</span><span class="p">(</span><span class="n">results</span><span class="p">[,</span><span class="w"> </span><span class="n">grep</span><span class="p">(</span><span class="s2">"Recover"</span><span class="p">,</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">results</span><span class="p">))])</span><span class="w">
</span><span class="n">results</span><span class="o">$</span><span class="n">log2_ratio</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">log2</span><span class="p">(</span><span class="n">results</span><span class="o">$</span><span class="n">sum_Recover</span><span class="o">/</span><span class="n">results</span><span class="o">$</span><span class="n">sum_Death</span><span class="p">)</span><span class="w">
</span><span class="n">results</span><span class="o">$</span><span class="n">predicted_outcome</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">results</span><span class="o">$</span><span class="n">log2_ratio</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">1.5</span><span class="p">,</span><span class="w"> </span><span class="s2">"Recover"</span><span class="p">,</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">results</span><span class="o">$</span><span class="n">log2_ratio</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">-1.5</span><span class="p">,</span><span class="w"> </span><span class="s2">"Death"</span><span class="p">,</span><span class="w"> </span><span class="s2">"uncertain"</span><span class="p">))</span><span class="w">
</span><span class="n">results</span><span class="p">[,</span><span class="w"> </span><span class="o">-</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">16</span><span class="p">)]</span><span class="w">

</span><span class="n">results</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">results_last_week</span><span class="w">
</span></code></pre>
</div>

<p>Combining the table with predicted outcomes from this and last week with the original data.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">result</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">result</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">5</span><span class="p">)],</span><span class="w"> </span><span class="n">results_last_week</span><span class="p">[,</span><span class="w"> </span><span class="n">ncol</span><span class="p">(</span><span class="n">results_last_week</span><span class="p">),</span><span class="w"> </span><span class="n">drop</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">],</span><span class="w"> </span><span class="n">by.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"case_ID"</span><span class="p">,</span><span class="w"> </span><span class="n">by.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"row.names"</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">result</span><span class="p">)[</span><span class="m">2</span><span class="o">:</span><span class="m">3</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"predicted_outcome_xgboost"</span><span class="p">,</span><span class="w"> </span><span class="s2">"predicted_outcome_last_week"</span><span class="p">)</span><span class="w">

</span><span class="n">results_combined</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">result</span><span class="p">,</span><span class="w"> </span><span class="n">fluH7N9.china.2013</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">fluH7N9.china.2013</span><span class="o">$</span><span class="n">case.ID</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">result</span><span class="o">$</span><span class="n">case_ID</span><span class="p">),</span><span class="w"> </span><span class="p">],</span><span class="w"> 
                          </span><span class="n">by.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"case_ID"</span><span class="p">,</span><span class="w"> </span><span class="n">by.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"case.ID"</span><span class="p">)</span><span class="w">
</span><span class="n">results_combined</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">results_combined</span><span class="p">[,</span><span class="w"> </span><span class="o">-</span><span class="nf">c</span><span class="p">(</span><span class="m">6</span><span class="p">,</span><span class="m">7</span><span class="p">)]</span><span class="w">
</span></code></pre>
</div>

<p>For plotting with ggplot2, the dataframe needs to be gathered.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">tidyr</span><span class="p">)</span><span class="w">
</span><span class="n">results_combined_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">results_combined</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">gather</span><span class="p">(</span><span class="n">group_dates</span><span class="p">,</span><span class="w"> </span><span class="n">date</span><span class="p">,</span><span class="w"> </span><span class="n">date.of.onset</span><span class="o">:</span><span class="n">date.of.hospitalisation</span><span class="p">)</span><span class="w">

</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">group_dates</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">group_dates</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"date.of.onset"</span><span class="p">,</span><span class="w"> </span><span class="s2">"date.of.hospitalisation"</span><span class="p">))</span><span class="w">

</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">group_dates</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mapvalues</span><span class="p">(</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">group_dates</span><span class="p">,</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"date.of.onset"</span><span class="p">,</span><span class="w"> </span><span class="s2">"date.of.hospitalisation"</span><span class="p">),</span><span class="w"> 
                                             </span><span class="n">to</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Date of onset"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Date of hospitalisation"</span><span class="p">))</span><span class="w">

</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">gender</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mapvalues</span><span class="p">(</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">gender</span><span class="p">,</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"f"</span><span class="p">,</span><span class="w"> </span><span class="s2">"m"</span><span class="p">),</span><span class="w"> 
                                             </span><span class="n">to</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Female"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Male"</span><span class="p">))</span><span class="w">
</span><span class="n">levels</span><span class="p">(</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">gender</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">levels</span><span class="p">(</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">gender</span><span class="p">),</span><span class="w"> </span><span class="s2">"unknown"</span><span class="p">)</span><span class="w">
</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">gender</span><span class="p">[</span><span class="nf">is.na</span><span class="p">(</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">gender</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"unknown"</span><span class="w">

</span><span class="n">results_combined_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">results_combined_gather</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">gather</span><span class="p">(</span><span class="n">group_pred</span><span class="p">,</span><span class="w"> </span><span class="n">prediction</span><span class="p">,</span><span class="w"> </span><span class="n">predicted_outcome_xgboost</span><span class="o">:</span><span class="n">predicted_outcome_last_week</span><span class="p">)</span><span class="w">

</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">group_pred</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mapvalues</span><span class="p">(</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">group_pred</span><span class="p">,</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"predicted_outcome_xgboost"</span><span class="p">,</span><span class="w"> </span><span class="s2">"predicted_outcome_last_week"</span><span class="p">),</span><span class="w"> 
                                             </span><span class="n">to</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Predicted outcome from XGBoost"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Predicted outcome from last week"</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<p>Setting a custom theme for plotting:</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span><span class="w">
</span><span class="n">my_theme</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">base_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">,</span><span class="w"> </span><span class="n">base_family</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"sans"</span><span class="p">){</span><span class="w">
  </span><span class="n">theme_minimal</span><span class="p">(</span><span class="n">base_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">base_size</span><span class="p">,</span><span class="w"> </span><span class="n">base_family</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">base_family</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="w">
    </span><span class="n">axis.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">),</span><span class="w">
    </span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">45</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">),</span><span class="w">
    </span><span class="n">axis.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">14</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.grid.major</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_line</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.grid.minor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
    </span><span class="n">panel.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"aliceblue"</span><span class="p">),</span><span class="w">
    </span><span class="n">strip.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"lightgrey"</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="w">
    </span><span class="n">strip.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bold"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">),</span><span class="w">
    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bottom"</span><span class="p">,</span><span class="w">
    </span><span class="n">legend.justification</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"top"</span><span class="p">,</span><span class="w"> 
    </span><span class="n">legend.box</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"horizontal"</span><span class="p">,</span><span class="w">
    </span><span class="n">legend.box.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey50"</span><span class="p">),</span><span class="w">
    </span><span class="n">legend.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
    </span><span class="n">panel.border</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.spacing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">unit</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="s2">"lines"</span><span class="p">)</span><span class="w">
  </span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">province</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mapvalues</span><span class="p">(</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">province</span><span class="p">,</span><span class="w"> 
                                                </span><span class="n">from</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Anhui"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Beijing"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Fujian"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Guangdong"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Hebei"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Henan"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Hunan"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Jiangxi"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Shandong"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Taiwan"</span><span class="p">),</span><span class="w"> 
                                                </span><span class="n">to</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="s2">"Other"</span><span class="p">,</span><span class="w"> </span><span class="m">10</span><span class="p">))</span><span class="w">

</span><span class="n">levels</span><span class="p">(</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">gender</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">levels</span><span class="p">(</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">gender</span><span class="p">),</span><span class="w"> </span><span class="s2">"unknown"</span><span class="p">)</span><span class="w">
</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">gender</span><span class="p">[</span><span class="nf">is.na</span><span class="p">(</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">gender</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"unknown"</span><span class="w">

</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">province</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">province</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Jiangsu"</span><span class="p">,</span><span class="w">  </span><span class="s2">"Shanghai"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Zhejiang"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Other"</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">results_combined_gather</span><span class="p">,</span><span class="w"> </span><span class="n">group_dates</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Date of onset"</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">date</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">age</span><span class="p">),</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">prediction</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">stat_density2d</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">..level..</span><span class="p">),</span><span class="w"> </span><span class="n">geom</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"polygon"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_jitter</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">prediction</span><span class="p">,</span><span class="w"> </span><span class="n">shape</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gender</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_rug</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">prediction</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="w">
    </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Predicted outcome"</span><span class="p">,</span><span class="w">
    </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Predicted outcome"</span><span class="p">,</span><span class="w">
    </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Level"</span><span class="p">,</span><span class="w">
    </span><span class="n">shape</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Gender"</span><span class="p">,</span><span class="w">
    </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Date of onset in 2013"</span><span class="p">,</span><span class="w">
    </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Age"</span><span class="p">,</span><span class="w">
    </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"2013 Influenza A H7N9 cases in China"</span><span class="p">,</span><span class="w">
    </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Predicted outcome of cases with unknown outcome"</span><span class="p">,</span><span class="w">
    </span><span class="n">caption</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="w">
  </span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_grid</span><span class="p">(</span><span class="n">group_pred</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">province</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_shape_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">15</span><span class="p">,</span><span class="w"> </span><span class="m">16</span><span class="p">,</span><span class="w"> </span><span class="m">17</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_color_brewer</span><span class="p">(</span><span class="n">palette</span><span class="o">=</span><span class="s2">"Set1"</span><span class="p">,</span><span class="w"> </span><span class="n">na.value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey50"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_brewer</span><span class="p">(</span><span class="n">palette</span><span class="o">=</span><span class="s2">"Set1"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<p><img src="flu_outcome_ML_2_post_files/figure-markdown_github/unnamed-chunk-37-1.png" style="display: block; margin: auto;" /></p>

<p>There is much less uncertainty in the XGBoost data, even tough I used slightly different methods for classifying uncertainty: In last week’s analysis I based uncertainty on the ratio of combined prediction values from all analyses, this week uncertainty is based on the prediction value from one analysis.</p>

<p><br /></p>

<p>For a comparison of feature selection methods see <a href="https://shiring.github.io/machine_learning/2017/01/15/rfe_ga_post">here</a>.</p>

<hr />

<p>If you are interested in more machine learning posts, check out <a href="https://shiring.github.io/categories.html#machine_learning-ref">the category listing for <strong>machine_learning</strong></a>.</p>

<hr />

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sessionInfo</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## R version 3.3.2 (2016-10-31)
## Platform: x86_64-w64-mingw32/x64 (64-bit)
## Running under: Windows 7 x64 (build 7601) Service Pack 1
## 
## locale:
## [1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252 LC_NUMERIC=C                           LC_TIME=English_United States.1252    
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] tidyr_0.6.0     plyr_1.8.4      xgboost_0.6-0   caret_6.0-73    ggplot2_2.2.0   lattice_0.20-34 mice_2.25       Rcpp_0.12.8     dplyr_0.5.0     outbreaks_1.0.0
## 
## loaded via a namespace (and not attached):
##  [1] Ckmeans.1d.dp_3.4.6-4 RColorBrewer_1.1-2    compiler_3.3.2        nloptr_1.0.4          class_7.3-14          iterators_1.0.8       tools_3.3.2           rpart_4.1-10          digest_0.6.10         lme4_1.1-12           evaluate_0.10         tibble_1.2            gtable_0.2.0          nlme_3.1-128          mgcv_1.8-16           Matrix_1.2-7.1        foreach_1.4.3         DBI_0.5-1             parallel_3.3.2        yaml_2.1.14           SparseM_1.74          e1071_1.6-7           stringr_1.1.0         knitr_1.15            MatrixModels_0.4-1    stats4_3.3.2          grid_3.3.2            nnet_7.3-12           data.table_1.9.6      R6_2.2.0              survival_2.40-1       rmarkdown_1.1         minqa_1.2.4           reshape2_1.4.2        car_2.1-3             magrittr_1.5          scales_0.4.1          codetools_0.2-15      ModelMetrics_1.1.0    htmltools_0.3.5       MASS_7.3-45           splines_3.3.2         assertthat_0.1        pbkrtest_0.4-6        colorspace_1.3-0      labeling_0.3          quantreg_5.29         stringi_1.1.2         lazyeval_0.2.0        munsell_0.4.3         chron_2.3-47
</code></pre>
</div>

  </div>

  <h1><a href="/machine_learning/2016/11/27/flu_outcome_ML_post">Can we predict flu deaths with Machine Learning and R?</a></h1>
  <p class="author">
    <span class="date">2016-11-27 00:00:00 +0100</span>
  </p>
  <div class="content">
    <p><strong>Edited on 26 December 2016</strong></p>

<hr />

<p>Among the many R packages, there is the <a href="https://mran.microsoft.com/web/packages/outbreaks/outbreaks.pdf">outbreaks</a> package. It contains datasets on epidemics, on of which is from the 2013 outbreak of <a href="http://www.who.int/influenza/human_animal_interface/faq_H7N9/en/">influenza A H7N9</a> in <a href="http://www.who.int/influenza/human_animal_interface/influenza_h7n9/ChinaH7N9JointMissionReport2013u.pdf?ua=1">China</a>, as analysed by Kucharski et al. (2014):</p>

<blockquote>
  <p>A. Kucharski, H. Mills, A. Pinsent, C. Fraser, M. Van Kerkhove, C. A. Donnelly, and S. Riley. 2014. Distinguishing between reservoir exposure and human-to-human transmission for emerging pathogens using case onset data. PLOS Currents Outbreaks. Mar 7, edition 1. doi: 10.1371/currents.outbreaks.e1473d9bfc99d080ca242139a06c455f.</p>
</blockquote>

<blockquote>
  <p>A. Kucharski, H. Mills, A. Pinsent, C. Fraser, M. Van Kerkhove, C. A. Donnelly, and S. Riley. 2014. Data from: Distinguishing between reservoir exposure and human-to-human transmission for emerging pathogens using case onset data. Dryad Digital Repository. <a href="http://dx.doi.org/10.5061/dryad.2g43n">http://dx.doi.org/10.5061/dryad.2g43n</a>.</p>
</blockquote>

<p>I will be using their data as an example to test whether we can use Machine Learning algorithms for predicting disease outcome.</p>

<p>To do so, I selected and extracted features from the raw data, including age, days between onset and outcome, gender, whether the patients were hospitalised, etc. Missing values were imputed and different model algorithms were used to predict outcome (death or recovery). The prediction accuracy, sensitivity and specificity.
The thus prepared dataset was devided into training and testing subsets. The test subset contained all cases with an unknown outcome. Before I applied the models to the test data, I further split the training data into validation subsets.</p>

<p>The tested modeling algorithms were similarly successful at predicting the outcomes of the validation data. To decide on final classifications, I compared predictions from all models and defined the outcome “Death” or “Recovery” as a function of all models, whereas classifications with a low prediction probability were flagged as “uncertain”. Accounting for this uncertainty led to a 100% correct classification of the validation test set.</p>

<p>The training cases with unknown outcome were then classified based on the same algorithms. From 57 unknown cases, 14 were classified as “Recovery”, 10 as “Death” and 33 as uncertain.</p>

<p><br /></p>

<p>In a <a href="https://shiring.github.io/machine_learning/2016/12/02/flu_outcome_ML_2_post">Part 2</a> I am looking at how extreme gradient boosting performs on this dataset.</p>

<p><br /></p>

<h1 id="the-data">The data</h1>

<p>The dataset contains case ID, date of onset, date of hospitalisation, date of outcome, gender, age, province and of course the outcome: Death or Recovery.
I can already see that there are a couple of missing values in the data, which I will deal with later.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># install and load package
</span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="n">require</span><span class="p">(</span><span class="s2">"outbreaks"</span><span class="p">))</span><span class="w"> </span><span class="n">install.packages</span><span class="p">(</span><span class="s2">"outbreaks"</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">outbreaks</span><span class="p">)</span><span class="w">
</span><span class="n">fluH7N9.china.2013_backup</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">fluH7N9.china.2013</span><span class="w"> </span><span class="c1"># back up original dataset in case something goes awry along the way
</span><span class="w">
</span><span class="c1"># convert ? to NAs
</span><span class="n">fluH7N9.china.2013</span><span class="o">$</span><span class="n">age</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">fluH7N9.china.2013</span><span class="o">$</span><span class="n">age</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"?"</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kc">NA</span><span class="w">

</span><span class="c1"># create a new column with case ID
</span><span class="n">fluH7N9.china.2013</span><span class="o">$</span><span class="n">case.ID</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="s2">"case"</span><span class="p">,</span><span class="w"> </span><span class="n">fluH7N9.china.2013</span><span class="o">$</span><span class="n">case.ID</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">fluH7N9.china.2013</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   case.ID date.of.onset date.of.hospitalisation date.of.outcome outcome gender age province
## 1  case_1    2013-02-19                    &lt;NA&gt;      2013-03-04   Death      m  87 Shanghai
## 2  case_2    2013-02-27              2013-03-03      2013-03-10   Death      m  27 Shanghai
## 3  case_3    2013-03-09              2013-03-19      2013-04-09   Death      f  35    Anhui
## 4  case_4    2013-03-19              2013-03-27            &lt;NA&gt;    &lt;NA&gt;      f  45  Jiangsu
## 5  case_5    2013-03-19              2013-03-30      2013-05-15 Recover      f  48  Jiangsu
## 6  case_6    2013-03-21              2013-03-28      2013-04-26   Death      f  32  Jiangsu
</code></pre>
</div>

<p><br /></p>

<p>Before I start preparing the data for Machine Learning, I want to get an idea of the distribution of the data points and their different variables by plotting.</p>

<p>Most provinces have only a handful of cases, so I am combining them into the category “other” and keep only Jiangsu, Shanghai and Zhejian and separate provinces.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># gather for plotting with ggplot2
</span><span class="n">library</span><span class="p">(</span><span class="n">tidyr</span><span class="p">)</span><span class="w">
</span><span class="n">fluH7N9.china.2013_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">fluH7N9.china.2013</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">gather</span><span class="p">(</span><span class="n">Group</span><span class="p">,</span><span class="w"> </span><span class="n">Date</span><span class="p">,</span><span class="w"> </span><span class="n">date.of.onset</span><span class="o">:</span><span class="n">date.of.outcome</span><span class="p">)</span><span class="w">

</span><span class="c1"># rearrange group order
</span><span class="n">fluH7N9.china.2013_gather</span><span class="o">$</span><span class="n">Group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">fluH7N9.china.2013_gather</span><span class="o">$</span><span class="n">Group</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"date.of.onset"</span><span class="p">,</span><span class="w"> </span><span class="s2">"date.of.hospitalisation"</span><span class="p">,</span><span class="w"> </span><span class="s2">"date.of.outcome"</span><span class="p">))</span><span class="w">

</span><span class="c1"># rename groups
</span><span class="n">library</span><span class="p">(</span><span class="n">plyr</span><span class="p">)</span><span class="w">
</span><span class="n">fluH7N9.china.2013_gather</span><span class="o">$</span><span class="n">Group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mapvalues</span><span class="p">(</span><span class="n">fluH7N9.china.2013_gather</span><span class="o">$</span><span class="n">Group</span><span class="p">,</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"date.of.onset"</span><span class="p">,</span><span class="w"> </span><span class="s2">"date.of.hospitalisation"</span><span class="p">,</span><span class="w"> </span><span class="s2">"date.of.outcome"</span><span class="p">),</span><span class="w"> 
          </span><span class="n">to</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Date of onset"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Date of hospitalisation"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Date of outcome"</span><span class="p">))</span><span class="w">

</span><span class="c1"># renaming provinces
</span><span class="n">fluH7N9.china.2013_gather</span><span class="o">$</span><span class="n">province</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mapvalues</span><span class="p">(</span><span class="n">fluH7N9.china.2013_gather</span><span class="o">$</span><span class="n">province</span><span class="p">,</span><span class="w"> 
                                                </span><span class="n">from</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Anhui"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Beijing"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Fujian"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Guangdong"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Hebei"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Henan"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Hunan"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Jiangxi"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Shandong"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Taiwan"</span><span class="p">),</span><span class="w"> 
                                                </span><span class="n">to</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="s2">"Other"</span><span class="p">,</span><span class="w"> </span><span class="m">10</span><span class="p">))</span><span class="w">

</span><span class="c1"># add a level for unknown gender
</span><span class="n">levels</span><span class="p">(</span><span class="n">fluH7N9.china.2013_gather</span><span class="o">$</span><span class="n">gender</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">levels</span><span class="p">(</span><span class="n">fluH7N9.china.2013_gather</span><span class="o">$</span><span class="n">gender</span><span class="p">),</span><span class="w"> </span><span class="s2">"unknown"</span><span class="p">)</span><span class="w">
</span><span class="n">fluH7N9.china.2013_gather</span><span class="o">$</span><span class="n">gender</span><span class="p">[</span><span class="nf">is.na</span><span class="p">(</span><span class="n">fluH7N9.china.2013_gather</span><span class="o">$</span><span class="n">gender</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"unknown"</span><span class="w">

</span><span class="c1"># rearrange province order so that Other is the last
</span><span class="n">fluH7N9.china.2013_gather</span><span class="o">$</span><span class="n">province</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">fluH7N9.china.2013_gather</span><span class="o">$</span><span class="n">province</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Jiangsu"</span><span class="p">,</span><span class="w">  </span><span class="s2">"Shanghai"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Zhejiang"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Other"</span><span class="p">))</span><span class="w">

</span><span class="c1"># convert age to numeric
</span><span class="n">fluH7N9.china.2013_gather</span><span class="o">$</span><span class="n">age</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">fluH7N9.china.2013_gather</span><span class="o">$</span><span class="n">age</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># preparing my ggplot2 theme of choice
</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span><span class="w">
</span><span class="n">my_theme</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">base_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">,</span><span class="w"> </span><span class="n">base_family</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"sans"</span><span class="p">){</span><span class="w">
  </span><span class="n">theme_minimal</span><span class="p">(</span><span class="n">base_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">base_size</span><span class="p">,</span><span class="w"> </span><span class="n">base_family</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">base_family</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="w">
    </span><span class="n">axis.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">),</span><span class="w">
    </span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">45</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">),</span><span class="w">
    </span><span class="n">axis.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">14</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.grid.major</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_line</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.grid.minor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
    </span><span class="n">panel.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"aliceblue"</span><span class="p">),</span><span class="w">
    </span><span class="n">strip.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"lightgrey"</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="w">
    </span><span class="n">strip.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bold"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">),</span><span class="w">
    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bottom"</span><span class="p">,</span><span class="w">
    </span><span class="n">legend.justification</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"top"</span><span class="p">,</span><span class="w"> 
    </span><span class="n">legend.box</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"horizontal"</span><span class="p">,</span><span class="w">
    </span><span class="n">legend.box.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey50"</span><span class="p">),</span><span class="w">
    </span><span class="n">legend.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
    </span><span class="n">panel.border</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w">
  </span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># plotting raw data
</span><span class="w">
</span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">fluH7N9.china.2013_gather</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Date</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">age</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">outcome</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">stat_density2d</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">..level..</span><span class="p">),</span><span class="w"> </span><span class="n">geom</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"polygon"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_jitter</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">outcome</span><span class="p">,</span><span class="w"> </span><span class="n">shape</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gender</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1.5</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_rug</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">outcome</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="w">
    </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Outcome"</span><span class="p">,</span><span class="w">
    </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Outcome"</span><span class="p">,</span><span class="w">
    </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Level"</span><span class="p">,</span><span class="w">
    </span><span class="n">shape</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Gender"</span><span class="p">,</span><span class="w">
    </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Date in 2013"</span><span class="p">,</span><span class="w">
    </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Age"</span><span class="p">,</span><span class="w">
    </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"2013 Influenza A H7N9 cases in China"</span><span class="p">,</span><span class="w">
    </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Dataset from 'outbreaks' package (Kucharski et al. 2014)"</span><span class="p">,</span><span class="w">
    </span><span class="n">caption</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="w">
  </span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_grid</span><span class="p">(</span><span class="n">Group</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">province</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_shape_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">15</span><span class="p">,</span><span class="w"> </span><span class="m">16</span><span class="p">,</span><span class="w"> </span><span class="m">17</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_color_brewer</span><span class="p">(</span><span class="n">palette</span><span class="o">=</span><span class="s2">"Set1"</span><span class="p">,</span><span class="w"> </span><span class="n">na.value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey50"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_brewer</span><span class="p">(</span><span class="n">palette</span><span class="o">=</span><span class="s2">"Set1"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="flu_outcome_ML_post_files/figure-markdown_github/unnamed-chunk-4-1.png" style="display: block; margin: auto;" /></p>

<p>This plot shows the dates of onset, hospitalisation and outcome (if known) of each data point. Outcome is marked by color and age shown on the y-axis. Gender is marked by point shape.</p>

<p>The density distribution of date by age for the cases seems to indicate that older people died more frequently in the Jiangsu and Zhejiang province than in Shanghai and in other provinces.</p>

<p>When we look at the distribution of points along the time axis, it suggests that their might be a positive correlation between the likelihood of death and an early onset or early outcome.</p>

<p><br /></p>

<p>I also want to know how many cases there are for each gender and province and compare the genders’ age distribution.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">fluH7N9.china.2013_gather_2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">fluH7N9.china.2013_gather</span><span class="p">[,</span><span class="w"> </span><span class="m">-4</span><span class="p">]</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">gather</span><span class="p">(</span><span class="n">group_2</span><span class="p">,</span><span class="w"> </span><span class="n">value</span><span class="p">,</span><span class="w"> </span><span class="n">gender</span><span class="o">:</span><span class="n">province</span><span class="p">)</span><span class="w">

</span><span class="n">fluH7N9.china.2013_gather_2</span><span class="o">$</span><span class="n">value</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mapvalues</span><span class="p">(</span><span class="n">fluH7N9.china.2013_gather_2</span><span class="o">$</span><span class="n">value</span><span class="p">,</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"m"</span><span class="p">,</span><span class="w"> </span><span class="s2">"f"</span><span class="p">,</span><span class="w"> </span><span class="s2">"unknown"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Other"</span><span class="p">),</span><span class="w"> 
          </span><span class="n">to</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Male"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Female"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Unknown gender"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Other province"</span><span class="p">))</span><span class="w">

</span><span class="n">fluH7N9.china.2013_gather_2</span><span class="o">$</span><span class="n">value</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">fluH7N9.china.2013_gather_2</span><span class="o">$</span><span class="n">value</span><span class="p">,</span><span class="w"> 
                                            </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Female"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Male"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Unknown gender"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Jiangsu"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Shanghai"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Zhejiang"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Other province"</span><span class="p">))</span><span class="w">

</span><span class="n">p</span><span class="m">1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">fluH7N9.china.2013_gather_2</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">value</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">outcome</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">outcome</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_bar</span><span class="p">(</span><span class="n">position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"dodge"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_brewer</span><span class="p">(</span><span class="n">palette</span><span class="o">=</span><span class="s2">"Set1"</span><span class="p">,</span><span class="w"> </span><span class="n">na.value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey50"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_color_brewer</span><span class="p">(</span><span class="n">palette</span><span class="o">=</span><span class="s2">"Set1"</span><span class="p">,</span><span class="w"> </span><span class="n">na.value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey50"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="w">
    </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w">
    </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w">
    </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w">
    </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Count"</span><span class="p">,</span><span class="w">
    </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"2013 Influenza A H7N9 cases in China"</span><span class="p">,</span><span class="w">
    </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Gender and Province numbers of flu cases"</span><span class="p">,</span><span class="w">
    </span><span class="n">caption</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="w">
  </span><span class="p">)</span><span class="w">

</span><span class="n">p</span><span class="m">2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">fluH7N9.china.2013_gather</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">age</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">outcome</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">outcome</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_density</span><span class="p">(</span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_rug</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_color_brewer</span><span class="p">(</span><span class="n">palette</span><span class="o">=</span><span class="s2">"Set1"</span><span class="p">,</span><span class="w"> </span><span class="n">na.value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey50"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_brewer</span><span class="p">(</span><span class="n">palette</span><span class="o">=</span><span class="s2">"Set1"</span><span class="p">,</span><span class="w"> </span><span class="n">na.value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey50"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="w">
    </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w">
    </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w">
    </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Age"</span><span class="p">,</span><span class="w">
    </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Density"</span><span class="p">,</span><span class="w">
    </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w">
    </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Age distribution of flu cases"</span><span class="p">,</span><span class="w">
    </span><span class="n">caption</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="w">
  </span><span class="p">)</span><span class="w">

</span><span class="n">library</span><span class="p">(</span><span class="n">gridExtra</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">grid</span><span class="p">)</span><span class="w">

</span><span class="n">grid.arrange</span><span class="p">(</span><span class="n">p</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="flu_outcome_ML_post_files/figure-markdown_github/unnamed-chunk-5-1.png" style="display: block; margin: auto;" /></p>

<p>In the dataset, there are more male than female cases and correspondingly, we see more deaths, recoveries and unknown outcomes in men than in women. This is potentially a problem later on for modeling because the inherent likelihoods for outcome are not directly comparable between the sexes.</p>

<p>Most unknown outcomes were recorded in Zhejiang. Similarly to gender, we don’t have an equal distribution of data points across provinces either.</p>

<p>When we look at the age distribution it is obvious that people who died tended to be slightly older than those who recovered. The density curve of unknown outcomes is more similar to that of death than of recovery, suggesting that among these people there might have been more deaths than recoveries.</p>

<p><br /></p>

<p>And lastly, I want to plot how many days passed between onset, hospitalisation and outcome for each case.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">fluH7N9.china.2013_gather</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Date</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">age</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">outcome</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">shape</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gender</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1.5</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.6</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_path</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">case.ID</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_wrap</span><span class="p">(</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">province</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_shape_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">15</span><span class="p">,</span><span class="w"> </span><span class="m">16</span><span class="p">,</span><span class="w"> </span><span class="m">17</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_color_brewer</span><span class="p">(</span><span class="n">palette</span><span class="o">=</span><span class="s2">"Set1"</span><span class="p">,</span><span class="w"> </span><span class="n">na.value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey50"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_brewer</span><span class="p">(</span><span class="n">palette</span><span class="o">=</span><span class="s2">"Set1"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="w">
    </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Outcome"</span><span class="p">,</span><span class="w">
    </span><span class="n">shape</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Gender"</span><span class="p">,</span><span class="w">
    </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Date in 2013"</span><span class="p">,</span><span class="w">
    </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Age"</span><span class="p">,</span><span class="w">
    </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"2013 Influenza A H7N9 cases in China"</span><span class="p">,</span><span class="w">
    </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Dataset from 'outbreaks' package (Kucharski et al. 2014)"</span><span class="p">,</span><span class="w">
    </span><span class="n">caption</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"\nTime from onset of flu to outcome."</span><span class="w">
  </span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="flu_outcome_ML_post_files/figure-markdown_github/unnamed-chunk-6-1.png" style="display: block; margin: auto;" /></p>

<p>This plot shows that there are many missing values in the dates, so it is hard to draw a general conclusion.</p>

<p><br /></p>

<h1 id="features">Features</h1>

<p>In Machine Learning-speak features are the variables used for model training. Using the right features dramatically influences the accuracy of the model.</p>

<p>Because we don’t have many features, I am keeping age as it is, but I am also generating new features:</p>

<ul>
  <li>from the date information I am calculating the days between onset and outcome and between onset and hospitalisation</li>
  <li>I am converting gender into numeric values with 1 for female and 0 for male</li>
  <li>similarly, I am converting provinces to binary classifiers (yes == 1, no == 0) for Shanghai, Zhejiang, Jiangsu and other provinces</li>
  <li>the same binary classification is given for whether a case was hospitalised, and whether they had an early onset or early outcome (earlier than the median date)</li>
</ul>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># preparing the data frame for modeling
# 
</span><span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span><span class="w">

</span><span class="n">dataset</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">fluH7N9.china.2013</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">mutate</span><span class="p">(</span><span class="n">hospital</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.factor</span><span class="p">(</span><span class="n">ifelse</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">date.of.hospitalisation</span><span class="p">),</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">)),</span><span class="w">
         </span><span class="n">gender_f</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.factor</span><span class="p">(</span><span class="n">ifelse</span><span class="p">(</span><span class="n">gender</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"f"</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">)),</span><span class="w">
         </span><span class="n">province_Jiangsu</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.factor</span><span class="p">(</span><span class="n">ifelse</span><span class="p">(</span><span class="n">province</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Jiangsu"</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">)),</span><span class="w">
         </span><span class="n">province_Shanghai</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.factor</span><span class="p">(</span><span class="n">ifelse</span><span class="p">(</span><span class="n">province</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Shanghai"</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">)),</span><span class="w">
         </span><span class="n">province_Zhejiang</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.factor</span><span class="p">(</span><span class="n">ifelse</span><span class="p">(</span><span class="n">province</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Zhejiang"</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">)),</span><span class="w">
         </span><span class="n">province_other</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.factor</span><span class="p">(</span><span class="n">ifelse</span><span class="p">(</span><span class="n">province</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Zhejiang"</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">province</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Jiangsu"</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">province</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Shanghai"</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">)),</span><span class="w">
         </span><span class="n">days_onset_to_outcome</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">gsub</span><span class="p">(</span><span class="s2">" days"</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w">
                                      </span><span class="n">as.Date</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">date.of.outcome</span><span class="p">),</span><span class="w"> </span><span class="n">format</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"%Y-%m-%d"</span><span class="p">)</span><span class="w"> </span><span class="o">-</span><span class="w"> 
                                        </span><span class="n">as.Date</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">date.of.onset</span><span class="p">),</span><span class="w"> </span><span class="n">format</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"%Y-%m-%d"</span><span class="p">)))),</span><span class="w">
         </span><span class="n">days_onset_to_hospital</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">gsub</span><span class="p">(</span><span class="s2">" days"</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w">
                                      </span><span class="n">as.Date</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">date.of.hospitalisation</span><span class="p">),</span><span class="w"> </span><span class="n">format</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"%Y-%m-%d"</span><span class="p">)</span><span class="w"> </span><span class="o">-</span><span class="w"> 
                                        </span><span class="n">as.Date</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">date.of.onset</span><span class="p">),</span><span class="w"> </span><span class="n">format</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"%Y-%m-%d"</span><span class="p">)))),</span><span class="w">
         </span><span class="n">age</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">age</span><span class="p">)),</span><span class="w">
         </span><span class="n">early_onset</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.factor</span><span class="p">(</span><span class="n">ifelse</span><span class="p">(</span><span class="n">date.of.onset</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="n">summary</span><span class="p">(</span><span class="n">fluH7N9.china.2013</span><span class="o">$</span><span class="n">date.of.onset</span><span class="p">)[[</span><span class="m">3</span><span class="p">]],</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">)),</span><span class="w">
         </span><span class="n">early_outcome</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.factor</span><span class="p">(</span><span class="n">ifelse</span><span class="p">(</span><span class="n">date.of.outcome</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="n">summary</span><span class="p">(</span><span class="n">fluH7N9.china.2013</span><span class="o">$</span><span class="n">date.of.outcome</span><span class="p">)[[</span><span class="m">3</span><span class="p">]],</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">)))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="n">select</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">-</span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="o">:</span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="m">6</span><span class="p">,</span><span class="w"> </span><span class="m">8</span><span class="p">))</span><span class="w">
</span><span class="n">rownames</span><span class="p">(</span><span class="n">dataset</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">dataset</span><span class="o">$</span><span class="n">case.ID</span><span class="w">
</span><span class="n">dataset</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">dataset</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">dataset</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##        outcome age hospital gender_f province_Jiangsu province_Shanghai province_Zhejiang province_other days_onset_to_outcome days_onset_to_hospital early_onset early_outcome
## case_1   Death  87        0        0                0                 1                 0              0                    13                     NA           1             1
## case_2   Death  27        1        0                0                 1                 0              0                    11                      4           1             1
## case_3   Death  35        1        1                0                 0                 0              1                    31                     10           1             1
## case_4    &lt;NA&gt;  45        1        1                1                 0                 0              0                    NA                      8           1          &lt;NA&gt;
## case_5 Recover  48        1        1                1                 0                 0              0                    57                     11           1             0
## case_6   Death  32        1        1                1                 0                 0              0                    36                      7           1             1
</code></pre>
</div>

<p><br /></p>

<h2 id="imputing-missing-values">Imputing missing values</h2>

<p>When looking at the dataset I created for modeling, it is obvious that we have quite a few missing values.</p>

<p>The missing values from the outcome column are what I want to predict but for the rest I would either have to remove the entire row from the data or impute the missing information. I decided to try the latter with the <a href="https://www.r-bloggers.com/imputing-missing-data-with-r-mice-package">mice</a> package.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># impute missing data
</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">mice</span><span class="p">)</span><span class="w">

</span><span class="n">dataset_impute</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mice</span><span class="p">(</span><span class="n">dataset</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w">  </span><span class="n">print</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span><span class="n">dataset_impute</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Multiply imputed dataset
## Call:
## mice(data = dataset[, -1], printFlag = FALSE)
## Number of multiple imputations:  5
## Missing cells per column:
##                    age               hospital               gender_f       province_Jiangsu      province_Shanghai      province_Zhejiang         province_other  days_onset_to_outcome days_onset_to_hospital            early_onset          early_outcome 
##                      2                      0                      2                      0                      0                      0                      0                     67                     74                     10                     65 
## Imputation methods:
##                    age               hospital               gender_f       province_Jiangsu      province_Shanghai      province_Zhejiang         province_other  days_onset_to_outcome days_onset_to_hospital            early_onset          early_outcome 
##                  "pmm"                     ""               "logreg"                     ""                     ""                     ""                     ""                  "pmm"                  "pmm"               "logreg"               "logreg" 
## VisitSequence:
##                    age               gender_f  days_onset_to_outcome days_onset_to_hospital            early_onset          early_outcome 
##                      1                      3                      8                      9                     10                     11 
## PredictorMatrix:
##                        age hospital gender_f province_Jiangsu province_Shanghai province_Zhejiang province_other days_onset_to_outcome days_onset_to_hospital early_onset early_outcome
## age                      0        1        1                1                 1                 1              1                     1                      1           1             1
## hospital                 0        0        0                0                 0                 0              0                     0                      0           0             0
## gender_f                 1        1        0                1                 1                 1              1                     1                      1           1             1
## province_Jiangsu         0        0        0                0                 0                 0              0                     0                      0           0             0
## province_Shanghai        0        0        0                0                 0                 0              0                     0                      0           0             0
## province_Zhejiang        0        0        0                0                 0                 0              0                     0                      0           0             0
## province_other           0        0        0                0                 0                 0              0                     0                      0           0             0
## days_onset_to_outcome    1        1        1                1                 1                 1              1                     0                      1           1             1
## days_onset_to_hospital   1        1        1                1                 1                 1              1                     1                      0           1             1
## early_onset              1        1        1                1                 1                 1              1                     1                      1           0             1
## early_outcome            1        1        1                1                 1                 1              1                     1                      1           1             0
## Random generator seed value:  NA
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># recombine imputed data frame with the outcome column
</span><span class="w">
</span><span class="n">dataset_complete</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">dataset</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">drop</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">],</span><span class="w"> </span><span class="n">mice</span><span class="o">::</span><span class="n">complete</span><span class="p">(</span><span class="n">dataset_impute</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"row.names"</span><span class="p">,</span><span class="w"> </span><span class="n">all</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">rownames</span><span class="p">(</span><span class="n">dataset_complete</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">dataset_complete</span><span class="o">$</span><span class="n">Row.names</span><span class="w">
</span><span class="n">dataset_complete</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">dataset_complete</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<h1 id="test-train-and-validation-datasets">Test, train and validation datasets</h1>

<p>For building the model, I am separating the imputed data frame into training and test data. Test data are the 57 cases with unknown outcome.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">summary</span><span class="p">(</span><span class="n">dataset</span><span class="o">$</span><span class="n">outcome</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   Death Recover    NA's 
##      32      47      57
</code></pre>
</div>

<p><br /></p>

<p>The training data will be further devided for validation of the models: 70% of the training data will be kept for model building and the remaining 30% will be used for model testing.</p>

<p>I am using the <a href="http://topepo.github.io/caret/index.html">caret</a> package for modeling.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">train_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">which</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">dataset_complete</span><span class="o">$</span><span class="n">outcome</span><span class="p">))</span><span class="w">
</span><span class="n">train_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">dataset_complete</span><span class="p">[</span><span class="o">-</span><span class="n">train_index</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="n">test_data</span><span class="w">  </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">dataset_complete</span><span class="p">[</span><span class="n">train_index</span><span class="p">,</span><span class="w"> </span><span class="m">-1</span><span class="p">]</span><span class="w">

</span><span class="n">library</span><span class="p">(</span><span class="n">caret</span><span class="p">)</span><span class="w">
</span><span class="n">set.seed</span><span class="p">(</span><span class="m">27</span><span class="p">)</span><span class="w">
</span><span class="n">val_index</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">createDataPartition</span><span class="p">(</span><span class="n">train_data</span><span class="o">$</span><span class="n">outcome</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">,</span><span class="w"> </span><span class="n">list</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span><span class="n">val_train_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">train_data</span><span class="p">[</span><span class="n">val_index</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="n">val_test_data</span><span class="w">  </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">train_data</span><span class="p">[</span><span class="o">-</span><span class="n">val_index</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="n">val_train_X</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">val_train_data</span><span class="p">[,</span><span class="m">-1</span><span class="p">]</span><span class="w">
</span><span class="n">val_test_X</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="m">-1</span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<h2 id="decision-trees">Decision trees</h2>

<p>To get an idea about how each feature contributes to the prediction of the outcome, I first built a decision tree based on the training data using <a href="https://cran.r-project.org/web/packages/rpart/rpart.pdf">rpart</a> and <a href="https://cran.r-project.org/web/packages/rattle/vignettes/rattle.pdf">rattle</a>.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">rpart</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">rattle</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">rpart.plot</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">RColorBrewer</span><span class="p">)</span><span class="w">

</span><span class="n">set.seed</span><span class="p">(</span><span class="m">27</span><span class="p">)</span><span class="w">
</span><span class="n">fit</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rpart</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                    </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">train_data</span><span class="p">,</span><span class="w">
                    </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"class"</span><span class="p">,</span><span class="w">
                    </span><span class="n">control</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rpart.control</span><span class="p">(</span><span class="n">xval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">minbucket</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">cp</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">),</span><span class="w"> </span><span class="n">parms</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">split</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"information"</span><span class="p">))</span><span class="w">

</span><span class="n">fancyRpartPlot</span><span class="p">(</span><span class="n">fit</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="flu_outcome_ML_post_files/figure-markdown_github/unnamed-chunk-12-1.png" style="display: block; margin: auto;" /></p>

<p>This randomly generated decision tree shows that cases with an early outcome were most likely to die when they were 68 or older, when they also had an early onset and when they were sick for fewer than 13 days. If a person was not among the first cases and was younger than 52, they had a good chance of recovering, but if they were 82 or older, they were more likely to die from the flu.</p>

<p><br /></p>

<h2 id="feature-importance">Feature Importance</h2>

<p>Not all of the features I created will be equally important to the model. The decision tree already gave me an idea of which features might be most important but I also want to estimate feature importance using a Random Forest approach with repeated cross validation.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># prepare training scheme
</span><span class="n">control</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">)</span><span class="w">

</span><span class="c1"># train the model
</span><span class="n">set.seed</span><span class="p">(</span><span class="m">27</span><span class="p">)</span><span class="w">
</span><span class="n">model</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">train_data</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"rf"</span><span class="p">,</span><span class="w"> </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">control</span><span class="p">)</span><span class="w">

</span><span class="c1"># estimate variable importance
</span><span class="n">importance</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">varImp</span><span class="p">(</span><span class="n">model</span><span class="p">,</span><span class="w"> </span><span class="n">scale</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># prepare for plotting
</span><span class="n">importance_df_1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">importance</span><span class="o">$</span><span class="n">importance</span><span class="w">
</span><span class="n">importance_df_1</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">importance_df_1</span><span class="p">)</span><span class="w">

</span><span class="n">importance_df_1</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mapvalues</span><span class="p">(</span><span class="n">importance_df_1</span><span class="o">$</span><span class="n">group</span><span class="p">,</span><span class="w"> 
                                           </span><span class="n">from</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"age"</span><span class="p">,</span><span class="w"> </span><span class="s2">"hospital2"</span><span class="p">,</span><span class="w"> </span><span class="s2">"gender_f2"</span><span class="p">,</span><span class="w"> </span><span class="s2">"province_Jiangsu2"</span><span class="p">,</span><span class="w"> </span><span class="s2">"province_Shanghai2"</span><span class="p">,</span><span class="w"> </span><span class="s2">"province_Zhejiang2"</span><span class="p">,</span><span class="w">
                                                    </span><span class="s2">"province_other2"</span><span class="p">,</span><span class="w"> </span><span class="s2">"days_onset_to_outcome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"days_onset_to_hospital"</span><span class="p">,</span><span class="w"> </span><span class="s2">"early_onset2"</span><span class="p">,</span><span class="w"> </span><span class="s2">"early_outcome2"</span><span class="w"> </span><span class="p">),</span><span class="w"> 
                                           </span><span class="n">to</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Age"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Hospital"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Female"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Jiangsu"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Shanghai"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Zhejiang"</span><span class="p">,</span><span class="w">
                                                    </span><span class="s2">"Other province"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Days onset to outcome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Days onset to hospital"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Early onset"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Early outcome"</span><span class="w"> </span><span class="p">))</span><span class="w">
</span><span class="n">f</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">importance_df_1</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">importance_df_1</span><span class="o">$</span><span class="n">Overall</span><span class="p">,</span><span class="w"> </span><span class="n">decreasing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">),</span><span class="w"> </span><span class="s2">"group"</span><span class="p">]</span><span class="w">

</span><span class="n">importance_df_2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">importance_df_1</span><span class="w">
</span><span class="n">importance_df_2</span><span class="o">$</span><span class="n">Overall</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span><span class="w">

</span><span class="n">importance_df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rbind</span><span class="p">(</span><span class="n">importance_df_1</span><span class="p">,</span><span class="w"> </span><span class="n">importance_df_2</span><span class="p">)</span><span class="w">

</span><span class="c1"># setting factor levels
</span><span class="n">importance_df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">within</span><span class="p">(</span><span class="n">importance_df</span><span class="p">,</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">f</span><span class="p">))</span><span class="w">
</span><span class="n">importance_df_1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">within</span><span class="p">(</span><span class="n">importance_df_1</span><span class="p">,</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">f</span><span class="p">))</span><span class="w">

</span><span class="n">ggplot</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">importance_df_1</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Overall</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">group</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_path</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">importance_df</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Overall</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">group</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_color_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="n">brewer.pal</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="s2">"Set1"</span><span class="p">)[</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="m">11</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"none"</span><span class="p">,</span><span class="w">
        </span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="w">
    </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Importance"</span><span class="p">,</span><span class="w">
    </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w">
    </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"2013 Influenza A H7N9 cases in China"</span><span class="p">,</span><span class="w">
    </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Scaled feature importance"</span><span class="p">,</span><span class="w">
    </span><span class="n">caption</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"\nDetermined with Random Forest and
    repeated cross validation (10 repeats, 10 times)"</span><span class="w">
  </span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="flu_outcome_ML_post_files/figure-markdown_github/unnamed-chunk-14-1.png" style="display: block; margin: auto;" /></p>

<p>This tells me that age is the most important determining factor for predicting disease outcome, followed by days between onset an outcome, early outcome and days between onset and hospitalisation.</p>

<p><br /></p>

<h2 id="feature-plot">Feature Plot</h2>

<p>Before I start actually building models, I want to check whether the distribution of feature values is comparable between training, validation and test datasets.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># prepare for plotting
</span><span class="n">dataset_complete_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">dataset_complete</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">mutate</span><span class="p">(</span><span class="n">set</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">rownames</span><span class="p">(</span><span class="n">dataset_complete</span><span class="p">)</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">test_data</span><span class="p">),</span><span class="w"> </span><span class="s2">"Test Data"</span><span class="p">,</span><span class="w"> 
                               </span><span class="n">ifelse</span><span class="p">(</span><span class="n">rownames</span><span class="p">(</span><span class="n">dataset_complete</span><span class="p">)</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">val_train_data</span><span class="p">),</span><span class="w"> </span><span class="s2">"Validation Train Data"</span><span class="p">,</span><span class="w">
                                      </span><span class="n">ifelse</span><span class="p">(</span><span class="n">rownames</span><span class="p">(</span><span class="n">dataset_complete</span><span class="p">)</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">val_test_data</span><span class="p">),</span><span class="w"> </span><span class="s2">"Validation Test Data"</span><span class="p">,</span><span class="w"> </span><span class="s2">"NA"</span><span class="p">))),</span><span class="w">
         </span><span class="n">case_ID</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">.</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">gather</span><span class="p">(</span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">value</span><span class="p">,</span><span class="w"> </span><span class="n">age</span><span class="o">:</span><span class="n">early_outcome</span><span class="p">)</span><span class="w">

</span><span class="n">dataset_complete_gather</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mapvalues</span><span class="p">(</span><span class="n">dataset_complete_gather</span><span class="o">$</span><span class="n">group</span><span class="p">,</span><span class="w"> 
                                           </span><span class="n">from</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"age"</span><span class="p">,</span><span class="w"> </span><span class="s2">"hospital"</span><span class="p">,</span><span class="w"> </span><span class="s2">"gender_f"</span><span class="p">,</span><span class="w"> </span><span class="s2">"province_Jiangsu"</span><span class="p">,</span><span class="w"> </span><span class="s2">"province_Shanghai"</span><span class="p">,</span><span class="w"> </span><span class="s2">"province_Zhejiang"</span><span class="p">,</span><span class="w">
                                                    </span><span class="s2">"province_other"</span><span class="p">,</span><span class="w"> </span><span class="s2">"days_onset_to_outcome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"days_onset_to_hospital"</span><span class="p">,</span><span class="w"> </span><span class="s2">"early_onset"</span><span class="p">,</span><span class="w"> </span><span class="s2">"early_outcome"</span><span class="w"> </span><span class="p">),</span><span class="w"> 
                                           </span><span class="n">to</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Age"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Hospital"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Female"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Jiangsu"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Shanghai"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Zhejiang"</span><span class="p">,</span><span class="w">
                                                    </span><span class="s2">"Other province"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Days onset to outcome"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Days onset to hospital"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Early onset"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Early outcome"</span><span class="w"> </span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dataset_complete_gather</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">value</span><span class="p">),</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">outcome</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">outcome</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_density</span><span class="p">(</span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.2</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_rug</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_color_brewer</span><span class="p">(</span><span class="n">palette</span><span class="o">=</span><span class="s2">"Set1"</span><span class="p">,</span><span class="w"> </span><span class="n">na.value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey50"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_brewer</span><span class="p">(</span><span class="n">palette</span><span class="o">=</span><span class="s2">"Set1"</span><span class="p">,</span><span class="w"> </span><span class="n">na.value</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey50"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_wrap</span><span class="p">(</span><span class="n">set</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">11</span><span class="p">,</span><span class="w"> </span><span class="n">scales</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"free"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="w">
    </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Value"</span><span class="p">,</span><span class="w">
    </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Density"</span><span class="p">,</span><span class="w">
    </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"2013 Influenza A H7N9 cases in China"</span><span class="p">,</span><span class="w">
    </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Features for classifying outcome"</span><span class="p">,</span><span class="w">
    </span><span class="n">caption</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"\nDensity distribution of all features used for classification of flu outcome."</span><span class="w">
  </span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">subset</span><span class="p">(</span><span class="n">dataset_complete_gather</span><span class="p">,</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Age"</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Days onset to hospital"</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Days onset to outcome"</span><span class="p">),</span><span class="w"> 
       </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">outcome</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="o">=</span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">value</span><span class="p">),</span><span class="w"> </span><span class="n">fill</span><span class="o">=</span><span class="n">set</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">geom_boxplot</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_brewer</span><span class="p">(</span><span class="n">palette</span><span class="o">=</span><span class="s2">"Set1"</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"div "</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_wrap</span><span class="p">(</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">scales</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"free"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="w">
    </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w">
    </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Outcome"</span><span class="p">,</span><span class="w">
    </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Value"</span><span class="p">,</span><span class="w">
    </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"2013 Influenza A H7N9 cases in China"</span><span class="p">,</span><span class="w">
    </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Features for classifying outcome"</span><span class="p">,</span><span class="w">
    </span><span class="n">caption</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"\nBoxplot of the features age, days from onset to hospitalisation and days from onset to outcome."</span><span class="w">
  </span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="flu_outcome_ML_post_files/figure-markdown_github/unnamed-chunk-16-1.png" style="display: block; margin: auto;" /></p>

<p><img src="flu_outcome_ML_post_files/figure-markdown_github/unnamed-chunk-17-1.png" style="display: block; margin: auto;" /></p>

<p>Luckily, the distributions looks reasonably similar between the validation and test data, except for a few outliers.</p>

<p><br /></p>

<hr />

<h1 id="comparing-machine-learning-algorithms">Comparing Machine Learning algorithms</h1>

<p>Before I try to predict the outcome of the unknown cases, I am testing the models’ accuracy with the validation datasets on a couple of algorithms. I have chosen only a few more well known algorithms, but <a href="http://topepo.github.io/caret/index.html">caret</a> implements many more.</p>

<p>I have chose to not do any preprocessing because I was worried that the different data distributions with continuous variables (e.g. age) and binary variables (i.e. 0, 1 classification of e.g. hospitalisation) would lead to problems.</p>

<p><br /></p>

<h2 id="random-forest">Random Forest</h2>

<p><a href="https://www.stat.berkeley.edu/~breiman/RandomForests/cc_home.htm">Random Forests</a> predictions are based on the generation of multiple classification trees.</p>

<p>This model classified 14 out of 23 cases correctly.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">set.seed</span><span class="p">(</span><span class="m">27</span><span class="p">)</span><span class="w">
</span><span class="n">model_rf</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">caret</span><span class="o">::</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                             </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_train_data</span><span class="p">,</span><span class="w">
                             </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"rf"</span><span class="p">,</span><span class="w">
                             </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w">
                             </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span><span class="n">model_rf</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Random Forest 
## 
## 56 samples
## 11 predictors
##  2 classes: 'Death', 'Recover' 
## 
## No pre-processing
## Resampling: Cross-Validated (10 fold, repeated 10 times) 
## Summary of sample sizes: 50, 50, 51, 51, 51, 50, ... 
## Resampling results across tuning parameters:
## 
##   mtry  Accuracy   Kappa    
##    2    0.7801905  0.5338408
##    6    0.7650952  0.4926366
##   11    0.7527619  0.4638073
## 
## Accuracy was used to select the optimal model using  the largest value.
## The final value used for the model was mtry = 2.
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">confusionMatrix</span><span class="p">(</span><span class="n">predict</span><span class="p">(</span><span class="n">model_rf</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]),</span><span class="w"> </span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Confusion Matrix and Statistics
## 
##           Reference
## Prediction Death Recover
##    Death       4       4
##    Recover     5      10
##                                           
##                Accuracy : 0.6087          
##                  95% CI : (0.3854, 0.8029)
##     No Information Rate : 0.6087          
##     P-Value [Acc &gt; NIR] : 0.5901          
##                                           
##                   Kappa : 0.1619          
##  Mcnemar's Test P-Value : 1.0000          
##                                           
##             Sensitivity : 0.4444          
##             Specificity : 0.7143          
##          Pos Pred Value : 0.5000          
##          Neg Pred Value : 0.6667          
##              Prevalence : 0.3913          
##          Detection Rate : 0.1739          
##    Detection Prevalence : 0.3478          
##       Balanced Accuracy : 0.5794          
##                                           
##        'Positive' Class : Death           
## 
</code></pre>
</div>

<p><br /></p>

<h2 id="elastic-net-regularized-generalized-linear-models">Elastic-Net Regularized Generalized Linear Models</h2>

<p><a href="https://en.wikipedia.org/wiki/Elastic_net_regularization">Lasso or elastic net regularization for generalized linear model regression</a> are based on linear regression models and is useful when we have feature correlation in our model.</p>

<p>This model classified 13 out of 23 cases correctly.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">set.seed</span><span class="p">(</span><span class="m">27</span><span class="p">)</span><span class="w">
</span><span class="n">model_glmnet</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">caret</span><span class="o">::</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                             </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_train_data</span><span class="p">,</span><span class="w">
                             </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"glmnet"</span><span class="p">,</span><span class="w">
                             </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w">
                             </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span><span class="n">model_glmnet</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## glmnet 
## 
## 56 samples
## 11 predictors
##  2 classes: 'Death', 'Recover' 
## 
## No pre-processing
## Resampling: Cross-Validated (10 fold, repeated 10 times) 
## Summary of sample sizes: 50, 50, 51, 51, 51, 50, ... 
## Resampling results across tuning parameters:
## 
##   alpha  lambda        Accuracy   Kappa    
##   0.10   0.0005154913  0.7211905  0.4218952
##   0.10   0.0051549131  0.7189524  0.4189835
##   0.10   0.0515491312  0.7469524  0.4704687
##   0.55   0.0005154913  0.7211905  0.4218952
##   0.55   0.0051549131  0.7236190  0.4280528
##   0.55   0.0515491312  0.7531905  0.4801031
##   1.00   0.0005154913  0.7228571  0.4245618
##   1.00   0.0051549131  0.7278571  0.4361809
##   1.00   0.0515491312  0.7678571  0.5094194
## 
## Accuracy was used to select the optimal model using  the largest value.
## The final values used for the model were alpha = 1 and lambda = 0.05154913.
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">confusionMatrix</span><span class="p">(</span><span class="n">predict</span><span class="p">(</span><span class="n">model_glmnet</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]),</span><span class="w"> </span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Confusion Matrix and Statistics
## 
##           Reference
## Prediction Death Recover
##    Death       3       4
##    Recover     6      10
##                                           
##                Accuracy : 0.5652          
##                  95% CI : (0.3449, 0.7681)
##     No Information Rate : 0.6087          
##     P-Value [Acc &gt; NIR] : 0.7418          
##                                           
##                   Kappa : 0.0496          
##  Mcnemar's Test P-Value : 0.7518          
##                                           
##             Sensitivity : 0.3333          
##             Specificity : 0.7143          
##          Pos Pred Value : 0.4286          
##          Neg Pred Value : 0.6250          
##              Prevalence : 0.3913          
##          Detection Rate : 0.1304          
##    Detection Prevalence : 0.3043          
##       Balanced Accuracy : 0.5238          
##                                           
##        'Positive' Class : Death           
## 
</code></pre>
</div>

<p><br /></p>

<h2 id="k-nearest-neighbors">k-Nearest Neighbors</h2>

<p><a href="https://en.wikipedia.org/wiki/K-nearest_neighbors_algorithm">k-nearest neighbors</a> predicts based on point distances with predefined constants.</p>

<p>This model classified 14 out of 23 cases correctly.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">set.seed</span><span class="p">(</span><span class="m">27</span><span class="p">)</span><span class="w">
</span><span class="n">model_kknn</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">caret</span><span class="o">::</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                             </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_train_data</span><span class="p">,</span><span class="w">
                             </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"kknn"</span><span class="p">,</span><span class="w">
                             </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w">
                             </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span><span class="n">model_kknn</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## k-Nearest Neighbors 
## 
## 56 samples
## 11 predictors
##  2 classes: 'Death', 'Recover' 
## 
## No pre-processing
## Resampling: Cross-Validated (10 fold, repeated 10 times) 
## Summary of sample sizes: 50, 50, 51, 51, 51, 50, ... 
## Resampling results across tuning parameters:
## 
##   kmax  Accuracy   Kappa    
##   5     0.6531905  0.2670442
##   7     0.7120476  0.4031836
##   9     0.7106190  0.4004564
## 
## Tuning parameter 'distance' was held constant at a value of 2
## Tuning parameter 'kernel' was held constant at a value of optimal
## Accuracy was used to select the optimal model using  the largest value.
## The final values used for the model were kmax = 7, distance = 2 and kernel = optimal.
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">confusionMatrix</span><span class="p">(</span><span class="n">predict</span><span class="p">(</span><span class="n">model_kknn</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]),</span><span class="w"> </span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Confusion Matrix and Statistics
## 
##           Reference
## Prediction Death Recover
##    Death       3       3
##    Recover     6      11
##                                           
##                Accuracy : 0.6087          
##                  95% CI : (0.3854, 0.8029)
##     No Information Rate : 0.6087          
##     P-Value [Acc &gt; NIR] : 0.5901          
##                                           
##                   Kappa : 0.1266          
##  Mcnemar's Test P-Value : 0.5050          
##                                           
##             Sensitivity : 0.3333          
##             Specificity : 0.7857          
##          Pos Pred Value : 0.5000          
##          Neg Pred Value : 0.6471          
##              Prevalence : 0.3913          
##          Detection Rate : 0.1304          
##    Detection Prevalence : 0.2609          
##       Balanced Accuracy : 0.5595          
##                                           
##        'Positive' Class : Death           
## 
</code></pre>
</div>

<p><br /></p>

<h2 id="penalized-discriminant-analysis">Penalized Discriminant Analysis</h2>

<p><a href="https://web.stanford.edu/~hastie/Papers/pda.pdf">Penalized Discriminant Analysis</a> is the penalized linear discriminant analysis and is also useful for when we have highly correlated features.</p>

<p>This model classified 14 out of 23 cases correctly.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">set.seed</span><span class="p">(</span><span class="m">27</span><span class="p">)</span><span class="w">
</span><span class="n">model_pda</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">caret</span><span class="o">::</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                             </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_train_data</span><span class="p">,</span><span class="w">
                             </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"pda"</span><span class="p">,</span><span class="w">
                             </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w">
                             </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span><span class="n">model_pda</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Penalized Discriminant Analysis 
## 
## 56 samples
## 11 predictors
##  2 classes: 'Death', 'Recover' 
## 
## No pre-processing
## Resampling: Cross-Validated (10 fold, repeated 10 times) 
## Summary of sample sizes: 50, 50, 51, 51, 51, 50, ... 
## Resampling results across tuning parameters:
## 
##   lambda  Accuracy   Kappa    
##   0e+00         NaN        NaN
##   1e-04   0.7255238  0.4373766
##   1e-01   0.7235238  0.4342554
## 
## Accuracy was used to select the optimal model using  the largest value.
## The final value used for the model was lambda = 1e-04.
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">confusionMatrix</span><span class="p">(</span><span class="n">predict</span><span class="p">(</span><span class="n">model_pda</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]),</span><span class="w"> </span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Confusion Matrix and Statistics
## 
##           Reference
## Prediction Death Recover
##    Death       4       4
##    Recover     5      10
##                                           
##                Accuracy : 0.6087          
##                  95% CI : (0.3854, 0.8029)
##     No Information Rate : 0.6087          
##     P-Value [Acc &gt; NIR] : 0.5901          
##                                           
##                   Kappa : 0.1619          
##  Mcnemar's Test P-Value : 1.0000          
##                                           
##             Sensitivity : 0.4444          
##             Specificity : 0.7143          
##          Pos Pred Value : 0.5000          
##          Neg Pred Value : 0.6667          
##              Prevalence : 0.3913          
##          Detection Rate : 0.1739          
##    Detection Prevalence : 0.3478          
##       Balanced Accuracy : 0.5794          
##                                           
##        'Positive' Class : Death           
## 
</code></pre>
</div>

<p><br /></p>

<h2 id="stabilized-linear-discriminant-analysis">Stabilized Linear Discriminant Analysis</h2>

<p><a href="https://books.google.de/books?id=RYaMCwAAQBAJ&amp;pg=PA89&amp;lpg=PA89&amp;dq=%22Stabilized+Linear+Discriminant+Analysis%22&amp;source=bl&amp;ots=YwY0mLEeXx&amp;sig=74Uf3plf0Ma8CT1vh64Wc9MzFqI&amp;hl=de&amp;sa=X&amp;ved=0ahUKEwjXgsOh7sPQAhUMBsAKHc_mAw4Q6AEINjAD#v=onepage&amp;q=%22Stabilized%20Linear%20Discriminant%20Analysis%22&amp;f=false">Stabilized Linear Discriminant Analysis</a> is designed for high-dimensional data and correlated co-variables.</p>

<p>This model classified 15 out of 23 cases correctly.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">set.seed</span><span class="p">(</span><span class="m">27</span><span class="p">)</span><span class="w">
</span><span class="n">model_slda</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">caret</span><span class="o">::</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                             </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_train_data</span><span class="p">,</span><span class="w">
                             </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"slda"</span><span class="p">,</span><span class="w">
                             </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w">
                             </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span><span class="n">model_slda</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Stabilized Linear Discriminant Analysis 
## 
## 56 samples
## 11 predictors
##  2 classes: 'Death', 'Recover' 
## 
## No pre-processing
## Resampling: Cross-Validated (10 fold, repeated 10 times) 
## Summary of sample sizes: 50, 50, 51, 51, 51, 50, ... 
## Resampling results:
## 
##   Accuracy   Kappa    
##   0.6886667  0.3512234
## 
## 
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">confusionMatrix</span><span class="p">(</span><span class="n">predict</span><span class="p">(</span><span class="n">model_slda</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]),</span><span class="w"> </span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Confusion Matrix and Statistics
## 
##           Reference
## Prediction Death Recover
##    Death       3       2
##    Recover     6      12
##                                           
##                Accuracy : 0.6522          
##                  95% CI : (0.4273, 0.8362)
##     No Information Rate : 0.6087          
##     P-Value [Acc &gt; NIR] : 0.4216          
##                                           
##                   Kappa : 0.2069          
##  Mcnemar's Test P-Value : 0.2888          
##                                           
##             Sensitivity : 0.3333          
##             Specificity : 0.8571          
##          Pos Pred Value : 0.6000          
##          Neg Pred Value : 0.6667          
##              Prevalence : 0.3913          
##          Detection Rate : 0.1304          
##    Detection Prevalence : 0.2174          
##       Balanced Accuracy : 0.5952          
##                                           
##        'Positive' Class : Death           
## 
</code></pre>
</div>

<p><br /></p>

<h2 id="nearest-shrunken-centroids">Nearest Shrunken Centroids</h2>

<p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC124443/">Nearest Shrunken Centroids</a> computes a standardized centroid for each class and shrinks each centroid toward the overall centroid for all classes.</p>

<p>This model classified 15 out of 23 cases correctly.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">set.seed</span><span class="p">(</span><span class="m">27</span><span class="p">)</span><span class="w">
</span><span class="n">model_pam</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">caret</span><span class="o">::</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                             </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_train_data</span><span class="p">,</span><span class="w">
                             </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"pam"</span><span class="p">,</span><span class="w">
                             </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w">
                             </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span><span class="n">model_pam</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Nearest Shrunken Centroids 
## 
## 56 samples
## 11 predictors
##  2 classes: 'Death', 'Recover' 
## 
## No pre-processing
## Resampling: Cross-Validated (10 fold, repeated 10 times) 
## Summary of sample sizes: 50, 50, 51, 51, 51, 50, ... 
## Resampling results across tuning parameters:
## 
##   threshold  Accuracy   Kappa    
##   0.1200215  0.7283333  0.4418904
##   1.7403111  0.6455714  0.2319674
##   3.3606007  0.5904762  0.0000000
## 
## Accuracy was used to select the optimal model using  the largest value.
## The final value used for the model was threshold = 0.1200215.
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">confusionMatrix</span><span class="p">(</span><span class="n">predict</span><span class="p">(</span><span class="n">model_pam</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]),</span><span class="w"> </span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Confusion Matrix and Statistics
## 
##           Reference
## Prediction Death Recover
##    Death       4       3
##    Recover     5      11
##                                           
##                Accuracy : 0.6522          
##                  95% CI : (0.4273, 0.8362)
##     No Information Rate : 0.6087          
##     P-Value [Acc &gt; NIR] : 0.4216          
##                                           
##                   Kappa : 0.2397          
##  Mcnemar's Test P-Value : 0.7237          
##                                           
##             Sensitivity : 0.4444          
##             Specificity : 0.7857          
##          Pos Pred Value : 0.5714          
##          Neg Pred Value : 0.6875          
##              Prevalence : 0.3913          
##          Detection Rate : 0.1739          
##    Detection Prevalence : 0.3043          
##       Balanced Accuracy : 0.6151          
##                                           
##        'Positive' Class : Death           
## 
</code></pre>
</div>

<p><br /></p>

<h2 id="single-c50-tree">Single C5.0 Tree</h2>

<p><a href="http://www.bowdoin.edu/~echown/courses/370/tutorial.html">C5.0</a> is another tree-based modeling algorithm.</p>

<p>This model classified 15 out of 23 cases correctly.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">set.seed</span><span class="p">(</span><span class="m">27</span><span class="p">)</span><span class="w">
</span><span class="n">model_C5.0Tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">caret</span><span class="o">::</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                             </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_train_data</span><span class="p">,</span><span class="w">
                             </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"C5.0Tree"</span><span class="p">,</span><span class="w">
                             </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w">
                             </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span><span class="n">model_C5.0Tree</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Single C5.0 Tree 
## 
## 56 samples
## 11 predictors
##  2 classes: 'Death', 'Recover' 
## 
## No pre-processing
## Resampling: Cross-Validated (10 fold, repeated 10 times) 
## Summary of sample sizes: 50, 50, 51, 51, 51, 50, ... 
## Resampling results:
## 
##   Accuracy   Kappa    
##   0.7103333  0.4047334
## 
## 
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">confusionMatrix</span><span class="p">(</span><span class="n">predict</span><span class="p">(</span><span class="n">model_C5.0Tree</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]),</span><span class="w"> </span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Confusion Matrix and Statistics
## 
##           Reference
## Prediction Death Recover
##    Death       5       4
##    Recover     4      10
##                                           
##                Accuracy : 0.6522          
##                  95% CI : (0.4273, 0.8362)
##     No Information Rate : 0.6087          
##     P-Value [Acc &gt; NIR] : 0.4216          
##                                           
##                   Kappa : 0.2698          
##  Mcnemar's Test P-Value : 1.0000          
##                                           
##             Sensitivity : 0.5556          
##             Specificity : 0.7143          
##          Pos Pred Value : 0.5556          
##          Neg Pred Value : 0.7143          
##              Prevalence : 0.3913          
##          Detection Rate : 0.2174          
##    Detection Prevalence : 0.3913          
##       Balanced Accuracy : 0.6349          
##                                           
##        'Positive' Class : Death           
## 
</code></pre>
</div>

<p><br /></p>

<h2 id="partial-least-squares">Partial Least Squares</h2>

<p><a href="https://en.wikipedia.org/wiki/Partial_least_squares_regression">Partial least squares regression</a> combined principal component analysis and multiple regression and is useful for modeling with correlated features.</p>

<p>This model classified 15 out of 23 cases correctly.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">set.seed</span><span class="p">(</span><span class="m">27</span><span class="p">)</span><span class="w">
</span><span class="n">model_pls</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">caret</span><span class="o">::</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                             </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_train_data</span><span class="p">,</span><span class="w">
                             </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"pls"</span><span class="p">,</span><span class="w">
                             </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w">
                             </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span><span class="n">model_pls</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Partial Least Squares 
## 
## 56 samples
## 11 predictors
##  2 classes: 'Death', 'Recover' 
## 
## No pre-processing
## Resampling: Cross-Validated (10 fold, repeated 10 times) 
## Summary of sample sizes: 50, 50, 51, 51, 51, 50, ... 
## Resampling results across tuning parameters:
## 
##   ncomp  Accuracy   Kappa    
##   1      0.6198571  0.2112982
##   2      0.6376190  0.2436222
##   3      0.6773810  0.3305780
## 
## Accuracy was used to select the optimal model using  the largest value.
## The final value used for the model was ncomp = 3.
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">confusionMatrix</span><span class="p">(</span><span class="n">predict</span><span class="p">(</span><span class="n">model_pls</span><span class="p">,</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]),</span><span class="w"> </span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Confusion Matrix and Statistics
## 
##           Reference
## Prediction Death Recover
##    Death       3       2
##    Recover     6      12
##                                           
##                Accuracy : 0.6522          
##                  95% CI : (0.4273, 0.8362)
##     No Information Rate : 0.6087          
##     P-Value [Acc &gt; NIR] : 0.4216          
##                                           
##                   Kappa : 0.2069          
##  Mcnemar's Test P-Value : 0.2888          
##                                           
##             Sensitivity : 0.3333          
##             Specificity : 0.8571          
##          Pos Pred Value : 0.6000          
##          Neg Pred Value : 0.6667          
##              Prevalence : 0.3913          
##          Detection Rate : 0.1304          
##    Detection Prevalence : 0.2174          
##       Balanced Accuracy : 0.5952          
##                                           
##        'Positive' Class : Death           
## 
</code></pre>
</div>

<p><br /></p>

<h2 id="comparing-accuracy-of-models">Comparing accuracy of models</h2>

<p>All models were similarly accurate.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># Create a list of models
</span><span class="n">models</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">rf</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">model_rf</span><span class="p">,</span><span class="w"> </span><span class="n">glmnet</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">model_glmnet</span><span class="p">,</span><span class="w"> </span><span class="n">kknn</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">model_kknn</span><span class="p">,</span><span class="w"> </span><span class="n">pda</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">model_pda</span><span class="p">,</span><span class="w"> </span><span class="n">slda</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">model_slda</span><span class="p">,</span><span class="w">
               </span><span class="n">pam</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">model_pam</span><span class="p">,</span><span class="w"> </span><span class="n">C</span><span class="m">5.0</span><span class="n">Tree</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">model_C5.0Tree</span><span class="p">,</span><span class="w"> </span><span class="n">pls</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">model_pls</span><span class="p">)</span><span class="w">

</span><span class="c1"># Resample the models
</span><span class="n">resample_results</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">resamples</span><span class="p">(</span><span class="n">models</span><span class="p">)</span><span class="w">

</span><span class="c1"># Generate a summary
</span><span class="n">summary</span><span class="p">(</span><span class="n">resample_results</span><span class="p">,</span><span class="w"> </span><span class="n">metric</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Kappa"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Accuracy"</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## 
## Call:
## summary.resamples(object = resample_results, metric = c("Kappa", "Accuracy"))
## 
## Models: rf, glmnet, kknn, pda, slda, pam, C5.0Tree, pls 
## Number of resamples: 100 
## 
## Kappa 
##             Min. 1st Qu. Median   Mean 3rd Qu. Max. NA's
## rf       -0.3636  0.3214 0.5714 0.5338  0.6800    1    0
## glmnet   -0.3636  0.2768 0.5455 0.5094  0.6667    1    0
## kknn     -0.3636  0.1667 0.5035 0.4032  0.6282    1    0
## pda      -0.6667  0.2431 0.5455 0.4374  0.6667    1    0
## slda     -0.5217  0.0000 0.4308 0.3512  0.6667    1    0
## pam      -0.5000  0.2292 0.5455 0.4419  0.6667    1    0
## C5.0Tree -0.4286  0.1667 0.4167 0.4047  0.6154    1    0
## pls      -0.6667  0.0000 0.3333 0.3306  0.6282    1    0
## 
## Accuracy 
##            Min. 1st Qu. Median   Mean 3rd Qu. Max. NA's
## rf       0.4000  0.6667 0.8000 0.7802  0.8393    1    0
## glmnet   0.4000  0.6500 0.8000 0.7679  0.8333    1    0
## kknn     0.3333  0.6000 0.7571 0.7120  0.8333    1    0
## pda      0.2000  0.6000 0.8000 0.7255  0.8333    1    0
## slda     0.2000  0.6000 0.6905 0.6887  0.8333    1    0
## pam      0.2000  0.6000 0.8000 0.7283  0.8333    1    0
## C5.0Tree 0.2000  0.6000 0.7143 0.7103  0.8333    1    0
## pls      0.1429  0.5929 0.6667 0.6774  0.8333    1    0
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">bwplot</span><span class="p">(</span><span class="n">resample_results</span><span class="w"> </span><span class="p">,</span><span class="w"> </span><span class="n">metric</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Kappa"</span><span class="p">,</span><span class="s2">"Accuracy"</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<p><img src="flu_outcome_ML_post_files/figure-markdown_github/unnamed-chunk-35-1.png" style="display: block; margin: auto;" /></p>

<p><br /></p>

<h3 id="combined-results-of-predicting-validation-test-samples">Combined results of predicting validation test samples</h3>

<p>To compare the predictions from all models, I summed up the prediction probabilities for Death and Recovery from all models and calculated the log2 of the ratio between the summed probabilities for Recovery by the summed probabilities for Death. All cases with a log2 ratio bigger than 1.5 were defined as Recover, cases with a log2 ratio below -1.5 were defined as Death, and the remaining cases were defined as uncertain.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">results</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">randomForest</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_rf</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
                      </span><span class="n">glmnet</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_glmnet</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
                      </span><span class="n">kknn</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_kknn</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
                      </span><span class="n">pda</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_pda</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
                      </span><span class="n">slda</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_slda</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
                      </span><span class="n">pam</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_pam</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
                      </span><span class="n">C</span><span class="m">5.0</span><span class="n">Tree</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_C5.0Tree</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
                      </span><span class="n">pls</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_pls</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">val_test_data</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">))</span><span class="w">

</span><span class="n">results</span><span class="o">$</span><span class="n">sum_Death</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rowSums</span><span class="p">(</span><span class="n">results</span><span class="p">[,</span><span class="w"> </span><span class="n">grep</span><span class="p">(</span><span class="s2">"Death"</span><span class="p">,</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">results</span><span class="p">))])</span><span class="w">
</span><span class="n">results</span><span class="o">$</span><span class="n">sum_Recover</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rowSums</span><span class="p">(</span><span class="n">results</span><span class="p">[,</span><span class="w"> </span><span class="n">grep</span><span class="p">(</span><span class="s2">"Recover"</span><span class="p">,</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">results</span><span class="p">))])</span><span class="w">
</span><span class="n">results</span><span class="o">$</span><span class="n">log2_ratio</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">log2</span><span class="p">(</span><span class="n">results</span><span class="o">$</span><span class="n">sum_Recover</span><span class="o">/</span><span class="n">results</span><span class="o">$</span><span class="n">sum_Death</span><span class="p">)</span><span class="w">
</span><span class="n">results</span><span class="o">$</span><span class="n">true_outcome</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">val_test_data</span><span class="o">$</span><span class="n">outcome</span><span class="w">
</span><span class="n">results</span><span class="o">$</span><span class="n">pred_outcome</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">results</span><span class="o">$</span><span class="n">log2_ratio</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">1.5</span><span class="p">,</span><span class="w"> </span><span class="s2">"Recover"</span><span class="p">,</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">results</span><span class="o">$</span><span class="n">log2_ratio</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">-1.5</span><span class="p">,</span><span class="w"> </span><span class="s2">"Death"</span><span class="p">,</span><span class="w"> </span><span class="s2">"uncertain"</span><span class="p">))</span><span class="w">
</span><span class="n">results</span><span class="o">$</span><span class="n">prediction</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">results</span><span class="o">$</span><span class="n">pred_outcome</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">results</span><span class="o">$</span><span class="n">true_outcome</span><span class="p">,</span><span class="w"> </span><span class="s2">"CORRECT"</span><span class="p">,</span><span class="w"> 
                             </span><span class="n">ifelse</span><span class="p">(</span><span class="n">results</span><span class="o">$</span><span class="n">pred_outcome</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"uncertain"</span><span class="p">,</span><span class="w"> </span><span class="s2">"uncertain"</span><span class="p">,</span><span class="w"> </span><span class="s2">"wrong"</span><span class="p">))</span><span class="w">
</span><span class="n">results</span><span class="p">[,</span><span class="w"> </span><span class="o">-</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">16</span><span class="p">)]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##          sum_Death sum_Recover  log2_ratio true_outcome pred_outcome prediction
## case_123 2.2236374    5.776363  1.37723972        Death    uncertain  uncertain
## case_127 0.7522674    7.247733  3.26821230      Recover      Recover    CORRECT
## case_128 2.4448101    5.555190  1.18411381      Recover    uncertain  uncertain
## case_14  2.9135620    5.086438  0.80387172      Recover    uncertain  uncertain
## case_19  2.2113377    5.788662  1.38831062        Death    uncertain  uncertain
## case_2   3.6899508    4.310049  0.22410277        Death    uncertain  uncertain
## case_20  4.5621189    3.437881 -0.40818442      Recover    uncertain  uncertain
## case_21  4.9960238    3.003976 -0.73390698      Recover    uncertain  uncertain
## case_34  6.1430233    1.856977 -1.72599312        Death        Death    CORRECT
## case_37  0.8717595    7.128241  3.03154401      Recover      Recover    CORRECT
## case_5   1.7574945    6.242505  1.82860496      Recover      Recover    CORRECT
## case_51  3.8917736    4.108226  0.07808791      Recover    uncertain  uncertain
## case_55  3.1548368    4.845163  0.61897986      Recover    uncertain  uncertain
## case_6   3.5163388    4.483661  0.35060317        Death    uncertain  uncertain
## case_61  5.0803241    2.919676 -0.79911232        Death    uncertain  uncertain
## case_65  1.1282313    6.871769  2.60661863      Recover      Recover    CORRECT
## case_74  5.3438427    2.656157 -1.00853699      Recover    uncertain  uncertain
## case_78  3.2183947    4.781605  0.57115378        Death    uncertain  uncertain
## case_79  1.9521617    6.047838  1.63134700      Recover      Recover    CORRECT
## case_8   3.6106045    4.389395  0.28178184        Death    uncertain  uncertain
## case_87  4.9712879    3.028712 -0.71491522        Death    uncertain  uncertain
## case_91  1.8648837    6.135116  1.71800508      Recover      Recover    CORRECT
## case_94  2.1198087    5.880191  1.47192901      Recover    uncertain  uncertain
</code></pre>
</div>

<p>All predictions based on all models were either correct or uncertain.</p>

<p><br /></p>

<h1 id="predicting-unknown-outcomes">Predicting unknown outcomes</h1>

<p>The above models will now be used to predict the outcome of cases with unknown fate.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">set.seed</span><span class="p">(</span><span class="m">27</span><span class="p">)</span><span class="w">
</span><span class="n">model_rf</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">caret</span><span class="o">::</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                             </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">train_data</span><span class="p">,</span><span class="w">
                             </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"rf"</span><span class="p">,</span><span class="w">
                             </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w">
                             </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span><span class="n">model_glmnet</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">caret</span><span class="o">::</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                             </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">train_data</span><span class="p">,</span><span class="w">
                             </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"glmnet"</span><span class="p">,</span><span class="w">
                             </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w">
                             </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span><span class="n">model_kknn</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">caret</span><span class="o">::</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                             </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">train_data</span><span class="p">,</span><span class="w">
                             </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"kknn"</span><span class="p">,</span><span class="w">
                             </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w">
                             </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span><span class="n">model_pda</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">caret</span><span class="o">::</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                             </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">train_data</span><span class="p">,</span><span class="w">
                             </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"pda"</span><span class="p">,</span><span class="w">
                             </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w">
                             </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span><span class="n">model_slda</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">caret</span><span class="o">::</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                             </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">train_data</span><span class="p">,</span><span class="w">
                             </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"slda"</span><span class="p">,</span><span class="w">
                             </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w">
                             </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span><span class="n">model_pam</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">caret</span><span class="o">::</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                             </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">train_data</span><span class="p">,</span><span class="w">
                             </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"pam"</span><span class="p">,</span><span class="w">
                             </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w">
                             </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span><span class="n">model_C5.0Tree</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">caret</span><span class="o">::</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                             </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">train_data</span><span class="p">,</span><span class="w">
                             </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"C5.0Tree"</span><span class="p">,</span><span class="w">
                             </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w">
                             </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span><span class="n">model_pls</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">caret</span><span class="o">::</span><span class="n">train</span><span class="p">(</span><span class="n">outcome</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">,</span><span class="w">
                             </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">train_data</span><span class="p">,</span><span class="w">
                             </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"pls"</span><span class="p">,</span><span class="w">
                             </span><span class="n">preProcess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w">
                             </span><span class="n">trControl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">trainControl</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"repeatedcv"</span><span class="p">,</span><span class="w"> </span><span class="n">number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">repeats</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">verboseIter</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">))</span><span class="w">

</span><span class="n">models</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">rf</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">model_rf</span><span class="p">,</span><span class="w"> </span><span class="n">glmnet</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">model_glmnet</span><span class="p">,</span><span class="w"> </span><span class="n">kknn</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">model_kknn</span><span class="p">,</span><span class="w"> </span><span class="n">pda</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">model_pda</span><span class="p">,</span><span class="w"> </span><span class="n">slda</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">model_slda</span><span class="p">,</span><span class="w">
               </span><span class="n">pam</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">model_pam</span><span class="p">,</span><span class="w"> </span><span class="n">C</span><span class="m">5.0</span><span class="n">Tree</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">model_C5.0Tree</span><span class="p">,</span><span class="w"> </span><span class="n">pls</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">model_pls</span><span class="p">)</span><span class="w">

</span><span class="c1"># Resample the models
</span><span class="n">resample_results</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">resamples</span><span class="p">(</span><span class="n">models</span><span class="p">)</span><span class="w">

</span><span class="n">bwplot</span><span class="p">(</span><span class="n">resample_results</span><span class="w"> </span><span class="p">,</span><span class="w"> </span><span class="n">metric</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Kappa"</span><span class="p">,</span><span class="s2">"Accuracy"</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<p><img src="flu_outcome_ML_post_files/figure-markdown_github/unnamed-chunk-37-1.png" style="display: block; margin: auto;" /></p>

<p>Here again, the accuracy is similar in all models.</p>

<p><br /></p>

<p>The final results are calculate as described above.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">results</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">randomForest</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_rf</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">test_data</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
                      </span><span class="n">glmnet</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_glmnet</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">test_data</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
                      </span><span class="n">kknn</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_kknn</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">test_data</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
                      </span><span class="n">pda</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_pda</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">test_data</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
                      </span><span class="n">slda</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_slda</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">test_data</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
                      </span><span class="n">pam</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_pam</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">test_data</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
                      </span><span class="n">C</span><span class="m">5.0</span><span class="n">Tree</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_C5.0Tree</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">test_data</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">),</span><span class="w">
                      </span><span class="n">pls</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predict</span><span class="p">(</span><span class="n">model_pls</span><span class="p">,</span><span class="w"> </span><span class="n">newdata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">test_data</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"prob"</span><span class="p">))</span><span class="w">

</span><span class="n">results</span><span class="o">$</span><span class="n">sum_Death</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rowSums</span><span class="p">(</span><span class="n">results</span><span class="p">[,</span><span class="w"> </span><span class="n">grep</span><span class="p">(</span><span class="s2">"Death"</span><span class="p">,</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">results</span><span class="p">))])</span><span class="w">
</span><span class="n">results</span><span class="o">$</span><span class="n">sum_Recover</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rowSums</span><span class="p">(</span><span class="n">results</span><span class="p">[,</span><span class="w"> </span><span class="n">grep</span><span class="p">(</span><span class="s2">"Recover"</span><span class="p">,</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">results</span><span class="p">))])</span><span class="w">
</span><span class="n">results</span><span class="o">$</span><span class="n">log2_ratio</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">log2</span><span class="p">(</span><span class="n">results</span><span class="o">$</span><span class="n">sum_Recover</span><span class="o">/</span><span class="n">results</span><span class="o">$</span><span class="n">sum_Death</span><span class="p">)</span><span class="w">
</span><span class="n">results</span><span class="o">$</span><span class="n">predicted_outcome</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">results</span><span class="o">$</span><span class="n">log2_ratio</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">1.5</span><span class="p">,</span><span class="w"> </span><span class="s2">"Recover"</span><span class="p">,</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">results</span><span class="o">$</span><span class="n">log2_ratio</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">-1.5</span><span class="p">,</span><span class="w"> </span><span class="s2">"Death"</span><span class="p">,</span><span class="w"> </span><span class="s2">"uncertain"</span><span class="p">))</span><span class="w">
</span><span class="n">results</span><span class="p">[,</span><span class="w"> </span><span class="o">-</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">16</span><span class="p">)]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##          sum_Death sum_Recover  log2_ratio predicted_outcome
## case_100 3.6707808    4.329219  0.23801986         uncertain
## case_101 6.3194220    1.680578 -1.91083509             Death
## case_102 6.4960471    1.503953 -2.11080274             Death
## case_103 2.3877223    5.612278  1.23295134         uncertain
## case_104 4.1254604    3.874540 -0.09053024         uncertain
## case_105 2.1082161    5.891784  1.48268174         uncertain
## case_108 6.5941436    1.405856 -2.22973606             Death
## case_109 2.2780779    5.721922  1.32868278         uncertain
## case_110 1.7853361    6.214664  1.79948072           Recover
## case_112 4.5102439    3.489756 -0.37007925         uncertain
## case_113 4.7047554    3.295245 -0.51373420         uncertain
## case_114 1.6397105    6.360290  1.95565132           Recover
## case_115 1.2351519    6.764848  2.45336902           Recover
## case_118 1.4675571    6.532443  2.15420601           Recover
## case_120 1.9322149    6.067785  1.65091447           Recover
## case_122 1.1165786    6.883421  2.62404109           Recover
## case_126 5.4933927    2.506607 -1.13196145         uncertain
## case_130 4.3035732    3.696427 -0.21940367         uncertain
## case_132 5.0166184    2.983382 -0.74976671         uncertain
## case_136 2.6824566    5.317543  0.98720505         uncertain
## case_15  5.8545370    2.145463 -1.44826607         uncertain
## case_16  6.5063794    1.493621 -2.12304122             Death
## case_22  4.2929861    3.707014 -0.21172398         uncertain
## case_28  6.6129447    1.387055 -2.25326754             Death
## case_31  6.2625800    1.737420 -1.84981057             Death
## case_32  6.0986316    1.901368 -1.68144746             Death
## case_38  1.7581540    6.241846  1.82791134           Recover
## case_39  2.4765188    5.523481  1.15726428         uncertain
## case_4   5.1309910    2.869009 -0.83868503         uncertain
## case_40  6.6069297    1.393070 -2.24571191             Death
## case_41  5.5499144    2.450086 -1.17963336         uncertain
## case_42  2.0709160    5.929084  1.51754019           Recover
## case_47  6.1988812    1.801119 -1.78311450             Death
## case_48  5.6500772    2.349923 -1.26565724         uncertain
## case_52  5.9096543    2.090346 -1.49933214         uncertain
## case_54  2.6763066    5.323693  0.99218409         uncertain
## case_56  2.6826414    5.317359  0.98705557         uncertain
## case_62  6.0722552    1.927745 -1.65531833             Death
## case_63  3.9541381    4.045862  0.03308379         uncertain
## case_66  5.8320010    2.167999 -1.42762690         uncertain
## case_67  2.1732059    5.826794  1.42287745         uncertain
## case_68  4.3464174    3.653583 -0.25051491         uncertain
## case_69  1.8902845    6.109715  1.69250181           Recover
## case_70  4.5986242    3.401376 -0.43508393         uncertain
## case_71  2.7966938    5.203306  0.89570626         uncertain
## case_80  2.7270986    5.272901  0.95123017         uncertain
## case_84  1.9485253    6.051475  1.63490411           Recover
## case_85  3.0076953    4.992305  0.73104755         uncertain
## case_86  2.0417802    5.958220  1.54505376           Recover
## case_88  0.7285419    7.271458  3.31916083           Recover
## case_9   2.2188620    5.781138  1.38153353         uncertain
## case_90  1.3973262    6.602674  2.24038155           Recover
## case_92  5.0994993    2.900501 -0.81405366         uncertain
## case_93  4.3979048    3.602095 -0.28798006         uncertain
## case_95  3.5561792    4.443821  0.32147260         uncertain
## case_96  1.3359082    6.664092  2.31858744           Recover
## case_99  5.0776686    2.922331 -0.79704644         uncertain
</code></pre>
</div>

<p>From 57 cases, 14 were defined as Recover, 10 as Death and 33 as uncertain.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">results_combined</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">results</span><span class="p">[,</span><span class="w"> </span><span class="o">-</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">16</span><span class="p">)],</span><span class="w"> </span><span class="n">fluH7N9.china.2013</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">fluH7N9.china.2013</span><span class="o">$</span><span class="n">case.ID</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">results</span><span class="p">)),</span><span class="w"> </span><span class="p">],</span><span class="w"> 
                          </span><span class="n">by.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"row.names"</span><span class="p">,</span><span class="w"> </span><span class="n">by.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"case.ID"</span><span class="p">)</span><span class="w">
</span><span class="n">results_combined</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">results_combined</span><span class="p">[,</span><span class="w"> </span><span class="o">-</span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="m">8</span><span class="p">,</span><span class="w"> </span><span class="m">9</span><span class="p">)]</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">results_combined_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">results_combined</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">gather</span><span class="p">(</span><span class="n">group_dates</span><span class="p">,</span><span class="w"> </span><span class="n">date</span><span class="p">,</span><span class="w"> </span><span class="n">date.of.onset</span><span class="o">:</span><span class="n">date.of.hospitalisation</span><span class="p">)</span><span class="w">

</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">group_dates</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">group_dates</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"date.of.onset"</span><span class="p">,</span><span class="w"> </span><span class="s2">"date.of.hospitalisation"</span><span class="p">))</span><span class="w">

</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">group_dates</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mapvalues</span><span class="p">(</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">group_dates</span><span class="p">,</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"date.of.onset"</span><span class="p">,</span><span class="w"> </span><span class="s2">"date.of.hospitalisation"</span><span class="p">),</span><span class="w"> 
                                             </span><span class="n">to</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Date of onset"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Date of hospitalisation"</span><span class="p">))</span><span class="w">

</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">gender</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mapvalues</span><span class="p">(</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">gender</span><span class="p">,</span><span class="w"> </span><span class="n">from</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"f"</span><span class="p">,</span><span class="w"> </span><span class="s2">"m"</span><span class="p">),</span><span class="w"> 
                                             </span><span class="n">to</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Female"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Male"</span><span class="p">))</span><span class="w">
</span><span class="n">levels</span><span class="p">(</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">gender</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">levels</span><span class="p">(</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">gender</span><span class="p">),</span><span class="w"> </span><span class="s2">"unknown"</span><span class="p">)</span><span class="w">
</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">gender</span><span class="p">[</span><span class="nf">is.na</span><span class="p">(</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">gender</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"unknown"</span><span class="w">
</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">age</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">results_combined_gather</span><span class="o">$</span><span class="n">age</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">results_combined_gather</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">date</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">log2_ratio</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">predicted_outcome</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_jitter</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">age</span><span class="p">),</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_rug</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_grid</span><span class="p">(</span><span class="n">gender</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">group_dates</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="w">
    </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Predicted outcome"</span><span class="p">,</span><span class="w">
    </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Age"</span><span class="p">,</span><span class="w">
    </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Date in 2013"</span><span class="p">,</span><span class="w">
    </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"log2 ratio of prediction Recover vs Death"</span><span class="p">,</span><span class="w">
    </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"2013 Influenza A H7N9 cases in China"</span><span class="p">,</span><span class="w">
    </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Predicted outcome"</span><span class="p">,</span><span class="w">
    </span><span class="n">caption</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="w">
  </span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_color_brewer</span><span class="p">(</span><span class="n">palette</span><span class="o">=</span><span class="s2">"Set1"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_brewer</span><span class="p">(</span><span class="n">palette</span><span class="o">=</span><span class="s2">"Set1"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="flu_outcome_ML_post_files/figure-markdown_github/unnamed-chunk-41-1.png" style="display: block; margin: auto;" /></p>

<p>The comparison of date of onset, data of hospitalisation, gender and age with predicted outcome shows that predicted deaths were associated with older age than predicted Recoveries. Date of onset does not show an obvious bias in either direction.</p>

<p><br /></p>

<h1 id="conclusions">Conclusions</h1>

<p>This dataset posed a couple of difficulties to begin with, like unequal distribution of data points across variables and missing data. This makes the modeling inherently prone to flaws. However, real life data isn’t perfect either, so I went ahead and tested the modeling success anyway.</p>

<p>By accounting for uncertain classification with low predictions probability, the validation data could be classified accurately. However, for a more accurate model, these few cases don’t give enough information to reliably predict the outcome. More cases, more information (i.e. more features) and fewer missing data would improve the modeling outcome.</p>

<p>Also, this example is only applicable for this specific case of flu. In order to be able to draw more general conclusions about flu outcome, other cases and additional information, for example on medical parameters like preexisting medical conditions, disase parameters, demographic information, etc. would be necessary.</p>

<p>All in all, this dataset served as a nice example of the possibilities (and pitfalls) of machine learning applications and showcases a basic workflow for building prediction models with R.</p>

<p><br /></p>

<p>For a comparison of feature selection methods see <a href="https://shiring.github.io/machine_learning/2017/01/15/rfe_ga_post">here</a>.</p>

<p>If you see any mistakes or have tips and tricks for improvement, please don’t hesitate to let me know! Thanks. :-)</p>

<hr />

<p>If you are interested in more machine learning posts, check out <a href="https://shiring.github.io/categories.html#machine_learning-ref">the category listing for <strong>machine_learning</strong></a>.</p>

<hr />

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sessionInfo</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## R version 3.3.2 (2016-10-31)
## Platform: x86_64-w64-mingw32/x64 (64-bit)
## Running under: Windows 7 x64 (build 7601) Service Pack 1
## 
## locale:
## [1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252 LC_NUMERIC=C                           LC_TIME=English_United States.1252    
## 
## attached base packages:
## [1] grid      stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] pls_2.5-0           C50_0.1.0-24        pamr_1.55           survival_2.40-1     cluster_2.0.5       ipred_0.9-5         mda_0.4-9           class_7.3-14        kknn_1.3.1          glmnet_2.0-5        foreach_1.4.3       Matrix_1.2-7.1      randomForest_4.6-12 RColorBrewer_1.1-2  rpart.plot_2.1.0    rattle_4.1.0        rpart_4.1-10        caret_6.0-73        lattice_0.20-34     mice_2.25           Rcpp_0.12.7         dplyr_0.5.0         gridExtra_2.2.1     ggplot2_2.2.0       plyr_1.8.4          tidyr_0.6.0         outbreaks_1.0.0    
## 
## loaded via a namespace (and not attached):
##  [1] RGtk2_2.20.31      assertthat_0.1     digest_0.6.10      R6_2.2.0           MatrixModels_0.4-1 stats4_3.3.2       evaluate_0.10      e1071_1.6-7        lazyeval_0.2.0     minqa_1.2.4        SparseM_1.74       car_2.1-3          nloptr_1.0.4       partykit_1.1-1     rmarkdown_1.1      labeling_0.3       splines_3.3.2      lme4_1.1-12        stringr_1.1.0      igraph_1.0.1       munsell_0.4.3      compiler_3.3.2     mgcv_1.8-16        htmltools_0.3.5    nnet_7.3-12        tibble_1.2         prodlim_1.5.7      codetools_0.2-15   MASS_7.3-45        ModelMetrics_1.1.0 nlme_3.1-128       gtable_0.2.0       DBI_0.5-1          magrittr_1.5       scales_0.4.1       stringi_1.1.2      reshape2_1.4.2     Formula_1.2-1      lava_1.4.5         iterators_1.0.8    tools_3.3.2        parallel_3.3.2     pbkrtest_0.4-6     yaml_2.1.14        colorspace_1.3-0   knitr_1.15         quantreg_5.29
</code></pre>
</div>

  </div>

  <h1><a href="/text_analysis/2016/11/20/gilmore_girls_part2_post">Analysing the Gilmore Girls' coffee addiction with R</a></h1>
  <p class="author">
    <span class="date">2016-11-20 00:00:00 +0100</span>
  </p>
  <div class="content">
    <p>Last week’s post showed how to create a <a href="https://shiring.github.io/text_analysis/2016/11/13/gilmore_girls_part1_post">Gilmore Girls character network</a>.</p>

<p>In this week’s short post, I want to explore the Gilmore Girls’ famous coffee addiction by analysing the same episode transcripts that were also used last week.</p>

<p>I am also showcasing how to use the <a href="https://blog.rstudio.org/2016/11/14/ggplot2-2-2-0/">recently updated ggplot2 2.2.0</a>.</p>

<p>The transcripts were prepared as described in last week’s post. Full R code can be found below.</p>

<p><br /></p>

<h2 id="how-often-do-they-talk-about-coffee">How often do they talk about coffee?</h2>

<p>The first question I want to explore is how often Lorelai, Rory, Luke, Sookie, Lane and Paris talked about coffee. For this, I subsetted the lines of these characters to where they talk about coffee (i.e. lines that include the word “coffee” at least once). I then created a plot showing how often they talk about coffee per episode and season.</p>

<p><br /></p>

<p><img src="gilmore_girls_part2_post_files/figure-markdown_github/unnamed-chunk-6-1.png" style="display: block; margin: auto;" /></p>

<p><br /></p>

<p>The differences in coffee mention between characters are of course somewhat biased by differences in the total number of lines of each character in the respective episodes, as well by the length of the episodes.</p>

<p>This led me to wonder what proportion of each episode was occupied by which character.</p>

<p><br /></p>

<p><img src="gilmore_girls_part2_post_files/figure-markdown_github/unnamed-chunk-9-1.png" style="display: block; margin: auto;" /></p>

<p><br /></p>

<p><img src="gilmore_girls_part2_post_files/figure-markdown_github/unnamed-chunk-11-1.png" style="display: block; margin: auto;" /></p>

<hr />

<p><br /></p>

<h1 id="r-code">R code</h1>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># set names
</span><span class="n">names</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"LORELAI"</span><span class="p">,</span><span class="w"> </span><span class="s2">"RORY"</span><span class="p">,</span><span class="w"> </span><span class="s2">"LUKE"</span><span class="p">,</span><span class="w"> </span><span class="s2">"SOOKIE"</span><span class="p">,</span><span class="w"> </span><span class="s2">"LANE"</span><span class="p">,</span><span class="w"> </span><span class="s2">"PARIS"</span><span class="p">)</span><span class="w">

</span><span class="c1"># extract lines for each character in names
</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">names</span><span class="p">){</span><span class="w">
  </span><span class="n">assign</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">),</span><span class="w"> </span><span class="n">transcripts_2</span><span class="p">[</span><span class="n">grep</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="n">name</span><span class="p">),</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">),</span><span class="w"> </span><span class="p">])</span><span class="w">
</span><span class="p">}</span><span class="w">

</span><span class="c1"># extract lines not from characters in names
</span><span class="n">lines_OTHER</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">transcripts_2</span><span class="p">[</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"LORELAI|RORY|LUKE|SOOKIE|LANE|PARIS"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="c1"># update names to include "other"
</span><span class="n">names</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">names</span><span class="p">,</span><span class="w"> </span><span class="s2">"OTHER"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># which lines mention coffee.
</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">names</span><span class="p">){</span><span class="w">
  </span><span class="n">df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">))</span><span class="w">
  </span><span class="n">assign</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"coffee"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">),</span><span class="w"> </span><span class="n">df</span><span class="p">[</span><span class="n">grep</span><span class="p">(</span><span class="s2">"coffee"</span><span class="p">,</span><span class="w"> </span><span class="n">df</span><span class="o">$</span><span class="n">dialogue</span><span class="p">),</span><span class="w"> </span><span class="p">])</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># calculate number of lines in total and with coffee mentions per episode
</span><span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span><span class="w">

</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">names</span><span class="p">){</span><span class="w">
  </span><span class="n">lines_p_ep</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">))</span><span class="w">
  </span><span class="n">lines_p_ep</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">lines_p_ep</span><span class="o">$</span><span class="n">episode</span><span class="p">))</span><span class="w">
  
  </span><span class="n">df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"coffee"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">))</span><span class="w">
  </span><span class="n">df_2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">df</span><span class="o">$</span><span class="n">episode</span><span class="p">))</span><span class="w">
  
  </span><span class="c1"># calculate coffee lines per episode
</span><span class="w">  </span><span class="n">coffee_lines_p_ep</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">full_join</span><span class="p">(</span><span class="n">lines_p_ep</span><span class="p">,</span><span class="w"> </span><span class="n">df_2</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">)</span><span class="w">
  </span><span class="n">coffee_lines_p_ep</span><span class="o">$</span><span class="n">ratio_p_ep</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">coffee_lines_p_ep</span><span class="o">$</span><span class="n">Freq.y</span><span class="o">/</span><span class="n">coffee_lines_p_ep</span><span class="o">$</span><span class="n">Freq.x</span><span class="w">
  
  </span><span class="n">coffee_lines_p_ep</span><span class="o">$</span><span class="n">Season</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"_.*"</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">coffee_lines_p_ep</span><span class="o">$</span><span class="n">Var1</span><span class="p">)</span><span class="w">
  </span><span class="n">coffee_lines_p_ep</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="n">name</span><span class="p">)</span><span class="w">
  
  </span><span class="n">assign</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"coffee_lines_p_ep"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">),</span><span class="w"> </span><span class="n">coffee_lines_p_ep</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">

</span><span class="n">coffee_df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rbind</span><span class="p">(</span><span class="n">coffee_lines_p_ep_LORELAI</span><span class="p">,</span><span class="w"> </span><span class="n">coffee_lines_p_ep_RORY</span><span class="p">,</span><span class="w"> </span><span class="n">coffee_lines_p_ep_LUKE</span><span class="p">,</span><span class="w"> </span><span class="n">coffee_lines_p_ep_SOOKIE</span><span class="p">,</span><span class="w"> 
                   </span><span class="n">coffee_lines_p_ep_LANE</span><span class="p">,</span><span class="w"> </span><span class="n">coffee_lines_p_ep_PARIS</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span><span class="w">

</span><span class="n">my_theme</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">base_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">,</span><span class="w"> </span><span class="n">base_family</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"sans"</span><span class="p">){</span><span class="w">
  </span><span class="n">theme_grey</span><span class="p">(</span><span class="n">base_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">base_size</span><span class="p">,</span><span class="w"> </span><span class="n">base_family</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">base_family</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="w">
    </span><span class="n">axis.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">),</span><span class="w">
    </span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">45</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">),</span><span class="w">
    </span><span class="n">axis.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">14</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.grid.major</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_line</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.grid.minor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
    </span><span class="n">panel.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"aliceblue"</span><span class="p">),</span><span class="w">
    </span><span class="n">strip.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"lightgrey"</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="w">
    </span><span class="n">strip.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bold"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">),</span><span class="w">
    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bottom"</span><span class="p">,</span><span class="w">
    </span><span class="n">legend.justification</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"top"</span><span class="p">,</span><span class="w"> 
    </span><span class="n">legend.box</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"horizontal"</span><span class="p">,</span><span class="w">
    </span><span class="n">legend.box.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey50"</span><span class="p">),</span><span class="w">
    </span><span class="n">legend.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
    </span><span class="n">panel.border</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w">
  </span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">tidyr</span><span class="p">)</span><span class="w">

</span><span class="n">coffee_df_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">coffee_df</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">value</span><span class="p">,</span><span class="w"> </span><span class="n">Freq.x</span><span class="o">:</span><span class="n">ratio_p_ep</span><span class="p">)</span><span class="w">

</span><span class="n">coffee_df_gather</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">coffee_df_gather</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Freq.x"</span><span class="p">),</span><span class="w"> </span><span class="s2">"group"</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"Lines per episode"</span><span class="w">
</span><span class="n">coffee_df_gather</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">coffee_df_gather</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Freq.y"</span><span class="p">),</span><span class="w"> </span><span class="s2">"group"</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"Coffee lines per episode"</span><span class="w">
</span><span class="n">coffee_df_gather</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">coffee_df_gather</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"ratio_p_ep"</span><span class="p">),</span><span class="w"> </span><span class="s2">"group"</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"Ratio of coffee lines per episode"</span><span class="w">

</span><span class="n">coffee_df_gather</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">coffee_df_gather</span><span class="o">$</span><span class="n">character</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">names</span><span class="p">)</span><span class="w">
</span><span class="n">coffee_df_gather</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">coffee_df_gather</span><span class="o">$</span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Lines per episode"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Coffee lines per episode"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Ratio of coffee lines per episode"</span><span class="p">))</span><span class="w">

</span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">coffee_df_gather</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">character</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">value</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Season</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Season</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_boxplot</span><span class="p">(</span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="w">
    </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w">
    </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number or ratio of lines"</span><span class="p">,</span><span class="w">
    </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Lorelai &amp; Luke are the coffee queen and king"</span><span class="p">,</span><span class="w">
    </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Coffee mentions per episode and season"</span><span class="p">,</span><span class="w">
    </span><span class="n">caption</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"\nThese boxplots show the number of total lines per episode and season, lines with coffee mentions per epsiode and season,
     as well as the ratio between the two for the main characters Lorelai, Rory, Luke, Sookie, Lane and Paris. Lorelai consistently had
     the most lines per episode, followed closely by Rory and Luke. Sookie, Lane and Paris had roughly the same numbers of lines per episode,
     although their variance is higher. The same trend is reflected by the number of lines with coffee mentions, except that Rory and Luke seem
     to talk similarly often about coffee. Interestingly, Lorelai's (verbal) coffee obsession seems to have decreased slightly over the seasons.
     The ratio of coffee lines divided by the total number of lines for each episode reflects that even though Lorelai, Rory and Luke talk a lot about coffee,
     they also talk a lot in general, so their ratio of coffee mentions is only slightly higher than Sookie's, Lane's and Paris'. 
     While the latter's mean ratio is lower. they have a much higher variance with a few episodes where they seem to have talked relatively much about coffee.
     During the first seasons, Luke is the character with the highest ratio of coffee vs general talk but this is decreasing over time and becomes more
     similar to Lorelai and Rory's coffee talk ratio."</span><span class="w">
  </span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_wrap</span><span class="p">(</span><span class="o">~</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">scales</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"free"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">guides</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">guide_legend</span><span class="p">(</span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">byrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_brewer</span><span class="p">(</span><span class="n">palette</span><span class="o">=</span><span class="s2">"Dark2"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_color_brewer</span><span class="p">(</span><span class="n">palette</span><span class="o">=</span><span class="s2">"Dark2"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># calculate number of lines in total
</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">names</span><span class="p">){</span><span class="w">
  </span><span class="n">lines_p_ep</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">))</span><span class="w">
  </span><span class="n">lines_p_ep</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">lines_p_ep</span><span class="o">$</span><span class="n">episode</span><span class="p">))</span><span class="w">

  </span><span class="n">lines_p_ep</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="n">name</span><span class="p">)</span><span class="w">
  
  </span><span class="n">assign</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines_p_ep"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">),</span><span class="w"> </span><span class="n">lines_p_ep</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">

</span><span class="c1"># combine and convert to wide format to make calculating proportions easier
</span><span class="n">lines_df_wide</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">spread</span><span class="p">(</span><span class="n">rbind</span><span class="p">(</span><span class="n">lines_p_ep_LORELAI</span><span class="p">,</span><span class="w"> </span><span class="n">lines_p_ep_RORY</span><span class="p">,</span><span class="w"> </span><span class="n">lines_p_ep_LUKE</span><span class="p">,</span><span class="w"> </span><span class="n">lines_p_ep_SOOKIE</span><span class="p">,</span><span class="w"> 
                              </span><span class="n">lines_p_ep_LANE</span><span class="p">,</span><span class="w"> </span><span class="n">lines_p_ep_PARIS</span><span class="p">,</span><span class="w"> </span><span class="n">lines_p_ep_OTHER</span><span class="p">),</span><span class="w"> </span><span class="n">Var1</span><span class="p">,</span><span class="w"> </span><span class="n">Freq</span><span class="p">)</span><span class="w">
</span><span class="n">rownames</span><span class="p">(</span><span class="n">lines_df_wide</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">lines_df_wide</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="p">]</span><span class="w">
</span><span class="n">lines_df_wide</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">lines_df_wide</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]))</span><span class="w">

</span><span class="c1"># calculate proportions and percentage
</span><span class="n">prop_lines_df_wide</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sweep</span><span class="p">(</span><span class="n">lines_df_wide</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">rowSums</span><span class="p">(</span><span class="n">lines_df_wide</span><span class="p">),</span><span class="w"> </span><span class="n">FUN</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"/"</span><span class="p">)</span><span class="w">
</span><span class="n">percent_lines_df_wide</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">prop_lines_df_wide</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">100</span><span class="w">
</span><span class="n">percent_lines_df_wide</span><span class="o">$</span><span class="n">Episode</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">percent_lines_df_wide</span><span class="p">)</span><span class="w">
</span><span class="n">percent_lines_df_wide</span><span class="o">$</span><span class="n">Season</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"_.*"</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">percent_lines_df_wide</span><span class="o">$</span><span class="n">Episode</span><span class="p">)</span><span class="w">

</span><span class="c1"># gather for plotting
</span><span class="n">percent_lines_df_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">percent_lines_df_wide</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">Character</span><span class="p">,</span><span class="w"> </span><span class="n">Percent</span><span class="p">,</span><span class="w"> </span><span class="n">LANE</span><span class="o">:</span><span class="n">SOOKIE</span><span class="p">)</span><span class="w">

</span><span class="n">percent_lines_df_gather</span><span class="o">$</span><span class="n">Character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">percent_lines_df_gather</span><span class="o">$</span><span class="n">Character</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">names</span><span class="p">)</span><span class="w">
</span><span class="n">percent_lines_df_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">percent_lines_df_gather</span><span class="p">,</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">lines_LORELAI</span><span class="p">,</span><span class="w"> </span><span class="n">select</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">episode</span><span class="p">,</span><span class="w"> </span><span class="n">episode_running_nr</span><span class="p">)),</span><span class="w"> </span><span class="n">by.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Episode"</span><span class="p">,</span><span class="w"> </span><span class="n">by.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"episode"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">percent_lines_df_gather</span><span class="o">$</span><span class="n">Episode</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">percent_lines_df_gather</span><span class="o">$</span><span class="n">Episode</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">percent_lines_df_gather</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">percent_lines_df_gather</span><span class="o">$</span><span class="n">episode_running_nr</span><span class="p">,</span><span class="w"> </span><span class="n">decreasing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="s2">"Episode"</span><span class="p">])</span><span class="w">
</span><span class="n">percent_lines_df_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">percent_lines_df_gather</span><span class="p">[</span><span class="o">!</span><span class="n">duplicated</span><span class="p">(</span><span class="n">percent_lines_df_gather</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">percent_lines_df_gather</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Character</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Percent</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Season</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_boxplot</span><span class="p">(</span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="w">
    </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w">
    </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Percent"</span><span class="p">,</span><span class="w">
    </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Who's the biggest talker? (a)"</span><span class="p">,</span><span class="w">
    </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Percentage of lines per episode spoken by main characters"</span><span class="p">,</span><span class="w">
    </span><span class="n">caption</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"\nThe boxplot shows the percentages of lines spoken by the main characters per episode and season.
    In the first three seasons the majority of lines are clearly spoken by Lorelai and Rory, followed with some distance by Luke.
    During the later seasons, the gap between Lorelai, Rory and the rest is not as pronounced any more."</span><span class="w">
  </span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_wrap</span><span class="p">(</span><span class="o">~</span><span class="w"> </span><span class="n">Season</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="n">scales</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"free"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">guides</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">guide_legend</span><span class="p">(</span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">byrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_brewer</span><span class="p">(</span><span class="n">palette</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Dark2"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_color_brewer</span><span class="p">(</span><span class="n">palette</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Dark2"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># calculate number of lines in total
</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">names</span><span class="p">){</span><span class="w">
  </span><span class="n">lines_p_season</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">))</span><span class="w">
  </span><span class="n">lines_p_season</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">lines_p_season</span><span class="o">$</span><span class="n">season</span><span class="p">))</span><span class="w">

  </span><span class="n">lines_p_season</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="n">name</span><span class="p">)</span><span class="w">
  
  </span><span class="n">assign</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines_p_season"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">),</span><span class="w"> </span><span class="n">lines_p_season</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">

</span><span class="c1"># combine and convert to wide format to make calculating proportions easier
</span><span class="n">lines_df_wide</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">spread</span><span class="p">(</span><span class="n">rbind</span><span class="p">(</span><span class="n">lines_p_season_LORELAI</span><span class="p">,</span><span class="w"> </span><span class="n">lines_p_season_RORY</span><span class="p">,</span><span class="w"> </span><span class="n">lines_p_season_LUKE</span><span class="p">,</span><span class="w"> </span><span class="n">lines_p_season_SOOKIE</span><span class="p">,</span><span class="w"> 
                              </span><span class="n">lines_p_season_LANE</span><span class="p">,</span><span class="w"> </span><span class="n">lines_p_season_PARIS</span><span class="p">,</span><span class="w"> </span><span class="n">lines_p_season_OTHER</span><span class="p">),</span><span class="w"> </span><span class="n">Var1</span><span class="p">,</span><span class="w"> </span><span class="n">Freq</span><span class="p">)</span><span class="w">
</span><span class="n">rownames</span><span class="p">(</span><span class="n">lines_df_wide</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">lines_df_wide</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="p">]</span><span class="w">
</span><span class="n">lines_df_wide</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">lines_df_wide</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]))</span><span class="w">

</span><span class="c1"># calculate proportions and percentage
</span><span class="n">prop_lines_df_wide</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sweep</span><span class="p">(</span><span class="n">lines_df_wide</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">rowSums</span><span class="p">(</span><span class="n">lines_df_wide</span><span class="p">),</span><span class="w"> </span><span class="n">FUN</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"/"</span><span class="p">)</span><span class="w">
</span><span class="n">percent_lines_df_wide</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">prop_lines_df_wide</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">100</span><span class="w">
</span><span class="n">percent_lines_df_wide</span><span class="o">$</span><span class="n">Season</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">percent_lines_df_wide</span><span class="p">)</span><span class="w">

</span><span class="c1"># gather for plotting
</span><span class="n">percent_lines_df_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">percent_lines_df_wide</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">gather</span><span class="p">(</span><span class="n">Character</span><span class="p">,</span><span class="w"> </span><span class="n">Percent</span><span class="p">,</span><span class="w"> </span><span class="n">LANE</span><span class="o">:</span><span class="n">SOOKIE</span><span class="p">)</span><span class="w">

</span><span class="n">percent_lines_df_gather</span><span class="o">$</span><span class="n">Character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">percent_lines_df_gather</span><span class="o">$</span><span class="n">Character</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">names</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># plot
</span><span class="n">bp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">percent_lines_df_gather</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Percent</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Character</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_bar</span><span class="p">(</span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"identity"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">theme_minimal</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_brewer</span><span class="p">(</span><span class="n">palette</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Spectral"</span><span class="p">)</span><span class="w">

</span><span class="n">pie</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bp</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">coord_polar</span><span class="p">(</span><span class="s2">"y"</span><span class="p">,</span><span class="w"> </span><span class="n">start</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="w">
  </span><span class="n">axis.title.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
  </span><span class="n">axis.title.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
  </span><span class="n">panel.border</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
  </span><span class="n">panel.grid</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
  </span><span class="n">axis.ticks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
  </span><span class="n">plot.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">14</span><span class="p">,</span><span class="w"> </span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bold"</span><span class="p">),</span><span class="w">
  </span><span class="n">legend.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
  </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bottom"</span><span class="p">,</span><span class="w">
  </span><span class="n">legend.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">8</span><span class="p">),</span><span class="w">
  </span><span class="n">legend.justification</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"top"</span><span class="p">,</span><span class="w"> 
    </span><span class="n">legend.box</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"horizontal"</span><span class="p">,</span><span class="w">
    </span><span class="n">legend.box.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey50"</span><span class="p">),</span><span class="w">
    </span><span class="n">legend.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
  </span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">guides</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">guide_legend</span><span class="p">(</span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">byrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">))</span><span class="w">

</span><span class="n">pie</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">facet_wrap</span><span class="p">(</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">Season</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">4</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="w">
    </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Who's the biggest talker? (b)"</span><span class="p">,</span><span class="w">
    </span><span class="n">subtitle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Percent of lines per season spoken by main characters"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<hr />

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sessionInfo</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## R version 3.3.2 (2016-10-31)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: macOS Sierra 10.12
## 
## locale:
## [1] de_DE.UTF-8/de_DE.UTF-8/de_DE.UTF-8/C/de_DE.UTF-8/de_DE.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] ggplot2_2.2.0         dplyr_0.5.0           splitstackshape_1.4.2
## [4] data.table_1.9.6      tidyr_0.6.0          
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.8        knitr_1.15         magrittr_1.5      
##  [4] munsell_0.4.3      colorspace_1.3-1   R6_2.2.0          
##  [7] stringr_1.1.0      plyr_1.8.4         tools_3.3.2       
## [10] grid_3.3.2         gtable_0.2.0       DBI_0.5-1         
## [13] htmltools_0.3.5    lazyeval_0.2.0     yaml_2.1.14       
## [16] assertthat_0.1     digest_0.6.10      tibble_1.2        
## [19] RColorBrewer_1.1-2 evaluate_0.10      rmarkdown_1.1     
## [22] labeling_0.3       stringi_1.1.2      scales_0.4.1      
## [25] chron_2.3-47
</code></pre>
</div>

  </div>

  <h1><a href="/text_analysis/2016/11/13/gilmore_girls_part1_post">Creating a Gilmore Girls character network with R</a></h1>
  <p class="author">
    <span class="date">2016-11-13 00:00:00 +0100</span>
  </p>
  <div class="content">
    <p>With the impending (and by many - including me - much awaited) <a href="http://www.imdb.com/title/tt5435008/">Gilmore Girls Revival</a>, I wanted to take a somewhat different look at our beloved characters from Stars Hollow.</p>

<p>I had recently read a few cool examples of how to create <a href="http://www.rdatamining.com/examples/social-network-analysis">co-occurrence networks</a> and wanted to combine this with an analysis similar to what <a href="http://varianceexplained.org/">David Robinson</a> did for <a href="http://varianceexplained.org/r/love-actually-network/">Love Actually</a>.</p>

<p>Fortunately, there are people out there, who have invested their time (and a lot of it, I imagine) to write up transcripts for every Gilmore Girls episode. I chose <a href="http://www.crazy-internet-people.com/site/gilmoregirls">www.crazy-internet-people.com</a>’s list of transcripts.</p>

<p>Based on these transcripts I calculated the main character’s number of lines per episode and from there the co-occurrence matrix with other characters. This told me with which other major characters they appeared together in episodes and how often. This network nicely illustrates that Lorelai has the most lines of all characters (node size reflects total number of lines in all episodes a character had), followed by Rory. No surprise there but interestingly, the third place goes to Luke and fourth to Emily. Most interaction happened between Lorelai and Rory of course (edge width reflects number of co-occurences in episodes), but Lorelai and Luke, Rory and Luke, Emily and Rory and Lorelai and Sookie follow suit. What the network also shows is that Lorelai has major connections with most other characters - more so than Rory.</p>

<p><img src="gilmore_girls_part1_post_files/figure-markdown_github/unnamed-chunk-11-1.png" style="display: block; margin: auto;" /></p>

<p>A cluster dendrogram shows us the character co-occurrence in a slightly different way: The further down in the dendrogram tree two nodes split, the more episodes these characters had in common. Again, no surprise here that Lorelai and Rory have the most closely connected nodes, closely followed by Luke. We can also see quite nicely that the couples Sookie and Jackson, Lane and Zack and Emily and Richard share a lot of episodes.</p>

<p><img src="gilmore_girls_part1_post_files/figure-markdown_github/unnamed-chunk-12-1.png" style="display: block; margin: auto;" /></p>

<p><br /></p>

<p>Of course, these numbers also reflect the total number of episodes that characters were in, so that there is an inherent bias for characters with short occurrences in many episodes being more strongly connected to e.g. Lorelai than characters with fewer but more important plots. It would have been interesting to calculate the co-occurrence per scene instead of episode but unfortunately, this information was not given in the transcripts (if someone knows transcripts that denote scene number, please contact me).</p>

<p>I also wanted to see in which episodes these 20 characters appeared. Of course, Lorelai and Rory appeared in every episode but for other characters, there are clear gaps.</p>

<p><br /></p>

<p><img src="gilmore_girls_part1_post_files/figure-markdown_github/unnamed-chunk-14-1.png" style="display: block; margin: auto;" /></p>

<p><br /></p>

<p>And finally, I wanted to know how many lines per episode each of them spoke: This boxplot shows the median number of lines per episode for each character (middle line of the boxes), as well as the lower and upper quartiles (outer edges of the boxes) and the outlier episodes (dots).</p>

<p><br /></p>

<p><img src="gilmore_girls_part1_post_files/figure-markdown_github/unnamed-chunk-15-1.png" style="display: block; margin: auto;" /></p>

<p><br /></p>

<p>There will be a <a href="https://shiring.github.io/text_analysis/2016/11/20/gilmore_girls_part2_post">part 2 next week</a>, where I will explore the Gilmore Girls a bit more. No spoilers, but among other things, I’ll be looking at their coffee consumption through the data lens…</p>

<p><br /></p>

<p>For a detailed description, plus R code for the plots see further below or find the <a href="https://github.com/ShirinG/blog_posts_prep/blob/master/character_network/gilmore_girls.Rmd">R Markdown on Github</a>.</p>

<p>If you don’t care about the show and have not (unlike me) watched every episode at least twice, maybe you’ll be interested in using my R code to recreate a similar character network for other TV shows, movies or books (and if you do, please share them with me)!</p>

<p><br /></p>

<hr />

<h2 id="obtaining-all-episode-transcripts">Obtaining all episode transcripts</h2>

<p>The transcript URLs from <a href="http://www.crazy-internet-people.com/site/gilmoregirls">www.crazy-internet-people.com</a> have the following scheme: “<a href="http://www.crazy-internet-people.com/site/gilmoregirls/pages/">http://www.crazy-internet-people.com/site/gilmoregirls/pages/</a>, s#/, s#s/, §.html” (#: number of season, of which there are seven; §: running number of episode, from 1 to 153). Following this scheme, I looped over all seasons and episodes to read the lines for each HTML directly into R via their respective URLs.</p>

<p>Because the raw HTML looked a bit messy, I had to do some tidying of the text:</p>

<ul>
  <li>First, I grabbed only lines with a character name at the beginning, which indicates the character who is speaking (these were all in caps).</li>
  <li>Then, I had to remove the remaining HTML tags/ descriptors.</li>
  <li>After this, I had the transcript text remaining, which I could transform into a data frame</li>
  <li>and add season, episode number and running episode number to each line of text.</li>
  <li>And finally, I combined all transcripts into one object.</li>
</ul>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="k">for</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">7</span><span class="p">){</span><span class="w">                                            </span><span class="c1"># there are 7 seasons
</span><span class="w">
  </span><span class="k">if</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">1</span><span class="p">){</span><span class="w">                                             </span><span class="c1"># all seasons except the first have 22 episodes (the first has 21)
</span><span class="w">
    </span><span class="k">for</span><span class="p">(</span><span class="n">j</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">21</span><span class="p">){</span><span class="w">

      </span><span class="n">cat</span><span class="p">(</span><span class="s2">"\nSeason"</span><span class="p">,</span><span class="w"> </span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="s2">", Episode"</span><span class="p">,</span><span class="w"> </span><span class="n">j</span><span class="p">,</span><span class="w"> </span><span class="s2">"\n"</span><span class="p">)</span><span class="w">            </span><span class="c1"># to see the progress I am printing the season and episode number
</span><span class="w">
      </span><span class="n">thepage</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">readLines</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">"http://www.crazy-internet-people.com/site/gilmoregirls/pages/s"</span><span class="p">,</span><span class="w"> </span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="s2">"/s"</span><span class="p">,</span><span class="w"> </span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="s2">"s/"</span><span class="p">,</span><span class="w"> </span><span class="n">j</span><span class="p">,</span><span class="w"> </span><span class="s2">".html"</span><span class="p">))</span><span class="w">

      </span><span class="n">thepage</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">thepage</span><span class="p">[</span><span class="n">grep</span><span class="p">(</span><span class="s2">"^[[:upper:]]+:"</span><span class="p">,</span><span class="w"> </span><span class="n">thepage</span><span class="p">)]</span><span class="w">  </span><span class="c1"># grabbing character lines only
</span><span class="w">      </span><span class="n">thepage</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"\t"</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">thepage</span><span class="p">)</span><span class="w">                  </span><span class="c1"># removing HTML tags
</span><span class="w">      </span><span class="n">thepage</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"&lt;.*&gt;"</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">thepage</span><span class="p">)</span><span class="w">                </span><span class="c1"># removing some more HTML tags
</span><span class="w">
      </span><span class="n">thepage</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">thepage</span><span class="p">)</span><span class="w">
      </span><span class="n">thepage</span><span class="o">$</span><span class="n">season</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">i</span><span class="w">                                 </span><span class="c1"># add season number
</span><span class="w">      </span><span class="n">thepage</span><span class="o">$</span><span class="n">episode</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="n">j</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)</span><span class="w">           </span><span class="c1"># add episode number
</span><span class="w">      </span><span class="n">thepage</span><span class="o">$</span><span class="n">episode_running_nr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">j</span><span class="w">                     </span><span class="c1"># add running epsiode number
</span><span class="w">
      </span><span class="k">if</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">j</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">1</span><span class="p">){</span><span class="w">                                </span><span class="c1"># combine all transcripts into one object
</span><span class="w">        </span><span class="n">transcripts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">thepage</span><span class="w">
      </span><span class="p">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="p">{</span><span class="w">
        </span><span class="n">transcripts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rbind</span><span class="p">(</span><span class="n">transcripts</span><span class="p">,</span><span class="w"> </span><span class="n">thepage</span><span class="p">)</span><span class="w">
      </span><span class="p">}</span><span class="w">
    </span><span class="p">}</span><span class="w">

    </span><span class="p">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="p">{</span><span class="w">                                              </span><span class="c1"># repeat for seasons 2 to 7
</span><span class="w">
      </span><span class="k">for</span><span class="p">(</span><span class="n">j</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">22</span><span class="p">){</span><span class="w">

        </span><span class="n">cat</span><span class="p">(</span><span class="s2">"\nSeason"</span><span class="p">,</span><span class="w"> </span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="s2">", Episode"</span><span class="p">,</span><span class="w"> </span><span class="n">j</span><span class="p">,</span><span class="w"> </span><span class="s2">"\n"</span><span class="p">)</span><span class="w">

        </span><span class="k">if</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">2</span><span class="p">){</span><span class="w">                                       </span><span class="c1"># to get the running episode number, 
</span><span class="w">                                                          </span><span class="c1"># I have to add the number of episodes from previous seasons
</span><span class="w">          </span><span class="n">n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">j</span><span class="m">+21</span><span class="w">
        </span><span class="p">}</span><span class="w">

        </span><span class="k">if</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">3</span><span class="p">){</span><span class="w">
          </span><span class="n">n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">j</span><span class="m">+21+22</span><span class="w">
        </span><span class="p">}</span><span class="w">

        </span><span class="k">if</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">4</span><span class="p">){</span><span class="w">
          </span><span class="n">n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">j</span><span class="m">+21+22+22</span><span class="w">
        </span><span class="p">}</span><span class="w">

        </span><span class="k">if</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">5</span><span class="p">){</span><span class="w">
          </span><span class="n">n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">j</span><span class="m">+21+22+22+22</span><span class="w">
        </span><span class="p">}</span><span class="w">

        </span><span class="k">if</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">6</span><span class="p">){</span><span class="w">
          </span><span class="n">n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">j</span><span class="m">+21+22+22+22+22</span><span class="w">
        </span><span class="p">}</span><span class="w">

        </span><span class="k">if</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">7</span><span class="p">){</span><span class="w">
          </span><span class="n">n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">j</span><span class="m">+21+22+22+22+22+22</span><span class="w">
        </span><span class="p">}</span><span class="w">

        </span><span class="c1"># rinse and repeat
</span><span class="w">        
        </span><span class="n">thepage</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">readLines</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">"http://www.crazy-internet-people.com/site/gilmoregirls/pages/s"</span><span class="p">,</span><span class="w"> </span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="s2">"/s"</span><span class="p">,</span><span class="w"> </span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="s2">"s/"</span><span class="p">,</span><span class="w"> </span><span class="n">n</span><span class="p">,</span><span class="w"> </span><span class="s2">".html"</span><span class="p">))</span><span class="w">

        </span><span class="n">thepage</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">thepage</span><span class="p">[</span><span class="n">grep</span><span class="p">(</span><span class="s2">"^[[:upper:]]{2,}:"</span><span class="p">,</span><span class="w"> </span><span class="n">thepage</span><span class="p">)]</span><span class="w">
        </span><span class="n">thepage</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"\t"</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">thepage</span><span class="p">)</span><span class="w">
        </span><span class="n">thepage</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"&lt;.*&gt;"</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">thepage</span><span class="p">)</span><span class="w">

        </span><span class="n">thepage</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">thepage</span><span class="p">)</span><span class="w">
        </span><span class="n">thepage</span><span class="o">$</span><span class="n">season</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">i</span><span class="w">
        </span><span class="n">thepage</span><span class="o">$</span><span class="n">episode</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="n">j</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)</span><span class="w">
        </span><span class="n">thepage</span><span class="o">$</span><span class="n">episode_running_nr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">n</span><span class="w">

        </span><span class="k">if</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">j</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">1</span><span class="p">){</span><span class="w">
          </span><span class="n">transcripts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">thepage</span><span class="w">
        </span><span class="p">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="p">{</span><span class="w">
          </span><span class="n">transcripts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rbind</span><span class="p">(</span><span class="n">transcripts</span><span class="p">,</span><span class="w"> </span><span class="n">thepage</span><span class="p">)</span><span class="w">
        </span><span class="p">}</span><span class="w">
        </span><span class="p">}</span><span class="w">
  </span><span class="p">}</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<p>Some of the lines were empty, so I removed those.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">transcripts</span><span class="o">$</span><span class="n">thepage</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">transcripts</span><span class="o">$</span><span class="n">thepage</span><span class="p">)</span><span class="w">  </span><span class="c1"># convert to character vector
</span><span class="n">transcripts</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">transcripts</span><span class="p">[</span><span class="o">!</span><span class="n">transcripts</span><span class="o">$</span><span class="n">thepage</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">  </span><span class="c1"># remove empty lines
</span></code></pre>
</div>

<p><br /></p>

<p>This is how the data frame looked like at this point:</p>

<ul>
  <li>each row contains one line of dialogue with the character name in caps coming before their lines.</li>
</ul>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">head</span><span class="p">(</span><span class="n">transcripts</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##                                          thepage season episode
## 1 LORELAI: Please, Luke. Please, please, please.      1     1_1
## 2 LUKE: How many cups have you had this morning?      1     1_1
## 3                                 LORELAI: None.      1     1_1
## 4                                  LUKE: Plus...      1     1_1
## 5            LORELAI: Five, but yours is better.      1     1_1
## 6                      LUKE: You have a problem.      1     1_1
##   episode_running_nr
## 1                  1
## 2                  1
## 3                  1
## 4                  1
## 5                  1
## 6                  1
</code></pre>
</div>

<p><br /></p>

<p>To be able to count the characters, I separated the character names from their lines. This was done by splitting the first column after the first colon, using the <strong>tidyr</strong> package.</p>

<p>I also removed all leading and trailing whitespace from the character names, changed all letters in the character column to all caps and changed “ands” and apostrophes to the proper encoding. And I also had to manually correct quite a few misspelled character names.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># separate first column after first colon
</span><span class="n">library</span><span class="p">(</span><span class="n">tidyr</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">separate</span><span class="p">(</span><span class="n">transcripts</span><span class="p">,</span><span class="w"> </span><span class="s2">"thepage"</span><span class="p">,</span><span class="w"> </span><span class="n">into</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"character"</span><span class="p">,</span><span class="w"> </span><span class="s2">"dialogue"</span><span class="p">),</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">":"</span><span class="p">,</span><span class="w"> </span><span class="n">extra</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"merge"</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"right"</span><span class="p">)</span><span class="w">

</span><span class="c1"># remove leading and trailing whitespace
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"^\\s+|\\s+$"</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">

</span><span class="c1"># convert all character names to all upper case
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">toupper</span><span class="p">(</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">

</span><span class="c1"># fix misspelled character names
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"ZACK"</span><span class="p">,</span><span class="w"> </span><span class="s2">"ZACH"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"LORLEAI"</span><span class="p">,</span><span class="w"> </span><span class="s2">"LORELAI"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"LOREALI"</span><span class="p">,</span><span class="w"> </span><span class="s2">"LORELAI"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"LORELI"</span><span class="p">,</span><span class="w"> </span><span class="s2">"LORELAI"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"LORLAI"</span><span class="p">,</span><span class="w"> </span><span class="s2">"LORELAI"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"LORELA$"</span><span class="p">,</span><span class="w"> </span><span class="s2">"LORELAI"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"LORLELAI"</span><span class="p">,</span><span class="w"> </span><span class="s2">"LORELAI"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"^ORELAI"</span><span class="p">,</span><span class="w"> </span><span class="s2">"LORELAI"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"LOREAI"</span><span class="p">,</span><span class="w"> </span><span class="s2">"LORELAI"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"^ORY"</span><span class="p">,</span><span class="w"> </span><span class="s2">"RORY"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"LUK$"</span><span class="p">,</span><span class="w"> </span><span class="s2">"LUKE"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"BABETE"</span><span class="p">,</span><span class="w"> </span><span class="s2">"BABETTE"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"BABETTER"</span><span class="p">,</span><span class="w"> </span><span class="s2">"BABETTE"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"BARBETTE"</span><span class="p">,</span><span class="w"> </span><span class="s2">"BABETTE"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"BABETTE/MISS PATTY"</span><span class="p">,</span><span class="w"> </span><span class="s2">"BABETTE AND MISS PATTY"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"JACKSON/SOOKIE"</span><span class="p">,</span><span class="w"> </span><span class="s2">"JACKSON AND SOOKIE"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"LORELAI/SOOKIE"</span><span class="p">,</span><span class="w"> </span><span class="s2">"LORELAI AND SOOKIE"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"LORELAI/RORY"</span><span class="p">,</span><span class="w"> </span><span class="s2">"LORELAI AND RORY"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"TAYOR"</span><span class="p">,</span><span class="w"> </span><span class="s2">"TAYLOR"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"TRISTIN"</span><span class="p">,</span><span class="w"> </span><span class="s2">"TRISTAN"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"MICHE$"</span><span class="p">,</span><span class="w"> </span><span class="s2">"MICHEL"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"MICHELL"</span><span class="p">,</span><span class="w"> </span><span class="s2">"MICHEL"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"SOOKI$"</span><span class="p">,</span><span class="w"> </span><span class="s2">"SOOKIE"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"SOOKEI"</span><span class="p">,</span><span class="w"> </span><span class="s2">"SOOKIE"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"SOOKIES"</span><span class="p">,</span><span class="w"> </span><span class="s2">"SOOKIE"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"Mrs.KIM"</span><span class="p">,</span><span class="w"> </span><span class="s2">"MRS KIM"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"MRS.KIM"</span><span class="p">,</span><span class="w"> </span><span class="s2">"MRS KIM"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"MRS KIM"</span><span class="p">,</span><span class="w"> </span><span class="s2">"MRS KIM"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"RICHRAD"</span><span class="p">,</span><span class="w"> </span><span class="s2">"RICHARD"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"RMILY"</span><span class="p">,</span><span class="w"> </span><span class="s2">"EMILY"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"CHRISTOHPER"</span><span class="p">,</span><span class="w"> </span><span class="s2">"CHRISTOPHER"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"CHRISTOPER"</span><span class="p">,</span><span class="w"> </span><span class="s2">"CHRISTOPHER"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"CHRSTOPHER"</span><span class="p">,</span><span class="w"> </span><span class="s2">"CHRISTOPHER"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"CHRIS$"</span><span class="p">,</span><span class="w"> </span><span class="s2">"CHRISTOPHER"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"CHERRY"</span><span class="p">,</span><span class="w"> </span><span class="s2">"SHERRY"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"LINDAY"</span><span class="p">,</span><span class="w"> </span><span class="s2">"LINDSAY"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">

</span><span class="c1"># substitute &amp;#146; with apostrophe
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"&amp;#146;"</span><span class="p">,</span><span class="w"> </span><span class="s2">"'"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">

</span><span class="c1"># some ANDs are written as &amp;AMP; so they will be changed as well
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"&amp;AMP;"</span><span class="p">,</span><span class="w"> </span><span class="s2">"AND"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">

</span><span class="c1"># and finally I want ANDs to be written as semicolons
</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">" AND "</span><span class="p">,</span><span class="w"> </span><span class="s2">";"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">)</span><span class="w">

</span><span class="c1">#  and remove disclaimer lines
</span><span class="n">transcripts_2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">transcripts_2</span><span class="p">[</span><span class="o">-</span><span class="n">which</span><span class="p">(</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"DISCLAIMER"</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w"> 
</span><span class="n">head</span><span class="p">(</span><span class="n">transcripts_2</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   character                                  dialogue season episode
## 1   LORELAI     Please, Luke. Please, please, please.      1     1_1
## 2      LUKE  How many cups have you had this morning?      1     1_1
## 3   LORELAI                                     None.      1     1_1
## 4      LUKE                                   Plus...      1     1_1
## 5   LORELAI                Five, but yours is better.      1     1_1
## 6      LUKE                       You have a problem.      1     1_1
##   episode_running_nr
## 1                  1
## 2                  1
## 3                  1
## 4                  1
## 5                  1
## 6                  1
</code></pre>
</div>

<p>This is how the data frame looked like after tidying.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">nrow</span><span class="p">(</span><span class="n">transcripts_2</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 116954
</code></pre>
</div>

<p>In total there are now 116,983 lines.</p>

<p><br /></p>

<h2 id="how-many-characters-are-there-and-how-many-lines-do-they-have">How many characters are there and how many lines do they have?</h2>

<p>To find out how many characters there were in Gilmore Girls during 153 episodes, I couldn’t simply count them because there are combined characters (e.g. Lorelai and Rory speaking together) and voice overs among them.</p>

<p>First, I want to duplicate all lines with two speakers, to make them count for each character. I also want to only count lines where there is only one character, so I removed all character fields with multiple, generic or unspecific characters. And I don’t want to have voice overs either.</p>

<p>Most of the characters, however, were still not recurring characters, so I filtered out all those characters that only occurred in one episode.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># separating all rows where multiple characters spoke into one line per character with duplicate line text
</span><span class="n">library</span><span class="p">(</span><span class="n">splitstackshape</span><span class="p">)</span><span class="w">
</span><span class="n">transcripts_2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">cSplit</span><span class="p">(</span><span class="n">transcripts_2</span><span class="p">,</span><span class="w"> </span><span class="n">splitCols</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"character"</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">";"</span><span class="p">,</span><span class="w"> </span><span class="n">direction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"long"</span><span class="p">)</span><span class="w">

</span><span class="c1"># manually removing characters I don't want to keep
</span><span class="n">characters</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">episode</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">" VOICE"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"^ALL"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"AS A GROUP"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"CROWD"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"RADIO"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"BIKERS"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"ANNOUNCER"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"BOTH"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"WOMAN"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"VOICE"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"BARTENDER"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"OFFICER"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"GIRL"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"GIRLS"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"BOYS"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"EVERYONE"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"SUPERVISOR"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"PHOTOGRAPHER"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"RECEPTIONIST"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"CUSTOMER"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"TV"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"VET"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"KID"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"MOM"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"DOCTOR"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"BOUNCER"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"LYRICS"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"SPEAKER"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"BOY"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"TEACHER"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"EMPLOYEE"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"MAID"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"CASHIER"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"MAN"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"NURSE"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"PLAYER"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"WAITER"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"WAITRESS"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"DRIVER"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"GRANDMA"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"LADIES"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"OPERATOR"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"TOURIST"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"CHEF"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"GUEST"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"HOSTESS"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"JUDGE"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"LADY"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"SECRETARY"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"PRIEST"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"STUDENT"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"PROFESSOR"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">subset</span><span class="p">(</span><span class="o">!</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"ANSWERING MACHINE"</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="p">))</span><span class="w">

</span><span class="c1"># find out which characters had the most lines and which to remove
</span><span class="n">library</span><span class="p">(</span><span class="n">reshape2</span><span class="p">)</span><span class="w">
</span><span class="n">characters_2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">dcast</span><span class="p">(</span><span class="n">characters</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">Var2</span><span class="p">)</span><span class="w">
</span><span class="n">rownames</span><span class="p">(</span><span class="n">characters_2</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">characters_2</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="p">]</span><span class="w">
</span><span class="n">characters_2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">characters_2</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]</span><span class="w">

</span><span class="n">characters_3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rowSums</span><span class="p">(</span><span class="n">characters_2</span><span class="p">)</span><span class="w">
</span><span class="n">major_characters</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">characters_3</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">characters_3</span><span class="p">,</span><span class="w"> </span><span class="n">decreasing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)]</span><span class="w">

</span><span class="n">characters_2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">characters_2</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w">

</span><span class="n">characters_to_remove</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">characters_2</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">rowSums</span><span class="p">(</span><span class="n">characters_2</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="w"> </span><span class="p">])</span><span class="w">

</span><span class="n">characters</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">characters</span><span class="p">[</span><span class="o">-</span><span class="n">which</span><span class="p">(</span><span class="n">characters</span><span class="o">$</span><span class="n">Var1</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">characters_to_remove</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="nf">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">characters</span><span class="o">$</span><span class="n">Var1</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 153
</code></pre>
</div>

<p>This leaves us with these 153 characters:</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">unique</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">characters</span><span class="o">$</span><span class="n">Var1</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   [1] "AK"               "ALEX"             "ANDREW"          
##   [4] "ANNA"             "APRIL"            "ASHER"           
##   [7] "AUDREY"           "AUNT JUN"         "BABETTE"         
##  [10] "BARBARA"          "BILL"             "BILLY"           
##  [13] "BOBBI"            "BOBBY"            "BOOTSY"          
##  [16] "BRAD"             "BRIAN"            "BRUCE"           
##  [19] "CAESAR"           "CAESER"           "CARL"            
##  [22] "CAROL"            "CARRIE"           "CHARLESTON"      
##  [25] "CHARLIE"          "CHRISTOPHER"      "CLARA"           
##  [28] "CLAUDE"           "COLIN"            "DAVE"            
##  [31] "DEAN"             "DENNIS"           "DEREK"           
##  [34] "DOUG"             "DOYLE"            "DR SCHULTZ"      
##  [37] "DRELLA"           "ED"               "EMCEE"           
##  [40] "EMILY"            "FINN"             "FLOYD"           
##  [43] "FRAN"             "FRANCIE"          "FRANCINE"        
##  [46] "FRED"             "GIGI"             "GIL"             
##  [49] "GLENN"            "GUY"              "GYPSY"           
##  [52] "HARRY"            "HEADMASTER"       "HENRY"           
##  [55] "HONOR"            "JACK"             "JACKSON"         
##  [58] "JACOB"            "JAMIE"            "JANET"           
##  [61] "JASON"            "JESS"             "JIMMY"           
##  [64] "JOE"              "JOEL"             "JOHN"            
##  [67] "JONI"             "JORDAN"           "JULIET"          
##  [70] "KAREN"            "KIRK"             "KYLE"            
##  [73] "KYON"             "LANCE"            "LANE"            
##  [76] "LAURA"            "LINDSAY"          "LIZ"             
##  [79] "LIZA"             "LOGAN"            "LORELAI"         
##  [82] "LOU"              "LOUISE"           "LUCY"            
##  [85] "LUKE"             "LULU"             "MADELINE"        
##  [88] "MAGGIE"           "MARCIA"           "MARILYN"         
##  [91] "MARSHALL"         "MARTY"            "MAUREEN"         
##  [94] "MAX"              "MEGAN"            "MIA"             
##  [97] "MICHEL"           "MISS PATTY"       "MITCHUM"         
## [100] "MOREY"            "MR. HUNTER"       "MRS KIM"         
## [103] "MRS. CASSINI"     "MRS. KIM"         "MRS. LISTER"     
## [106] "MRS. O'MALLEY"    "MUSIC"            "NANCY"           
## [109] "NATALIE"          "NICK"             "NICOLE"          
## [112] "NORA"             "OLIVIA"           "PARIS"           
## [115] "PATTY"            "PAUL ANKA"        "PHILLIP"         
## [118] "PRINCIPAL"        "RABBI"            "RACHEL"          
## [121] "RAJ"              "REVEREND"         "REVEREND SKINNER"
## [124] "RICHARD"          "ROB"              "ROBERT"          
## [127] "RORY"             "ROSEMARY"         "RUNE"            
## [130] "SANDRA"           "SARAH"            "SHANE"           
## [133] "SHEILA"           "SHERRY"           "SHIRA"           
## [136] "SIMON"            "SOOKIE"           "SOPHIE"          
## [139] "STRAUB"           "SUE"              "SUSAN"           
## [142] "T.J.'S BROTHER"   "TANNA"            "TAYLOR"          
## [145] "TJ"               "TOBIN"            "TOM"             
## [148] "TRISTAN"          "TRIX"             "TROUBADOUR"      
## [151] "VIVIAN"           "YOUNG CHUI"       "ZACH"
</code></pre>
</div>

<p><br /></p>

<h2 id="character-co-occurrence-network">Character co-occurrence network</h2>

<p>The major part was to calculate the co-occurence matrix for characters per episode (i.e. which characters co-occurred together in episodes and how often). Because the network would get too big had I used too many characters, I restricted the network to the 20 main characters with the most lines over all episodes.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># extracting only those characters with the most lines
</span><span class="n">transcripts_3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">transcripts_2</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">transcripts_2</span><span class="o">$</span><span class="n">character</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">major_characters</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">])),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="n">transcripts_3</span><span class="o">$</span><span class="n">season_name</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="s2">"Season"</span><span class="p">,</span><span class="w"> </span><span class="n">transcripts_3</span><span class="o">$</span><span class="n">season</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">" "</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># create lines per episode matrix for each of these characters
</span><span class="n">library</span><span class="p">(</span><span class="n">reshape2</span><span class="p">)</span><span class="w">
</span><span class="n">speaker_scene_matrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">transcripts_3</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
  </span><span class="n">acast</span><span class="p">(</span><span class="n">character</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">episode</span><span class="p">,</span><span class="w"> </span><span class="n">fun.aggregate</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">length</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>The network was calculated from the co-occurrence matrix as a weighted network. Node colors reflect the character’s gender.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># calculate co-occurrence matrix
</span><span class="n">data_matrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.matrix</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">speaker_scene_matrix</span><span class="p">))</span><span class="w">
</span><span class="n">total_occurrences</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">colSums</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">speaker_scene_matrix</span><span class="p">))</span><span class="w">

</span><span class="n">co_occurrence</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">t</span><span class="p">(</span><span class="n">data_matrix</span><span class="p">)</span><span class="w"> </span><span class="o">%*%</span><span class="w"> </span><span class="n">data_matrix</span><span class="w">

</span><span class="c1"># plot the network graph
</span><span class="n">library</span><span class="p">(</span><span class="n">igraph</span><span class="p">)</span><span class="w">
</span><span class="n">g</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">graph.adjacency</span><span class="p">(</span><span class="n">co_occurrence</span><span class="p">,</span><span class="w">
                         </span><span class="n">weighted</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w">
                         </span><span class="n">diag</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">,</span><span class="w">
                         </span><span class="n">mode</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"upper"</span><span class="p">)</span><span class="w">

</span><span class="n">g</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">simplify</span><span class="p">(</span><span class="n">g</span><span class="p">,</span><span class="w"> </span><span class="n">remove.multiple</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">,</span><span class="w"> </span><span class="n">remove.loops</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">edge.attr.comb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">weight</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"sum"</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ignore"</span><span class="p">))</span><span class="w">

</span><span class="n">females</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"EMILY"</span><span class="p">,</span><span class="w"> </span><span class="s2">"LANE"</span><span class="p">,</span><span class="w"> </span><span class="s2">"LORELAI"</span><span class="p">,</span><span class="w"> </span><span class="s2">"MISS PATTY"</span><span class="p">,</span><span class="w"> </span><span class="s2">"PARIS"</span><span class="p">,</span><span class="w"> </span><span class="s2">"RORY"</span><span class="p">,</span><span class="w"> </span><span class="s2">"SOOKIE"</span><span class="p">)</span><span class="w"> 
        
</span><span class="n">V</span><span class="p">(</span><span class="n">g</span><span class="p">)</span><span class="o">$</span><span class="n">gender</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">V</span><span class="p">(</span><span class="n">g</span><span class="p">)</span><span class="o">$</span><span class="n">name</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">females</span><span class="p">,</span><span class="w"> </span><span class="s2">"female"</span><span class="p">,</span><span class="w"> </span><span class="s2">"male"</span><span class="p">)</span><span class="w">

</span><span class="n">plot</span><span class="p">(</span><span class="n">g</span><span class="p">,</span><span class="w">
     </span><span class="n">vertex.label.family</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Helvetica"</span><span class="p">,</span><span class="w">
     </span><span class="n">vertex.label.font</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w">
     </span><span class="n">vertex.shape</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"sphere"</span><span class="p">,</span><span class="w">
     </span><span class="n">vertex.size</span><span class="o">=</span><span class="n">total_occurrences</span><span class="o">/</span><span class="m">800</span><span class="p">,</span><span class="w">
     </span><span class="n">vertex.label.cex</span><span class="o">=</span><span class="m">0.8</span><span class="p">,</span><span class="w">
     </span><span class="n">vertex.color</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="w"> </span><span class="s2">"pink"</span><span class="p">,</span><span class="w"> </span><span class="s2">"skyblue"</span><span class="p">)[</span><span class="m">1</span><span class="o">+</span><span class="p">(</span><span class="n">V</span><span class="p">(</span><span class="n">g</span><span class="p">)</span><span class="o">$</span><span class="n">gender</span><span class="o">==</span><span class="s2">"male"</span><span class="p">)],</span><span class="w">
     </span><span class="n">vertex.label.color</span><span class="o">=</span><span class="s2">"black"</span><span class="p">,</span><span class="w">
     </span><span class="n">vertex.frame.color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">,</span><span class="w">
     </span><span class="n">edge.width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">E</span><span class="p">(</span><span class="n">g</span><span class="p">)</span><span class="o">$</span><span class="n">weight</span><span class="o">/</span><span class="m">100000</span><span class="p">,</span><span class="w">
     </span><span class="n">edge.curved</span><span class="o">=</span><span class="m">.1</span><span class="p">,</span><span class="w">
     </span><span class="n">layout</span><span class="o">=</span><span class="n">layout_in_circle</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<p>The cluster dendrogram shows us the character co-occurrence in a slightly different way:</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">norm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">speaker_scene_matrix</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">rowSums</span><span class="p">(</span><span class="n">speaker_scene_matrix</span><span class="p">)</span><span class="w">

</span><span class="n">h</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">hclust</span><span class="p">(</span><span class="n">dist</span><span class="p">(</span><span class="n">norm</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"manhattan"</span><span class="p">))</span><span class="w">

</span><span class="n">plot</span><span class="p">(</span><span class="n">h</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<p>And finally, I wanted to plot in which episodes these 20 characters appeared:</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span><span class="w">

</span><span class="n">my_theme</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">base_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">,</span><span class="w"> </span><span class="n">base_family</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"sans"</span><span class="p">){</span><span class="w">
  </span><span class="n">theme_grey</span><span class="p">(</span><span class="n">base_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">base_size</span><span class="p">,</span><span class="w"> </span><span class="n">base_family</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">base_family</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="w">
    </span><span class="n">axis.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">),</span><span class="w">
    </span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">90</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="w">
    </span><span class="n">axis.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">14</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.grid.major</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_line</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.grid.minor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
    </span><span class="n">panel.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"aliceblue"</span><span class="p">),</span><span class="w">
    </span><span class="n">strip.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"lightgrey"</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="w">
    </span><span class="n">strip.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bold"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">),</span><span class="w">
    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"none"</span><span class="p">,</span><span class="w">
    </span><span class="n">legend.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
    </span><span class="n">panel.margin</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">unit</span><span class="p">(</span><span class="m">.5</span><span class="p">,</span><span class="w"> </span><span class="s2">"lines"</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.border</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w">
  </span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">cbPalette</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"#999999"</span><span class="p">,</span><span class="w"> </span><span class="s2">"#E69F00"</span><span class="p">,</span><span class="w"> </span><span class="s2">"#56B4E9"</span><span class="p">,</span><span class="w"> </span><span class="s2">"#009E73"</span><span class="p">,</span><span class="w"> </span><span class="s2">"#F0E442"</span><span class="p">,</span><span class="w"> </span><span class="s2">"#0072B2"</span><span class="p">,</span><span class="w"> </span><span class="s2">"#D55E00"</span><span class="p">,</span><span class="w"> </span><span class="s2">"#CC79A7"</span><span class="p">)</span><span class="w">

</span><span class="n">ggplot</span><span class="p">(</span><span class="n">transcripts_3</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">episode_running_nr</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">character</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">season_name</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_path</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">episode_running_nr</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_wrap</span><span class="p">(</span><span class="o">~</span><span class="w"> </span><span class="n">season_name</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">scales</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"free_x"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_color_hue</span><span class="p">(</span><span class="n">l</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">50</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w">
       </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Episode running number"</span><span class="p">,</span><span class="w">
       </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Gilmore Girls major character's occurence in episodes"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<p>And I wanted to know how many lines per episode each of them spoke:</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># setting factor levels
</span><span class="n">f</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">characters</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">characters</span><span class="o">$</span><span class="n">Freq</span><span class="p">,</span><span class="w"> </span><span class="n">decreasing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">]</span><span class="w">
</span><span class="n">characters</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">within</span><span class="p">(</span><span class="n">characters</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">Var1</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">f</span><span class="p">))</span><span class="w">

</span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">characters</span><span class="p">,</span><span class="w"> </span><span class="n">Var1</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">major_characters</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">])),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Var1</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Freq</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_boxplot</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"navy"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.8</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of Lines"</span><span class="p">,</span><span class="w">
       </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"How many lines do the main characters have per epsisode?"</span><span class="p">,</span><span class="w"> 
       </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Characters"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<hr />

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sessionInfo</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## R version 3.3.2 (2016-10-31)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: macOS Sierra 10.12
## 
## locale:
## [1] de_DE.UTF-8/de_DE.UTF-8/de_DE.UTF-8/C/de_DE.UTF-8/de_DE.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] ggplot2_2.1.0         igraph_1.0.1          reshape2_1.4.2       
## [4] splitstackshape_1.4.2 data.table_1.9.6      tidyr_0.6.0          
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.7      knitr_1.14       magrittr_1.5     munsell_0.4.3   
##  [5] colorspace_1.2-7 stringr_1.1.0    plyr_1.8.4       tools_3.3.2     
##  [9] grid_3.3.2       gtable_0.2.0     htmltools_0.3.5  yaml_2.1.13     
## [13] assertthat_0.1   digest_0.6.10    tibble_1.2       formatR_1.4     
## [17] evaluate_0.10    rmarkdown_1.1    labeling_0.3     stringi_1.1.2   
## [21] scales_0.4.0     chron_2.3-47
</code></pre>
</div>

  </div>

  <h1><a href="/text_analysis/2016/11/06/sysk_earwax">Is 'Yeah' Josh and Chuck's favorite word?</a></h1>
  <p class="author">
    <span class="date">2016-11-06 00:00:00 +0100</span>
  </p>
  <div class="content">
    <h3 id="text-mining-and-sentiment-analysis-of-a-stuff-you-should-know-podcast">Text mining and sentiment analysis of a Stuff You Should Know Podcast</h3>

<p><br /></p>

<p><a href="http://www.stuffyoushouldknow.com/">Stuff You Should Know</a> (or SYSK) is one of the many great podcasts from <a href="http://www.howstuffworks.com/">How Stuff Works</a>. The two SYSK hosts <a href="http://www.stuffyoushouldknow.com/about-josh.htm">Josh</a> and <a href="http://www.stuffyoushouldknow.com/about-chuck.htm">Chuck</a> have taught me so many fascinating things over the years, and today I want to use one of their podcasts to learn a little bit about text analysis in R.</p>

<p>Initially, I wanted to explore all SYSK podcasts. Unfortunately however, I could only find a transcript for the episode <a href="http://www.stuffyoushouldknow.com/podcasts/earwax-live-with-it-2.htm">Earwax: Live With It</a>, posted on March 19, 2015.</p>

<p><br /></p>

<p>The complete R code can be found at the end of this post or as an R-Markdown on <a href="https://github.com/ShirinG/blog_posts_prep/blob/master/sysk/sysk_earwax.Rmd">Github</a>.</p>

<p><br /></p>

<h2 id="the-podcast-transcript">The podcast transcript</h2>

<p>I copied the episode transcript from its <a href="http://www.stuffyoushouldknow.com/podcasts/earwax-live-with-it-2.htm">web page</a> and saved it as a tab delimited text file. The file can be <a href="https://github.com/ShirinG/blog_posts_prep/blob/master/sysk/sysk_earwax.transcript.txt">downloaded from Github</a>.</p>

<p><br /></p>

<h2 id="separating-josh-and-chuck">Separating Josh and Chuck</h2>

<p>Of course, I wouldn’t actually want to separate Josh and Chuck. But for comparison’s sake in this analysis, I am creating two separate files for lines of dialogue spoken by either Josh or Chuck. I am also keeping the combination of both for background information.</p>

<p><br /></p>

<h2 id="how-emotional-are-josh-and-chuck">How emotional are Josh and Chuck?</h2>

<h3 id="sentiment-analysis">Sentiment analysis</h3>

<p>The first thing I want to explore is a sentiment analysis of the lines spoken by Josh and Chuck. Sentiment analysis categorizes text data into positive and negative sentiments and gives information about the emotional state or attitude of the speaker or the contents of a text.</p>

<p>I am using the package <a href="https://cran.r-project.org/web/packages/syuzhet/index.html">syuzhet</a> for sentiment analysis.</p>

<p><br /></p>

<h3 id="nrc-sentiments">NRC sentiments</h3>

<p><a href="http://saifmohammad.com/WebPages/NRC-Emotion-Lexicon.htm">Saif Mohammad’s NRC Emotion Lexicon</a> is a collection of words that were manually categorized based on their association with the emotions anger, anticipation, disgust, fear, joy, sadness, surprise, and trust, and with positive and negative sentiments.</p>

<p><br /></p>

<h4 id="what-sentiment-did-the-podcast-have">What sentiment did the podcast have?</h4>

<p>Before I go ahead with the sentiment analysis I want to get an idea of the podcast’s words’ association with the NRC categories.</p>

<p><img src="sysk_earwax_files/figure-markdown_github/unnamed-chunk-6-1.png" style="display: block; margin: auto;" /></p>

<p>As can be seen by the words and their associated emotions/ sentiments, sentiment analysis is not perfect. Most words make a lot of intuitive sense with their category (e.g. gross, fungus and spider in disgust), but a few I find to be really strange (like, why would waffle be associated with anger?). Still, the majority of categorisations make sense, so let’s go ahead with the sentiment analysis.</p>

<p><br /></p>

<h4 id="does-the-podcasts-sentiment-change-over-time">Does the podcast’s sentiment change over time?</h4>

<p>Sentiment analysis for each line of dialogue produces a matrix with one column per sentiment/ emotion and one row per line. If any of the words in a line of dialogue could be associated with a given category, this category would get a value of <em>1</em> in the matrix. If there was no association with a category, its value would be <em>0</em>. The lines of dialogue are sorted according to the original input text, in this case this means that they represent the order in which they were spoken in the podcast.
Because the plot would get too big with all categories, I split the data into positive and negative sentiments and emotions.</p>

<p>For analysing positive and negative sentiments, syuzhet implements four different methods, each of which uses a slightly different scale. But all of them assign negative values to indicate negative sentiment and positive values to indicate positive sentiments.</p>

<p>All of the methods rely on a precomputed lexicon of word-sentiment score associations. The emotional or sentiment valence is then computed based on the scores of the words from each line of dialogue.</p>

<p><img src="sysk_earwax_files/figure-markdown_github/unnamed-chunk-14-1.png" style="display: block; margin: auto;" /></p>

<p>The upper two plots show on the x-axis the progression of dialogue over time with each point being a line of dialogue. The sentiment score on the y-axis shows the intensity of the sentiment or emotions in the respective line of dialogue, i.e. the more words in a line were associated with the given category, the higher the line’s sentiment score.</p>

<p>From a first glance at these emotions and sentiments, it seems to me that the podcast is more positive than negative but we can get a better overview of positive and negative sentiments by scoring only positive and negative sentiments.</p>

<p>In the third graph we can see quite well that the trend goes towards positive scores, meaning the podcast is overall upbeat. While there are different peaks in both positive and negative directions in Chuck’s and Josh’s lines, there is no overall bias for one being more positive (or negative) than the other.</p>

<p>Finally, I am looking at the sentiment percentage values to get an idea about the percentage of positive versus negative scores along the podcast’s trajectory. Here, the podcast was divided into 20 bins and the mean sentiment valence calculated for each. This last plot shows a clear trend of increasing positivity towards the end of the podcast in Chuck’s lines. Josh on the other hand doesn’t change very much over the progression of the podcast. Interesting…</p>

<p><br /></p>

<h2 id="quantitative-text-analysis">Quantitative text analysis</h2>

<h3 id="building-a-corpus">Building a corpus</h3>

<p>In text analysis, a corpus refers to a collection of documents. Here, I am using the <a href="https://cran.r-project.org/web/packages/tm/index.html">tm</a> package to create my corpus from the character vectors of Josh’s and Chuck’s lines. <a href="https://cran.r-project.org/web/packages/SnowballC/index.html">SnowballC</a> is used for word stemming.</p>

<p>Before I can analyse the text data meaningfully, however, I have to do some pre-processing:</p>

<ol>
  <li>
    <p>Removing punctuation</p>

    <p>Here, I am removing all punctuation marks. Before I do that, I will transform all hyphens to a space, because the text includes some words which are connected by hyphens and would otherwise be connected if I simply removed the hyphen with the <em>removePunctuation</em> function.</p>
  </li>
  <li>
    <p>Transforming to lower case</p>

    <p>R character string processing is case-sensitive, so everything will be converted to lower case.</p>
  </li>
  <li>
    <p>Stripping numbers</p>

    <p>Numbers are usually not very meaningful for text analysis (if we are not specifically interested in dates, for example), so they are removed as well.</p>
  </li>
  <li>
    <p>Removing stopwords</p>

    <p>Stopwords are collections of very common words which by themselves don’t tell us very much about the content of a text (e.g. and, but, the, however, etc.). The tm package includes a list of stopwords for the English language.</p>
  </li>
  <li>
    <p>Stripping whitespace</p>

    <p>I’m also removing superfluous whitespace around words.</p>
  </li>
  <li>
    <p>Stemming</p>

    <p>Finally, I’m stemming the words in the corpus. This means that words with a common root are shortened to this root. Even though stemming algorithms are not perfect, they allow us to compare conjugated words from the same origin.</p>
  </li>
</ol>

<p><br /></p>

<h3 id="creating-the-document-term-matrix">Creating the Document Term Matrix</h3>

<p>The document term matrix (DTM) lists the number of occurrences of each word per document in the corpus. Here, each <em>document</em> in the corpus represents one line of dialogue from the original transcript.</p>

<p>By restricting the DTM to words with a minimum number of letters and an occurrence in at least a minimum number of lines of dialogue (cutoff), we exclude less specific terms.</p>

<p><br /></p>

<h3 id="is-yeah-josh-and-chucks-favorite-word">Is “Yeah” Josh and Chucks favorite word?</h3>

<h4 id="most-frequent-words">Most frequent words</h4>

<p>By summing up the occurrences of each word over all documents we get the word count frequencies.</p>

<p>When accounting for stem words and the cutoff I set for the DTM to evaluate, Josh and Chuck spoke roughly the same number of words (Josh: 880, Chuck: 865) and had almost the same number of dialogue lines (Josh: 202, Chuck: 193). So, good job on neither one dominating the discussion. ;-)</p>

<p><img src="sysk_earwax_files/figure-markdown_github/unnamed-chunk-24-1.png" style="display: block; margin: auto;" /></p>

<p>The lefthand plot shows the number of words spoken per line of dialogue. The background barplot shows the mean number of words spoken per line, the boxplot shows all individual data points (each point represents one line of dialogue and its corresponding word count). While the total number of words and of dialogue lines were basically the same, Chuck’s lines had a stronger deviation around the median with few very long lines. Josh on the other hand seems to have spoken lines with a more consistent length.</p>

<p>The righthand plot shows the most common words and how often they were used overall (red) and by Josh and Chuck separately (green and blue). The most frequent words include (not surprisingly) <em>“ear”</em> and <em>“earwax”</em>, but funnily also <em>“yeah”</em>. To be honest, while listening to the podcast I never noticed <em>yeah</em> being said exceptionally often but I guess the data doesn’t lie…</p>

<p><img src="sysk_earwax_files/figure-markdown_github/unnamed-chunk-25-1.png" style="display: block; margin: auto;" /></p>

<p>Wordclouds are another way to visualize the frequency of words. The frequency is indicated both by the size of the words (bigger words are more frequent than smaller words) and their color.</p>

<p><br /></p>

<h4 id="word-association">Word association</h4>

<p>Associations among words bigger than 60% were plotted in a heatmap to find words that most often co-occured in the same line of dialogue.</p>

<p><img src="sysk_earwax_files/figure-markdown_github/unnamed-chunk-27-1.png" style="display: block; margin: auto;" /><img src="sysk_earwax_files/figure-markdown_github/unnamed-chunk-27-2.png" style="display: block; margin: auto;" /></p>

<p>Among the most conspicuous associations were cotton and swab in Josh’s lines (this was probably a hyphenated word to begin with: <em>cotton-swab</em>) and between secret and gland in Chuck’s line (probably secretory gland).</p>

<p><br /></p>

<h3 id="hierarchical-clustering">Hierarchical clustering</h3>

<p>Hierarchical clustering can be used to classify words by sorting them into clusters according to similarity of occurence (i.e. their frequency or count).</p>

<p><img src="sysk_earwax_files/figure-markdown_github/unnamed-chunk-29-1.png" style="display: block; margin: auto;" /></p>

<p>We already knew that the words <em>“earwax”</em>, <em>“ear”</em> and <em>“yeah”</em> were the most common, so they were clustered accordingly.</p>

<p><br /></p>

<h3 id="knowledge-for-the-masses">Knowledge for the masses</h3>

<h4 id="shorter-words-are-more-frequent-than-longer-words">Shorter words are more frequent than longer words</h4>

<p><img src="sysk_earwax_files/figure-markdown_github/unnamed-chunk-33-1.png" style="display: block; margin: auto;" /></p>

<p>Most words have 4 or 5 letters, only a handful are longer than 7 letters. We don’t have words with fewer than 3 letters because they were excluded in the beginning when obtaining the DTM.</p>

<p>As we can see above, there is only a very small correlation between the length of words and how often they are used. As expected from what we intuitively know, the most common words tend to be shorter while long words are used only occasionally because they are often more specific terms. And there is no real difference between Josh or Chuck when it comes to the length (and complexity?) of the words they use. In general the words they use are rather on the short site, which makes sense as a big part of what makes their podcast so great is that they convey information in a down-to-earth, understandable way.</p>

<p>The frequency plot of all the letters in the alphabet shows that vocals are more common than consonants.</p>

<p><br /></p>

<p><img src="sysk_earwax_files/figure-markdown_github/unnamed-chunk-35-1.png" style="display: block; margin: auto;" /></p>

<p>The plot above shows how often each letter occurs at which position in a word in all the words used by Josh and Chuck. For example, the letter <em>j</em> occured only once and at position one (so 100% of <em>js</em> are at this position). The other letters are more equally distributed but according to the fact that fewer words were longer than 6 letters, there are fewer letters at positions 7 to 10.</p>

<p><br /></p>

<h2 id="conclusion">Conclusion</h2>

<p>This little excursion into text analysis gave an interesting different look at a podcast one would normally “evaluate” intuitively while listening. This hard cold look through the data lens can highlight aspects that probably would be overlooked otherwise; for example, I never noticed that Josh and Chuck used the word <em>yeah</em> that much!</p>

<p>It would be very interesting to broaden the analysis to more podcasts to see if the yeah-thing was just a fluke of this episode or whether it’s a recurrent thing, maybe by trying speech-to-text-conversion tools.</p>

<p><br /></p>

<hr />

<h1 id="r-code">R code</h1>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># setting my custom theme of choice
</span><span class="n">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">ggrepel</span><span class="p">)</span><span class="w">

</span><span class="n">my_theme</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">base_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">,</span><span class="w"> </span><span class="n">base_family</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"sans"</span><span class="p">){</span><span class="w">
  </span><span class="n">theme_grey</span><span class="p">(</span><span class="n">base_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">base_size</span><span class="p">,</span><span class="w"> </span><span class="n">base_family</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">base_family</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="w">
    </span><span class="n">axis.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">),</span><span class="w">
    </span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">90</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="w">
    </span><span class="n">axis.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">14</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.grid.major</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_line</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.grid.minor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
    </span><span class="n">panel.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"aliceblue"</span><span class="p">),</span><span class="w">
    </span><span class="n">strip.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"lightgrey"</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="w">
    </span><span class="n">strip.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bold"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"navy"</span><span class="p">),</span><span class="w">
    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bottom"</span><span class="p">,</span><span class="w">
    </span><span class="n">legend.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
    </span><span class="n">panel.margin</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">unit</span><span class="p">(</span><span class="m">.5</span><span class="p">,</span><span class="w"> </span><span class="s2">"lines"</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.border</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w">
  </span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># reading lines of transcript
</span><span class="n">raw</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">readLines</span><span class="p">(</span><span class="s2">"sysk_earwax.transcript.txt"</span><span class="p">)</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">raw</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "Josh: Hey, and welcome to the podcast. I'm Josh Clark. There is Charles W. \"Chuck\" Bryant, there is Jeri. Yeah, it's Stuff You Should Know."
## [2] ""                                                                                                                                             
## [3] "Chuck: He just shrugged."                                                                                                                     
## [4] ""                                                                                                                                             
## [5] "Josh: Yeah, like \"eh, what are we going to do? That's what we are.\""                                                                        
## [6] ""
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># extracting lines beginning with Josh:/ Chuck: from transcript by looping over the names
</span><span class="n">names</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Josh"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Chuck"</span><span class="p">)</span><span class="w">

</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">names</span><span class="p">){</span><span class="w">
  </span><span class="c1"># grep lines from Josh or Chuck
</span><span class="w">  </span><span class="n">assign</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">),</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">raw</span><span class="p">[</span><span class="n">grep</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">"^"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="s2">":"</span><span class="p">),</span><span class="w"> </span><span class="n">raw</span><span class="p">)]))</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># and removing the beginning of each line that indicates who's speaking
</span><span class="n">lines_Josh</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"^Josh: "</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">lines_Josh</span><span class="p">)</span><span class="w">
</span><span class="n">lines_Chuck</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"^Chuck: "</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">lines_Chuck</span><span class="p">)</span><span class="w">

</span><span class="c1"># Combining Josh's and Chuck's lines
</span><span class="n">lines_bg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">lines_Josh</span><span class="p">,</span><span class="w"> </span><span class="n">lines_Chuck</span><span class="p">)</span><span class="w">

</span><span class="n">names</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Josh"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Chuck"</span><span class="p">,</span><span class="w"> </span><span class="s2">"bg"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># get NRC sentiments for 
# a) each line of dialogue and
# b) each word spoken in the podcast
</span><span class="n">library</span><span class="p">(</span><span class="n">syuzhet</span><span class="p">)</span><span class="w">

</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">names</span><span class="p">){</span><span class="w">
  </span><span class="n">assign</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"get_nrc_sentiment"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">),</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">line_number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">))),</span><span class="w"> 
                                                                 </span><span class="n">get_nrc_sentiment</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)))))</span><span class="w">
  
  </span><span class="n">get_tokens</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">get_tokens</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)))</span><span class="w">
  </span><span class="n">assign</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"get_nrc_sentiments_tokens"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">),</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">word</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">get_tokens</span><span class="p">,</span><span class="w">
                                                                         </span><span class="n">get_nrc_sentiment</span><span class="p">(</span><span class="n">get_tokens</span><span class="p">)))</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># gather word sentiments for plotting
</span><span class="n">library</span><span class="p">(</span><span class="n">tidyr</span><span class="p">)</span><span class="w">

</span><span class="n">get_nrc_sentiments_tokens_bg_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">get_nrc_sentiments_tokens_bg</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">word</span><span class="p">,</span><span class="w"> </span><span class="n">sentiment</span><span class="p">,</span><span class="w"> </span><span class="n">anger</span><span class="o">:</span><span class="n">positive</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">get_nrc_sentiments_tokens_bg_gather</span><span class="p">)[</span><span class="m">2</span><span class="o">:</span><span class="m">3</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"sentiment"</span><span class="p">,</span><span class="w"> </span><span class="s2">"value"</span><span class="p">)</span><span class="w">
</span><span class="n">get_nrc_sentiments_tokens_bg_gather</span><span class="o">$</span><span class="n">name</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"bg"</span><span class="w">
</span><span class="n">get_nrc_sentiments_tokens_bg_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">get_nrc_sentiments_tokens_bg_gather</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">get_nrc_sentiments_tokens_bg_gather</span><span class="o">$</span><span class="n">value</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="m">0</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="n">get_nrc_sentiments_tokens_bg_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">get_nrc_sentiments_tokens_bg_gather</span><span class="p">[</span><span class="o">!</span><span class="n">duplicated</span><span class="p">(</span><span class="n">get_nrc_sentiments_tokens_bg_gather</span><span class="o">$</span><span class="n">word</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">get_nrc_sentiments_tokens_bg_gather</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">value</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">value</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">word</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">word</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_line</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_wrap</span><span class="p">(</span><span class="o">~</span><span class="w"> </span><span class="n">sentiment</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">theme</span><span class="p">(</span><span class="w">
    </span><span class="n">axis.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
    </span><span class="n">axis.ticks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
    </span><span class="n">axis.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"none"</span><span class="p">,</span><span class="w">
    </span><span class="n">panel.grid.major</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">())</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_text_repel</span><span class="p">(</span><span class="n">segment.color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"aliceblue"</span><span class="p">,</span><span class="w"> </span><span class="n">segment.alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Sentiment categories of words used in podcast"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">get_nrc_sentiment_Josh_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">get_nrc_sentiment_Josh</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">line_number</span><span class="p">,</span><span class="w"> </span><span class="n">sentiment</span><span class="p">,</span><span class="w"> </span><span class="n">anger</span><span class="o">:</span><span class="n">positive</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">get_nrc_sentiment_Josh_gather</span><span class="p">)[</span><span class="m">2</span><span class="o">:</span><span class="m">3</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"sentiment"</span><span class="p">,</span><span class="w"> </span><span class="s2">"value"</span><span class="p">)</span><span class="w">
</span><span class="n">get_nrc_sentiment_Josh_gather</span><span class="o">$</span><span class="n">name</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"Josh"</span><span class="w">

</span><span class="n">get_nrc_sentiment_Chuck_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">get_nrc_sentiment_Chuck</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">line_number</span><span class="p">,</span><span class="w"> </span><span class="n">sentiment</span><span class="p">,</span><span class="w"> </span><span class="n">anger</span><span class="o">:</span><span class="n">positive</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">get_nrc_sentiment_Chuck_gather</span><span class="p">)[</span><span class="m">2</span><span class="o">:</span><span class="m">3</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"sentiment"</span><span class="p">,</span><span class="w"> </span><span class="s2">"value"</span><span class="p">)</span><span class="w">
</span><span class="n">get_nrc_sentiment_Chuck_gather</span><span class="o">$</span><span class="n">name</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"Chuck"</span><span class="w">

</span><span class="n">get_nrc_sentiment_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rbind</span><span class="p">(</span><span class="n">get_nrc_sentiment_Josh_gather</span><span class="p">,</span><span class="w"> </span><span class="n">get_nrc_sentiment_Chuck_gather</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># split into positive and negative emotions/ sentiments
</span><span class="n">get_nrc_sentiment_gather_pos</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">get_nrc_sentiment_gather</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">get_nrc_sentiment_gather</span><span class="o">$</span><span class="n">sentiment</span><span class="w"> </span><span class="o">%in%</span><span class="w"> 
                                                                 </span><span class="nf">c</span><span class="p">(</span><span class="s2">"anticipation"</span><span class="p">,</span><span class="w"> </span><span class="s2">"joy"</span><span class="p">,</span><span class="w"> </span><span class="s2">"positive"</span><span class="p">,</span><span class="w"> </span><span class="s2">"surprise"</span><span class="p">,</span><span class="w"> </span><span class="s2">"trust"</span><span class="p">)),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="n">get_nrc_sentiment_gather_neg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">get_nrc_sentiment_gather</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">get_nrc_sentiment_gather</span><span class="o">$</span><span class="n">sentiment</span><span class="w"> </span><span class="o">%in%</span><span class="w"> 
                                                                 </span><span class="nf">c</span><span class="p">(</span><span class="s2">"anger"</span><span class="p">,</span><span class="w"> </span><span class="s2">"disgust"</span><span class="p">,</span><span class="w"> </span><span class="s2">"fear"</span><span class="p">,</span><span class="w"> </span><span class="s2">"negative"</span><span class="p">,</span><span class="w"> </span><span class="s2">"sadness"</span><span class="p">)),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="n">library</span><span class="p">(</span><span class="n">RColorBrewer</span><span class="p">)</span><span class="w">

</span><span class="n">p</span><span class="m">1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">get_nrc_sentiment_gather_pos</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">line_number</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">value</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_line</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">brewer.pal</span><span class="p">(</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="s2">"Set1"</span><span class="p">)[</span><span class="m">3</span><span class="p">])</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_grid</span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">sentiment</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">theme</span><span class="p">(</span><span class="w">
    </span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">),</span><span class="w">
    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"none"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Dialogue line number"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Sentiment valence"</span><span class="p">,</span><span class="w">
       </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Sentiments during podcast progression (positive sentiments)"</span><span class="p">)</span><span class="w">

</span><span class="n">p</span><span class="m">2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">get_nrc_sentiment_gather_neg</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">line_number</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">value</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_line</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">brewer.pal</span><span class="p">(</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="s2">"Set1"</span><span class="p">)[</span><span class="m">1</span><span class="p">])</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_grid</span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">sentiment</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">theme</span><span class="p">(</span><span class="w">
    </span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">),</span><span class="w">
    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"none"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Dialogue line number"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Sentiment valence"</span><span class="p">,</span><span class="w">
       </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Sentiments during podcast progression (negative sentiments)"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># get sentiment scores
</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">names</span><span class="p">){</span><span class="w">
  </span><span class="n">sentiment</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">line_number</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">))),</span><span class="w"> 
                              </span><span class="n">syuzhet</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">get_sentiment</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)),</span><span class="w"> </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"syuzhet"</span><span class="p">),</span><span class="w">
                              </span><span class="n">bing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">get_sentiment</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)),</span><span class="w"> </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bing"</span><span class="p">),</span><span class="w">
                              </span><span class="n">afinn</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">get_sentiment</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)),</span><span class="w"> </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"afinn"</span><span class="p">),</span><span class="w">
                              </span><span class="n">nrc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">get_sentiment</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)),</span><span class="w"> </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"nrc"</span><span class="p">))</span><span class="w">
  </span><span class="n">assign</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"sentiment"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">),</span><span class="w"> </span><span class="n">sentiment</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># gather for plotting
</span><span class="n">sentiment_Josh_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sentiment_Josh</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">line_number</span><span class="p">,</span><span class="w"> </span><span class="n">analysis</span><span class="p">,</span><span class="w"> </span><span class="n">syuzhet</span><span class="o">:</span><span class="n">nrc</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">sentiment_Josh_gather</span><span class="p">)[</span><span class="m">2</span><span class="o">:</span><span class="m">3</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"algorithm"</span><span class="p">,</span><span class="w"> </span><span class="s2">"value"</span><span class="p">)</span><span class="w">
</span><span class="n">sentiment_Josh_gather</span><span class="o">$</span><span class="n">name</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"Josh"</span><span class="w">

</span><span class="n">sentiment_Chuck_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sentiment_Chuck</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">line_number</span><span class="p">,</span><span class="w"> </span><span class="n">analysis</span><span class="p">,</span><span class="w"> </span><span class="n">syuzhet</span><span class="o">:</span><span class="n">nrc</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">sentiment_Chuck_gather</span><span class="p">)[</span><span class="m">2</span><span class="o">:</span><span class="m">3</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"algorithm"</span><span class="p">,</span><span class="w"> </span><span class="s2">"value"</span><span class="p">)</span><span class="w">
</span><span class="n">sentiment_Chuck_gather</span><span class="o">$</span><span class="n">name</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"Chuck"</span><span class="w">

</span><span class="n">sentiment_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rbind</span><span class="p">(</span><span class="n">sentiment_Josh_gather</span><span class="p">,</span><span class="w"> </span><span class="n">sentiment_Chuck_gather</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">p</span><span class="m">3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sentiment_gather</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">line_number</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">value</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">algorithm</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_hline</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">yintercept</span><span class="o">=</span><span class="m">0</span><span class="p">),</span><span class="w"> </span><span class="n">linetype</span><span class="o">=</span><span class="s2">"dashed"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_line</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_grid</span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">algorithm</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">theme</span><span class="p">(</span><span class="w">
    </span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">),</span><span class="w">
    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"none"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Dialogue line number"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Sentiment valence"</span><span class="p">,</span><span class="w">
       </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Sentiment during podcast progression according to different lexicons"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># get sentiment percent values
</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">names</span><span class="p">){</span><span class="w">
  </span><span class="n">sentiment_percent_vals</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">bin</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">,</span><span class="w">
                              </span><span class="n">syuzhet</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">get_percentage_values</span><span class="p">(</span><span class="n">get_sentiment</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)),</span><span class="w"> </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"syuzhet"</span><span class="p">),</span><span class="w"> </span><span class="n">bins</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">),</span><span class="w">
                              </span><span class="n">bing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">get_percentage_values</span><span class="p">(</span><span class="n">get_sentiment</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)),</span><span class="w"> </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bing"</span><span class="p">),</span><span class="w"> </span><span class="n">bins</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">),</span><span class="w">
                              </span><span class="n">afinn</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">get_percentage_values</span><span class="p">(</span><span class="n">get_sentiment</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)),</span><span class="w"> </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"afinn"</span><span class="p">),</span><span class="w"> </span><span class="n">bins</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">),</span><span class="w">
                              </span><span class="n">nrc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">get_percentage_values</span><span class="p">(</span><span class="n">get_sentiment</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)),</span><span class="w"> </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"nrc"</span><span class="p">),</span><span class="w"> </span><span class="n">bins</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">))</span><span class="w">
  </span><span class="n">assign</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"sentiment_percent_vals"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">),</span><span class="w"> </span><span class="n">sentiment_percent_vals</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># gather for plotting
</span><span class="n">sentiment_percent_vals_Josh_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sentiment_percent_vals_Josh</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">bin</span><span class="p">,</span><span class="w"> </span><span class="n">analysis</span><span class="p">,</span><span class="w"> </span><span class="n">syuzhet</span><span class="o">:</span><span class="n">nrc</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">sentiment_percent_vals_Josh_gather</span><span class="p">)[</span><span class="m">2</span><span class="o">:</span><span class="m">3</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"algorithm"</span><span class="p">,</span><span class="w"> </span><span class="s2">"value"</span><span class="p">)</span><span class="w">
</span><span class="n">sentiment_percent_vals_Josh_gather</span><span class="o">$</span><span class="n">name</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"Josh"</span><span class="w">

</span><span class="n">sentiment_percent_vals_Chuck_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sentiment_percent_vals_Chuck</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">bin</span><span class="p">,</span><span class="w"> </span><span class="n">analysis</span><span class="p">,</span><span class="w"> </span><span class="n">syuzhet</span><span class="o">:</span><span class="n">nrc</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">sentiment_percent_vals_Chuck_gather</span><span class="p">)[</span><span class="m">2</span><span class="o">:</span><span class="m">3</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"algorithm"</span><span class="p">,</span><span class="w"> </span><span class="s2">"value"</span><span class="p">)</span><span class="w">
</span><span class="n">sentiment_percent_vals_Chuck_gather</span><span class="o">$</span><span class="n">name</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"Chuck"</span><span class="w">

</span><span class="n">sentiment_percent_vals_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rbind</span><span class="p">(</span><span class="n">sentiment_percent_vals_Josh_gather</span><span class="p">,</span><span class="w"> </span><span class="n">sentiment_percent_vals_Chuck_gather</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">p</span><span class="m">4</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sentiment_percent_vals_gather</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bin</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">value</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">algorithm</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_hline</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">yintercept</span><span class="o">=</span><span class="m">0</span><span class="p">),</span><span class="w"> </span><span class="n">linetype</span><span class="o">=</span><span class="s2">"dashed"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_line</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_grid</span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">algorithm</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">theme</span><span class="p">(</span><span class="w">
    </span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">),</span><span class="w">
    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"none"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Bin"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Sentiment percent values"</span><span class="p">,</span><span class="w">
       </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Sentiment percent values during podcast progression"</span><span class="p">)</span><span class="w">

</span><span class="n">library</span><span class="p">(</span><span class="n">gridExtra</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">grid</span><span class="p">)</span><span class="w">

</span><span class="n">grid.arrange</span><span class="p">(</span><span class="n">p</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">tm</span><span class="p">)</span><span class="w">
</span><span class="n">data</span><span class="p">(</span><span class="s2">"crude"</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">SnowballC</span><span class="p">)</span><span class="w">

</span><span class="n">transform_to_space</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">content_transformer</span><span class="p">(</span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">pattern</span><span class="p">)</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="n">pattern</span><span class="p">,</span><span class="w"> </span><span class="s2">" "</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="p">))</span><span class="w">

</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">names</span><span class="p">){</span><span class="w">
  </span><span class="n">corpus</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">Corpus</span><span class="p">(</span><span class="n">VectorSource</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">))))</span><span class="w">

  </span><span class="c1"># Removing punctuation
</span><span class="w">  </span><span class="n">corpus</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">tm_map</span><span class="p">(</span><span class="n">corpus</span><span class="p">,</span><span class="w"> </span><span class="n">transform_to_space</span><span class="p">,</span><span class="w"> </span><span class="s2">"-"</span><span class="p">)</span><span class="w">
  </span><span class="n">corpus</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">tm_map</span><span class="p">(</span><span class="n">corpus</span><span class="p">,</span><span class="w"> </span><span class="n">removePunctuation</span><span class="p">)</span><span class="w">
  
  </span><span class="c1"># Transforming to lower case
</span><span class="w">  </span><span class="n">corpus</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">tm_map</span><span class="p">(</span><span class="n">corpus</span><span class="p">,</span><span class="w"> </span><span class="n">content_transformer</span><span class="p">(</span><span class="n">tolower</span><span class="p">))</span><span class="w">
    
  </span><span class="c1"># Stripping numbers
</span><span class="w">  </span><span class="n">corpus</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">tm_map</span><span class="p">(</span><span class="n">corpus</span><span class="p">,</span><span class="w"> </span><span class="n">removeNumbers</span><span class="p">)</span><span class="w">
    
  </span><span class="c1"># Removing stopwords
</span><span class="w">  </span><span class="n">corpus</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">tm_map</span><span class="p">(</span><span class="n">corpus</span><span class="p">,</span><span class="w"> </span><span class="n">removeWords</span><span class="p">,</span><span class="w"> </span><span class="n">stopwords</span><span class="p">(</span><span class="s2">"english"</span><span class="p">))</span><span class="w">
  
  </span><span class="c1"># Stripping whitespace
</span><span class="w">  </span><span class="n">corpus</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">tm_map</span><span class="p">(</span><span class="n">corpus</span><span class="p">,</span><span class="w"> </span><span class="n">stripWhitespace</span><span class="p">)</span><span class="w">
    
  </span><span class="c1"># Stemming
</span><span class="w">  </span><span class="n">corpus</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">tm_map</span><span class="p">(</span><span class="n">corpus</span><span class="p">,</span><span class="w"> </span><span class="n">stemDocument</span><span class="p">)</span><span class="w">

  </span><span class="c1"># Converting said to say because this isn't included in stemDocument
</span><span class="w">  </span><span class="n">corpus</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">tm_map</span><span class="p">(</span><span class="n">corpus</span><span class="p">,</span><span class="w"> </span><span class="n">gsub</span><span class="p">,</span><span class="w"> </span><span class="n">pattern</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"said"</span><span class="p">,</span><span class="w"> </span><span class="n">replacement</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"say"</span><span class="p">)</span><span class="w">
  
  </span><span class="c1"># There are a couple of words not included among the stopwords that I would like to remove
</span><span class="w">  </span><span class="n">myStopwords</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"one"</span><span class="p">,</span><span class="w"> </span><span class="s2">"two"</span><span class="p">,</span><span class="w"> </span><span class="s2">"three"</span><span class="p">,</span><span class="w"> </span><span class="s2">"just"</span><span class="p">,</span><span class="w"> </span><span class="s2">"your"</span><span class="p">,</span><span class="w"> </span><span class="s2">"that"</span><span class="p">,</span><span class="w"> </span><span class="s2">"can"</span><span class="p">)</span><span class="w">
  </span><span class="n">corpus</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">tm_map</span><span class="p">(</span><span class="n">corpus</span><span class="p">,</span><span class="w"> </span><span class="n">removeWords</span><span class="p">,</span><span class="w"> </span><span class="n">myStopwords</span><span class="p">)</span><span class="w">
  
  </span><span class="c1"># Assigning to object and treat as text document
</span><span class="w">  </span><span class="n">assign</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"corpus"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">),</span><span class="w"> </span><span class="n">tm_map</span><span class="p">(</span><span class="n">corpus</span><span class="p">,</span><span class="w"> </span><span class="n">PlainTextDocument</span><span class="p">))</span><span class="w">
</span><span class="p">}</span><span class="w">

</span><span class="n">corpus_Josh</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## &lt;&lt;VCorpus&gt;&gt;
## Metadata:  corpus specific: 0, document level (indexed): 0
## Content:  documents: 227
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">corpus_Chuck</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## &lt;&lt;VCorpus&gt;&gt;
## Metadata:  corpus specific: 0, document level (indexed): 0
## Content:  documents: 226
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">corpus_bg</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## &lt;&lt;VCorpus&gt;&gt;
## Metadata:  corpus specific: 0, document level (indexed): 0
## Content:  documents: 453
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1">#  but here I will keep them all to make the numbers comparable to background later
</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">names</span><span class="p">){</span><span class="w">
  </span><span class="n">assign</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"dtm"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">),</span><span class="w"> </span><span class="n">DocumentTermMatrix</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"corpus"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)),</span><span class="w"> 
         </span><span class="n">control</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">wordLengths</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="m">100</span><span class="p">),</span><span class="w"> </span><span class="n">bounds</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">global</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"lines"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">))))))))</span><span class="w">
</span><span class="p">}</span><span class="w">

</span><span class="n">dtm_Josh</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## &lt;&lt;DocumentTermMatrix (documents: 227, terms: 126)&gt;&gt;
## Non-/sparse entries: 801/27801
## Sparsity           : 97%
## Maximal term length: 10
## Weighting          : term frequency (tf)
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">dtm_Chuck</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## &lt;&lt;DocumentTermMatrix (documents: 226, terms: 122)&gt;&gt;
## Non-/sparse entries: 779/26793
## Sparsity           : 97%
## Maximal term length: 10
## Weighting          : term frequency (tf)
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">dtm_bg</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## &lt;&lt;DocumentTermMatrix (documents: 453, terms: 260)&gt;&gt;
## Non-/sparse entries: 1945/115835
## Sparsity           : 98%
## Maximal term length: 10
## Weighting          : term frequency (tf)
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># get word frequencies from DTM
</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">names</span><span class="p">){</span><span class="w">
  </span><span class="n">assign</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"freq_dtm"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">),</span><span class="w"> </span><span class="n">colSums</span><span class="p">(</span><span class="n">as.matrix</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"dtm"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)))))</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># find most frequent terms and strongest associations with "ear"
</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">names</span><span class="p">){</span><span class="w">
  </span><span class="n">print</span><span class="p">(</span><span class="n">findFreqTerms</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"dtm"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)),</span><span class="w"> </span><span class="n">lowfreq</span><span class="o">=</span><span class="m">40</span><span class="p">))</span><span class="w">
  </span><span class="n">print</span><span class="p">(</span><span class="n">findAssocs</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"dtm"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)),</span><span class="w"> </span><span class="s2">"ear"</span><span class="p">,</span><span class="w"> </span><span class="m">0.4</span><span class="p">))</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "ear"    "earwax" "yeah"  
## $ear
## canal candl 
##  0.44  0.44 
## 
## [1] "ear"  "yeah"
## $ear
## canal  want clean 
##  0.54  0.44  0.42 
## 
## [1] "dont"   "ear"    "earwax" "like"   "right"  "say"    "think"  "yeah"  
## $ear
## canal candl 
##  0.49  0.42
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">length</span><span class="p">(</span><span class="n">freq_dtm_Josh</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 126
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">length</span><span class="p">(</span><span class="n">freq_dtm_Chuck</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 122
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">length</span><span class="p">(</span><span class="n">freq_dtm_bg</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 260
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># get number of words per line
</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">names</span><span class="p">){</span><span class="w">
  </span><span class="n">words_per_doc</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">wordcount</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rowSums</span><span class="p">(</span><span class="n">as.matrix</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"dtm"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)))))</span><span class="w">
  </span><span class="n">words_per_doc</span><span class="o">$</span><span class="n">name</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">nrow</span><span class="p">(</span><span class="n">words_per_doc</span><span class="p">))</span><span class="w">
  
  </span><span class="n">assign</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"words_per_doc"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">),</span><span class="w"> </span><span class="n">words_per_doc</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">

</span><span class="n">words_per_doc</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rbind</span><span class="p">(</span><span class="n">words_per_doc_Josh</span><span class="p">,</span><span class="w"> </span><span class="n">words_per_doc_Chuck</span><span class="p">)</span><span class="w">
</span><span class="n">words_per_doc</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">words_per_doc</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">words_per_doc</span><span class="o">$</span><span class="n">wordcount</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="n">word_count</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">names</span><span class="p">[</span><span class="m">-3</span><span class="p">]),</span><span class="w"> 
                         </span><span class="n">number_all_words</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="n">words_per_doc</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">words_per_doc</span><span class="o">$</span><span class="n">name</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Josh"</span><span class="p">),</span><span class="w"> </span><span class="m">1</span><span class="p">]),</span><span class="w"> 
                                       </span><span class="nf">sum</span><span class="p">(</span><span class="n">words_per_doc</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">words_per_doc</span><span class="o">$</span><span class="n">name</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Chuck"</span><span class="p">),</span><span class="w"> </span><span class="m">1</span><span class="p">])),</span><span class="w">
                         </span><span class="n">number_lines</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">nrow</span><span class="p">(</span><span class="n">words_per_doc</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">words_per_doc</span><span class="o">$</span><span class="n">name</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Josh"</span><span class="p">),</span><span class="w"> </span><span class="p">]),</span><span class="w"> 
                                       </span><span class="n">nrow</span><span class="p">(</span><span class="n">words_per_doc</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">words_per_doc</span><span class="o">$</span><span class="n">name</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Chuck"</span><span class="p">),</span><span class="w"> </span><span class="p">])))</span><span class="w">
</span><span class="n">word_count</span><span class="o">$</span><span class="n">words_per_line</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">word_count</span><span class="o">$</span><span class="n">number_all_words</span><span class="o">/</span><span class="n">word_count</span><span class="o">$</span><span class="n">number_lines</span><span class="w">
</span><span class="n">word_count</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##    name number_all_words number_lines words_per_line
## 1  Josh              880          202       4.356436
## 2 Chuck              865          193       4.481865
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">p</span><span class="m">1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">words_per_doc</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">wordcount</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">name</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_bar</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">word_count</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">words_per_line</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">name</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"identity"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">.5</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_boxplot</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">scale_fill_brewer</span><span class="p">(</span><span class="n">palette</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Set1"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="n">axis.title.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
        </span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">),</span><span class="w">
        </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"none"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of words"</span><span class="p">,</span><span class="w">
       </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of words spoken\n per line of dialogue"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># create merged data frame for plotting
</span><span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span><span class="w">

</span><span class="n">freq_df_Josh</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">words</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">freq_dtm_Josh</span><span class="p">),</span><span class="w"> </span><span class="n">freq</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">freq_dtm_Josh</span><span class="p">)</span><span class="w">
</span><span class="n">freq_df_Chuck</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">words</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">freq_dtm_Chuck</span><span class="p">),</span><span class="w"> </span><span class="n">freq</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">freq_dtm_Chuck</span><span class="p">)</span><span class="w">

</span><span class="n">freq_df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">full_join</span><span class="p">(</span><span class="n">freq_df_Josh</span><span class="p">,</span><span class="w"> </span><span class="n">freq_df_Chuck</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"words"</span><span class="p">)</span><span class="w">

</span><span class="n">freq_df_bg</span><span class="w">  </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">words</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">freq_dtm_bg</span><span class="p">),</span><span class="w"> </span><span class="n">freq</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">freq_dtm_bg</span><span class="p">)</span><span class="w">
</span><span class="n">freq_df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">full_join</span><span class="p">(</span><span class="n">freq_df</span><span class="p">,</span><span class="w"> </span><span class="n">freq_df_bg</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"words"</span><span class="p">)</span><span class="w">

</span><span class="n">freq_df</span><span class="p">[</span><span class="nf">is.na</span><span class="p">(</span><span class="n">freq_df</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">freq_df</span><span class="p">)[</span><span class="m">2</span><span class="o">:</span><span class="m">4</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">names</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">freq_df</span><span class="p">)[</span><span class="m">4</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"background"</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># subsetting only those words which occur at least 15 times in either Josh's, Chuck's or their combined lines
</span><span class="n">freq_df_subs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">freq_df</span><span class="p">[</span><span class="n">apply</span><span class="p">(</span><span class="n">freq_df</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">any</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">15</span><span class="p">)),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="c1"># gather for plotting
</span><span class="n">freq_df_subs_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">freq_df_subs</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">words</span><span class="p">,</span><span class="w"> </span><span class="n">count</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">freq_df_subs_gather</span><span class="p">)[</span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"name"</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># setting factor levels
</span><span class="n">f</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.factor</span><span class="p">(</span><span class="n">freq_df_bg</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">freq_df_bg</span><span class="o">$</span><span class="n">freq</span><span class="p">,</span><span class="w"> </span><span class="n">decreasing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="s2">"words"</span><span class="p">])</span><span class="w">
</span><span class="n">freq_df_subs_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">within</span><span class="p">(</span><span class="n">freq_df_subs_gather</span><span class="p">,</span><span class="w"> </span><span class="n">words</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">words</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">f</span><span class="p">))</span><span class="w">

</span><span class="n">p</span><span class="m">2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">freq_df_subs_gather</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">words</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">count</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">name</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_bar</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">freq_df_subs_gather</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"background"</span><span class="p">),</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">.3</span><span class="p">,</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"identity"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_bar</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">freq_df_subs_gather</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="s2">"background"</span><span class="p">),</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"identity"</span><span class="p">,</span><span class="w"> </span><span class="n">position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">position_dodge</span><span class="p">(</span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">.8</span><span class="p">),</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">scale_fill_brewer</span><span class="p">(</span><span class="n">palette</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Set1"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="n">axis.title.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
        </span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">45</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="w">
        </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">.9</span><span class="p">,</span><span class="w"> </span><span class="m">.85</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Frequency of use (count)"</span><span class="p">,</span><span class="w">
       </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Frequency of the most common words in Josh and Chuck's podcast"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">grid.arrange</span><span class="p">(</span><span class="n">p</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">widths</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0.3</span><span class="p">,</span><span class="w"> </span><span class="m">0.7</span><span class="p">),</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">wordcloud</span><span class="p">)</span><span class="w">

</span><span class="c1">#setting the same seed makes the look consistent
</span><span class="n">set.seed</span><span class="p">(</span><span class="m">42</span><span class="p">)</span><span class="w">

</span><span class="n">layout</span><span class="p">(</span><span class="n">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="n">nrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">byrow</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="n">heights</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0.1</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">))</span><span class="w">
</span><span class="n">par</span><span class="p">(</span><span class="n">mar</span><span class="o">=</span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">))</span><span class="w">
</span><span class="n">plot.new</span><span class="p">()</span><span class="w">
</span><span class="n">text</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="o">=</span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="s2">"Josh:"</span><span class="p">,</span><span class="w"> </span><span class="n">cex</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span><span class="n">plot.new</span><span class="p">()</span><span class="w">
</span><span class="n">text</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="o">=</span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="s2">"Chuck:"</span><span class="p">,</span><span class="w"> </span><span class="n">cex</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span><span class="n">wordcloud</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">freq_dtm_Josh</span><span class="p">),</span><span class="w"> </span><span class="n">freq_dtm_Josh</span><span class="p">,</span><span class="w"> </span><span class="n">min.freq</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">8</span><span class="p">,</span><span class="w"> </span><span class="n">colors</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rev</span><span class="p">(</span><span class="n">brewer.pal</span><span class="p">(</span><span class="m">8</span><span class="p">,</span><span class="w"> </span><span class="s2">"Spectral"</span><span class="p">)))</span><span class="w">
</span><span class="n">wordcloud</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">freq_dtm_Chuck</span><span class="p">),</span><span class="w"> </span><span class="n">freq_dtm_Chuck</span><span class="p">,</span><span class="w"> </span><span class="n">min.freq</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">8</span><span class="p">,</span><span class="w"> </span><span class="n">colors</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rev</span><span class="p">(</span><span class="n">brewer.pal</span><span class="p">(</span><span class="m">8</span><span class="p">,</span><span class="w"> </span><span class="s2">"Spectral"</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span><span class="w">

</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">names</span><span class="p">){</span><span class="w">
  
  </span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"freq_dtm"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">))))){</span><span class="w">
      </span><span class="n">associations</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">findAssocs</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"dtm"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)),</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"freq_dtm"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)))[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span><span class="m">0</span><span class="p">))</span><span class="w">
      </span><span class="n">associations</span><span class="o">$</span><span class="n">word</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">associations</span><span class="p">)</span><span class="w">
  
  </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">1</span><span class="p">){</span><span class="w">
    </span><span class="n">associations_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">associations</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">)]</span><span class="w">
  </span><span class="p">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">associations_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">full_join</span><span class="p">(</span><span class="n">associations_table</span><span class="p">,</span><span class="w"> </span><span class="n">associations</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"word"</span><span class="p">)</span><span class="w">
  </span><span class="p">}</span><span class="w">
  
  </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"freq_dtm"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">))))){</span><span class="w">
    </span><span class="n">associations_table</span><span class="p">[</span><span class="nf">is.na</span><span class="p">(</span><span class="n">associations_table</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span><span class="w">
    </span><span class="n">assign</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"associations_table"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">),</span><span class="w"> </span><span class="n">associations_table</span><span class="p">)</span><span class="w">
  </span><span class="p">}</span><span class="w">
  </span><span class="p">}</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">reshape2</span><span class="p">)</span><span class="w">

</span><span class="n">cor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.6</span><span class="w">

</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">names</span><span class="p">){</span><span class="w">
  </span><span class="n">top_words</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"associations_table"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">))</span><span class="w">
  </span><span class="n">top_words</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">top_words</span><span class="p">[</span><span class="n">apply</span><span class="p">(</span><span class="n">top_words</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">any</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">cor</span><span class="p">)),</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">top_words</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">any</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">cor</span><span class="p">)))]</span><span class="w">
  
  </span><span class="n">associations_table_melt</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">melt</span><span class="p">(</span><span class="n">top_words</span><span class="p">)</span><span class="w">

  </span><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">associations_table_melt</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">word</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">variable</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">value</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
    </span><span class="n">geom_tile</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="m">.9</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="o">=</span><span class="m">.9</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
    </span><span class="n">scale_fill_gradient2</span><span class="p">(</span><span class="n">low</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"white"</span><span class="p">,</span><span class="w"> </span><span class="n">high</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"red"</span><span class="p">,</span><span class="w">  
                         </span><span class="n">limit</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">1</span><span class="p">),</span><span class="w"> </span><span class="n">space</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Lab"</span><span class="p">,</span><span class="w"> 
                         </span><span class="n">name</span><span class="o">=</span><span class="s2">""</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
    </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> 
    </span><span class="n">theme</span><span class="p">(</span><span class="w">
      </span><span class="n">axis.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
      </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"right"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
    </span><span class="n">coord_fixed</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
    </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">paste0</span><span class="p">(</span><span class="s2">"Word pair associations &gt; "</span><span class="p">,</span><span class="w"> </span><span class="n">cor</span><span class="p">,</span><span class="w"> </span><span class="s2">" in "</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="s2">"'s lines"</span><span class="p">))</span><span class="w">
  
  </span><span class="n">print</span><span class="p">(</span><span class="n">p</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">ggdendro</span><span class="p">)</span><span class="w">

</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">name</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">names</span><span class="p">){</span><span class="w">
  </span><span class="c1">#convert dtm to matrix
</span><span class="w">  </span><span class="n">m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.matrix</span><span class="p">(</span><span class="n">subset</span><span class="p">(</span><span class="n">get</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"freq_df"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)),</span><span class="w"> </span><span class="n">freq</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">10</span><span class="p">))</span><span class="w">
  </span><span class="n">d</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">dist</span><span class="p">(</span><span class="n">m</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"maximum"</span><span class="p">)</span><span class="w">
  </span><span class="n">hc</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">hclust</span><span class="p">(</span><span class="n">d</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"average"</span><span class="p">)</span><span class="w">
  
  </span><span class="n">assign</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">"hc"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">),</span><span class="w"> </span><span class="n">dendro_data</span><span class="p">(</span><span class="n">as.dendrogram</span><span class="p">(</span><span class="n">hc</span><span class="p">)))</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">ylim</span><span class="p">(</span><span class="m">-5</span><span class="p">,</span><span class="w"> </span><span class="m">50</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_segment</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">segment</span><span class="p">(</span><span class="n">hc_bg</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">y</span><span class="m">+40</span><span class="p">,</span><span class="w"> </span><span class="n">xend</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">xend</span><span class="p">,</span><span class="w"> </span><span class="n">yend</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">yend</span><span class="m">+40</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_segment</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">segment</span><span class="p">(</span><span class="n">hc_Josh</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">x</span><span class="m">+5</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">y</span><span class="p">,</span><span class="w"> </span><span class="n">xend</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">xend</span><span class="m">+5</span><span class="p">,</span><span class="w"> </span><span class="n">yend</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">yend</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_segment</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">segment</span><span class="p">(</span><span class="n">hc_Chuck</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">x</span><span class="m">+22</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">y</span><span class="p">,</span><span class="w"> </span><span class="n">xend</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">xend</span><span class="m">+22</span><span class="p">,</span><span class="w"> </span><span class="n">yend</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">yend</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_text</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">label</span><span class="p">(</span><span class="n">hc_bg</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">y</span><span class="m">+37</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">label</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">label</span><span class="p">),</span><span class="w"> </span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">45</span><span class="p">,</span><span class="w"> </span><span class="n">lineheight</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.6</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_text</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">label</span><span class="p">(</span><span class="n">hc_Josh</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">x</span><span class="m">+5</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">y</span><span class="m">-3</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">label</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">label</span><span class="p">),</span><span class="w"> </span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">45</span><span class="p">,</span><span class="w"> </span><span class="n">lineheight</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.6</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_text</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">label</span><span class="p">(</span><span class="n">hc_Chuck</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">x</span><span class="m">+22</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">y</span><span class="m">-3</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">label</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">label</span><span class="p">),</span><span class="w"> </span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">45</span><span class="p">,</span><span class="w"> </span><span class="n">lineheight</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.6</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">ggplot2</span><span class="o">::</span><span class="n">annotate</span><span class="p">(</span><span class="s2">"text"</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">48</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Background:"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">ggplot2</span><span class="o">::</span><span class="n">annotate</span><span class="p">(</span><span class="s2">"text"</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">4.9</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">26</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Josh:"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">ggplot2</span><span class="o">::</span><span class="n">annotate</span><span class="p">(</span><span class="s2">"text"</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">21.5</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">26</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Chuck:"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">theme</span><span class="p">(</span><span class="w">
    </span><span class="n">axis.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
    </span><span class="n">axis.ticks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
    </span><span class="n">axis.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"none"</span><span class="p">,</span><span class="w">
    </span><span class="n">panel.grid.major</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">())</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Hierarchical clustering dendrograms of words with at least 10 occurences"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">freq_df</span><span class="o">$</span><span class="n">characters</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">nchar</span><span class="p">(</span><span class="n">freq_df</span><span class="o">$</span><span class="n">words</span><span class="p">)</span><span class="w">
  
</span><span class="n">p</span><span class="m">1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">freq_df</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">characters</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_histogram</span><span class="p">(</span><span class="n">binwidth</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"navy"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_vline</span><span class="p">(</span><span class="n">xintercept</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">mean</span><span class="p">(</span><span class="n">freq_df</span><span class="o">$</span><span class="n">characters</span><span class="p">),</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">),</span><span class="w">
             </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"red"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">.5</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Letters"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Words"</span><span class="p">,</span><span class="w">
       </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Histogram of\nword length"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">freq_df_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">freq_df</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">words</span><span class="p">,</span><span class="w"> </span><span class="n">count</span><span class="p">,</span><span class="w"> </span><span class="n">Josh</span><span class="o">:</span><span class="n">background</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">freq_df_gather</span><span class="p">)[</span><span class="m">3</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"name"</span><span class="w">

</span><span class="n">p</span><span class="m">2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">subset</span><span class="p">(</span><span class="n">freq_df_gather</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="s2">"background"</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">characters</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">count</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">words</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"deepskyblue4"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_smooth</span><span class="p">(</span><span class="n">method</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"lm"</span><span class="p">,</span><span class="w"> </span><span class="n">se</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_grid</span><span class="p">(</span><span class="o">~</span><span class="w"> </span><span class="n">name</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Frequency of use (count)"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Word length"</span><span class="p">,</span><span class="w">
       </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Correlation between word length and frequency\nof use in Josh's and Chuck's lines"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_text_repel</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">freq_df_gather</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="s2">"background"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">count</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">25</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">stringr</span><span class="p">)</span><span class="w">
</span><span class="nb">letters</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">unlist</span><span class="p">(</span><span class="n">sapply</span><span class="p">(</span><span class="n">freq_df</span><span class="o">$</span><span class="n">words</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">str_split</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">)))</span><span class="w">

</span><span class="n">library</span><span class="p">(</span><span class="n">qdap</span><span class="p">)</span><span class="w">
</span><span class="n">letters_tab</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">dist_tab</span><span class="p">(</span><span class="nb">letters</span><span class="p">)</span><span class="w">

</span><span class="c1"># setting factor levels
</span><span class="n">f</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.factor</span><span class="p">(</span><span class="n">rev</span><span class="p">(</span><span class="n">letters_tab</span><span class="o">$</span><span class="n">interval</span><span class="p">))</span><span class="w">
</span><span class="n">letters_tab</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">within</span><span class="p">(</span><span class="n">letters_tab</span><span class="p">,</span><span class="w"> </span><span class="n">interval</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">interval</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">f</span><span class="p">))</span><span class="w">

</span><span class="n">p</span><span class="m">3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">letters_tab</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">interval</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">percent</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"identity"</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"navy"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.8</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">coord_flip</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Letter"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Percentage"</span><span class="p">,</span><span class="w">
       </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Frequency of letters"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">grid.arrange</span><span class="p">(</span><span class="n">p</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">widths</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0.2</span><span class="p">,</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="m">0.3</span><span class="p">),</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">letters_df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="nb">letters</span><span class="p">)</span><span class="w">
</span><span class="n">letters_df</span><span class="o">$</span><span class="n">position</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.factor</span><span class="p">(</span><span class="n">unlist</span><span class="p">(</span><span class="n">regmatches</span><span class="p">(</span><span class="n">rownames</span><span class="p">(</span><span class="n">letters_df</span><span class="p">),</span><span class="w"> </span><span class="n">gregexpr</span><span class="p">(</span><span class="s2">"[0-9]+"</span><span class="p">,</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">letters_df</span><span class="p">)))))</span><span class="w">

</span><span class="n">letters_df_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">letters_df</span><span class="p">))</span><span class="w">
</span><span class="n">total_number_letter</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">letters_df</span><span class="o">$</span><span class="nb">letters</span><span class="p">))</span><span class="w">

</span><span class="n">letters_df_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">left_join</span><span class="p">(</span><span class="n">letters_df_table</span><span class="p">,</span><span class="w"> </span><span class="n">total_number_letter</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"letters"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">))</span><span class="w">
</span><span class="n">letters_df_table</span><span class="o">$</span><span class="n">percent</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">letters_df_table</span><span class="o">$</span><span class="n">Freq.x</span><span class="o">*</span><span class="m">100</span><span class="o">/</span><span class="n">letters_df_table</span><span class="o">$</span><span class="n">Freq.y</span><span class="w">
</span><span class="n">letters_df_table</span><span class="o">$</span><span class="n">percent</span><span class="p">[</span><span class="n">letters_df_table</span><span class="o">$</span><span class="n">percent</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">0</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kc">Inf</span><span class="w">

</span><span class="n">letters_df_table</span><span class="o">$</span><span class="n">Freq.x</span><span class="p">[</span><span class="n">letters_df_table</span><span class="o">$</span><span class="n">Freq.x</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">0</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kc">Inf</span><span class="w">

</span><span class="n">letters_df_table_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">letters_df_table</span><span class="p">[,</span><span class="w"> </span><span class="m">-4</span><span class="p">]</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="nb">letters</span><span class="p">,</span><span class="w"> </span><span class="n">position</span><span class="p">,</span><span class="w"> </span><span class="n">Freq.x</span><span class="o">:</span><span class="n">percent</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">letters_df_table_gather</span><span class="p">)[</span><span class="m">3</span><span class="o">:</span><span class="m">4</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"measure"</span><span class="p">,</span><span class="w"> </span><span class="s2">"value"</span><span class="p">)</span><span class="w">

</span><span class="n">letters_df_table_gather</span><span class="o">$</span><span class="n">measure</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">letters_df_table_gather</span><span class="o">$</span><span class="n">measure</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"Freq.x"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Count"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Percent"</span><span class="p">)</span><span class="w">

</span><span class="c1"># setting factor levels
</span><span class="n">f</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.factor</span><span class="p">(</span><span class="n">rev</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">))</span><span class="w">
</span><span class="n">letters_df_table_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">within</span><span class="p">(</span><span class="n">letters_df_table_gather</span><span class="p">,</span><span class="w"> </span><span class="n">position</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">position</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">f</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">letters_df_table_gather</span><span class="p">,</span><span class="w"> </span><span class="o">!</span><span class="nf">is.infinite</span><span class="p">(</span><span class="n">value</span><span class="p">)),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">letters</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">position</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">value</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_tile</span><span class="p">(</span><span class="n">width</span><span class="o">=</span><span class="m">.9</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="o">=</span><span class="m">.9</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_grid</span><span class="p">(</span><span class="n">measure</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">.</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_gradient</span><span class="p">(</span><span class="n">low</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"moccasin"</span><span class="p">,</span><span class="w"> </span><span class="n">high</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"red"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">coord_flip</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">theme</span><span class="p">(</span><span class="w">
    </span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">),</span><span class="w">
    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"right"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">coord_fixed</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Position"</span><span class="p">,</span><span class="w">
       </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"How often does each letter occur\nat which position in a word?"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sessionInfo</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## R version 3.3.2 (2016-10-31)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: macOS Sierra 10.12
## 
## locale:
## [1] de_DE.UTF-8/de_DE.UTF-8/de_DE.UTF-8/C/de_DE.UTF-8/de_DE.UTF-8
## 
## attached base packages:
## [1] grid      stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] qdap_2.2.5             qdapTools_1.3.1        qdapRegex_0.6.0       
##  [4] qdapDictionaries_1.0.6 stringr_1.1.0          ggdendro_0.1-20       
##  [7] reshape2_1.4.2         wordcloud_2.5          dplyr_0.5.0           
## [10] SnowballC_0.5.1        tm_0.6-2               NLP_0.1-9             
## [13] gridExtra_2.2.1        RColorBrewer_1.1-2     tidyr_0.6.0           
## [16] syuzhet_1.0.0          ggrepel_0.6.3          ggplot2_2.1.0         
## 
## loaded via a namespace (and not attached):
##  [1] gtools_3.5.0        venneuler_1.1-0     slam_0.1-37        
##  [4] rJava_0.9-8         reports_0.1.4       colorspace_1.2-7   
##  [7] htmltools_0.3.5     yaml_2.1.13         chron_2.3-47       
## [10] XML_3.98-1.4        DBI_0.5-1           plyr_1.8.4         
## [13] munsell_0.4.3       gtable_0.2.0        evaluate_0.10      
## [16] labeling_0.3        knitr_1.14          gender_0.5.1       
## [19] parallel_3.3.2      xlsxjars_0.6.1      Rcpp_0.12.7        
## [22] scales_0.4.0        formatR_1.4         gdata_2.17.0       
## [25] plotrix_3.6-3       xlsx_0.5.7          openNLPdata_1.5.3-2
## [28] digest_0.6.10       stringi_1.1.2       tools_3.3.2        
## [31] bitops_1.0-6        magrittr_1.5        lazyeval_0.2.0     
## [34] RCurl_1.95-4.8      tibble_1.2          MASS_7.3-45        
## [37] data.table_1.9.6    assertthat_0.1      rmarkdown_1.1      
## [40] openNLP_0.2-6       R6_2.2.0            igraph_1.0.1
</code></pre>
</div>

  </div>

  <h1><a href="/genome/2016/11/01/AnnotationDbi_part2">Exploring the human genome (Part 2) - Transcripts</a></h1>
  <p class="author">
    <span class="date">2016-11-01 00:00:00 +0100</span>
  </p>
  <div class="content">
    <h1 id="how-many-transcripts-and-proteins-do-genes-have">How many transcripts and proteins do genes have?</h1>

<p>In <a href="https://shiring.github.io/genome/2016/10/23/AnnotationDbi">Exploring the human genome (Part 1) - Gene Annotations</a> I examined Ensembl, Entrez and HGNC gene annotations with <a href="http://www.bioconductor.org/packages/release/bioc/html/AnnotationDbi.html">AnnotationDbi</a> via three R packages: <a href="http://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html">org.Hs.eg.db</a>, <a href="https://bioconductor.org/packages/release/data/annotation/html/EnsDb.Hsapiens.v79.html">EnsDb.Hsapiens.v79</a> and <a href="http://bioconductor.org/packages/release/data/annotation/html/TxDb.Hsapiens.UCSC.hg38.knownGene.html">TxDb.Hsapiens.UCSC.hg38.knownGene</a>.</p>

<p>Now, I want to know how many transcripts there are for genes in these databases.</p>

<p><br /></p>

<h3 id="what-is-a-transcript">What is a transcript?</h3>

<p>While a gene is defined as a unit of DNA information which encodes for the production of a protein, it is really more a concept than an actual physical unit. Human genes conists of <a href="https://en.wikipedia.org/wiki/Exon">exons</a> and <a href="https://en.wikipedia.org/wiki/Intron">introns</a>, which can often be transcribed in different combinations - a process called <a href="https://en.wikipedia.org/wiki/Alternative_splicing">alternative splicing</a>.</p>

<blockquote>
  <p>“While the concept of a gene has been helpful in defining the relationship of a portion of a genome to a phenotype, this traditional term may not be as useful as it once was. Currently, “gene” has come to refer principally to a genomic region producing a polyadenylated mRNA that encodes a protein. However, the recent emergence of a large collection of unannotated transcripts with apparently little protein coding capacity, collectively called transcripts of unknown function (TUFs), has begun to blur the physical boundaries and genomic organization of genic regions with noncoding transcripts often overlapping protein-coding genes on the same (sense) and opposite strand (antisense). Moreover, they are often located in intergenic regions, making the genic portions of the human genome an interleaved network of both annotated polyadenylated and nonpolyadenylated transcripts, including splice variants with novel 5′ ends extending hundreds of kilobase. This complex transcriptional organization and other recently observed features of genomes argue for the reconsideration of the term “gene” and suggests that transcripts may be used to define the operational unit of a genome.” <a href="http://genome.cshlp.org/content/17/6/682.full">Thomas Gingeras, Genome Res. 2007</a></p>
</blockquote>

<p><br /></p>

<h2 id="orghsegdb">org.Hs.eg.db</h2>

<p>With org.Hs.eg.db I am using Entrez and Ensembl IDs to obtain Ensembl transcript IDs, which show splice variants of a gene, including protein-coding and non-coding transcripts.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">AnnotationDbi</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">)</span><span class="w">

</span><span class="n">ENTREZID_org</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">)</span><span class="w">
</span><span class="n">ENSEMBL_org</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENSEMBL"</span><span class="p">)</span><span class="w">

</span><span class="c1"># Summarize number of transcripts per gene Entrez ID
</span><span class="n">org_trans_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">AnnotationDbi</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ENTREZID_org</span><span class="p">,</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"ENSEMBLTRANS"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">)</span><span class="w">
</span><span class="n">org_transcript_num_table_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">org_trans_entrez</span><span class="o">$</span><span class="n">ENTREZID</span><span class="p">))</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">org_transcript_num_table_entrez</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Entrez"</span><span class="p">,</span><span class="w"> </span><span class="s2">"orgDb"</span><span class="p">)</span><span class="w">

</span><span class="c1"># Summarize number of transcripts per gene Ensembl ID
</span><span class="n">org_trans_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">AnnotationDbi</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ENSEMBL_org</span><span class="p">,</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"ENSEMBLTRANS"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENSEMBL"</span><span class="p">)</span><span class="w">
</span><span class="n">org_transcript_num_table_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">org_trans_ensembl</span><span class="o">$</span><span class="n">ENSEMBL</span><span class="p">))</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">org_transcript_num_table_ensembl</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Ensembl"</span><span class="p">,</span><span class="w"> </span><span class="s2">"orgDb"</span><span class="p">)</span><span class="w">

</span><span class="c1"># how many NAs are in each column?
</span><span class="n">sapply</span><span class="p">(</span><span class="n">org_trans_entrez</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##     ENTREZID ENSEMBLTRANS 
##            0        52081
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sapply</span><span class="p">(</span><span class="n">org_trans_ensembl</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##      ENSEMBL ENSEMBLTRANS 
##            0        17678
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">head</span><span class="p">(</span><span class="n">org_trans_entrez</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   ENTREZID    ENSEMBLTRANS
## 1        1            &lt;NA&gt;
## 2        2            &lt;NA&gt;
## 3        3 ENST00000543404
## 4        3 ENST00000566278
## 5        3 ENST00000545343
## 6        3 ENST00000544183
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">head</span><span class="p">(</span><span class="n">org_trans_ensembl</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##           ENSEMBL    ENSEMBLTRANS
## 1 ENSG00000121410            &lt;NA&gt;
## 2 ENSG00000175899            &lt;NA&gt;
## 3 ENSG00000256069 ENST00000543404
## 4 ENSG00000256069 ENST00000566278
## 5 ENSG00000256069 ENST00000545343
## 6 ENSG00000256069 ENST00000544183
</code></pre>
</div>

<p>Strangely, some genes, like A1BG (Entrez ID 1) are listed with one gene ID but no Ensembl transcript ID. This is weird, especially since I happen to know for this particular gene that it has several transcripts. Also, because each gene must have at least one transcript, all NAs are counted as 1 transcript in the summary table.</p>

<p>Let’s check other databases…</p>

<p><br /></p>

<h2 id="txdbhsapiensucschg38knowngene">TxDb.Hsapiens.UCSC.hg38.knownGene</h2>

<p>TxDb.Hsapiens.UCSC.hg38.knownGene only has Entrez IDs to identify genes and UCSC transcript ID to identify transcripts.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="p">)</span><span class="w">

</span><span class="n">ENTREZID_TxDb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">
</span><span class="n">TxDb_trans</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">AnnotationDbi</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ENTREZID_TxDb</span><span class="p">,</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"TXID"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">

</span><span class="c1"># Summarize number of transcripts per gene Entrez ID
</span><span class="n">TxDb_transcript_num_table_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">TxDb_trans</span><span class="o">$</span><span class="n">GENEID</span><span class="p">))</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">TxDb_transcript_num_table_entrez</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Entrez"</span><span class="p">,</span><span class="w"> </span><span class="s2">"TxDb"</span><span class="p">)</span><span class="w">

</span><span class="c1"># how many NAs are in each column?
</span><span class="n">sapply</span><span class="p">(</span><span class="n">TxDb_trans</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GENEID   TXID 
##      0      0
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">head</span><span class="p">(</span><span class="n">TxDb_trans</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   GENEID   TXID
## 1      1 166436
## 2      1 166437
## 3      1 166438
## 4      1 166439
## 5      1 166440
## 6      1 166441
</code></pre>
</div>

<p>Here, we don’t have NAs in the data and in contrast to org.Hs.eg.db we find at least six transcripts for A1BG.</p>

<p><br /></p>

<h2 id="ensdbhsapiensv79">EnsDb.Hsapiens.v79</h2>

<p>Like org.Hs.eg.db EnsDb.Hsapiens.v79 has Entrez and Ensembl IDs in reference to Ensembl transcript IDs.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">)</span><span class="w">

</span><span class="n">ENSEMBL_EnsDb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">
</span><span class="n">ENTREZ_EnsDb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">)</span><span class="w">

</span><span class="n">EnsDb_trans_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ENSEMBL_EnsDb</span><span class="p">,</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"TXID"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">
</span><span class="n">EnsDb_trans_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ENTREZ_EnsDb</span><span class="p">,</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"TXID"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">)</span><span class="w">

</span><span class="c1"># somehow there are empty fields in the Entrez ID column, replacing them with NA
</span><span class="n">EnsDb_trans_entrez</span><span class="p">[</span><span class="n">EnsDb_trans_entrez</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">""</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kc">NA</span><span class="w">

</span><span class="c1"># how many NAs are in each column?
</span><span class="n">sapply</span><span class="p">(</span><span class="n">EnsDb_trans_entrez</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## ENTREZID     TXID 
##    47828        0
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># and removing NA rows
</span><span class="n">EnsDb_trans_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">EnsDb_trans_entrez</span><span class="p">[</span><span class="o">!</span><span class="nf">is.na</span><span class="p">(</span><span class="n">EnsDb_trans_entrez</span><span class="o">$</span><span class="n">ENTREZID</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># order by Entrez ID to compare with other databases
</span><span class="n">EnsDb_trans_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">EnsDb_trans_entrez</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">EnsDb_trans_entrez</span><span class="o">$</span><span class="n">ENTREZID</span><span class="p">),]</span><span class="w">

</span><span class="n">head</span><span class="p">(</span><span class="n">EnsDb_trans_ensembl</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##            GENEID            TXID
## 1 ENSG00000000003 ENST00000373020
## 2 ENSG00000000003 ENST00000496771
## 3 ENSG00000000003 ENST00000494424
## 4 ENSG00000000003 ENST00000612152
## 5 ENSG00000000003 ENST00000614008
## 6 ENSG00000000005 ENST00000373031
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">head</span><span class="p">(</span><span class="n">EnsDb_trans_entrez</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##        ENTREZID            TXID
## 95915         1 ENST00000596924
## 95916         1 ENST00000263100
## 95917         1 ENST00000595014
## 95918         1 ENST00000598345
## 95919         1 ENST00000600966
## 135994       10 ENST00000286479
</code></pre>
</div>

<p>Here, we find five transcripts for A1BG.</p>

<p><br /></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># Summarize number of transcripts per gene Entrez ID
</span><span class="n">EnsDb_transcript_num_table_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">EnsDb_trans_entrez</span><span class="o">$</span><span class="n">ENTREZID</span><span class="p">))</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">EnsDb_transcript_num_table_entrez</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Entrez"</span><span class="p">,</span><span class="w"> </span><span class="s2">"EnsDb"</span><span class="p">)</span><span class="w">

</span><span class="c1"># Summarize number of transcripts per gene Ensembl ID
</span><span class="n">EnsDb_transcript_num_table_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">EnsDb_trans_ensembl</span><span class="o">$</span><span class="n">GENEID</span><span class="p">))</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">EnsDb_transcript_num_table_ensembl</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Ensembl"</span><span class="p">,</span><span class="w"> </span><span class="s2">"EnsDb"</span><span class="p">)</span><span class="w">

</span><span class="c1"># In the Entrez column, there are some with multiple entries
# divide entries with multiple gene names into one row per gene/ entry
</span><span class="n">head</span><span class="p">(</span><span class="n">EnsDb_transcript_num_table_entrez</span><span class="p">[</span><span class="n">grep</span><span class="p">(</span><span class="s2">";"</span><span class="p">,</span><span class="w"> </span><span class="n">EnsDb_transcript_num_table_entrez</span><span class="o">$</span><span class="n">Entrez</span><span class="p">),</span><span class="w"> </span><span class="p">])</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##                                     Entrez EnsDb
## 13                     100033415;100033421     1
## 15                     100033417;100033419     2
## 18                     100033421;100033415     1
## 40                     100033446;100033449     1
## 42 100033448;100033803;100033817;100033810     2
## 43                     100033449;100033446     1
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">splitstackshape</span><span class="p">)</span><span class="w">
</span><span class="n">out</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">cSplit</span><span class="p">(</span><span class="n">EnsDb_transcript_num_table_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">splitCols</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Entrez"</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">";"</span><span class="p">,</span><span class="w"> </span><span class="n">direction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"long"</span><span class="p">),</span><span class="w"> </span><span class="n">stringsAsFactors</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span><span class="n">out</span><span class="o">$</span><span class="n">Entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">out</span><span class="o">$</span><span class="n">Entrez</span><span class="p">)</span><span class="w">

</span><span class="c1"># remove duplicates and take the mean
</span><span class="n">library</span><span class="p">(</span><span class="n">plyr</span><span class="p">)</span><span class="w">
</span><span class="n">EnsDb_transcript_num_table_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ddply</span><span class="p">(</span><span class="n">out</span><span class="p">,</span><span class="w"> </span><span class="s2">"Entrez"</span><span class="p">,</span><span class="w"> </span><span class="n">numcolwise</span><span class="p">(</span><span class="n">mean</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<h2 id="comparison-of-transcript-numbers-per-database-package">Comparison of transcript numbers per database package</h2>

<p>The majority of genes have only one transcript. The number of genes with more transcripts decreases with number of transcripts; this can be seen in the plots below.</p>

<h3 id="entrez-ids">Entrez IDs</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># merging datasets by Entrez ID
</span><span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span><span class="w">
</span><span class="n">transcript_num_table_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">full_join</span><span class="p">(</span><span class="n">org_transcript_num_table_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">TxDb_transcript_num_table_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Entrez"</span><span class="p">)</span><span class="w">
</span><span class="n">transcript_num_table_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">full_join</span><span class="p">(</span><span class="n">transcript_num_table_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">EnsDb_transcript_num_table_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Entrez"</span><span class="p">)</span><span class="w">

</span><span class="c1"># gather for plotting
</span><span class="n">library</span><span class="p">(</span><span class="n">tidyr</span><span class="p">)</span><span class="w">
</span><span class="n">transcript_num_table_gather_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">transcript_num_table_entrez</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">DB</span><span class="p">,</span><span class="w"> </span><span class="n">count</span><span class="p">,</span><span class="w"> </span><span class="n">orgDb</span><span class="o">:</span><span class="n">EnsDb</span><span class="p">)</span><span class="w">

</span><span class="c1"># How many counts are NA?
</span><span class="n">sapply</span><span class="p">(</span><span class="n">transcript_num_table_gather_entrez</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Entrez     DB  count 
##      0      0  71066
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># removing rows with NA counts
</span><span class="n">transcript_num_table_gather_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">transcript_num_table_gather_entrez</span><span class="p">[</span><span class="o">!</span><span class="nf">is.na</span><span class="p">(</span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">count</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="c1"># because there are only a handful of genes with many transcripts, they can't be plotted together with genes with few transcripts
# separating them in high and low
</span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">count</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">25</span><span class="p">,</span><span class="w"> </span><span class="s2">"high"</span><span class="p">,</span><span class="w"> </span><span class="s2">"low"</span><span class="p">)</span><span class="w">

</span><span class="c1"># setting factor levels
</span><span class="n">f</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"low"</span><span class="p">,</span><span class="w"> </span><span class="s2">"high"</span><span class="p">)</span><span class="w">
</span><span class="n">transcript_num_table_gather_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">within</span><span class="p">(</span><span class="n">transcript_num_table_gather_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">f</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># setting my custom theme of choice
</span><span class="n">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span><span class="w">

</span><span class="n">my_theme</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">base_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">,</span><span class="w"> </span><span class="n">base_family</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"sans"</span><span class="p">){</span><span class="w">
  </span><span class="n">theme_grey</span><span class="p">(</span><span class="n">base_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">base_size</span><span class="p">,</span><span class="w"> </span><span class="n">base_family</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">base_family</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="w">
    </span><span class="n">axis.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">),</span><span class="w">
    </span><span class="n">axis.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">14</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.grid.major</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_line</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.grid.minor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
    </span><span class="n">panel.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"aliceblue"</span><span class="p">),</span><span class="w">
    </span><span class="n">strip.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"lightgrey"</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="w">
    </span><span class="n">strip.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bold"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">,</span><span class="w"> </span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"navy"</span><span class="p">),</span><span class="w">
    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bottom"</span><span class="p">,</span><span class="w">
    </span><span class="n">panel.margin</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">unit</span><span class="p">(</span><span class="m">.05</span><span class="p">,</span><span class="w"> </span><span class="s2">"lines"</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.border</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w">
  </span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">transcript_num_table_gather_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">count</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_histogram</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"deepskyblue4"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Histogram of number of transcripts per gene (Entrez ID)"</span><span class="p">,</span><span class="w"> 
       </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of transcripts per gene"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Count"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_wrap</span><span class="p">(</span><span class="n">DB</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">scales</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"free"</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">

</span><span class="n">ann_text_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">20</span><span class="p">,</span><span class="w"> </span><span class="m">300</span><span class="p">,</span><span class="w"> </span><span class="m">20</span><span class="p">,</span><span class="w"> </span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="m">20</span><span class="p">,</span><span class="w"> </span><span class="m">100</span><span class="p">),</span><span class="w">
                       </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">4000</span><span class="p">,</span><span class="w"> </span><span class="m">350</span><span class="p">,</span><span class="w"> </span><span class="m">40000</span><span class="p">,</span><span class="w"> </span><span class="m">15</span><span class="p">,</span><span class="w"> </span><span class="m">5000</span><span class="p">,</span><span class="w"> </span><span class="m">280</span><span class="p">),</span><span class="w">
                       </span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"low"</span><span class="p">,</span><span class="w"> </span><span class="s2">"high"</span><span class="p">),</span><span class="w"> </span><span class="m">3</span><span class="p">),</span><span class="w">
                       </span><span class="n">DB</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"EnsDb"</span><span class="p">,</span><span class="w"> </span><span class="s2">"orgDb"</span><span class="p">,</span><span class="w"> </span><span class="s2">"TxDb"</span><span class="p">),</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w">
                       </span><span class="n">labs</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"low"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> 
                                                     </span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"EnsDb"</span><span class="p">))),</span><span class="w"> 
                                </span><span class="n">paste</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"high"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> 
                                                     </span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"EnsDb"</span><span class="p">))),</span><span class="w">
                                </span><span class="n">paste</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"low"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> 
                                                     </span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"orgDb"</span><span class="p">))),</span><span class="w"> 
                                </span><span class="n">paste</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"high"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> 
                                                     </span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"orgDb"</span><span class="p">))),</span><span class="w">
                                </span><span class="n">paste</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"low"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> 
                                                     </span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"TxDb"</span><span class="p">))),</span><span class="w"> 
                                </span><span class="n">paste</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"high"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> 
                                                     </span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"TxDb"</span><span class="p">)))))</span><span class="w">

</span><span class="n">p</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">geom_text</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ann_text_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">labs</span><span class="p">,</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">8</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="AnnotationDbi_part2_blog_files/figure-markdown_github/transcript_histogram-1.png" style="display: block; margin: auto;" /></p>

<p>This time, orgDb has the highest number of Entrez ID gene entries with corresponding transcript information, the majority of which have fewer than 25 transcripts (59984 genes); only 152 genes have more than 25 transcripts. EnsDb and TxDB have comparable numbers of gene entries, also for genes with few (23971 and 24669 genes) and many transcripts (566 and 552 genes).</p>

<p><br /></p>

<h3 id="ensembl-ids">Ensembl IDs</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># merging datasets by Ensembl ID
</span><span class="n">transcript_num_table_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">full_join</span><span class="p">(</span><span class="n">org_transcript_num_table_ensembl</span><span class="p">,</span><span class="w"> </span><span class="n">EnsDb_transcript_num_table_ensembl</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Ensembl"</span><span class="p">)</span><span class="w">

</span><span class="c1"># gather for plotting
</span><span class="n">transcript_num_table_gather_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">transcript_num_table_ensembl</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">DB</span><span class="p">,</span><span class="w"> </span><span class="n">count</span><span class="p">,</span><span class="w"> </span><span class="n">orgDb</span><span class="o">:</span><span class="n">EnsDb</span><span class="p">)</span><span class="w">

</span><span class="c1"># How many counts are NA?
</span><span class="n">sapply</span><span class="p">(</span><span class="n">transcript_num_table_gather_ensembl</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Ensembl      DB   count 
##       0       0   38659
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># removing rows with NA counts
</span><span class="n">transcript_num_table_gather_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">transcript_num_table_gather_ensembl</span><span class="p">[</span><span class="o">!</span><span class="nf">is.na</span><span class="p">(</span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">count</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="c1"># because there are only a handful of genes with many transcripts, they can't be plotted together with genes with few transcripts
# separating them in high and low
</span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">count</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">25</span><span class="p">,</span><span class="w"> </span><span class="s2">"high"</span><span class="p">,</span><span class="w"> </span><span class="s2">"low"</span><span class="p">)</span><span class="w">

</span><span class="c1"># setting factor levels
</span><span class="n">f</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"low"</span><span class="p">,</span><span class="w"> </span><span class="s2">"high"</span><span class="p">)</span><span class="w">
</span><span class="n">transcript_num_table_gather_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">within</span><span class="p">(</span><span class="n">transcript_num_table_gather_ensembl</span><span class="p">,</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">f</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">transcript_num_table_gather_ensembl</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">count</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_histogram</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"deepskyblue4"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Histogram of number of transcripts per gene (Ensembl ID)"</span><span class="p">,</span><span class="w"> 
       </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of transcripts per gene"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Count"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_wrap</span><span class="p">(</span><span class="n">DB</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">scales</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"free"</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">

</span><span class="n">ann_text_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="m">75</span><span class="p">,</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="m">100</span><span class="p">),</span><span class="w">
                              </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">30000</span><span class="p">,</span><span class="w"> </span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="m">15000</span><span class="p">,</span><span class="w"> </span><span class="m">100</span><span class="p">),</span><span class="w">
                              </span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"low"</span><span class="p">,</span><span class="w"> </span><span class="s2">"high"</span><span class="p">),</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="w">
                              </span><span class="n">DB</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"EnsDb"</span><span class="p">,</span><span class="w"> </span><span class="s2">"orgDb"</span><span class="p">),</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w">
                              </span><span class="n">labs</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"low"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> 
                                                            </span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"EnsDb"</span><span class="p">))),</span><span class="w"> 
                                       </span><span class="n">paste</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"high"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> 
                                                            </span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"EnsDb"</span><span class="p">))),</span><span class="w">
                                       </span><span class="n">paste</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"low"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> 
                                                            </span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"orgDb"</span><span class="p">))),</span><span class="w"> 
                                       </span><span class="n">paste</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"high"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> 
                                                            </span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"orgDb"</span><span class="p">)))))</span><span class="w">

</span><span class="n">p</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">geom_text</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ann_text_ensembl</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">labs</span><span class="p">,</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">8</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="AnnotationDbi_part2_blog_files/figure-markdown_github/transcript_histogram2-1.png" style="display: block; margin: auto;" /></p>

<p>For Ensembl gene IDs EnsDb has the higher number of entries with transcript information; most of these genes (65374) have fewer than 25 transcripts. While orgDb has fewer Ensembl-transcript entries, proportionally more have a high number of transcripts (1035 vs 400), while 26980 genes have low transcript numbers.</p>

<p><br /></p>

<h2 id="comparison-of-transcript-numbers-per-gene">Comparison of transcript numbers per gene</h2>

<p>Now that we know what the distribution of transcript numbers per gene is in different database packages, I want to directly compare the numbers of transcripts that are identified for each gene.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># for comparison replacing NAs with 0
</span><span class="n">transcript_num_table_ensembl_NAtozero</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">transcript_num_table_ensembl</span><span class="w">
</span><span class="n">transcript_num_table_ensembl_NAtozero</span><span class="p">[</span><span class="nf">is.na</span><span class="p">(</span><span class="n">transcript_num_table_ensembl_NAtozero</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span><span class="w">

</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">transcript_num_table_entrez</span><span class="w">
</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">[</span><span class="nf">is.na</span><span class="p">(</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">ggrepel</span><span class="p">)</span><span class="w">

</span><span class="n">p</span><span class="m">1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">transcript_num_table_ensembl_NAtozero</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">orgDb</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_abline</span><span class="p">(</span><span class="n">linetype</span><span class="o">=</span><span class="s2">"dashed"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"deepskyblue4"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_smooth</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">            </span><span class="c1"># Add a loess smoothed fit curve with confidence region
</span><span class="w">  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of transcripts per Ensembl ID"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_text_repel</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">transcript_num_table_ensembl_NAtozero</span><span class="p">,</span><span class="w"> </span><span class="n">orgDb</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">140</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">EnsDb</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">150</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Ensembl</span><span class="p">))</span><span class="w">

</span><span class="n">p</span><span class="m">2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">orgDb</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">TxDb</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_abline</span><span class="p">(</span><span class="n">linetype</span><span class="o">=</span><span class="s2">"dashed"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"deepskyblue4"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_smooth</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of transcripts per Entrez ID"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_text_repel</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">,</span><span class="w"> </span><span class="n">orgDb</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">100</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">TxDb</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">200</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Entrez</span><span class="p">))</span><span class="w">

</span><span class="n">p</span><span class="m">3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">orgDb</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_abline</span><span class="p">(</span><span class="n">linetype</span><span class="o">=</span><span class="s2">"dashed"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"deepskyblue4"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_smooth</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of transcripts per Entrez ID"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_text_repel</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">,</span><span class="w"> </span><span class="n">orgDb</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">100</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">EnsDb</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">900</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Entrez</span><span class="p">))</span><span class="w">

</span><span class="n">p</span><span class="m">4</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">TxDb</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_abline</span><span class="p">(</span><span class="n">linetype</span><span class="o">=</span><span class="s2">"dashed"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"deepskyblue4"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_smooth</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of transcripts per Entrez ID"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_text_repel</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">,</span><span class="w"> </span><span class="n">TxDb</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">150</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">EnsDb</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">900</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Entrez</span><span class="p">))</span><span class="w">

</span><span class="n">library</span><span class="p">(</span><span class="n">gridExtra</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">grid</span><span class="p">)</span><span class="w">

</span><span class="n">grid.arrange</span><span class="p">(</span><span class="n">p</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="AnnotationDbi_part2_blog_files/figure-markdown_github/unnamed-chunk-2-1.png" style="display: block; margin: auto;" /></p>

<p>The plots above show scatterplots with fitted loess curves with confidence regions of the numbers of transcripts for each gene with either Entrez or Ensembl entries from different database packages; dashed lines show 1:1 ratio. For most extreme genes, the Ensembl/ Entrez ID is given in the plots.</p>

<p>While some genes do have the same number of transcripts in different database packages, the majority do not. Entrez IDs seem to give a better correlation than Ensembl IDs, except for a few extreme outliers.</p>

<p><br /></p>

<h2 id="is-there-a-correlation-between-number-of-transcripts-and-number-of-exons-of-a-gene">Is there a correlation between number of transcripts and number of exons of a gene?</h2>

<p>Because the number of exons a gene has exponentially increases the number of potential transcripts produced via alternative splicing, it would make sense if the number of exons correlates with the number of transcripts of a gene. However, not all combinations of exons will produce a functioning protein, so there won’t be as many transcripts as theoretically possible.</p>

<p>Exon IDs are only available in TxDb.Hsapiens.UCSC.hg38.knownGene and EnsDb.Hsapiens.v79.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">TxDb_exons</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">AnnotationDbi</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ENTREZID_TxDb</span><span class="p">,</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"EXONID"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">

</span><span class="n">EnsDb_exons_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ENSEMBL_EnsDb</span><span class="p">,</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"EXONID"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">
</span><span class="n">EnsDb_exons_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ENTREZ_EnsDb</span><span class="p">,</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"EXONID"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">)</span><span class="w">

</span><span class="c1"># summarize number of exons per gene
</span><span class="n">TxDb_exons_num_table_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">TxDb_exons</span><span class="o">$</span><span class="n">GENEID</span><span class="p">))</span><span class="w">

</span><span class="n">EnsDb_exons_num_table_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">EnsDb_exons_ensembl</span><span class="o">$</span><span class="n">GENEID</span><span class="p">))</span><span class="w">
</span><span class="n">EnsDb_exons_num_table_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">EnsDb_exons_entrez</span><span class="o">$</span><span class="n">ENTREZID</span><span class="p">))</span><span class="w">

</span><span class="c1"># there are over 98000 entries without corresponding Entrez ID, removing those
</span><span class="n">EnsDb_exons_num_table_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">EnsDb_exons_num_table_entrez</span><span class="p">[</span><span class="m">-1</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">EnsDb_trans_vs_exons_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">full_join</span><span class="p">(</span><span class="n">EnsDb_transcript_num_table_ensembl</span><span class="p">,</span><span class="w"> </span><span class="n">EnsDb_exons_num_table_ensembl</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Ensembl"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">))</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">EnsDb_trans_vs_exons_ensembl</span><span class="p">)[</span><span class="m">2</span><span class="o">:</span><span class="m">3</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"EnsDb_transcripts"</span><span class="p">,</span><span class="w"> </span><span class="s2">"EnsDb_exons"</span><span class="p">)</span><span class="w">

</span><span class="n">EnsDb_trans_vs_exons_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">full_join</span><span class="p">(</span><span class="n">EnsDb_transcript_num_table_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">EnsDb_exons_num_table_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Entrez"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">))</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">EnsDb_trans_vs_exons_entrez</span><span class="p">)[</span><span class="m">2</span><span class="o">:</span><span class="m">3</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"EnsDb_transcripts"</span><span class="p">,</span><span class="w"> </span><span class="s2">"EnsDb_exons"</span><span class="p">)</span><span class="w">

</span><span class="n">TxDb_trans_vs_exons_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">full_join</span><span class="p">(</span><span class="n">TxDb_transcript_num_table_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">TxDb_exons_num_table_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Entrez"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">))</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">TxDb_trans_vs_exons_entrez</span><span class="p">)[</span><span class="m">2</span><span class="o">:</span><span class="m">3</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"TxDb_transcripts"</span><span class="p">,</span><span class="w"> </span><span class="s2">"TxDb_exons"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">p</span><span class="m">1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">EnsDb_trans_vs_exons_ensembl</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb_transcripts</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb_exons</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_abline</span><span class="p">(</span><span class="n">linetype</span><span class="o">=</span><span class="s2">"dashed"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"deepskyblue4"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_smooth</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of transcripts vs number of exons\nper Ensembl ID in EnsDb"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_text_repel</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">EnsDb_trans_vs_exons_ensembl</span><span class="p">,</span><span class="w"> </span><span class="n">EnsDb_transcripts</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">150</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">EnsDb_exons</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">300</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Ensembl</span><span class="p">))</span><span class="w">

</span><span class="n">p</span><span class="m">2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">EnsDb_trans_vs_exons_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb_transcripts</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb_exons</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_abline</span><span class="p">(</span><span class="n">linetype</span><span class="o">=</span><span class="s2">"dashed"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"deepskyblue4"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_smooth</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of transcripts vs number of exons\nper Entrez ID in EnsDb"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_text_repel</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">EnsDb_trans_vs_exons_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">EnsDb_transcripts</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">300</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">EnsDb_exons</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">1000</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Entrez</span><span class="p">))</span><span class="w">

</span><span class="n">p</span><span class="m">3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">TxDb_trans_vs_exons_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">TxDb_transcripts</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">TxDb_exons</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_abline</span><span class="p">(</span><span class="n">linetype</span><span class="o">=</span><span class="s2">"dashed"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"deepskyblue4"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_smooth</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of transcripts vs number of exons\nper Entrez ID in TxDb"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_text_repel</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">TxDb_trans_vs_exons_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">TxDb_transcripts</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">200</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">TxDb_exons</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">400</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Entrez</span><span class="p">))</span><span class="w">

</span><span class="n">grid.arrange</span><span class="p">(</span><span class="n">p</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="AnnotationDbi_part2_blog_files/figure-markdown_github/unnamed-chunk-5-1.png" style="display: block; margin: auto;" /></p>

<p>Indeed, there is a reasonably good correlation between number of transcripts and number of exons ins genes. The scatterplots above show transcript vs exon number per gene in the three database packages with fitted loess curves with confidence regions. As expected, the increase is not exponential as theoretically possible, but stronger than 1:1. For Entrez IDs, in both EnsDb and TxDbm genes with up to about 100 transcripts show a steeper correlation than genes with more than 100 transcripts. For Ensembl IDs the cutoff is around 40 transcripts.</p>

<p><br /></p>

<h3 id="what-genes-have-many-exons-but-few-transcripts">What genes have many exons but few transcripts?</h3>

<p>Interestingly, there are a couple of genes with many exons but only few transcripts. To find out more about these genes, I will compare their proportion of gene biotypes.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># gene biotypes are only available in EnsDb
</span><span class="n">EnsDb_trans_vs_exons_ensembl</span><span class="o">$</span><span class="n">ratio</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">EnsDb_trans_vs_exons_ensembl</span><span class="o">$</span><span class="n">EnsDb_exons</span><span class="o">/</span><span class="n">EnsDb_trans_vs_exons_ensembl</span><span class="o">$</span><span class="n">EnsDb_transcripts</span><span class="w">
</span><span class="n">EnsDb_trans_vs_exons_ensembl_genes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">EnsDb_trans_vs_exons_ensembl</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">EnsDb_trans_vs_exons_ensembl</span><span class="o">$</span><span class="n">ratio</span><span class="p">,</span><span class="w"> </span><span class="n">decreasing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="s2">"Ensembl"</span><span class="p">][</span><span class="m">1</span><span class="o">:</span><span class="m">100</span><span class="p">])</span><span class="w">

</span><span class="n">EnsDb_trans_vs_exons_entrez</span><span class="o">$</span><span class="n">ratio</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">EnsDb_trans_vs_exons_entrez</span><span class="o">$</span><span class="n">EnsDb_exons</span><span class="o">/</span><span class="n">EnsDb_trans_vs_exons_entrez</span><span class="o">$</span><span class="n">EnsDb_transcripts</span><span class="w">
</span><span class="n">EnsDb_trans_vs_exons_entrez_genes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">EnsDb_trans_vs_exons_entrez</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">EnsDb_trans_vs_exons_entrez</span><span class="o">$</span><span class="n">ratio</span><span class="p">,</span><span class="w"> </span><span class="n">decreasing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="s2">"Entrez"</span><span class="p">][</span><span class="m">1</span><span class="o">:</span><span class="m">100</span><span class="p">])</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># get all gene biotypes for all gene IDs
</span><span class="n">biotypes_exons_trans_EnsDb_ensembl_all</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb_trans_vs_exons_ensembl_genes</span><span class="p">,</span><span class="w"> 
                                          </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEBIOTYPE"</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">

</span><span class="n">biotypes_exons_trans_EnsDb_entrez_all</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb_trans_vs_exons_entrez_genes</span><span class="p">,</span><span class="w"> 
                                          </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEBIOTYPE"</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># get all gene biotypes for all gene IDs
</span><span class="n">biotypes_EnsDb_ensembl_all</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">),</span><span class="w"> 
                                          </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEBIOTYPE"</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">

</span><span class="n">biotypes_EnsDb_entrez_all</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">),</span><span class="w"> 
                                          </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEBIOTYPE"</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># how many NAs are in each column?
</span><span class="n">sapply</span><span class="p">(</span><span class="n">biotypes_exons_trans_EnsDb_ensembl_all</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##      GENEID GENEBIOTYPE 
##           0           0
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sapply</span><span class="p">(</span><span class="n">biotypes_exons_trans_EnsDb_entrez_all</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##    ENTREZID GENEBIOTYPE 
##           0           0
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># summary table of number of genebiotypes
</span><span class="n">biotypes_exons_trans_EnsDb_ensembl_all_num_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">biotypes_exons_trans_EnsDb_ensembl_all</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">))</span><span class="w">
</span><span class="n">biotypes_exons_trans_EnsDb_entrez_all_num_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">biotypes_exons_trans_EnsDb_entrez_all</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># subset to keep only those biotypes that are among the top genes
</span><span class="n">biotypes_EnsDb_ensembl_all_subs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">biotypes_EnsDb_ensembl_all</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">biotypes_EnsDb_ensembl_all</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="w"> </span><span class="o">%in%</span><span class="w"> 
                                                                      </span><span class="n">unique</span><span class="p">(</span><span class="n">biotypes_exons_trans_EnsDb_ensembl_all</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">)),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="n">biotypes_EnsDb_entrez_all_subs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">biotypes_EnsDb_entrez_all</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">biotypes_EnsDb_entrez_all</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="w"> </span><span class="o">%in%</span><span class="w"> 
                                                                      </span><span class="n">unique</span><span class="p">(</span><span class="n">biotypes_exons_trans_EnsDb_entrez_all</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">)),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="c1"># summary table of biotypes background (all genes)
</span><span class="n">biotypes_EnsDb_ensembl_all_subs_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">biotypes_EnsDb_ensembl_all_subs</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">))</span><span class="w">
</span><span class="n">biotypes_EnsDb_entrez_all_subs_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">biotypes_EnsDb_entrez_all_subs</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">))</span><span class="w">

</span><span class="c1"># add background number of biotypes to number among top genes
</span><span class="n">biotypes_EnsDb_ensembl_top_num_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">biotypes_exons_trans_EnsDb_ensembl_all_num_table</span><span class="p">,</span><span class="w"> </span><span class="n">biotypes_EnsDb_ensembl_all_subs_table</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">,</span><span class="w"> </span><span class="n">all.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">biotypes_EnsDb_entrez_top_num_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">biotypes_exons_trans_EnsDb_entrez_all_num_table</span><span class="p">,</span><span class="w"> </span><span class="n">biotypes_EnsDb_entrez_all_subs_table</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">,</span><span class="w"> </span><span class="n">all.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># merge
</span><span class="n">biotypes_top_num_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">biotypes_EnsDb_ensembl_top_num_table</span><span class="p">,</span><span class="w"> </span><span class="n">biotypes_EnsDb_entrez_top_num_table</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">,</span><span class="w"> </span><span class="n">all</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">biotypes_top_num_table</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"GENEBIOTYPE"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Ensembl_top"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Ensembl_bg"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Entrez_top"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Entrez_bg"</span><span class="p">)</span><span class="w">

</span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Ensembl_top</span><span class="o">/</span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Ensembl_bg</span><span class="w">
</span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Entrez_top</span><span class="o">/</span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Entrez_bg</span><span class="w">

</span><span class="c1"># gather for plotting
</span><span class="n">biotypes_top_num_table_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">biotypes_top_num_table</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">6</span><span class="p">,</span><span class="w"> </span><span class="m">7</span><span class="p">)]</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">GENEBIOTYPE</span><span class="p">,</span><span class="w"> </span><span class="n">count</span><span class="p">,</span><span class="w"> </span><span class="n">Ensembl</span><span class="o">:</span><span class="n">Entrez</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">biotypes_top_num_table_gather</span><span class="p">)[</span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"Gene_ID"</span><span class="w">

</span><span class="c1"># replace NA with 0
</span><span class="n">biotypes_top_num_table_gather</span><span class="p">[</span><span class="nf">is.na</span><span class="p">(</span><span class="n">biotypes_top_num_table_gather</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">biotypes_top_num_table_gather</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">GENEBIOTYPE</span><span class="p">),</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">count</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Gene_ID</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"identity"</span><span class="p">,</span><span class="w"> </span><span class="n">position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"dodge"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"right"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">coord_flip</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Gene biotypes of genes with many exons but few transcripts"</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Ratio of biotype count vs background number"</span><span class="p">,</span><span class="w">
       </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Gene ID"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">scale_x_discrete</span><span class="p">(</span><span class="n">breaks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">biotypes_top_num_table_gather</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">,</span><span class="w">
                      </span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"_"</span><span class="p">,</span><span class="w"> </span><span class="s2">" "</span><span class="p">,</span><span class="w"> </span><span class="n">biotypes_top_num_table_gather</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_brewer</span><span class="p">(</span><span class="n">palette</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Set1"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="AnnotationDbi_part2_blog_files/figure-markdown_github/unnamed-chunk-12-1.png" style="display: block; margin: auto;" /></p>

<p>Compared to the background distribution of biotypes, the proportionally most strongly represented group are LRG genes. As defined by <a href="http://www.ensembl.org/common/Help/Glossary?db=core">Ensembl’s glossary</a>:</p>

<blockquote>
  <p>“An LRG (Locus Reference Genomic) is a fixed sequence, independent of the genome, specifically created for the diagnostic community to record DNA sequence variation on a fixed framework. Sequence variants in LSDBs (Locus Specific Databases) are reported using LRG sequences.”</p>
</blockquote>

<p>Other than this, there are only protein coding, processed and pseudogenes among genes with many exons and few transcripts. So, nothing exceptional about these genes’s biotypes.</p>

<p><br /></p>

<h2 id="what-types-of-genes-have-the-most-transcripts">What types of genes have the most transcripts?</h2>

<p>Is there a biological function associated with having many transcripts? In order to find out more about these genes, I am looking at</p>

<ol>
  <li>their biotypes and</li>
  <li>for over-represented gene ontology (GO) categories.</li>
</ol>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># mean number of transcripts over all database packages
</span><span class="n">mean_num_transcripts_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">mean_num_transcripts</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rowMeans</span><span class="p">(</span><span class="n">transcript_num_table_ensembl_NAtozero</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]),</span><span class="w">
               </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">transcript_num_table_ensembl_NAtozero</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="p">])</span><span class="w">

</span><span class="n">mean_num_transcripts_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">mean_num_transcripts</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rowMeans</span><span class="p">(</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]),</span><span class="w">
               </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="p">])</span><span class="w">

</span><span class="c1"># standard error of mean
</span><span class="n">mean_num_transcripts_ensembl</span><span class="o">$</span><span class="n">sem</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">transcript_num_table_ensembl_NAtozero</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">sd</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">/</span><span class="nf">sqrt</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span><span class="n">mean_num_transcripts_entrez</span><span class="o">$</span><span class="n">sem</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">sd</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">/</span><span class="nf">sqrt</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">

</span><span class="c1"># ranking the mean
</span><span class="n">mean_num_transcripts_ensembl</span><span class="o">$</span><span class="n">rank</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rank</span><span class="p">(</span><span class="n">mean_num_transcripts_ensembl</span><span class="o">$</span><span class="n">mean_num_transcripts</span><span class="p">)</span><span class="w">
</span><span class="n">mean_num_transcripts_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mean_num_transcripts_ensembl</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">mean_num_transcripts_ensembl</span><span class="o">$</span><span class="n">rank</span><span class="p">,</span><span class="w"> </span><span class="n">decreasing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="n">mean_num_transcripts_entrez</span><span class="o">$</span><span class="n">rank</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rank</span><span class="p">(</span><span class="n">mean_num_transcripts_entrez</span><span class="o">$</span><span class="n">mean_num_transcripts</span><span class="p">)</span><span class="w">
</span><span class="n">mean_num_transcripts_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mean_num_transcripts_entrez</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">mean_num_transcripts_entrez</span><span class="o">$</span><span class="n">rank</span><span class="p">,</span><span class="w"> </span><span class="n">decreasing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<p>I am picking the top 1000 genes with highest mean number of transcripts.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># picking top genes
</span><span class="n">mean_num_transcripts_ensembl_top</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mean_num_transcripts_ensembl</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">1000</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="n">mean_num_transcripts_entrez_top</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mean_num_transcripts_entrez</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">1000</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="c1"># distribution of mean number of transcripts
</span><span class="n">p</span><span class="m">1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">mean_num_transcripts_ensembl_top</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">mean_num_transcripts</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">geom_density</span><span class="p">(</span><span class="n">fill</span><span class="o">=</span><span class="s2">"navy"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Mean number of transcripts\nper Ensembl ID for the top genes"</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"mean number of transcripts pr gene"</span><span class="p">)</span><span class="w">
</span><span class="n">p</span><span class="m">2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">mean_num_transcripts_entrez_top</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">mean_num_transcripts</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">geom_density</span><span class="p">(</span><span class="n">fill</span><span class="o">=</span><span class="s2">"navy"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Mean number of transcripts\nper Entrez ID for the top genes"</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"mean number of transcripts per gene"</span><span class="p">)</span><span class="w">
</span><span class="n">grid.arrange</span><span class="p">(</span><span class="n">p</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="AnnotationDbi_part2_blog_files/figure-markdown_github/unnamed-chunk-15-1.png" style="display: block; margin: auto;" /></p>

<p>The density distribution of the mean number of transcripts per gene (mean over database packages) for the top genes with highest transcript numbers shows the stark difference between Ensembl and Entrez IDs.</p>

<p><br /></p>

<h3 id="gene-biotypes-of-top-genes-with-highest-mean-number-of-transcripts">Gene biotypes of top genes with highest mean number of transcripts</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">biotypes_EnsDb_ensembl_top</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">mean_num_transcripts_ensembl_top</span><span class="p">),</span><span class="w"> 
                                          </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEBIOTYPE"</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">

</span><span class="n">biotypes_EnsDb_entrez_top</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">mean_num_transcripts_entrez_top</span><span class="p">),</span><span class="w"> 
                                          </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEBIOTYPE"</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">)</span><span class="w">

</span><span class="c1"># how many NAs are in each column?
</span><span class="n">sapply</span><span class="p">(</span><span class="n">biotypes_EnsDb_ensembl_top</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##      GENEID GENEBIOTYPE 
##           0           0
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sapply</span><span class="p">(</span><span class="n">biotypes_EnsDb_entrez_top</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##    ENTREZID GENEBIOTYPE 
##           0           0
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># summary table of number of genebiotypes
</span><span class="n">biotypes_EnsDb_ensembl_top_num_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">biotypes_EnsDb_ensembl_top</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">))</span><span class="w">
</span><span class="n">biotypes_EnsDb_entrez_top_num_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">biotypes_EnsDb_entrez_top</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># subset to keep only those biotypes that are among the top genes
</span><span class="n">biotypes_EnsDb_ensembl_all_subs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">biotypes_EnsDb_ensembl_all</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">biotypes_EnsDb_ensembl_all</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="w"> </span><span class="o">%in%</span><span class="w"> 
                                                                      </span><span class="n">unique</span><span class="p">(</span><span class="n">biotypes_EnsDb_ensembl_top_num_table</span><span class="o">$</span><span class="n">Var1</span><span class="p">)),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="n">biotypes_EnsDb_entrez_all_subs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">biotypes_EnsDb_entrez_all</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">biotypes_EnsDb_entrez_all</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="w"> </span><span class="o">%in%</span><span class="w"> 
                                                                      </span><span class="n">unique</span><span class="p">(</span><span class="n">biotypes_EnsDb_entrez_top_num_table</span><span class="o">$</span><span class="n">Var1</span><span class="p">)),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="c1"># summary table of biotypes background (all genes)
</span><span class="n">biotypes_EnsDb_ensembl_all_subs_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">biotypes_EnsDb_ensembl_all_subs</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">))</span><span class="w">
</span><span class="n">biotypes_EnsDb_entrez_all_subs_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">biotypes_EnsDb_entrez_all_subs</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">))</span><span class="w">

</span><span class="c1"># add background number of biotypes to number among top genes
</span><span class="n">biotypes_EnsDb_ensembl_top_num_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">biotypes_EnsDb_ensembl_top_num_table</span><span class="p">,</span><span class="w"> </span><span class="n">biotypes_EnsDb_ensembl_all_subs_table</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">,</span><span class="w"> </span><span class="n">all.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">biotypes_EnsDb_entrez_top_num_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">biotypes_EnsDb_entrez_top_num_table</span><span class="p">,</span><span class="w"> </span><span class="n">biotypes_EnsDb_entrez_all_subs_table</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">,</span><span class="w"> </span><span class="n">all.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># merge
</span><span class="n">biotypes_top_num_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">biotypes_EnsDb_ensembl_top_num_table</span><span class="p">,</span><span class="w"> </span><span class="n">biotypes_EnsDb_entrez_top_num_table</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">,</span><span class="w"> </span><span class="n">all</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">biotypes_top_num_table</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"GENEBIOTYPE"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Ensembl_top"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Ensembl_bg"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Entrez_top"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Entrez_bg"</span><span class="p">)</span><span class="w">

</span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Ensembl_top</span><span class="o">/</span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Ensembl_bg</span><span class="w">
</span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Entrez_top</span><span class="o">/</span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Entrez_bg</span><span class="w">

</span><span class="c1"># gather for plotting
</span><span class="n">biotypes_top_num_table_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">biotypes_top_num_table</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">6</span><span class="p">,</span><span class="w"> </span><span class="m">7</span><span class="p">)]</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">GENEBIOTYPE</span><span class="p">,</span><span class="w"> </span><span class="n">count</span><span class="p">,</span><span class="w"> </span><span class="n">Ensembl</span><span class="o">:</span><span class="n">Entrez</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">biotypes_top_num_table_gather</span><span class="p">)[</span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"Gene_ID"</span><span class="w">

</span><span class="c1"># replace NA with 0
</span><span class="n">biotypes_top_num_table_gather</span><span class="p">[</span><span class="nf">is.na</span><span class="p">(</span><span class="n">biotypes_top_num_table_gather</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">biotypes_top_num_table_gather</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">GENEBIOTYPE</span><span class="p">),</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">count</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Gene_ID</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"identity"</span><span class="p">,</span><span class="w"> </span><span class="n">position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"dodge"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"right"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">coord_flip</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Gene biotypes of top genes\nwith highest transcript numbers"</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Ratio of biotype count vs background number"</span><span class="p">,</span><span class="w">
       </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Gene ID"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">scale_x_discrete</span><span class="p">(</span><span class="n">breaks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">biotypes_top_num_table_gather</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">,</span><span class="w">
                      </span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"_"</span><span class="p">,</span><span class="w"> </span><span class="s2">" "</span><span class="p">,</span><span class="w"> </span><span class="n">biotypes_top_num_table_gather</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_brewer</span><span class="p">(</span><span class="n">palette</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Set1"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="AnnotationDbi_part2_blog_files/figure-markdown_github/unnamed-chunk-19-1.png" style="display: block; margin: auto;" /></p>

<p>Again, compared to background, we find a high proportion of LRG, protein coding, processed and pseudogenes. But here, we also see a few other biotypes represented among genes with most transcripts. Especially interesting is that we find <a href="http://portals.broadinstitute.org/genome_bio/human_lincrnas/">lincRNAs</a> and <a href="https://en.wikipedia.org/wiki/MicroRNA">miRNAs</a> among them, not something I would have expected…</p>

<p><br /></p>

<h3 id="enrichment-analysis-of-top-genes-with-highest-mean-number-of-transcripts">Enrichment Analysis of top genes with highest mean number of transcripts</h3>

<p>Functional profiles for the top 100 genes with most transcripts were created with <a href="https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html">clusterProfiler</a>’s <strong>compareCluster</strong> function.</p>

<p><br /></p>

<h4 id="grouping-go-terms">Grouping GO terms</h4>

<p>With <strong>groupGO</strong>, we can see the distribution of GO categories among input genes.</p>

<p><img src="AnnotationDbi_part2_blog_files/figure-markdown_github/unnamed-chunk-22-1.png" style="display: block; margin: auto;" /></p>

<p><br /></p>

<h4 id="go-category-enrichment">GO category enrichment</h4>

<p><a href="http://bioconductor.org/packages/release/bioc/vignettes/DOSE/inst/doc/enrichmentAnalysis.html">Over-representation</a> of GO categories is calculated with <strong>enrichGO</strong>.</p>

<p><img src="AnnotationDbi_part2_blog_files/figure-markdown_github/unnamed-chunk-24-1.png" style="display: block; margin: auto;" /></p>

<p>Especially interesting to me is that the GO enrichment analysis of genes with many transcripts finds these genes to be overproportionally involved in immune system functions like <a href="http://www.cell.com/cell/abstract/S0092-8674(11)00300-X?_returnURL=http%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS009286741100300X%3Fshowall%3Dtrue">MCH class II</a> and <a href="https://www.ncbi.nlm.nih.gov/books/NBK27160/">antigen binding</a>. These functions are known for their high variability, which is necessary for efficient immune defenses against a wide variety of intruders (pathogens).</p>

<p>We also see an enrichment of genes involved in <a href="https://www.ncbi.nlm.nih.gov/books/NBK26937/">cell-cell-adhesion</a>, specifically <a href="https://en.wikipedia.org/wiki/Cadherin">cadherin binding</a>. These have also been known for their high variability via alternative splicing.</p>

<hr />

<p><br /></p>

<div class="highlighter-rouge"><pre class="highlight"><code>## R version 3.3.1 (2016-06-21)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: OS X 10.12 (Sierra)
## 
## locale:
## [1] de_DE.UTF-8/de_DE.UTF-8/de_DE.UTF-8/C/de_DE.UTF-8/de_DE.UTF-8
## 
## attached base packages:
##  [1] grid      parallel  stats4    stats     graphics  grDevices utils    
##  [8] datasets  methods   base     
## 
## other attached packages:
##  [1] clusterProfiler_3.2.0                  
##  [2] DOSE_3.0.0                             
##  [3] gridExtra_2.2.1                        
##  [4] ggrepel_0.5                            
##  [5] ggplot2_2.1.0                          
##  [6] tidyr_0.6.0                            
##  [7] dplyr_0.5.0                            
##  [8] plyr_1.8.4                             
##  [9] splitstackshape_1.4.2                  
## [10] data.table_1.9.6                       
## [11] EnsDb.Hsapiens.v79_1.1.0               
## [12] ensembldb_1.6.0                        
## [13] TxDb.Hsapiens.UCSC.hg38.knownGene_3.4.0
## [14] GenomicFeatures_1.26.0                 
## [15] GenomicRanges_1.26.0                   
## [16] GenomeInfoDb_1.10.0                    
## [17] org.Hs.eg.db_3.4.0                     
## [18] AnnotationDbi_1.36.0                   
## [19] IRanges_2.8.0                          
## [20] S4Vectors_0.12.0                       
## [21] Biobase_2.34.0                         
## [22] BiocGenerics_0.20.0                    
## 
## loaded via a namespace (and not attached):
##  [1] httr_1.2.1                    AnnotationHub_2.6.0          
##  [3] splines_3.3.1                 shiny_0.14.1                 
##  [5] assertthat_0.1                DO.db_2.9                    
##  [7] interactiveDisplayBase_1.12.0 Rsamtools_1.26.0             
##  [9] yaml_2.1.13                   RSQLite_1.0.0                
## [11] lattice_0.20-34               chron_2.3-47                 
## [13] digest_0.6.10                 RColorBrewer_1.1-2           
## [15] XVector_0.14.0                qvalue_2.6.0                 
## [17] colorspace_1.2-7              htmltools_0.3.5              
## [19] httpuv_1.3.3                  Matrix_1.2-7.1               
## [21] XML_3.98-1.4                  biomaRt_2.30.0               
## [23] zlibbioc_1.20.0               xtable_1.8-2                 
## [25] GO.db_3.4.0                   scales_0.4.0                 
## [27] BiocParallel_1.8.0            tibble_1.2                   
## [29] mgcv_1.8-15                   SummarizedExperiment_1.4.0   
## [31] lazyeval_0.2.0                magrittr_1.5                 
## [33] mime_0.5                      evaluate_0.10                
## [35] nlme_3.1-128                  BiocInstaller_1.24.0         
## [37] tools_3.3.1                   formatR_1.4                  
## [39] stringr_1.1.0                 munsell_0.4.3                
## [41] Biostrings_2.42.0             RCurl_1.95-4.8               
## [43] igraph_1.0.1                  bitops_1.0-6                 
## [45] labeling_0.3                  rmarkdown_1.1                
## [47] gtable_0.2.0                  DBI_0.5-1                    
## [49] reshape2_1.4.1                R6_2.2.0                     
## [51] GenomicAlignments_1.10.0      knitr_1.14                   
## [53] rtracklayer_1.34.0            fastmatch_1.0-4              
## [55] fgsea_1.0.0                   stringi_1.1.2                
## [57] GOSemSim_2.0.0                Rcpp_0.12.7
</code></pre>
</div>

  </div>

  <h1><a href="/genome/2016/10/23/AnnotationDbi">Exploring the human genome (Part 1) - Gene Annotations</a></h1>
  <p class="author">
    <span class="date">2016-10-23 00:00:00 +0200</span>
  </p>
  <div class="content">
    <p>When working with any type of genome data, we often look for annotation information about genes, e.g. what’s the gene’s full name, what’s its abbreviated symbol, what ID it has in other databases, what functions have been described, how many and which transcripts exist, etc.</p>

<p>However, when looking for this information we (luckily) find a number of different databases and packages to access their information in R. But with this often comes confusion, because they all vary - sometimes only slightly, other times quite strongly - in how many genes they list and what they consider to be a “gene”.</p>

<hr />

<p><br /></p>

<h1 id="what-exactly-is-a-gene">What exactly is a gene?</h1>

<p>The narrow traditional definition of a gene is that it is a hereditary unit of information, which meant that it is a unit of DNA which encodes for the production of a protein. The Human Genome Project has estimated that the human genome comprises 20000 to 25000 genes.</p>

<p>However, if we take the definition of gene more liberally, we could also consider chunks of DNA which code for other functional units, like miRNAs, pseudogenes or other regulatory elements.</p>

<hr />

<p><br /></p>

<h1 id="annotationdbi">AnnotationDbi</h1>

<p>The R package <a href="http://www.bioconductor.org/packages/release/bioc/html/AnnotationDbi.html">AnnotationDbi</a> provides connections to a large number of annotation resources for gene and genome information of many organisms. It can access feature data from a number of gene or genome centric packages.</p>

<p>Here, I want to explore some of these packages and compare the information they contain.</p>

<hr />

<p><br /></p>

<h3 id="how-many-genes-are-there-in-different-databases">How many genes are there in different databases?</h3>

<p>Because genes can have identifiers from many databases (but not necessarily from all and there is not always a one-to-one mapping between them), we compare the number of unique identifiers for Entrez and Ensembl, as well as for gene symbol.</p>

<blockquote>
  <p>“<a href="https://www.ncbi.nlm.nih.gov/gene">Entrez Gene</a> is NCBI’s database for gene-specific information. It does not include all known or predicted genes; instead Entrez Gene focuses on the genomes that have been completely sequenced, that have an active research community to contribute gene-specific information, or that are scheduled for intense sequence analysis.” <a href="http://nar.oxfordjournals.org/content/33/suppl_1/D54.full">Maglott et al., Nucleic Acids Research 2005</a></p>
</blockquote>

<blockquote>
  <p><a href="http://www.ensembl.org/index.html">Ensembl</a> is a genome browser but it also includes gene annotations.</p>
</blockquote>

<blockquote>
  <p>Gene symbols represent the official gene name abbreviation from <a href="http://www.genenames.org/">HGNC</a>.</p>
</blockquote>

<hr />

<p><br /></p>

<h2 id="orghsegdb">org.Hs.eg.db</h2>

<p>The first package I looked at was <a href="http://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html">org.Hs.eg.db</a>, which contains genome wide annotations for the human genome.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">AnnotationDbi</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">)</span><span class="w">

</span><span class="c1"># We can find out which keys and/or columns can be accessed by running
# keytypes(org.Hs.eg.db) or columns(org.Hs.eg.db)
</span><span class="w">
</span><span class="n">ENTREZID_org</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">)</span><span class="w">
</span><span class="n">ENSEMBL_org</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENSEMBL"</span><span class="p">)</span><span class="w">
</span><span class="n">HGNC_org</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"SYMBOL"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<h2 id="ensdbhsapiensv79">EnsDb.Hsapiens.v79</h2>

<p><a href="https://bioconductor.org/packages/release/data/annotation/html/EnsDb.Hsapiens.v79.html">EnsDb.Hsapiens.v79</a> contains annotation information from Ensembl. Here, we also find genes based on Ensembl, Entrez and HGNC IDs.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">)</span><span class="w">

</span><span class="c1"># keytypes(EnsDb.Hsapiens.v79)
</span><span class="w">
</span><span class="n">ENTREZID_EnsDb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">)</span><span class="w">
</span><span class="n">ENSEMBL_EnsDb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">
</span><span class="n">HGNC_EnsDb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"SYMBOL"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<h2 id="txdbhsapiensucschg38knowngene">TxDb.Hsapiens.UCSC.hg38.knownGene</h2>

<p><a href="http://bioconductor.org/packages/release/data/annotation/html/TxDb.Hsapiens.UCSC.hg38.knownGene.html">TxDb.Hsapiens.UCSC.hg38.knownGene</a> provides annotation information from UCSC, specifically on transcripts in form of TxDb objects. Here, we only find comparable information on Entrez IDs.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="p">)</span><span class="w">

</span><span class="c1"># keytypes(TxDb.Hsapiens.UCSC.hg38.knownGene)
</span><span class="w">
</span><span class="n">ENTREZID_TxDb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h1 id="comparison-of-gene-numbers-from-the-three-databases-accessed-via-the-three-r-packages">Comparison of gene numbers from the three databases accessed via the three R packages</h1>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># setting my custom theme of choice
</span><span class="n">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span><span class="w">

</span><span class="n">my_theme</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">base_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">,</span><span class="w"> </span><span class="n">base_family</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"sans"</span><span class="p">){</span><span class="w">
  </span><span class="n">theme_grey</span><span class="p">(</span><span class="n">base_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">base_size</span><span class="p">,</span><span class="w"> </span><span class="n">base_family</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">base_family</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="w">
    </span><span class="n">axis.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">14</span><span class="p">,</span><span class="w"> </span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">90</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="w">
    </span><span class="n">axis.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">14</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.grid.major</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_line</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.grid.minor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
    </span><span class="n">panel.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"white"</span><span class="p">),</span><span class="w">
    </span><span class="n">strip.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"cornsilk"</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"maroon"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w">
    </span><span class="n">strip.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bold"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">15</span><span class="p">,</span><span class="w"> </span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"maroon"</span><span class="p">),</span><span class="w">
    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bottom"</span><span class="w">
  </span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># preparing the gene table for plotting
</span><span class="n">gene_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">DB</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Entrez_orgDb"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Ensembl_orgDb"</span><span class="p">,</span><span class="w"> </span><span class="s2">"HGNC_orgDb"</span><span class="p">,</span><span class="w"> 
                                </span><span class="s2">"Entrez_EnsDb"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Ensembl_EnsDb"</span><span class="p">,</span><span class="w"> </span><span class="s2">"HGNC_EnsDb"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Entrez_TxDb"</span><span class="p">),</span><span class="w"> 
                         </span><span class="n">no_genes</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">ENTREZID_org</span><span class="p">)),</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">ENSEMBL_org</span><span class="p">)),</span><span class="w"> 
                                      </span><span class="nf">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">HGNC_org</span><span class="p">)),</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">ENTREZID_EnsDb</span><span class="p">)),</span><span class="w">
                                      </span><span class="nf">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">ENSEMBL_EnsDb</span><span class="p">)),</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">HGNC_EnsDb</span><span class="p">)),</span><span class="w"> 
                                      </span><span class="nf">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">ENTREZID_TxDb</span><span class="p">))))</span><span class="w">

</span><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">gene_table</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">DB</span><span class="p">),</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">no_genes</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">no_genes</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"identity"</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"maroon"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of genes in databases"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Count"</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">)</span><span class="w">

</span><span class="n">p</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">geom_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">6</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1.5</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="AnnotationDbi_blog_files/figure-markdown_github/gene_numbers-1.png" style="display: block; margin: auto;" /></p>

<p>When we compare the numbers of genes we can obtain from the three databases Ensembl, Entrez and HGNC via three R packages (org.Hs.eg.db, EnsDb.Hsapiens.v79 and TxDb.Hsapiens.UCSC.hg38.knownGene), we see that there are strong differences: There seem to be either around 25000 to 28000 genes or around 60000. The only database with consistent numbers is HGNC. The reason for these divergent numbers could be due to how they classify a “gene”. Everything that is around 25000 could be because they consider protein coding genes only. Will we check if that’s indeed the case in a minute.</p>

<hr />

<p><br /></p>

<h4 id="overlap-between-databases">Overlap between databases</h4>

<p>First, I want to know how well the database information from different packages overlaps. Venn diagrams show us the number of genes overlapping between the three databases.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">gplots</span><span class="p">)</span><span class="w">

</span><span class="n">venn</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">orgDb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">unique</span><span class="p">(</span><span class="n">ENTREZID_org</span><span class="p">),</span><span class="w"> </span><span class="n">EnsDb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">unique</span><span class="p">(</span><span class="n">ENTREZID_EnsDb</span><span class="p">),</span><span class="w"> </span><span class="n">TxDb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">unique</span><span class="p">(</span><span class="n">ENTREZID_TxDb</span><span class="p">)),</span><span class="w"> 
     </span><span class="n">show.plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">small</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">,</span><span class="w"> </span><span class="n">showSetLogicLabel</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">mtext</span><span class="p">(</span><span class="s2">"Entrez database"</span><span class="p">,</span><span class="w"> </span><span class="n">side</span><span class="o">=</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">cex</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1.2</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="AnnotationDbi_blog_files/figure-markdown_github/VennDiagram-1.png" style="display: block; margin: auto;" /></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">venn</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">orgDb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">unique</span><span class="p">(</span><span class="n">ENSEMBL_org</span><span class="p">),</span><span class="w"> </span><span class="n">EnsDb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">unique</span><span class="p">(</span><span class="n">ENSEMBL_EnsDb</span><span class="p">)),</span><span class="w"> 
     </span><span class="n">show.plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">small</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">,</span><span class="w"> </span><span class="n">showSetLogicLabel</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">mtext</span><span class="p">(</span><span class="s2">"Ensembl database"</span><span class="p">,</span><span class="w"> </span><span class="n">side</span><span class="o">=</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">cex</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1.2</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="AnnotationDbi_blog_files/figure-markdown_github/VennDiagram2-1.png" style="display: block; margin: auto;" /></p>

<p><img src="AnnotationDbi_blog_files/figure-markdown_github/VennDiagram3-1.png" style="display: block; margin: auto;" /></p>

<p>While there is a reasonably large overlap between packages, there are still many genes which are either in one or the other package.</p>

<p>For gene names this is worst, but it might be due to changing names of genes (one gene often has several symbols which can be used, the preferred one often changing with new knowledge about the gene’s function).</p>

<p>For Ensembl and Entrez it seems like one of the databases simply contains more gene entries than the other(s), while the majority of genes from the smaller database are included.</p>

<hr />

<p><br /></p>

<h3 id="gene-biotypes">Gene biotypes</h3>

<p>Gene biotype annotation tells us the general category of a gene. The biggest category is protein coding genes. They allow us to check whether the genes in different databases and package consider different categories as “genes”.</p>

<p>Gene biotype information is only available for genes in EnsemblDb! Thus, the classification of genes in the other packages is biased in that only genes in EnsemblDb could be considered.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># gene dataframe for org.Hs.eg.db
</span><span class="n">gene_dataframe_orgDb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">AnnotationDbi</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="o">=</span><span class="n">ENTREZID_org</span><span class="p">,</span><span class="w"> </span><span class="n">columns</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"ENSEMBL"</span><span class="p">,</span><span class="w"> </span><span class="s2">"SYMBOL"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="o">=</span><span class="s2">"ENTREZID"</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">gene_dataframe_orgDb</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Entrez"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Ensembl"</span><span class="p">,</span><span class="w"> </span><span class="s2">"HGNC"</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">gene_dataframe_orgDb</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">gene_dataframe_orgDb</span><span class="p">),</span><span class="w"> </span><span class="s2">"orgDb"</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)</span><span class="w">
</span><span class="c1"># additional column with HGNC added for merging
</span><span class="n">gene_dataframe_orgDb</span><span class="o">$</span><span class="n">HGNC</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gene_dataframe_orgDb</span><span class="o">$</span><span class="n">HGNC_orgDb</span><span class="w">

</span><span class="c1"># gene dataframe for EnsDb.Hsapiens.v79
</span><span class="n">gene_dataframe_EnsDb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="o">=</span><span class="n">ENSEMBL_EnsDb</span><span class="p">,</span><span class="w"> 
                                          </span><span class="n">columns</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"ENTREZID"</span><span class="p">,</span><span class="w"> </span><span class="s2">"SYMBOL"</span><span class="p">,</span><span class="w"> </span><span class="s2">"GENEBIOTYPE"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="o">=</span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">gene_dataframe_EnsDb</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Ensembl"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Entrez"</span><span class="p">,</span><span class="w"> </span><span class="s2">"HGNC"</span><span class="p">,</span><span class="w"> </span><span class="s2">"GENEBIOTYPE"</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">gene_dataframe_EnsDb</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">gene_dataframe_EnsDb</span><span class="p">),</span><span class="w"> </span><span class="s2">"EnsDb"</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"_"</span><span class="p">)</span><span class="w">
</span><span class="c1"># additional column with HGNC added for merging &amp; keeping one copy of it in the merged dataframe
</span><span class="n">gene_dataframe_EnsDb</span><span class="o">$</span><span class="n">HGNC</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gene_dataframe_EnsDb</span><span class="o">$</span><span class="n">HGNC_EnsDb</span><span class="w">

</span><span class="c1"># merging the two dataframes by HGNC
</span><span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span><span class="w">
</span><span class="n">gene_dataframe</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">full_join</span><span class="p">(</span><span class="n">gene_dataframe_EnsDb</span><span class="p">,</span><span class="w"> </span><span class="n">gene_dataframe_orgDb</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"HGNC"</span><span class="p">)</span><span class="w">
</span><span class="n">gene_dataframe</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gene_dataframe</span><span class="p">[,</span><span class="w"> </span><span class="m">-5</span><span class="p">]</span><span class="w">

</span><span class="c1"># making a dataframe with additional column for merging &amp; keeping one copy in the merged dataframe
</span><span class="n">ENTREZID_TxDb_df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">ENTREZID</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">ENTREZID_TxDb</span><span class="p">),</span><span class="w"> </span><span class="n">Entrez_TxDb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ENTREZID_TxDb</span><span class="p">)</span><span class="w">
</span><span class="n">gene_dataframe</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">left_join</span><span class="p">(</span><span class="n">gene_dataframe</span><span class="p">,</span><span class="w"> </span><span class="n">ENTREZID_TxDb_df</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Entrez_orgDb"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">))</span><span class="w"> 
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># calculating percentages of gene biotypes
</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">gene_dataframe</span><span class="p">))){</span><span class="w">
  </span><span class="n">key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">gene_dataframe</span><span class="p">)[</span><span class="n">i</span><span class="p">]</span><span class="w">
  
  </span><span class="c1"># remove duplicate and NA plots
</span><span class="w">  </span><span class="n">gene_dataframe_subs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gene_dataframe</span><span class="p">[</span><span class="o">!</span><span class="n">duplicated</span><span class="p">(</span><span class="n">gene_dataframe</span><span class="p">[,</span><span class="w"> </span><span class="n">key</span><span class="p">]),</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">key</span><span class="p">,</span><span class="w"> </span><span class="s2">"GENEBIOTYPE_EnsDb"</span><span class="p">)]</span><span class="w">
  </span><span class="n">gene_dataframe_subs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gene_dataframe_subs</span><span class="p">[</span><span class="o">!</span><span class="nf">is.na</span><span class="p">(</span><span class="n">gene_dataframe_subs</span><span class="p">[,</span><span class="w"> </span><span class="n">key</span><span class="p">]),</span><span class="w"> </span><span class="p">]</span><span class="w">
  
  </span><span class="n">cat</span><span class="p">(</span><span class="s2">"\nKey:"</span><span class="p">,</span><span class="w"> </span><span class="n">key</span><span class="p">,</span><span class="w"> </span><span class="s2">"has"</span><span class="p">,</span><span class="w"> </span><span class="n">nrow</span><span class="p">(</span><span class="n">gene_dataframe_subs</span><span class="p">),</span><span class="w"> </span><span class="s2">"unique rows without NAs.\nOf these,"</span><span class="p">,</span><span class="w"> 
      </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">gene_dataframe_subs</span><span class="o">$</span><span class="n">GENEBIOTYPE_EnsDb</span><span class="p">)),</span><span class="w"> </span><span class="s2">"don't have a gene biotype annotation\n"</span><span class="p">)</span><span class="w">
  
  </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">1</span><span class="p">){</span><span class="w">
    </span><span class="n">genebiotypes_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">gene_dataframe_subs</span><span class="o">$</span><span class="n">GENEBIOTYPE_EnsDb</span><span class="p">))</span><span class="w">
    </span><span class="n">genebiotypes_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rbind</span><span class="p">(</span><span class="n">genebiotypes_table</span><span class="p">,</span><span class="w">
                                      </span><span class="n">data.frame</span><span class="p">(</span><span class="n">Var1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"NA"</span><span class="p">,</span><span class="w"> 
                                                 </span><span class="n">Freq</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">gene_dataframe_subs</span><span class="o">$</span><span class="n">GENEBIOTYPE_EnsDb</span><span class="p">))))</span><span class="w">
    </span><span class="n">genebiotypes_table</span><span class="o">$</span><span class="n">percent</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">genebiotypes_table</span><span class="o">$</span><span class="n">Freq</span><span class="w"> </span><span class="o">/</span><span class="w"> 
                                                </span><span class="nf">sum</span><span class="p">(</span><span class="n">genebiotypes_table</span><span class="o">$</span><span class="n">Freq</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">4</span><span class="p">)</span><span class="w">
  </span><span class="p">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="p">{</span><span class="w">
      </span><span class="n">genebiotypes_table_pre</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">gene_dataframe_subs</span><span class="o">$</span><span class="n">GENEBIOTYPE_EnsDb</span><span class="p">))</span><span class="w">
      </span><span class="n">genebiotypes_table_pre</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rbind</span><span class="p">(</span><span class="n">genebiotypes_table_pre</span><span class="p">,</span><span class="w">
                                      </span><span class="n">data.frame</span><span class="p">(</span><span class="n">Var1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"NA"</span><span class="p">,</span><span class="w"> 
                                                 </span><span class="n">Freq</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">gene_dataframe_subs</span><span class="o">$</span><span class="n">GENEBIOTYPE_EnsDb</span><span class="p">))))</span><span class="w">
      </span><span class="n">genebiotypes_table_pre</span><span class="o">$</span><span class="n">percent</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">round</span><span class="p">(</span><span class="n">genebiotypes_table_pre</span><span class="o">$</span><span class="n">Freq</span><span class="w"> </span><span class="o">/</span><span class="w"> 
                                                </span><span class="nf">sum</span><span class="p">(</span><span class="n">genebiotypes_table_pre</span><span class="o">$</span><span class="n">Freq</span><span class="p">)</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">4</span><span class="p">)</span><span class="w">
      </span><span class="n">genebiotypes_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">full_join</span><span class="p">(</span><span class="n">genebiotypes_table</span><span class="p">,</span><span class="w"> </span><span class="n">genebiotypes_table_pre</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">)</span><span class="w">
  </span><span class="p">}</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## 
## Key: Ensembl_EnsDb has 65774 unique rows without NAs.
## Of these, 0 don't have a gene biotype annotation
## 
## Key: Entrez_EnsDb has 24263 unique rows without NAs.
## Of these, 0 don't have a gene biotype annotation
## 
## Key: HGNC_EnsDb has 59074 unique rows without NAs.
## Of these, 0 don't have a gene biotype annotation
## 
## Key: Entrez_orgDb has 60136 unique rows without NAs.
## Of these, 25740 don't have a gene biotype annotation
## 
## Key: Ensembl_orgDb has 28015 unique rows without NAs.
## Of these, 2456 don't have a gene biotype annotation
## 
## Key: HGNC_orgDb has 60083 unique rows without NAs.
## Of these, 25691 don't have a gene biotype annotation
## 
## Key: Entrez_TxDb has 25217 unique rows without NAs.
## Of these, 2087 don't have a gene biotype annotation
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># create unique colnames
</span><span class="n">colnames</span><span class="p">(</span><span class="n">genebiotypes_table</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">make.unique</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"Key"</span><span class="p">,</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">gene_dataframe</span><span class="p">),</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)))</span><span class="w">
</span><span class="n">genebiotypes_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">genebiotypes_table</span><span class="p">[,</span><span class="w"> </span><span class="o">-</span><span class="nf">c</span><span class="p">(</span><span class="m">8</span><span class="p">,</span><span class="w"> </span><span class="m">9</span><span class="p">)]</span><span class="w">

</span><span class="c1"># order by percentage of Ensembl EnsDb
</span><span class="n">genebiotypes_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">genebiotypes_table</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">genebiotypes_table</span><span class="o">$</span><span class="n">Ensembl_EnsDb.1</span><span class="p">,</span><span class="w"> </span><span class="n">decreasing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<h4 id="how-many-protein-coding-genes-are-in-each-database-and-annotation-package">How many protein coding genes are in each database and annotation package?</h4>

<p>Ensembl IDs from EnsDb.Hsapiens.v79 have the most protein coding genes, which is consistent with also having the highest number of annotations in general. The number of protein coding genes in the other databases/ packages is only slightly lower, however and fits well with the predicted 20000 to 25000.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">genebiotypes_table_protein_coding</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">genebiotypes_table</span><span class="p">[</span><span class="w">
  </span><span class="n">which</span><span class="p">(</span><span class="n">genebiotypes_table</span><span class="o">$</span><span class="n">Key</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"protein_coding"</span><span class="p">,</span><span class="w"> </span><span class="s2">"NA"</span><span class="p">)),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="n">genebiotypes_table_count</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">genebiotypes_table_protein_coding</span><span class="p">[</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">grep</span><span class="p">(</span><span class="s2">".1"</span><span class="p">,</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">genebiotypes_table_protein_coding</span><span class="p">))</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="m">1</span><span class="p">)]</span><span class="w">

</span><span class="n">library</span><span class="p">(</span><span class="n">tidyr</span><span class="p">)</span><span class="w">
</span><span class="n">genebiotypes_table_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">genebiotypes_table_count</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">Key</span><span class="p">,</span><span class="w"> </span><span class="n">Count</span><span class="p">,</span><span class="w"> </span><span class="n">Ensembl_EnsDb</span><span class="o">:</span><span class="n">Entrez_TxDb</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">genebiotypes_table_gather</span><span class="p">)[</span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"DB"</span><span class="w">

</span><span class="c1"># plot
</span><span class="n">bp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">genebiotypes_table_gather</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">DB</span><span class="p">),</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Count</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Count</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"identity"</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"maroon"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">my_theme</span><span class="p">()</span><span class="w">
</span><span class="n">bp</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of protein coding genes in databases"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Count"</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">6</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1.5</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="AnnotationDbi_blog_files/figure-markdown_github/protein_coding-1.png" style="display: block; margin: auto;" /></p>

<p>When looking at the percentage of protein coding genes among all gene annotations we can confirm that the database/ package combinations with lower gene numbers (Ensembl org.Db, Entrez EnsDb and Entrez TxDb) are mainly comprised of protein coding genes while databases/ packages with around 60000 gene entries consider other functional units as well.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">genebiotypes_table_percent</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">genebiotypes_table_protein_coding</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">grep</span><span class="p">(</span><span class="s2">".1"</span><span class="p">,</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">genebiotypes_table_protein_coding</span><span class="p">)))]</span><span class="w">

</span><span class="n">library</span><span class="p">(</span><span class="n">tidyr</span><span class="p">)</span><span class="w">
</span><span class="n">genebiotypes_table_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">genebiotypes_table_percent</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">Key</span><span class="p">,</span><span class="w"> </span><span class="n">Percent</span><span class="p">,</span><span class="w"> </span><span class="n">Ensembl_EnsDb.1</span><span class="o">:</span><span class="n">Entrez_TxDb.1</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">genebiotypes_table_gather</span><span class="p">)[</span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"DB"</span><span class="w">
</span><span class="n">genebiotypes_table_gather</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">".1"</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">genebiotypes_table_gather</span><span class="o">$</span><span class="n">DB</span><span class="p">)</span><span class="w">

</span><span class="c1"># plot
</span><span class="n">bp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">genebiotypes_table_gather</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Percent</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Key</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_bar</span><span class="p">(</span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"identity"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">theme_minimal</span><span class="p">()</span><span class="w">

</span><span class="n">pie</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bp</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">coord_polar</span><span class="p">(</span><span class="s2">"y"</span><span class="p">,</span><span class="w"> </span><span class="n">start</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">ggtitle</span><span class="p">(</span><span class="s2">"Percentage of protein coding genes in gene databases"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="w">
  </span><span class="n">axis.title.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
  </span><span class="n">axis.title.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
  </span><span class="n">panel.border</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
  </span><span class="n">panel.grid</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
  </span><span class="n">axis.ticks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
  </span><span class="n">plot.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">14</span><span class="p">,</span><span class="w"> </span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bold"</span><span class="p">),</span><span class="w">
  </span><span class="n">legend.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
  </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bottom"</span><span class="p">,</span><span class="w">
  </span><span class="n">legend.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">8</span><span class="p">)</span><span class="w">
  </span><span class="p">)</span><span class="w">
</span><span class="n">pie</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">facet_wrap</span><span class="p">(</span><span class="o">~</span><span class="n">DB</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">4</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="AnnotationDbi_blog_files/figure-markdown_github/protein_coding2-1.png" style="display: block; margin: auto;" /></p>

<p><br /></p>

<h4 id="percentages-of-gene-biotypes">Percentages of gene biotypes</h4>

<p>What are these other biotypes that most databases/ packages consider as genes? The second largest group following protein coding genes are processed pseudogenes, then come lincRNAs and other small RNAs.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">genebiotypes_table_percent</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">genebiotypes_table</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">grep</span><span class="p">(</span><span class="s2">".1"</span><span class="p">,</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">genebiotypes_table</span><span class="p">)))]</span><span class="w">
</span><span class="n">genebiotypes_table_percent</span><span class="o">$</span><span class="n">Key</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">genebiotypes_table_percent</span><span class="o">$</span><span class="n">Key</span><span class="p">,</span><span class="w"> 
                                         </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">paste0</span><span class="p">(</span><span class="n">genebiotypes_table_percent</span><span class="o">$</span><span class="n">Key</span><span class="p">))</span><span class="w">

</span><span class="c1"># gather dataframe for plotting
</span><span class="n">library</span><span class="p">(</span><span class="n">tidyr</span><span class="p">)</span><span class="w">
</span><span class="n">genebiotypes_table_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">genebiotypes_table_percent</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">Key</span><span class="p">,</span><span class="w"> </span><span class="n">Percent</span><span class="p">,</span><span class="w"> </span><span class="n">Ensembl_EnsDb.1</span><span class="o">:</span><span class="n">Entrez_TxDb.1</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">genebiotypes_table_gather</span><span class="p">)[</span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"DB"</span><span class="w">
</span><span class="n">genebiotypes_table_gather</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">".1"</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">genebiotypes_table_gather</span><span class="o">$</span><span class="n">DB</span><span class="p">)</span><span class="w">

</span><span class="c1"># plot
</span><span class="n">bp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">genebiotypes_table_gather</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Percent</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Key</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_bar</span><span class="p">(</span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"identity"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">theme_minimal</span><span class="p">()</span><span class="w">

</span><span class="n">pie</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bp</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">coord_polar</span><span class="p">(</span><span class="s2">"y"</span><span class="p">,</span><span class="w"> </span><span class="n">start</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">ggtitle</span><span class="p">(</span><span class="s2">"Percentage of gene biotypes in gene databases"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="w">
  </span><span class="n">axis.title.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
  </span><span class="n">axis.title.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
  </span><span class="n">panel.border</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
  </span><span class="n">panel.grid</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
  </span><span class="n">axis.ticks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
  </span><span class="n">plot.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">14</span><span class="p">,</span><span class="w"> </span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bold"</span><span class="p">),</span><span class="w">
  </span><span class="n">legend.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
  </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bottom"</span><span class="p">,</span><span class="w">
  </span><span class="n">legend.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">8</span><span class="p">)</span><span class="w">
  </span><span class="p">)</span><span class="w">
</span><span class="n">pie</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">facet_wrap</span><span class="p">(</span><span class="o">~</span><span class="n">DB</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">4</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="AnnotationDbi_blog_files/figure-markdown_github/genebiotypes_percent-1.png" style="display: block; margin: auto;" /></p>

<hr />

<p><br /></p>

<h1 id="what-does-this-mean-for-genetics-and-genomics-research">What does this mean for genetics and genomics research?</h1>

<p>Overall, Ensembl gene annotations from EnsDb.Hsapiens.v79 have the largest number of gene annotations.</p>

<p>However, this comparison shows that it can be critical for identifying genes and annotating them correctly in your analyses to consider the strengths and weaknesses of the different databases and packages you use to access the information. I believe it is well worthwhile the time and effort to consider what information a database can or cannot give you before using it on your data.</p>

<p><br /></p>

<p>In a <a href="https://shiring.github.io/genome/2016/11/01/AnnotationDbi_part2">part 2</a> I am looking at transcripts.</p>

<p><br /></p>

<hr />

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sessionInfo</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## R version 3.3.1 (2016-06-21)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: OS X 10.12 (Sierra)
## 
## locale:
## [1] de_DE.UTF-8/de_DE.UTF-8/de_DE.UTF-8/C/de_DE.UTF-8/de_DE.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] tidyr_0.6.0                            
##  [2] dplyr_0.5.0                            
##  [3] gplots_3.0.1                           
##  [4] ggplot2_2.1.0                          
##  [5] TxDb.Hsapiens.UCSC.hg38.knownGene_3.4.0
##  [6] EnsDb.Hsapiens.v79_1.1.0               
##  [7] ensembldb_1.6.0                        
##  [8] GenomicFeatures_1.26.0                 
##  [9] GenomicRanges_1.26.0                   
## [10] GenomeInfoDb_1.10.0                    
## [11] org.Hs.eg.db_3.4.0                     
## [12] AnnotationDbi_1.36.0                   
## [13] IRanges_2.8.0                          
## [14] S4Vectors_0.12.0                       
## [15] Biobase_2.34.0                         
## [16] BiocGenerics_0.20.0                    
## 
## loaded via a namespace (and not attached):
##  [1] SummarizedExperiment_1.4.0    gtools_3.5.0                 
##  [3] lattice_0.20-34               colorspace_1.2-7             
##  [5] htmltools_0.3.5               rtracklayer_1.34.0           
##  [7] yaml_2.1.13                   interactiveDisplayBase_1.12.0
##  [9] XML_3.98-1.4                  DBI_0.5-1                    
## [11] BiocParallel_1.8.0            plyr_1.8.4                   
## [13] stringr_1.1.0                 zlibbioc_1.20.0              
## [15] Biostrings_2.42.0             munsell_0.4.3                
## [17] gtable_0.2.0                  caTools_1.17.1               
## [19] evaluate_0.10                 labeling_0.3                 
## [21] knitr_1.14                    biomaRt_2.30.0               
## [23] httpuv_1.3.3                  BiocInstaller_1.24.0         
## [25] Rcpp_0.12.7                   KernSmooth_2.23-15           
## [27] xtable_1.8-2                  scales_0.4.0                 
## [29] formatR_1.4                   gdata_2.17.0                 
## [31] XVector_0.14.0                mime_0.5                     
## [33] Rsamtools_1.26.0              AnnotationHub_2.6.0          
## [35] digest_0.6.10                 stringi_1.1.2                
## [37] shiny_0.14.1                  grid_3.3.1                   
## [39] tools_3.3.1                   bitops_1.0-6                 
## [41] magrittr_1.5                  lazyeval_0.2.0               
## [43] RCurl_1.95-4.8                tibble_1.2                   
## [45] RSQLite_1.0.0                 Matrix_1.2-7.1               
## [47] assertthat_0.1                rmarkdown_1.1                
## [49] httr_1.2.1                    R6_2.2.0                     
## [51] GenomicAlignments_1.10.0
</code></pre>
</div>

  </div>

  <h1><a href="/maps/gpx/2016/10/16/roadtrip2016">USA/ Canada Roadtrip 2016</a></h1>
  <p class="author">
    <span class="date">2016-10-16 00:00:00 +0200</span>
  </p>
  <div class="content">
    <h1 id="mapping-gps-data-from-our-usa-canada-roadtrip">Mapping GPS data from our USA/ Canada Roadtrip</h1>

<p>This September we went on a roadtrip to the US and Canada. 
Of course, we had our trusty GPS to guide us along the way - the data from which I downloaded and used to play around with.</p>

<p>(If you want to see a few photos from the trip and don’t care about the rest, skip to the bottom…)</p>

<p><br /></p>

<h2 id="loading-the-data">Loading the data</h2>

<p>I used the <strong>XML</strong> package to load the gpx files following <a href="https://www.r-bloggers.com/stay-on-track-plotting-gps-tracks-with-r/">this blog post</a>.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">XML</span><span class="p">)</span><span class="w">

</span><span class="n">myfiles</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">list.files</span><span class="p">(</span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"../gpx/gpx_files"</span><span class="p">,</span><span class="w"> </span><span class="n">full.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">myfiles</span><span class="p">)){</span><span class="w">

  </span><span class="c1"># One of the files seems to be broken, but since the file is one with earlier data
</span><span class="w">  </span><span class="c1"># not from the trip, I will simply skip it
</span><span class="w">  </span><span class="n">tryCatch</span><span class="p">({</span><span class="w">
    </span><span class="n">pfile</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">htmlTreeParse</span><span class="p">(</span><span class="n">myfiles</span><span class="p">[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span><span class="n">useInternalNodes</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">)</span><span class="w">

  </span><span class="p">},</span><span class="w"> </span><span class="n">error</span><span class="o">=</span><span class="k">function</span><span class="p">(</span><span class="n">e</span><span class="p">){</span><span class="n">cat</span><span class="p">(</span><span class="s2">"ERROR\n"</span><span class="p">)})</span><span class="w">

  </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="n">exists</span><span class="p">(</span><span class="s2">"pfile"</span><span class="p">)){</span><span class="w">
    </span><span class="c1"># Get all elevations, times and coordinates via the respective xpath
</span><span class="w">    </span><span class="n">elevations</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">xpathSApply</span><span class="p">(</span><span class="n">pfile</span><span class="p">,</span><span class="w"> </span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"//trkpt/ele"</span><span class="p">,</span><span class="w"> </span><span class="n">xmlValue</span><span class="p">)))</span><span class="w">
    </span><span class="n">times</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">xpathSApply</span><span class="p">(</span><span class="n">pfile</span><span class="p">,</span><span class="w"> </span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"//trkpt/time"</span><span class="p">,</span><span class="w"> </span><span class="n">xmlValue</span><span class="p">)</span><span class="w">
    </span><span class="n">coords</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">xpathSApply</span><span class="p">(</span><span class="n">pfile</span><span class="p">,</span><span class="w"> </span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"//trkpt"</span><span class="p">,</span><span class="w"> </span><span class="n">xmlAttrs</span><span class="p">)</span><span class="w">
    </span><span class="c1"># Extract latitude and longitude from the coordinates
</span><span class="w">    </span><span class="n">lats</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">coords</span><span class="p">[</span><span class="s2">"lat"</span><span class="p">,]))</span><span class="w">
    </span><span class="n">lons</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">coords</span><span class="p">[</span><span class="s2">"lon"</span><span class="p">,]))</span><span class="w">
    </span><span class="c1"># Put everything in a dataframe and get rid of old variables
</span><span class="w">    </span><span class="n">geodf</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">lat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lats</span><span class="p">,</span><span class="w"> </span><span class="n">lon</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lons</span><span class="p">,</span><span class="w"> </span><span class="n">ele</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">elevations</span><span class="p">,</span><span class="w"> </span><span class="n">time</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">times</span><span class="p">)</span><span class="w">

    </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">1</span><span class="p">){</span><span class="w">
      </span><span class="n">geodata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">geodf</span><span class="w">
      </span><span class="p">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="p">{</span><span class="w">
      </span><span class="n">geodata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rbind</span><span class="p">(</span><span class="n">geodata</span><span class="p">,</span><span class="w"> </span><span class="n">geodf</span><span class="p">)</span><span class="w">
    </span><span class="p">}</span><span class="w">
  </span><span class="p">}</span><span class="w">
  </span><span class="n">rm</span><span class="p">(</span><span class="n">pfile</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span><span class="n">geodata_all</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">geodata</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<p>I then did some cleaning of the date and time column and removed all the old track that were not part of our trip.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># Transforming the time column
</span><span class="n">geodata_all</span><span class="o">$</span><span class="n">time</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">strptime</span><span class="p">(</span><span class="n">geodata_all</span><span class="o">$</span><span class="n">time</span><span class="p">,</span><span class="w"> </span><span class="n">format</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"%Y-%m-%dT%H:%M:%SZ"</span><span class="p">)</span><span class="w">

</span><span class="c1"># ordering by date
</span><span class="n">library</span><span class="p">(</span><span class="n">plyr</span><span class="p">)</span><span class="w">
</span><span class="n">geodata_all</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">arrange</span><span class="p">(</span><span class="n">geodata_all</span><span class="p">,</span><span class="w"> </span><span class="n">time</span><span class="p">)</span><span class="w">

</span><span class="c1"># removing all tracks from Germany
</span><span class="n">geodata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">geodata_all</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">geodata_all</span><span class="o">$</span><span class="n">lon</span><span class="p">))</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0</span><span class="p">),]</span><span class="w">

</span><span class="c1"># adding a column with day for separating the plotting by day lateron
</span><span class="n">geodata</span><span class="o">$</span><span class="n">day</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.factor</span><span class="p">(</span><span class="n">format</span><span class="p">(</span><span class="n">as.Date</span><span class="p">(</span><span class="n">geodata</span><span class="o">$</span><span class="n">time</span><span class="p">,</span><span class="n">format</span><span class="o">=</span><span class="s2">"%Y-%m-%d"</span><span class="p">),</span><span class="w"> </span><span class="s2">"%d"</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>
<hr />

<p><br /></p>

<p>We have 119225 data points with a range from 2016-09-09 19:23:53 to 2016-09-23 21:36:06.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">summary</span><span class="p">(</span><span class="n">geodata</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##       lat             lon              ele        
##  Min.   :36.99   Min.   :-88.68   Min.   :-26.78  
##  1st Qu.:40.09   1st Qu.:-84.86   1st Qu.:181.83  
##  Median :41.90   Median :-82.99   Median :198.65  
##  Mean   :41.27   Mean   :-82.76   Mean   :216.58  
##  3rd Qu.:42.33   3rd Qu.:-80.44   3rd Qu.:255.85  
##  Max.   :43.65   Max.   :-78.85   Max.   :438.02  
##                                                   
##       time                          day       
##  Min.   :2016-09-09 19:23:53   19     :17203  
##  1st Qu.:2016-09-12 12:14:47   12     :14455  
##  Median :2016-09-15 21:52:37   11     :14020  
##  Mean   :2016-09-15 22:35:18   10     :12061  
##  3rd Qu.:2016-09-19 15:59:09   16     :10856  
##  Max.   :2016-09-23 21:36:06   13     :10787  
##                                (Other):39843
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">str</span><span class="p">(</span><span class="n">geodata</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## 'data.frame':    119225 obs. of  5 variables:
##  $ lat : num  42 42 42 42 42 ...
##  $ lon : num  -87.9 -87.9 -87.9 -87.9 -87.9 ...
##  $ ele : num  122 122 122 122 122 ...
##  $ time: POSIXlt, format: "2016-09-09 19:23:53" "2016-09-09 19:23:53" ...
##  $ day : Factor w/ 15 levels "09","10","11",..: 1 1 1 1 1 1 1 1 1 1 ...
</code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="getting-additional-data-from-google-maps">Getting additional data from Google Maps</h2>

<p>Somehow the GPS didn’t record the last two days of our trip back to Chicago, so I added a few way points with the Google Maps Geocoding and Directions APIs (accessed via the <strong>googleway</strong> package).</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">googleway</span><span class="p">)</span><span class="w">

</span><span class="c1"># on th 26th we had lunch at cracker barrel
</span><span class="n">cracker_barrel</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">google_geocode</span><span class="p">(</span><span class="n">address</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"2101 N Kenyon Rd, Urbana, IL 61802, USA"</span><span class="p">,</span><span class="w">
                                 </span><span class="n">key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"your_key"</span><span class="p">,</span><span class="w">
                                 </span><span class="n">language</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"en"</span><span class="p">,</span><span class="w">
                                 </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="n">cracker_barrel_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">lat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">cracker_barrel</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                                  </span><span class="n">lon</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">cracker_barrel</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w">
                                  </span><span class="n">ele</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"NA"</span><span class="p">,</span><span class="w"> </span><span class="n">time</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.POSIXlt</span><span class="p">(</span><span class="s2">"2016-09-26 19:00:00"</span><span class="p">))</span><span class="w">

</span><span class="c1"># then we stayed the night in Aurora/ Naperville
</span><span class="n">super8</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">google_geocode</span><span class="p">(</span><span class="n">address</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"4228 Longmeadow Dr, Aurora, IL 60504, USA"</span><span class="p">,</span><span class="w">
                         </span><span class="n">key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"your_key"</span><span class="p">,</span><span class="w">
                         </span><span class="n">language</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"en"</span><span class="p">,</span><span class="w">
                         </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="n">super8_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">lat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">super8</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                          </span><span class="n">lon</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">super8</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w">
                          </span><span class="n">ele</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"NA"</span><span class="p">,</span><span class="w"> </span><span class="n">time</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.POSIXlt</span><span class="p">(</span><span class="s2">"2016-09-26 23:30:00"</span><span class="p">))</span><span class="w">

</span><span class="c1"># on the 27th we went back to the airport to fly home
</span><span class="n">ohare</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">google_geocode</span><span class="p">(</span><span class="n">address</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"10000 W O'Hare Ave, Chicago, IL 60666, USA"</span><span class="p">,</span><span class="w">
                        </span><span class="n">key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"your_key"</span><span class="p">,</span><span class="w">
                        </span><span class="n">language</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"en"</span><span class="p">,</span><span class="w">
                        </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="n">ohare_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">lat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ohare</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                         </span><span class="n">lon</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ohare</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w">
                         </span><span class="n">ele</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"NA"</span><span class="p">,</span><span class="w"> </span><span class="n">time</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.POSIXlt</span><span class="p">(</span><span class="s2">"2016-09-27 18:00:00"</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<p>Using only these three waypoints won’t plot nicely on a map because it will draw straight lines between the points.
So, I also want to get travel information from Google Maps between these points to get a few more waypoints along the actual route we drove.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1">## Route from Paducah to Cracker Barrel
</span><span class="w">
</span><span class="n">route_pad_cracker</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">google_directions</span><span class="p">(</span><span class="n">origin</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">geodata</span><span class="o">$</span><span class="n">lat</span><span class="p">[</span><span class="n">nrow</span><span class="p">(</span><span class="n">geodata</span><span class="p">)],</span><span class="w"> </span><span class="n">geodata</span><span class="o">$</span><span class="n">lon</span><span class="p">[</span><span class="n">nrow</span><span class="p">(</span><span class="n">geodata</span><span class="p">)]),</span><span class="w">
                  </span><span class="n">destination</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">cracker_barrel_data</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">cracker_barrel_data</span><span class="o">$</span><span class="n">lon</span><span class="p">),</span><span class="w">
                  </span><span class="n">mode</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"driving"</span><span class="p">,</span><span class="w">
                  </span><span class="n">key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"your_key"</span><span class="p">,</span><span class="w">
                  </span><span class="n">language</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"en"</span><span class="p">,</span><span class="w">
                  </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="n">route_pad_cracker_legs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">route_pad_cracker</span><span class="o">$</span><span class="n">routes</span><span class="o">$</span><span class="n">legs</span><span class="p">)</span><span class="w">
</span><span class="n">route_pad_cracker_legs_steps</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">route_pad_cracker_legs</span><span class="o">$</span><span class="n">steps</span><span class="p">)</span><span class="w">

</span><span class="c1">## Route from Cracker Barrel to Aurora
</span><span class="w">
</span><span class="n">route_cracker_aurora</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">google_directions</span><span class="p">(</span><span class="n">origin</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">cracker_barrel_data</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">cracker_barrel_data</span><span class="o">$</span><span class="n">lon</span><span class="p">),</span><span class="w">
                                          </span><span class="n">destination</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">super8_data</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">super8_data</span><span class="o">$</span><span class="n">lon</span><span class="p">),</span><span class="w">
                                          </span><span class="n">mode</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"driving"</span><span class="p">,</span><span class="w">
                                          </span><span class="n">key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"your_key"</span><span class="p">,</span><span class="w">
                                          </span><span class="n">language</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"en"</span><span class="p">,</span><span class="w">
                                          </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="n">route_cracker_aurora_legs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">route_cracker_aurora</span><span class="o">$</span><span class="n">routes</span><span class="o">$</span><span class="n">legs</span><span class="p">)</span><span class="w">
</span><span class="n">route_cracker_aurora_legs_steps</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">route_cracker_aurora_legs</span><span class="o">$</span><span class="n">steps</span><span class="p">)</span><span class="w">

</span><span class="c1">## Route from Cracker Aurora to O'Hare
</span><span class="w">
</span><span class="n">route_aurora_ohare</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">google_directions</span><span class="p">(</span><span class="n">origin</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">super8_data</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">super8_data</span><span class="o">$</span><span class="n">lon</span><span class="p">),</span><span class="w">
                                        </span><span class="n">destination</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">ohare_data</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">ohare_data</span><span class="o">$</span><span class="n">lon</span><span class="p">),</span><span class="w">
                                        </span><span class="n">mode</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"driving"</span><span class="p">,</span><span class="w">
                                        </span><span class="n">key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"your_key"</span><span class="p">,</span><span class="w">
                                        </span><span class="n">language</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"en"</span><span class="p">,</span><span class="w">
                                        </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="n">route_aurora_ohare_legs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">route_aurora_ohare</span><span class="o">$</span><span class="n">routes</span><span class="o">$</span><span class="n">legs</span><span class="p">)</span><span class="w">
</span><span class="n">route_aurora_ohare_legs_steps</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">route_aurora_ohare_legs</span><span class="o">$</span><span class="n">steps</span><span class="p">)</span><span class="w">

</span><span class="c1"># Combining the data sets
</span><span class="n">geodata_2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">lat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">geodata</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                                </span><span class="n">route_pad_cracker_legs</span><span class="o">$</span><span class="n">start_location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">route_pad_cracker_legs_steps</span><span class="o">$</span><span class="n">start_location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                                </span><span class="n">route_pad_cracker_legs</span><span class="o">$</span><span class="n">end_location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">cracker_barrel_data</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                                </span><span class="n">route_cracker_aurora_legs</span><span class="o">$</span><span class="n">start_location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">route_cracker_aurora_legs_steps</span><span class="o">$</span><span class="n">start_location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                                </span><span class="n">route_cracker_aurora_legs</span><span class="o">$</span><span class="n">end_location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">super8_data</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                                </span><span class="n">route_aurora_ohare_legs</span><span class="o">$</span><span class="n">start_location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">route_aurora_ohare_legs_steps</span><span class="o">$</span><span class="n">start_location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                                </span><span class="n">route_aurora_ohare_legs</span><span class="o">$</span><span class="n">end_location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">ohare_data</span><span class="o">$</span><span class="n">lat</span><span class="p">),</span><span class="w">
                      </span><span class="n">lon</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">geodata</span><span class="o">$</span><span class="n">lon</span><span class="p">,</span><span class="w"> </span><span class="n">route_pad_cracker_legs</span><span class="o">$</span><span class="n">start_location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w"> </span><span class="n">route_pad_cracker_legs_steps</span><span class="o">$</span><span class="n">start_location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w">
                              </span><span class="n">route_pad_cracker_legs</span><span class="o">$</span><span class="n">end_location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w"> </span><span class="n">cracker_barrel_data</span><span class="o">$</span><span class="n">lon</span><span class="p">,</span><span class="w">
                              </span><span class="n">route_cracker_aurora_legs</span><span class="o">$</span><span class="n">start_location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w"> </span><span class="n">route_cracker_aurora_legs_steps</span><span class="o">$</span><span class="n">start_location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w">
                              </span><span class="n">route_cracker_aurora_legs</span><span class="o">$</span><span class="n">end_location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w"> </span><span class="n">super8_data</span><span class="o">$</span><span class="n">lon</span><span class="p">,</span><span class="w">
                              </span><span class="n">route_aurora_ohare_legs</span><span class="o">$</span><span class="n">start_location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w"> </span><span class="n">route_aurora_ohare_legs_steps</span><span class="o">$</span><span class="n">start_location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w">
                              </span><span class="n">route_aurora_ohare_legs</span><span class="o">$</span><span class="n">end_location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w"> </span><span class="n">ohare_data</span><span class="o">$</span><span class="n">lon</span><span class="p">),</span><span class="w">
                      </span><span class="n">day</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">geodata</span><span class="o">$</span><span class="n">day</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="s2">"26"</span><span class="p">,</span><span class="w">
                                               </span><span class="nf">length</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">route_pad_cracker_legs</span><span class="o">$</span><span class="n">start_location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                                                        </span><span class="n">route_pad_cracker_legs_steps</span><span class="o">$</span><span class="n">start_location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                                                        </span><span class="n">route_pad_cracker_legs</span><span class="o">$</span><span class="n">end_location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">cracker_barrel_data</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                                                        </span><span class="n">route_cracker_aurora_legs</span><span class="o">$</span><span class="n">start_location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                                                        </span><span class="n">route_cracker_aurora_legs_steps</span><span class="o">$</span><span class="n">start_location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                                                        </span><span class="n">route_cracker_aurora_legs</span><span class="o">$</span><span class="n">end_location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">super8_data</span><span class="o">$</span><span class="n">lat</span><span class="p">))),</span><span class="w">
                              </span><span class="nf">rep</span><span class="p">(</span><span class="s2">"27"</span><span class="p">,</span><span class="w">
                                  </span><span class="nf">length</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">route_aurora_ohare_legs</span><span class="o">$</span><span class="n">start_location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">route_aurora_ohare_legs_steps</span><span class="o">$</span><span class="n">start_location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                                           </span><span class="n">route_aurora_ohare_legs</span><span class="o">$</span><span class="n">end_location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">ohare_data</span><span class="o">$</span><span class="n">lat</span><span class="p">)))))</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="plotting">Plotting</h2>

<p>I tried several packages for plotting but ended up with <strong>ggmap</strong> as the nicest looking solution.</p>

<p><br /></p>

<h3 id="overview-of-the-trip">Overview of the trip</h3>

<p>We flew to Chicago, went west across Michigan to Ann Arbor, Detroit, Toronto, then south again to Pittsburgh, from where we followed the Ohio river and headed to Louisville. From there, we went to Paducah, KY and back north again to Chicago. Each color shows a different day.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">ggmap</span><span class="p">)</span><span class="w">

</span><span class="n">bc_bbox</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">make_bbox</span><span class="p">(</span><span class="n">lat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">lon</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lon</span><span class="p">,</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">geodata_2</span><span class="p">,</span><span class="w"> </span><span class="n">f</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">.10</span><span class="p">)</span><span class="w">
</span><span class="n">map</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">get_map</span><span class="p">(</span><span class="n">location</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bc_bbox</span><span class="p">,</span><span class="w"> </span><span class="n">maptype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"roadmap"</span><span class="p">)</span><span class="w">

</span><span class="n">ggmap</span><span class="p">(</span><span class="n">map</span><span class="p">,</span><span class="w"> </span><span class="n">darken</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0.3</span><span class="p">,</span><span class="w"> </span><span class="s2">"white"</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">geom_point</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lon</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">day</span><span class="p">),</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">day</span><span class="p">)),</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">geodata_2</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_path</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lon</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">day</span><span class="p">)),</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">geodata_2</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">geodata_2</span><span class="o">$</span><span class="n">day</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"26"</span><span class="p">,</span><span class="w"> </span><span class="s2">"27"</span><span class="p">)),</span><span class="w"> </span><span class="p">])</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="o">=</span><span class="s2">"none"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">"Longitude"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="o">=</span><span class="s2">"Latitude"</span><span class="p">,</span><span class="w"> </span><span class="n">title</span><span class="o">=</span><span class="s2">"Roadtrip 2016"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="roadtrip2016_files/figure-markdown_github/unnamed-chunk-7-1.png" alt="" /></p>

<p><br /></p>

<h3 id="road-trip-with-stops">Road trip with stops</h3>

<p>First, I need to obtain the geocodes from the stops (again via Google Maps API).</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">chicago</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">google_geocode</span><span class="p">(</span><span class="n">address</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Chicago"</span><span class="p">,</span><span class="w">
                          </span><span class="n">key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"your_key"</span><span class="p">,</span><span class="w">
                          </span><span class="n">language</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"en"</span><span class="p">,</span><span class="w">
                          </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="n">stjospeh</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">google_geocode</span><span class="p">(</span><span class="n">address</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"St. Joseph, MI"</span><span class="p">,</span><span class="w">
                          </span><span class="n">key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"your_key"</span><span class="p">,</span><span class="w">
                          </span><span class="n">language</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"en"</span><span class="p">,</span><span class="w">
                          </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="n">annarbor</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">google_geocode</span><span class="p">(</span><span class="n">address</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Ann Arbor, MI"</span><span class="p">,</span><span class="w">
                           </span><span class="n">key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"your_key"</span><span class="p">,</span><span class="w">
                           </span><span class="n">language</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"en"</span><span class="p">,</span><span class="w">
                           </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="n">detroit</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">google_geocode</span><span class="p">(</span><span class="n">address</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Detroit, MI"</span><span class="p">,</span><span class="w">
                           </span><span class="n">key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"your_key"</span><span class="p">,</span><span class="w">
                           </span><span class="n">language</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"en"</span><span class="p">,</span><span class="w">
                           </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="n">burlington</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">google_geocode</span><span class="p">(</span><span class="n">address</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Burlington, ON"</span><span class="p">,</span><span class="w">
                          </span><span class="n">key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"your_key"</span><span class="p">,</span><span class="w">
                          </span><span class="n">language</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"en"</span><span class="p">,</span><span class="w">
                          </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="n">ashtabula</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">google_geocode</span><span class="p">(</span><span class="n">address</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Ashtabula"</span><span class="p">,</span><span class="w">
                             </span><span class="n">key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"your_key"</span><span class="p">,</span><span class="w">
                             </span><span class="n">language</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"en"</span><span class="p">,</span><span class="w">
                             </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="n">pittsburgh</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">google_geocode</span><span class="p">(</span><span class="n">address</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Pittsburgh"</span><span class="p">,</span><span class="w">
                            </span><span class="n">key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"your_key"</span><span class="p">,</span><span class="w">
                            </span><span class="n">language</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"en"</span><span class="p">,</span><span class="w">
                            </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="n">marietta</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">google_geocode</span><span class="p">(</span><span class="n">address</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Marietta, OH"</span><span class="p">,</span><span class="w">
                             </span><span class="n">key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"your_key"</span><span class="p">,</span><span class="w">
                             </span><span class="n">language</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"en"</span><span class="p">,</span><span class="w">
                             </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="n">maysville</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">google_geocode</span><span class="p">(</span><span class="n">address</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Maysville, KY"</span><span class="p">,</span><span class="w">
                             </span><span class="n">key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"your_key"</span><span class="p">,</span><span class="w">
                             </span><span class="n">language</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"en"</span><span class="p">,</span><span class="w">
                             </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="n">louisville</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">google_geocode</span><span class="p">(</span><span class="n">address</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Louisville, KY"</span><span class="p">,</span><span class="w">
                             </span><span class="n">key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"your_key"</span><span class="p">,</span><span class="w">
                             </span><span class="n">language</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"en"</span><span class="p">,</span><span class="w">
                             </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="n">paducah</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">google_geocode</span><span class="p">(</span><span class="n">address</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Paducah, KY"</span><span class="p">,</span><span class="w">
                             </span><span class="n">key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"your_key"</span><span class="p">,</span><span class="w">
                             </span><span class="n">language</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"en"</span><span class="p">,</span><span class="w">
                             </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="n">ann_text</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">chicago</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w">
                             </span><span class="n">stjospeh</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w">
                             </span><span class="n">annarbor</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w">
                             </span><span class="n">detroit</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w">
                             </span><span class="n">burlington</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w">
                             </span><span class="n">ashtabula</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w">
                             </span><span class="n">pittsburgh</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w">
                             </span><span class="n">marietta</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w">
                             </span><span class="n">maysville</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lng</span><span class="p">,</span><span class="w">
                             </span><span class="n">louisville</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lng</span><span class="p">[</span><span class="m">1</span><span class="p">],</span><span class="w"> </span><span class="c1">#Maps found a second entry for the location
</span><span class="w">                             </span><span class="n">paducah</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lng</span><span class="p">),</span><span class="w">
                       </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">chicago</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                             </span><span class="n">stjospeh</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                             </span><span class="n">annarbor</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                             </span><span class="n">detroit</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lat</span><span class="m">+0.1</span><span class="p">,</span><span class="w"> </span><span class="c1">#adjusting Detroit label to avoid overlapping with Ann Arbor
</span><span class="w">                             </span><span class="n">burlington</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                             </span><span class="n">ashtabula</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                             </span><span class="n">pittsburgh</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                             </span><span class="n">marietta</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                             </span><span class="n">maysville</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w">
                             </span><span class="n">louisville</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lat</span><span class="p">[</span><span class="m">1</span><span class="p">],</span><span class="w">
                             </span><span class="n">paducah</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">geometry</span><span class="o">$</span><span class="n">location</span><span class="o">$</span><span class="n">lat</span><span class="p">),</span><span class="w">
                       </span><span class="n">labs</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="n">chicago</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">formatted_address</span><span class="p">),</span><span class="w">
                                </span><span class="n">paste0</span><span class="p">(</span><span class="n">stjospeh</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">formatted_address</span><span class="p">),</span><span class="w">
                                </span><span class="n">paste0</span><span class="p">(</span><span class="n">annarbor</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">formatted_address</span><span class="p">),</span><span class="w">
                                </span><span class="n">paste0</span><span class="p">(</span><span class="n">detroit</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">formatted_address</span><span class="p">),</span><span class="w">
                                </span><span class="n">paste0</span><span class="p">(</span><span class="n">burlington</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">formatted_address</span><span class="p">),</span><span class="w">
                                </span><span class="n">paste0</span><span class="p">(</span><span class="n">ashtabula</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">formatted_address</span><span class="p">),</span><span class="w">
                                </span><span class="n">paste0</span><span class="p">(</span><span class="n">pittsburgh</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">formatted_address</span><span class="p">),</span><span class="w">
                                </span><span class="n">paste0</span><span class="p">(</span><span class="n">marietta</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">formatted_address</span><span class="p">),</span><span class="w">
                                </span><span class="n">paste0</span><span class="p">(</span><span class="n">maysville</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">formatted_address</span><span class="p">),</span><span class="w">
                                </span><span class="n">paste0</span><span class="p">(</span><span class="n">louisville</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">formatted_address</span><span class="p">[</span><span class="m">1</span><span class="p">]),</span><span class="w">
                                </span><span class="n">paste0</span><span class="p">(</span><span class="n">paducah</span><span class="o">$</span><span class="n">results</span><span class="o">$</span><span class="n">formatted_address</span><span class="p">)</span><span class="w">
                                </span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">map</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">get_map</span><span class="p">(</span><span class="n">location</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bc_bbox</span><span class="p">,</span><span class="w"> </span><span class="n">maptype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"watercolor"</span><span class="p">)</span><span class="w">

</span><span class="n">roadtrip_map</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggmap</span><span class="p">(</span><span class="n">map</span><span class="p">,</span><span class="w"> </span><span class="n">darken</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0.3</span><span class="p">,</span><span class="w"> </span><span class="s2">"white"</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">geom_point</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lon</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">day</span><span class="p">),</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">day</span><span class="p">)),</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">geodata_2</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_path</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lon</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">day</span><span class="p">)),</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">geodata_2</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">geodata_2</span><span class="o">$</span><span class="n">day</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"26"</span><span class="p">,</span><span class="w"> </span><span class="s2">"27"</span><span class="p">)),</span><span class="w"> </span><span class="p">],</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="o">=</span><span class="s2">"none"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">"Longitude"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="o">=</span><span class="s2">"Latitude"</span><span class="p">,</span><span class="w"> </span><span class="n">title</span><span class="o">=</span><span class="s2">"Roadtrip 2016"</span><span class="p">)</span><span class="w">
</span><span class="n">roadtrip_map</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">geom_text</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ann_text</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">labs</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">4</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<p><img src="roadtrip2016_files/figure-markdown_github/unnamed-chunk-10-1.png" alt="" /></p>

<h3 id="road-trip-statistics">Road trip statistics</h3>

<h4 id="how-many-kilometers-did-we-drive">How many kilometers did we drive?</h4>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># removing duplicate entries
</span><span class="n">geodata_2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">geodata_2</span><span class="p">[</span><span class="o">!</span><span class="n">duplicated</span><span class="p">(</span><span class="n">geodata_2</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="c1"># Shifting vectors for latitude and longitude to include end position
</span><span class="n">shift.vec</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="w"> </span><span class="p">(</span><span class="n">vec</span><span class="p">,</span><span class="w"> </span><span class="n">shift</span><span class="p">){</span><span class="w">
  </span><span class="k">if</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">vec</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="nf">abs</span><span class="p">(</span><span class="n">shift</span><span class="p">)){</span><span class="w">
    </span><span class="nf">rep</span><span class="p">(</span><span class="kc">NA</span><span class="w"> </span><span class="p">,</span><span class="nf">length</span><span class="p">(</span><span class="n">vec</span><span class="p">))</span><span class="w">
  </span><span class="p">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="n">shift</span><span class="w"> </span><span class="o">&gt;=</span><span class="w"> </span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
      </span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">shift</span><span class="p">),</span><span class="w"> </span><span class="n">vec</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">vec</span><span class="p">)</span><span class="o">-</span><span class="n">shift</span><span class="p">)])</span><span class="w"> </span><span class="p">}</span><span class="w">
    </span><span class="k">else</span><span class="w"> </span><span class="p">{</span><span class="w">
      </span><span class="nf">c</span><span class="p">(</span><span class="n">vec</span><span class="p">[(</span><span class="nf">abs</span><span class="p">(</span><span class="n">shift</span><span class="p">)</span><span class="m">+1</span><span class="p">)</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">vec</span><span class="p">)],</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="nf">abs</span><span class="p">(</span><span class="n">shift</span><span class="p">)))</span><span class="w">
    </span><span class="p">}</span><span class="w">
  </span><span class="p">}</span><span class="w">
</span><span class="p">}</span><span class="w">

</span><span class="n">geodata_2</span><span class="o">$</span><span class="n">lat.p1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">shift.vec</span><span class="p">(</span><span class="n">geodata_2</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="m">-1</span><span class="p">)</span><span class="w">
</span><span class="n">geodata_2</span><span class="o">$</span><span class="n">lon.p1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">shift.vec</span><span class="p">(</span><span class="n">geodata_2</span><span class="o">$</span><span class="n">lon</span><span class="p">,</span><span class="w"> </span><span class="m">-1</span><span class="p">)</span><span class="w">

</span><span class="c1"># Calculating distances between points (in metres) with the function pointDistance from the 'raster' package.
</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">raster</span><span class="p">)</span><span class="w">
</span><span class="n">geodata_2</span><span class="o">$</span><span class="n">dist.to.prev</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">geodata_2</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">FUN</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="k">function</span><span class="w"> </span><span class="p">(</span><span class="n">row</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
  </span><span class="n">pointDistance</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">"lat.p1"</span><span class="p">])),</span><span class="w">
                  </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">"lon.p1"</span><span class="p">]))),</span><span class="w">
                </span><span class="nf">c</span><span class="p">(</span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">"lat"</span><span class="p">])),</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">"lon"</span><span class="p">]))),</span><span class="w">
                </span><span class="n">lonlat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">)</span><span class="w"> </span><span class="c1"># Parameter 'lonlat' has to be TRUE!
</span><span class="p">})</span><span class="w">
</span></code></pre>
</div>

<p>We drove approximately 3236.95 km in 18 days. Not counting the ~6 days where we stayed in one place and didn’t drive much, that’s around 269.75 km/ day.</p>

<hr />

<p><br /></p>

<h3 id="identifying-local-time">Identifying local time</h3>

<p>The time stamps are in CEST - Central European Summer Time. We arrived in the Central Time Zone (CDT) but spent the majority of our trip in Estern Time Zone (EDT).</p>

<p><br /></p>

<h4 id="converting-to-cdt-and-edt">Converting to CDT and EDT</h4>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">lubridate</span><span class="p">)</span><span class="w">

</span><span class="n">geodata</span><span class="o">$</span><span class="n">time_CDT</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">with_tz</span><span class="p">(</span><span class="n">geodata</span><span class="o">$</span><span class="n">time</span><span class="p">,</span><span class="w"> </span><span class="n">tz</span><span class="o">=</span><span class="s2">"America/Chicago"</span><span class="p">)</span><span class="w">
</span><span class="n">geodata</span><span class="o">$</span><span class="n">time_EDT</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">with_tz</span><span class="p">(</span><span class="n">geodata</span><span class="o">$</span><span class="n">time</span><span class="p">,</span><span class="w"> </span><span class="n">tz</span><span class="o">=</span><span class="s2">"America/New_York"</span><span class="p">)</span><span class="w">

</span><span class="c1"># extracting day for EDT
</span><span class="n">geodata</span><span class="o">$</span><span class="n">day_EDT</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.factor</span><span class="p">(</span><span class="n">format</span><span class="p">(</span><span class="n">as.Date</span><span class="p">(</span><span class="n">geodata</span><span class="o">$</span><span class="n">time_EDT</span><span class="p">,</span><span class="n">format</span><span class="o">=</span><span class="s2">"%Y-%m-%d"</span><span class="p">),</span><span class="w"> </span><span class="s2">"%d"</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<h4 id="identifying-the-local-time-based-on-geocode">Identifying the local time based on geocode</h4>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># the days where we crossed time zones were the 10th and 23rd
</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">googleway</span><span class="p">)</span><span class="w">

</span><span class="n">geodata_timezone</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">geodata</span><span class="w">
</span><span class="n">geodata_timezone</span><span class="o">$</span><span class="n">local_time</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"NA"</span><span class="w">

</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="n">nrow</span><span class="p">(</span><span class="n">geodata</span><span class="p">)){</span><span class="w">
  </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">nchar</span><span class="p">(</span><span class="n">strsplit</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">i</span><span class="o">/</span><span class="m">100</span><span class="p">),</span><span class="w"> </span><span class="s2">"\\."</span><span class="p">)[[</span><span class="m">1</span><span class="p">]][</span><span class="m">2</span><span class="p">]))){</span><span class="w">
    </span><span class="n">print</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="w">
  </span><span class="p">}</span><span class="w">

  </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="n">geodata</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="s2">"day_EDT"</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"10"</span><span class="p">){</span><span class="w">
    </span><span class="n">data_row</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">geodata</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
    </span><span class="n">google_timezone_api</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">google_timezone</span><span class="p">(</span><span class="n">location</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">data_row</span><span class="o">$</span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">data_row</span><span class="o">$</span><span class="n">lon</span><span class="p">),</span><span class="w">
                                           </span><span class="n">timestamp</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.POSIXct</span><span class="p">(</span><span class="n">data_row</span><span class="o">$</span><span class="n">time_EDT</span><span class="p">),</span><span class="w"> </span><span class="c1"># because the majority of our trip was in EDT
</span><span class="w">                                           </span><span class="n">key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"AIzaSyBkqrk8mBqqvao-W-jKkL2VhHCGLWqZVCY"</span><span class="p">,</span><span class="w">
                                           </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w">
                                           </span><span class="n">language</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"en"</span><span class="p">)</span><span class="w">

    </span><span class="n">time_zone</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">google_timezone_api</span><span class="o">$</span><span class="n">timeZoneId</span><span class="w">

    </span><span class="n">local_time</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.POSIXlt</span><span class="p">(</span><span class="n">with_tz</span><span class="p">(</span><span class="n">data_row</span><span class="o">$</span><span class="n">time</span><span class="p">,</span><span class="w"> </span><span class="n">tz</span><span class="o">=</span><span class="n">paste</span><span class="p">(</span><span class="n">time_zone</span><span class="p">)))</span><span class="w">

    </span><span class="n">geodata_timezone</span><span class="o">$</span><span class="n">local_time</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">local_time</span><span class="p">)</span><span class="w">

    </span><span class="n">rm</span><span class="p">(</span><span class="n">data_row</span><span class="p">,</span><span class="w"> </span><span class="n">google_timezone_api</span><span class="p">,</span><span class="w"> </span><span class="n">time_zone</span><span class="p">)</span><span class="w">

  </span><span class="p">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="p">{</span><span class="w">

    </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="n">geodata</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="s2">"day_EDT"</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"09"</span><span class="p">){</span><span class="w">
      </span><span class="n">geodata_timezone</span><span class="o">$</span><span class="n">local_time</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">geodata_timezone</span><span class="o">$</span><span class="n">time_CDT</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="w">
    </span><span class="p">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="p">{</span><span class="w">
      </span><span class="n">geodata_timezone</span><span class="o">$</span><span class="n">local_time</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">geodata_timezone</span><span class="o">$</span><span class="n">time_EDT</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="w">
    </span><span class="p">}</span><span class="w">
  </span><span class="p">}</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>
<hr />

<p><br /></p>

<h3 id="plotting-elevation">Plotting elevation</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">theme_set</span><span class="p">(</span><span class="n">theme_bw</span><span class="p">())</span><span class="w"> </span><span class="c1"># Change the theme to my preference
</span><span class="w">
</span><span class="n">labels</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"9"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"9th, Chicago"</span><span class="p">,</span><span class="w"> </span><span class="s2">"10"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"10th, St. Joseph"</span><span class="p">,</span><span class="w"> </span><span class="s2">"11"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"11th, Ann Arbor"</span><span class="p">,</span><span class="w"> </span><span class="s2">"12"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"12th, Detroit"</span><span class="p">,</span><span class="w"> </span><span class="s2">"13"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"13th, Burlington"</span><span class="p">,</span><span class="w">
            </span><span class="s2">"14"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"14th, Burlington"</span><span class="p">,</span><span class="w"> </span><span class="s2">"15"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"15th, Toronto"</span><span class="p">,</span><span class="w"> </span><span class="s2">"16"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"16th, Niagara &amp; Ashtabula"</span><span class="p">,</span><span class="w"> </span><span class="s2">"17"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"17th, Pittsburgh"</span><span class="p">,</span><span class="w">
            </span><span class="s2">"18"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"18th, Pittsburgh"</span><span class="p">,</span><span class="w"> </span><span class="s2">"19"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"19th, Marietta"</span><span class="p">,</span><span class="w"> </span><span class="s2">"20"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"20th, Maysville"</span><span class="p">,</span><span class="w"> </span><span class="s2">"21"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"21st, Louisville"</span><span class="p">,</span><span class="w">
            </span><span class="s2">"22"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"22nd, Louisville"</span><span class="p">,</span><span class="w"> </span><span class="s2">"23"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"23rd, Paducah"</span><span class="p">)</span><span class="w">

</span><span class="n">ggplot</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.POSIXlt</span><span class="p">(</span><span class="n">local_time</span><span class="p">),</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ele</span><span class="p">),</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">geodata_timezone</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">facet_wrap</span><span class="p">(</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">day_EDT</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="o">=</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">scales</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"free_x"</span><span class="p">,</span><span class="w"> </span><span class="n">labeller</span><span class="o">=</span><span class="n">labeller</span><span class="p">(</span><span class="n">day_EDT</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">labels</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">"Local Time"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="o">=</span><span class="s2">"Elevation"</span><span class="p">,</span><span class="w"> </span><span class="n">title</span><span class="o">=</span><span class="s2">"Elevation per day and time"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="roadtrip2016_files/figure-markdown_github/unnamed-chunk-14-1.png" alt="" /></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">day_df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">unique</span><span class="p">(</span><span class="n">geodata_timezone</span><span class="o">$</span><span class="n">day_EDT</span><span class="p">),</span><span class="w">
                     </span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">unique</span><span class="p">(</span><span class="n">geodata_timezone</span><span class="o">$</span><span class="n">day_EDT</span><span class="p">),</span><span class="w">
                     </span><span class="n">xmin</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">geodata_timezone</span><span class="o">$</span><span class="n">day_EDT</span><span class="p">))),</span><span class="w">
                     </span><span class="n">xmax</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">geodata_timezone</span><span class="o">$</span><span class="n">day_EDT</span><span class="p">))),</span><span class="w">
                     </span><span class="n">ymin</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="o">-</span><span class="kc">Inf</span><span class="p">,</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">geodata_timezone</span><span class="o">$</span><span class="n">day_EDT</span><span class="p">))),</span><span class="w">
                     </span><span class="n">ymax</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="kc">Inf</span><span class="p">,</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">geodata_timezone</span><span class="o">$</span><span class="n">day_EDT</span><span class="p">))))</span><span class="w">

</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">day</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">unique</span><span class="p">(</span><span class="n">geodata_timezone</span><span class="o">$</span><span class="n">day_EDT</span><span class="p">)){</span><span class="w">
  </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="o">!</span><span class="n">day</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"23"</span><span class="p">){</span><span class="w">
    </span><span class="n">day_df</span><span class="p">[</span><span class="n">paste</span><span class="p">(</span><span class="n">day</span><span class="p">),</span><span class="w"> </span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">which</span><span class="p">(</span><span class="n">geodata_timezone</span><span class="o">$</span><span class="n">day_EDT</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">day</span><span class="p">)[</span><span class="m">1</span><span class="p">]</span><span class="w">
    </span><span class="n">day_df</span><span class="p">[</span><span class="n">paste</span><span class="p">(</span><span class="n">day</span><span class="p">),</span><span class="w"> </span><span class="m">3</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">which</span><span class="p">(</span><span class="n">geodata_timezone</span><span class="o">$</span><span class="n">day_EDT</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">day</span><span class="p">))</span><span class="m">+1</span><span class="p">)[</span><span class="m">1</span><span class="p">]</span><span class="m">-1</span><span class="w">
  </span><span class="p">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">day_df</span><span class="p">[</span><span class="n">paste</span><span class="p">(</span><span class="n">day</span><span class="p">),</span><span class="w"> </span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">which</span><span class="p">(</span><span class="n">geodata_timezone</span><span class="o">$</span><span class="n">day_EDT</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="n">day</span><span class="p">)[</span><span class="m">1</span><span class="p">]</span><span class="w">
    </span><span class="n">day_df</span><span class="p">[</span><span class="n">paste</span><span class="p">(</span><span class="n">day</span><span class="p">),</span><span class="w"> </span><span class="m">3</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">nrow</span><span class="p">(</span><span class="n">geodata_timezone</span><span class="p">)</span><span class="w">
  </span><span class="p">}</span><span class="w">
</span><span class="p">}</span><span class="w">

</span><span class="n">ggplot</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">seq_along</span><span class="p">(</span><span class="n">ele</span><span class="p">),</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ele</span><span class="p">),</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">geodata_timezone</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_rect</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">day_df</span><span class="p">,</span><span class="w"> </span><span class="n">inherit.aes</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w">
           </span><span class="n">aes</span><span class="p">(</span><span class="n">xmin</span><span class="o">=</span><span class="n">xmin</span><span class="p">,</span><span class="w"> </span><span class="n">xmax</span><span class="o">=</span><span class="n">xmax</span><span class="p">,</span><span class="w"> </span><span class="n">ymin</span><span class="o">=</span><span class="n">ymin</span><span class="p">,</span><span class="w"> </span><span class="n">ymax</span><span class="o">=</span><span class="n">ymax</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">group</span><span class="p">)),</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="o">=</span><span class="m">0.5</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="c1">#annotate("rect", xmin=0, xmax=0, ymin=-Inf, ymax=0, fill="red", alpha=0.5) +
</span><span class="w">  </span><span class="n">geom_path</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">"Index"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="o">=</span><span class="s2">"Elevation (line) and day (rectangles)"</span><span class="p">,</span><span class="w"> </span><span class="n">title</span><span class="o">=</span><span class="s2">"Elevation per day"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_y_continuous</span><span class="p">(</span><span class="n">limits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">min</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">geodata_timezone</span><span class="o">$</span><span class="n">ele</span><span class="p">,</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">geodata_timezone</span><span class="o">$</span><span class="n">day_EDT</span><span class="p">)))),</span><span class="w"> 
                                </span><span class="nf">max</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="n">geodata_timezone</span><span class="o">$</span><span class="n">ele</span><span class="p">,</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">geodata_timezone</span><span class="o">$</span><span class="n">day_EDT</span><span class="p">))))))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_discrete</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">"Days"</span><span class="p">,</span><span class="w"> </span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">labels</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="roadtrip2016_files/figure-markdown_github/unnamed-chunk-15-1.png" alt="" /></p>

<hr />

<p><br /></p>

<h2 id="extracting-speed-data-from-the-gpx-files">Extracting speed data from the gpx files</h2>

<p>Looking at the gpx file, there is also speed and course data. It is saved into the extensions tag. However, not all data points seem to have a speed tag.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">myfiles</span><span class="p">)){</span><span class="w">

  </span><span class="n">tryCatch</span><span class="p">({</span><span class="w">
    </span><span class="n">singleString</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="n">readLines</span><span class="p">(</span><span class="n">myfiles</span><span class="p">[</span><span class="n">i</span><span class="p">]),</span><span class="w"> </span><span class="n">collapse</span><span class="o">=</span><span class="s2">" "</span><span class="p">)</span><span class="w">
    </span><span class="n">pfile2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"&lt;/time&gt;&lt;extensions&gt;&lt;gpxtpx:TrackPointExtension&gt;&lt;gpxtpx:course&gt;"</span><span class="p">,</span><span class="w">
                   </span><span class="s2">"&lt;/time&gt;&lt;extensions&gt;&lt;gpxtpx:TrackPointExtension&gt;&lt;gpxtpx:speed&gt;0.00&lt;/gpxtpx:speed&gt;&lt;gpxtpx:course&gt;"</span><span class="p">,</span><span class="w"> </span><span class="n">singleString</span><span class="p">)</span><span class="w">

    </span><span class="n">pfile</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">htmlTreeParse</span><span class="p">(</span><span class="n">pfile2</span><span class="p">,</span><span class="w"> </span><span class="n">useInternalNodes</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">)</span><span class="w">

    </span><span class="c1"># Get all elevations, times and coordinates via the respective xpath
</span><span class="w">    </span><span class="n">times</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">xpathSApply</span><span class="p">(</span><span class="n">pfile</span><span class="p">,</span><span class="w"> </span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"//trkpt/time"</span><span class="p">,</span><span class="w"> </span><span class="n">xmlValue</span><span class="p">)</span><span class="w">
    </span><span class="n">coords</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">xpathSApply</span><span class="p">(</span><span class="n">pfile</span><span class="p">,</span><span class="w"> </span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"//trkpt"</span><span class="p">,</span><span class="w"> </span><span class="n">xmlAttrs</span><span class="p">)</span><span class="w">
    </span><span class="c1"># Extract latitude and longitude from the coordinates
</span><span class="w">    </span><span class="n">lats</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">coords</span><span class="p">[</span><span class="s2">"lat"</span><span class="p">,]))</span><span class="w">
    </span><span class="n">lons</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">coords</span><span class="p">[</span><span class="s2">"lon"</span><span class="p">,]))</span><span class="w">

    </span><span class="n">speed</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">xpathSApply</span><span class="p">(</span><span class="n">pfile</span><span class="p">,</span><span class="w"> </span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"//trkpt/extensions/*/speed"</span><span class="p">,</span><span class="w"> </span><span class="n">xmlValue</span><span class="p">)</span><span class="c1"># in m/s
</span><span class="w">    
    </span><span class="c1"># course gives the degrees where the car was headed at that point
</span><span class="w">    </span><span class="n">course</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">xpathSApply</span><span class="p">(</span><span class="n">pfile</span><span class="p">,</span><span class="w"> </span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"//trkpt/extensions/*/course"</span><span class="p">,</span><span class="w"> </span><span class="n">xmlValue</span><span class="p">)</span><span class="w">
    </span><span class="c1"># Put everything in a dataframe and get rid of old variables
</span><span class="w">    </span><span class="n">geodf</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">lat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lats</span><span class="p">,</span><span class="w"> </span><span class="n">lon</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lons</span><span class="p">,</span><span class="w"> </span><span class="n">time</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">times</span><span class="p">,</span><span class="w"> </span><span class="n">speed</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">speed</span><span class="p">,</span><span class="w"> </span><span class="n">course</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">course</span><span class="p">)</span><span class="w">

    </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="m">1</span><span class="p">){</span><span class="w">
      </span><span class="n">geodata_new</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">geodf</span><span class="w">
    </span><span class="p">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="p">{</span><span class="w">
      </span><span class="n">geodata_new</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rbind</span><span class="p">(</span><span class="n">geodata_new</span><span class="p">,</span><span class="w"> </span><span class="n">geodf</span><span class="p">)</span><span class="w">
    </span><span class="p">}</span><span class="w">

    </span><span class="n">rm</span><span class="p">(</span><span class="n">pfile</span><span class="p">,</span><span class="w"> </span><span class="n">pfile2</span><span class="p">,</span><span class="w"> </span><span class="n">singleString</span><span class="p">)</span><span class="w">
  </span><span class="p">},</span><span class="w"> </span><span class="n">error</span><span class="o">=</span><span class="k">function</span><span class="p">(</span><span class="n">e</span><span class="p">){</span><span class="n">cat</span><span class="p">(</span><span class="s2">"ERROR\n"</span><span class="p">)})</span><span class="w">

</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># Transforming the time column
</span><span class="n">geodata_new</span><span class="o">$</span><span class="n">time</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">strptime</span><span class="p">(</span><span class="n">geodata_new</span><span class="o">$</span><span class="n">time</span><span class="p">,</span><span class="w"> </span><span class="n">format</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"%Y-%m-%dT%H:%M:%SZ"</span><span class="p">)</span><span class="w">

</span><span class="c1"># ordering by date
</span><span class="n">geodata_new</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">arrange</span><span class="p">(</span><span class="n">geodata_new</span><span class="p">,</span><span class="w"> </span><span class="n">time</span><span class="p">)</span><span class="w">

</span><span class="c1"># removing all tracks from Germany
</span><span class="n">geodata_new_trip</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">geodata_new</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">geodata_new</span><span class="o">$</span><span class="n">lon</span><span class="p">))</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0</span><span class="p">),]</span><span class="w">

</span><span class="c1"># adding a column with day for separating the plotting by day lateron
</span><span class="n">geodata_new_trip</span><span class="o">$</span><span class="n">day</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.factor</span><span class="p">(</span><span class="n">format</span><span class="p">(</span><span class="n">as.Date</span><span class="p">(</span><span class="n">geodata_new_trip</span><span class="o">$</span><span class="n">time</span><span class="p">,</span><span class="n">format</span><span class="o">=</span><span class="s2">"%Y-%m-%d"</span><span class="p">),</span><span class="w"> </span><span class="s2">"%d"</span><span class="p">))</span><span class="w">

</span><span class="n">head</span><span class="p">(</span><span class="n">geodata_new_trip</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##            lat      lon                time speed course day
## 35664 41.98135 -87.8818 2016-09-09 19:23:53  0.00   0.00  09
## 35665 41.98135 -87.8818 2016-09-09 19:23:53  0.00   0.00  09
## 35666 41.98135 -87.8818 2016-09-09 19:23:53  0.00   0.00  09
## 35667 41.98135 -87.8818 2016-09-09 19:23:53  0.00   0.00  09
## 35668 41.98135 -87.8818 2016-09-09 19:23:53  0.00   0.00  09
## 35669 41.98135 -87.8818 2016-09-09 19:23:53  0.00   0.00  09
</code></pre>
</div>

<p><br /></p>

<p>Garmin shows speed as m/s, so I am converting it to km/h.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">yarrr</span><span class="p">)</span><span class="w">

</span><span class="n">geodata_new_trip</span><span class="o">$</span><span class="n">speed_kmh</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">geodata_new_trip</span><span class="o">$</span><span class="n">speed</span><span class="p">))</span><span class="w"> </span><span class="o">*</span><span class="m">3.6</span><span class="w">

</span><span class="n">speed_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">geodata_new_trip</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">geodata_new_trip</span><span class="o">$</span><span class="n">speed_kmh</span><span class="p">))</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0</span><span class="p">),]</span><span class="w">
</span><span class="n">speed_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">speed_data</span><span class="p">[</span><span class="o">!</span><span class="n">duplicated</span><span class="p">(</span><span class="n">speed_data</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="n">summary</span><span class="p">(</span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">speed_data</span><span class="o">$</span><span class="n">speed_kmh</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##   4.932  24.700  44.460  49.210  69.190 128.500
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">pirateplot</span><span class="p">(</span><span class="n">formula</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.numeric</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">speed_kmh</span><span class="p">))</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">as.factor</span><span class="p">(</span><span class="n">day</span><span class="p">),</span><span class="w">
           </span><span class="n">point.o</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">.1</span><span class="p">,</span><span class="w">
           </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">speed_data</span><span class="p">,</span><span class="w">
           </span><span class="n">ylab</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Speed (km/h)"</span><span class="p">,</span><span class="w">
           </span><span class="n">xlab</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Day in September"</span><span class="p">,</span><span class="w">
           </span><span class="n">main</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Speed per day"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="roadtrip2016_files/figure-markdown_github/unnamed-chunk-18-1.png" alt="" /></p>

<hr />

<p><br /></p>

<h2 id="photos">Photos</h2>

<p>And finally some photos from our trip. Their locations are marked on the map below:</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">myphotos</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">list.files</span><span class="p">(</span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"."</span><span class="p">,</span><span class="w"> </span><span class="n">full.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">myphotos</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">myphotos</span><span class="p">[</span><span class="n">grep</span><span class="p">(</span><span class="s2">"jpg"</span><span class="p">,</span><span class="w"> </span><span class="n">myphotos</span><span class="p">)]</span><span class="w"> 

</span><span class="n">library</span><span class="p">(</span><span class="n">exif</span><span class="p">)</span><span class="w">

</span><span class="n">photo_geotags</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"./"</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">myphotos</span><span class="p">),</span><span class="w"> </span><span class="n">lat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="nf">length</span><span class="p">(</span><span class="n">myphotos</span><span class="p">)),</span><span class="w"> </span><span class="n">lon</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">)</span><span class="w">

</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">file</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"./"</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">myphotos</span><span class="p">)){</span><span class="w">
  </span><span class="n">photodata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read_exif</span><span class="p">(</span><span class="n">file</span><span class="p">)</span><span class="w">
  </span><span class="n">lat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">photodata</span><span class="o">$</span><span class="n">latitude</span><span class="w">
  </span><span class="n">lon</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">photodata</span><span class="o">$</span><span class="n">longitude</span><span class="w">
  </span><span class="n">photo_geotags</span><span class="p">[</span><span class="n">file</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="o">$</span><span class="n">lat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">lat</span><span class="w">
  </span><span class="n">photo_geotags</span><span class="p">[</span><span class="n">file</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="o">$</span><span class="n">lon</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">lon</span><span class="w">
</span><span class="p">}</span><span class="w">

</span><span class="n">photo_geotags</span><span class="o">$</span><span class="n">text</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">seq</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="n">nrow</span><span class="p">(</span><span class="n">photo_geotags</span><span class="p">))</span><span class="w">

</span><span class="n">library</span><span class="p">(</span><span class="n">ggmap</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">ggrepel</span><span class="p">)</span><span class="w">

</span><span class="n">bc_bbox</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">make_bbox</span><span class="p">(</span><span class="n">lat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lat</span><span class="p">,</span><span class="w"> </span><span class="n">lon</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lon</span><span class="p">,</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">photo_geotags</span><span class="p">,</span><span class="w"> </span><span class="n">f</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">.10</span><span class="p">)</span><span class="w">
</span><span class="n">map</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">get_map</span><span class="p">(</span><span class="n">location</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">bc_bbox</span><span class="p">,</span><span class="w"> </span><span class="n">maptype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"roadmap"</span><span class="p">)</span><span class="w"> 

</span><span class="n">ggmap</span><span class="p">(</span><span class="n">map</span><span class="p">,</span><span class="w"> </span><span class="n">darken</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0.3</span><span class="p">,</span><span class="w"> </span><span class="s2">"white"</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lon</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">lat</span><span class="p">),</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">photo_geotags</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="o">=</span><span class="s2">"none"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">labs</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">"Longitude"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="o">=</span><span class="s2">"Latitude"</span><span class="p">,</span><span class="w"> </span><span class="n">title</span><span class="o">=</span><span class="s2">"Roadtrip 2016"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_text_repel</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">text</span><span class="p">),</span><span class="w"> </span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">photo_geotags</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>
<p><img src="roadtrip2016_files/figure-markdown_github/photos.png" alt="" /></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160910_103018.jpg" alt="1" />
<em>Photo number 1</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160910_145027.jpg" alt="2" />
<em>Photo number 2</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160910_150727.jpg" alt="3" />
<em>Photo number 3</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160910_184320.jpg" alt="4" />
<em>Photo number 4</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160911_082319.jpg" alt="5" />
<em>Photo number 5</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160911_083539.jpg" alt="6" />
<em>Photo number 6</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160912_092912.jpg" alt="7" />
<em>Photo number 7</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160912_092947.jpg" alt="8" />
<em>Photo number 8</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160912_113839.jpg" alt="9" />
<em>Photo number 9</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160912_164110.jpg" alt="10" />
<em>Photo number 10</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160913_133745.jpg" alt="11" />
<em>Photo number 11</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160913_135424.jpg" alt="12" />
<em>Photo number 12</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160915_113309.jpg" alt="13" />
<em>Photo number 13</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160915_121038.jpg" alt="14" />
<em>Photo number 14</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160915_160715.jpg" alt="15" />
<em>Photo number 15</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160916_113314.jpg" alt="16" />
<em>Photo number 16</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160916_143748.jpg" alt="17" />
<em>Photo number 17</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160917_133657.jpg" alt="18" />
<em>Photo number 18</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160918_102429.jpg" alt="19" />
<em>Photo number 19</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160918_105942.jpg" alt="20" />
<em>Photo number 20</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160919_135834.jpg" alt="21" />
<em>Photo number 21</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160919_135948_001.jpg" alt="22" />
<em>Photo number 22</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160919_180049.jpg" alt="23" />
<em>Photo number 23</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160919_180224.jpg" alt="24" />
<em>Photo number 24</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160920_113007.jpg" alt="25" />
<em>Photo number 25</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160920_133646.jpg" alt="26" />
<em>Photo number 26</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160920_133922.jpg" alt="27" />
<em>Photo number 27</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160921_123419.jpg" alt="28" />
<em>Photo number 28</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160921_123621.jpg" alt="29" />
<em>Photo number 29</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160921_192418.jpg" alt="30" />
<em>Photo number 30</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160921_192701.jpg" alt="31" />
<em>Photo number 31</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160922_192452.jpg" alt="32" />
<em>Photo number 32</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160922_215704.jpg" alt="33" />
<em>Photo number 33</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160924_142213.jpg" alt="34" />
<em>Photo number 34</em></p>

<p><img src="roadtrip2016_files/figure-markdown_github/20160926_140318.jpg" alt="35" />
<em>Photo number 35</em></p>

  </div>

  <h1><a href="/rna-seq/deseq2/teaching/2016/09/29/DESeq2-course">DESeq2 Course Work</a></h1>
  <p class="author">
    <span class="date">2016-09-29 00:00:00 +0200</span>
  </p>
  <div class="content">
    <hr />

<p><br /></p>

<p>The following workflow has been designed as teaching instructions for an introductory course to RNA-seq data analysis with DESeq2.</p>

<p>The course is designed for PhD students and will be given at the University of Münster from 10th to 21st of October 2016.</p>

<p>For questions or other comments, please <a href="http://shiring.github.io/about">contact me</a>.</p>

<hr />

<p><br /></p>

<p>Go to <a href="https://github.com/ShirinG/exprAnalysis/">exprAnalysis</a> or <a href="https://shiring.github.io/rna-seq/microarray/2016/09/28/exprAnalysis">this post</a> for installation instructions.</p>

<p>For all functions, use the help pages to find out more about parameters and usage.</p>

<p>See Vignette for additional information.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">exprAnalysis</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="input-data">Input data</h2>

<blockquote>
  <p>“As input, the DESeq2 package expects count data as obtained, e. g., from RNAseq or another high-throughput sequencing experiment, in the form of a matrix of integer values. The value in the i-th row and the j-th column of the matrix tells how many reads can be assigned to gene i in sample j.” Love et al., DESeq2 vignette</p>
</blockquote>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">data</span><span class="p">(</span><span class="s2">"countmatrix"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="count-data-analysis-with-deseq2">Count data analysis with DESeq2</h2>

<p>See <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.pdf">DESeq2 Vignette</a> for details.</p>

<ul>
  <li>read in saved count matrix</li>
  <li>define experimental design</li>
  <li>convert to DESeq data set</li>
</ul>

<h3 id="count-matrix-input">Count matrix input</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">design</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"(.*)(_[0-9])"</span><span class="p">,</span><span class="w"> </span><span class="s2">"\\1"</span><span class="p">,</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">countmatrix</span><span class="p">))</span><span class="w">
</span><span class="n">ExpDesign</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">row.names</span><span class="o">=</span><span class="n">colnames</span><span class="p">(</span><span class="n">countmatrix</span><span class="p">),</span><span class="w"> </span><span class="n">treatment</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">design</span><span class="p">)</span><span class="w">

</span><span class="n">data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">DESeq2</span><span class="o">::</span><span class="n">DESeqDataSetFromMatrix</span><span class="p">(</span><span class="n">countData</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">countmatrix</span><span class="p">,</span><span class="w"> </span><span class="n">colData</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ExpDesign</span><span class="p">,</span><span class="w"> </span><span class="n">design</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">~</span><span class="n">treatment</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## converting counts to integer mode
</code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="deseq2">DESeq2</h2>

<ul>
  <li>optional, but recommended: remove genes with zero counts over all samples</li>
  <li>run DESeq</li>
  <li>Extracting transformed values</li>
</ul>

<blockquote>
  <p>“While it is not necessary to pre-filter low count genes before running the DESeq2 functions, there are two reasons which make pre-filtering useful: by removing rows in which there are no reads or nearly no reads, we reduce the memory size of the dds data object and we increase the speed of the transformation and testing functions within DESeq2.” Love et al., DESeq2 vignette</p>
</blockquote>

<p>Note: the rlog transformation is provided for applications other than differential testing. For differential testing we recommend the DESeq function applied to raw counts, as described later in this workflow, which also takes into account the dependence of the variance of counts on the mean value during the dispersion estimation step.</p>

<p>For a quick first glance at the data, we can use <span style="color:red">pcaExplorer</span>.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data</span><span class="p">[</span><span class="n">rowSums</span><span class="p">(</span><span class="n">DESeq2</span><span class="o">::</span><span class="n">counts</span><span class="p">(</span><span class="n">data</span><span class="p">))</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="n">data_DESeq</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">DESeq2</span><span class="o">::</span><span class="n">DESeq</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## estimating size factors

## estimating dispersions

## gene-wise dispersion estimates

## mean-dispersion relationship

## final dispersion estimates

## fitting model and testing
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">expmatrix_DESeq</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">DESeq2</span><span class="o">::</span><span class="n">rlog</span><span class="p">(</span><span class="n">data_DESeq</span><span class="p">,</span><span class="w"> </span><span class="n">fitType</span><span class="o">=</span><span class="s2">"local"</span><span class="p">)</span><span class="w">
</span><span class="n">expmatrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">SummarizedExperiment</span><span class="o">::</span><span class="n">assay</span><span class="p">(</span><span class="n">expmatrix_DESeq</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="s2">"pcaExplorer"</span><span class="p">)</span><span class="w">
</span><span class="n">pcaExplorer</span><span class="p">(</span><span class="n">data_DESeq</span><span class="p">,</span><span class="w"> </span><span class="n">expmatrix_DESeq</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="dispersion-plot">Dispersion plot</h2>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">DESeq2</span><span class="o">::</span><span class="n">plotDispEsts</span><span class="p">(</span><span class="n">data_DESeq</span><span class="p">,</span><span class="w"> </span><span class="n">main</span><span class="o">=</span><span class="s2">"Dispersion Estimates"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="DESeq2_course_blog_files/figure-markdown_github/dispersion_plot-1.png?raw=true" alt="" /></p>

<hr />

<p><br /></p>

<h2 id="exploratory-analysis-of-all-genes">Exploratory analysis of all genes</h2>

<h3 id="variance-vs-mean-gene-expression-across-samples">Variance vs mean gene expression across samples</h3>

<p>Plots variance against mean gene expression across samples and calculates the correlation of a linear regression model.</p>

<p><strong>var_vs_mean()</strong> uses the R package matrixStats.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">var_vs_mean</span><span class="p">(</span><span class="n">countmatrix</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="DESeq2_course_blog_files/figure-markdown_github/var_vs_mean-1.png" alt="" /></p>

<div class="highlighter-rouge"><pre class="highlight"><code>## 
##  Pearson's product-moment correlation
## 
## data:  log2(dispersion[, 1] + 1) and log2(dispersion[, 2] + 1)
## t = 281.2, df = 9998, p-value &lt; 2.2e-16
## alternative hypothesis: true correlation is not equal to 0
## 95 percent confidence interval:
##  0.9399669 0.9443684
## sample estimates:
##       cor 
## 0.9422083
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">var_vs_mean</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="DESeq2_course_blog_files/figure-markdown_github/var_vs_mean-2.png" alt="" /></p>

<div class="highlighter-rouge"><pre class="highlight"><code>## 
##  Pearson's product-moment correlation
## 
## data:  log2(dispersion[, 1] + 1) and log2(dispersion[, 2] + 1)
## t = 12.133, df = 9998, p-value &lt; 2.2e-16
## alternative hypothesis: true correlation is not equal to 0
## 95 percent confidence interval:
##  0.1010951 0.1397267
## sample estimates:
##       cor 
## 0.1204565
</code></pre>
</div>

<hr />

<p><br /></p>

<h3 id="intersample-variances">Intersample variances</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">corrgram</span><span class="p">)</span><span class="w">

</span><span class="n">Ctrl_cor</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">expmatrix</span><span class="p">[,</span><span class="n">grep</span><span class="p">(</span><span class="s2">"Ctrl"</span><span class="p">,</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">))]</span><span class="w">

</span><span class="n">corrgram</span><span class="o">::</span><span class="n">corrgram</span><span class="p">(</span><span class="n">Ctrl_cor</span><span class="p">,</span><span class="w"> </span><span class="n">order</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">lower.panel</span><span class="o">=</span><span class="n">corrgram</span><span class="o">::</span><span class="n">panel.pie</span><span class="p">,</span><span class="w">
         </span><span class="n">upper.panel</span><span class="o">=</span><span class="n">corrgram</span><span class="o">::</span><span class="n">panel.pts</span><span class="p">,</span><span class="w"> </span><span class="n">text.panel</span><span class="o">=</span><span class="n">corrgram</span><span class="o">::</span><span class="n">panel.txt</span><span class="p">,</span><span class="w">
         </span><span class="n">main</span><span class="o">=</span><span class="s2">"Correlogram of controls"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="DESeq2_course_blog_files/figure-markdown_github/variance_sumOverlaps-1.png" alt="" /></p>

<h3 id="repeat-for-other-treatment-groups">Repeat for other treatment groups</h3>

<hr />

<p><br /></p>

<h2 id="principle-component-analysis">Principle Component Analysis</h2>

<p>Uses functions from the R package pcaGoPromoter.</p>

<p>You can only plot the principle components using:</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">groups</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.factor</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="s2">"Ctrl"</span><span class="p">,</span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="s2">"TolLPS"</span><span class="p">,</span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="s2">"TolS100A8"</span><span class="p">,</span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="s2">"ActLPS"</span><span class="p">,</span><span class="m">4</span><span class="p">)))</span><span class="w">
</span><span class="n">pca_plot</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">,</span><span class="w"> </span><span class="n">groups</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="DESeq2_course_blog_files/figure-markdown_github/pca_plot-1.png" alt="" /></p>

<p><br /></p>

<p>Or you can plot the principle components and calculate TF and GO term enrichments of genes (defaults to top 2.5%) with highest and lowest loadings. With this function, the ouput files are directly saved to .pdf and .txt (by default to working directory).</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">pca_plot_enrich</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">,</span><span class="w"> </span><span class="n">groups</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="heatmaps">Heatmaps</h2>

<p><strong>heatmaps()</strong> uses the R package gplots.</p>

<p>Here, of the 30 most highly expressed genes.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">select</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">order</span><span class="p">(</span><span class="n">rowMeans</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">),</span><span class="n">decreasing</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">30</span><span class="p">]</span><span class="w">
</span><span class="n">heatmaps</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">[</span><span class="n">select</span><span class="p">,],</span><span class="w"> </span><span class="n">samplecols</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"#E41A1C"</span><span class="p">,</span><span class="w"> </span><span class="s2">"#377EB8"</span><span class="p">,</span><span class="w"> </span><span class="s2">"#4DAF4A"</span><span class="p">,</span><span class="w"> </span><span class="s2">"#984EA3"</span><span class="p">),</span><span class="w"> </span><span class="n">each</span><span class="o">=</span><span class="m">4</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<p><img src="DESeq2_course_blog_files/figure-markdown_github/heatmaps-1.png" alt="" /></p>

<h3 id="heatmap-function-from-deseq2-using-pheatmap">Heatmap function from DESeq2, using pheatmap:</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">pheatmap</span><span class="p">)</span><span class="w">

</span><span class="n">sampleDists</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">dist</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">))</span><span class="w">
</span><span class="n">sampleDistMatrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.matrix</span><span class="p">(</span><span class="n">sampleDists</span><span class="p">)</span><span class="w">
</span><span class="n">rownames</span><span class="p">(</span><span class="n">sampleDistMatrix</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="n">expmatrix_DESeq</span><span class="o">$</span><span class="n">treatment</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">sampleDistMatrix</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kc">NULL</span><span class="w">
</span><span class="n">colors</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">grDevices</span><span class="o">::</span><span class="n">colorRampPalette</span><span class="p">(</span><span class="w"> </span><span class="n">rev</span><span class="p">(</span><span class="n">RColorBrewer</span><span class="o">::</span><span class="n">brewer.pal</span><span class="p">(</span><span class="m">9</span><span class="p">,</span><span class="w"> </span><span class="s2">"Blues"</span><span class="p">))</span><span class="w"> </span><span class="p">)(</span><span class="m">255</span><span class="p">)</span><span class="w">
</span><span class="n">pheatmap</span><span class="o">::</span><span class="n">pheatmap</span><span class="p">(</span><span class="n">sampleDistMatrix</span><span class="p">,</span><span class="w">
         </span><span class="n">clustering_distance_rows</span><span class="o">=</span><span class="n">sampleDists</span><span class="p">,</span><span class="w">
         </span><span class="n">clustering_distance_cols</span><span class="o">=</span><span class="n">sampleDists</span><span class="p">,</span><span class="w">
         </span><span class="n">col</span><span class="o">=</span><span class="n">colors</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="DESeq2_course_blog_files/figure-markdown_github/heatmaps2-1.png" alt="" /></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">treatment</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">SummarizedExperiment</span><span class="o">::</span><span class="n">colData</span><span class="p">(</span><span class="n">data_DESeq</span><span class="p">)[,</span><span class="nf">c</span><span class="p">(</span><span class="s2">"treatment"</span><span class="p">)],</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">SummarizedExperiment</span><span class="o">::</span><span class="n">colData</span><span class="p">(</span><span class="n">data_DESeq</span><span class="p">)))</span><span class="w">
</span><span class="n">pheatmap</span><span class="o">::</span><span class="n">pheatmap</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">[</span><span class="n">select</span><span class="p">,],</span><span class="w"> </span><span class="n">cluster_rows</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">show_rownames</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">cluster_cols</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">annotation_col</span><span class="o">=</span><span class="n">df</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="DESeq2_course_blog_files/figure-markdown_github/heatmaps2-2.png" alt="" /></p>

<hr />

<p><br /></p>

<h2 id="hierarchical-clustering-and-outlier-detection">Hierarchical Clustering and outlier detection</h2>

<p>Uses adjacency matrix function from the R package WGCNA and hierarchical clustering from the R package flashClust.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">datTraits</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">Ctrl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">12</span><span class="p">)),</span><span class="w"> </span><span class="n">TolPS</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">8</span><span class="p">)),</span><span class="w"> 
                        </span><span class="n">TolS100A8</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">8</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">)),</span><span class="w"> </span><span class="n">ActLPS</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">12</span><span class="p">),</span><span class="nf">rep</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">)),</span><span class="w"> 
                        </span><span class="n">Tol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">8</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">)),</span><span class="w"> 
                        </span><span class="n">ExPhenotype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">stats</span><span class="o">::</span><span class="n">rnorm</span><span class="p">(</span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="n">stats</span><span class="o">::</span><span class="n">rnorm</span><span class="p">(</span><span class="m">8</span><span class="p">,</span><span class="w"> </span><span class="m">25</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="n">stats</span><span class="o">::</span><span class="n">rnorm</span><span class="p">(</span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="m">50</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">)),</span><span class="w"> 
                        </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">))</span><span class="w">

</span><span class="n">datExpr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">wgcna_sample_dendrogram</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">,</span><span class="w"> </span><span class="n">datTraits</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## 
</code></pre>
</div>

<p><img src="DESeq2_course_blog_files/figure-markdown_github/wgcna_sample_dendrogram-1.png" alt="" width="700px" /></p>

<div class="highlighter-rouge"><pre class="highlight"><code>##  Flagging genes and samples with too many missing values...
##   ..step 1
## All genes are okay!
## All samples are okay!
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># Optional: Remove outlier samples and repeats: All genes flagged for removal are saved to the object "remove_genes"
#head(remove_genes)
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="differential-expression-analysis-using-deseq2">Differential expression analysis using DESeq2</h2>

<p>For raw read count data.</p>

<h3 id="contrast-de-groups">contrast DE groups:</h3>

<ul>
  <li>lfc = treatment &gt; Ctrl, - lfc = treatment &lt; Ctrl p-value &amp; p.adjust values of NA indicate outliers detected by Cook’s distance NA only for p.adjust means the gene is filtered by automatic independent filtering for having a low mean normalized count</li>
</ul>

<p>Information about which variables and tests were used can be found by calling the function <span style="color:red">mcols</span>, on the results object.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">DESeq2</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">ggrepel</span><span class="p">)</span><span class="w">

</span><span class="c1"># find possible contrasts with
</span><span class="n">DESeq2</span><span class="o">::</span><span class="n">resultsNames</span><span class="p">(</span><span class="n">data_DESeq</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "Intercept"          "treatmentActLPS"    "treatmentCtrl"     
## [4] "treatmentTolLPS"    "treatmentTolS100A8"
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">res</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">DESeq2</span><span class="o">::</span><span class="n">results</span><span class="p">(</span><span class="n">data_DESeq</span><span class="p">,</span><span class="w"> </span><span class="n">contrast</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="s2">"treatmentActLPS"</span><span class="p">,</span><span class="w"> </span><span class="s2">"treatmentCtrl"</span><span class="p">),</span><span class="w"> </span><span class="n">cooksCutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.99</span><span class="p">,</span><span class="w"> </span><span class="n">independentFiltering</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span><span class="n">pAdjustMethod</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"BH"</span><span class="p">)</span><span class="w">
</span><span class="n">summary</span><span class="p">(</span><span class="n">res</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## 
## out of 10000 with nonzero total read count
## adjusted p-value &lt; 0.05
## LFC &gt; 0 (up)     : 2608, 26% 
## LFC &lt; 0 (down)   : 2407, 24% 
## outliers [1]     : 175, 1.8% 
## low counts [2]   : 0, 0% 
## (mean count &lt; 5)
## [1] see 'cooksCutoff' argument of ?results
## [2] see 'independentFiltering' argument of ?results
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">mcols</span><span class="p">(</span><span class="n">res</span><span class="p">)</span><span class="o">$</span><span class="n">description</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "mean of normalized counts for all samples"               
## [2] "log2 fold change (MAP): treatmentActLPS vs treatmentCtrl"
## [3] "standard error: treatmentActLPS vs treatmentCtrl"        
## [4] "Wald statistic: treatmentActLPS vs treatmentCtrl"        
## [5] "Wald test p-value: treatmentActLPS vs treatmentCtrl"     
## [6] "BH adjusted p-values"
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># order results table by the smallest adjusted p value:
</span><span class="n">res</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">res</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">res</span><span class="o">$</span><span class="n">padj</span><span class="p">),]</span><span class="w">

</span><span class="n">results</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">dplyr</span><span class="o">::</span><span class="n">mutate</span><span class="p">(</span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">res</span><span class="p">),</span><span class="w"> </span><span class="n">sig</span><span class="o">=</span><span class="n">ifelse</span><span class="p">(</span><span class="n">res</span><span class="o">$</span><span class="n">padj</span><span class="o">&lt;</span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span><span class="s2">"FDR&lt;0.05"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Not Sig"</span><span class="p">)),</span><span class="w"> </span><span class="n">row.names</span><span class="o">=</span><span class="n">rownames</span><span class="p">(</span><span class="n">res</span><span class="p">))</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">results</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##               baseMean log2FoldChange      lfcSE     stat        pvalue
## CATSPER3     2469.0350       4.094351 0.08958149 45.70532  0.000000e+00
## GDA          5578.3392       4.200987 0.13525887 31.05886 8.660849e-212
## CCDC64B       544.6121       5.878983 0.19906885 29.53241 1.104831e-191
## GRB2         1751.2594       2.506512 0.09224234 27.17312 1.350120e-162
## LOC100507387  716.7431       9.537221 0.36844495 25.88506 9.810056e-148
## TMEM102      1346.0139       4.244962 0.16534946 25.67267 2.360945e-145
##                       padj      sig
## CATSPER3      0.000000e+00 FDR&lt;0.05
## GDA          4.254642e-208 FDR&lt;0.05
## CCDC64B      3.618323e-188 FDR&lt;0.05
## GRB2         3.316233e-159 FDR&lt;0.05
## LOC100507387 1.927676e-144 FDR&lt;0.05
## TMEM102      3.866048e-142 FDR&lt;0.05
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">DEgenes_DESeq</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">results</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="nf">abs</span><span class="p">(</span><span class="n">results</span><span class="o">$</span><span class="n">log2FoldChange</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">log2</span><span class="p">(</span><span class="m">1.5</span><span class="p">)</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">results</span><span class="o">$</span><span class="n">padj</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0.05</span><span class="p">),]</span><span class="w">

</span><span class="n">p</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ggplot2</span><span class="o">::</span><span class="n">ggplot</span><span class="p">(</span><span class="n">results</span><span class="p">,</span><span class="w"> </span><span class="n">ggplot2</span><span class="o">::</span><span class="n">aes</span><span class="p">(</span><span class="n">log2FoldChange</span><span class="p">,</span><span class="w"> </span><span class="o">-</span><span class="n">log10</span><span class="p">(</span><span class="n">pvalue</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">ggplot2</span><span class="o">::</span><span class="n">geom_point</span><span class="p">(</span><span class="n">ggplot2</span><span class="o">::</span><span class="n">aes</span><span class="p">(</span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sig</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">ggplot2</span><span class="o">::</span><span class="n">scale_color_manual</span><span class="p">(</span><span class="n">values</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"red"</span><span class="p">,</span><span class="w"> </span><span class="s2">"black"</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">ggplot2</span><span class="o">::</span><span class="n">ggtitle</span><span class="p">(</span><span class="s2">"Volcano Plot of DESeq2 analysis"</span><span class="p">)</span><span class="w">

</span><span class="n">p</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">ggrepel</span><span class="o">::</span><span class="n">geom_text_repel</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">results</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="p">],</span><span class="w"> </span><span class="n">ggplot2</span><span class="o">::</span><span class="n">aes</span><span class="p">(</span><span class="n">label</span><span class="o">=</span><span class="n">rownames</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="p">])))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Warning: Removed 175 rows containing missing values (geom_point).
</code></pre>
</div>

<p><img src="DESeq2_course_blog_files/figure-markdown_github/DEgenes_DESeq2-1.png" alt="" /></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># If there aren't too many DE genes:
#p + geom_text_repel(data = dplyr::filter(results, padj&lt;0.05), aes(label = rownames(results[1:10, ])))
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h3 id="ma-plot">MA-plot</h3>

<blockquote>
  <p>“These plots show the log2 fold changes from the treatment over the mean of normalized counts, i.e. the average of counts normalized by size factors. The left plot shows the “unshrunken” log2 fold changes, while the right plot, produced by the code above, shows the shrinkage of log2 fold changes resulting from the incorporation of zero-centered normal prior. The shrinkage is greater for the log2 fold change estimates from genes with low counts and high dispersion, as can be seen by the narrowing of spread of leftmost points in the right plot.” Love et al., DESeq2 vignette</p>
</blockquote>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">DESeq2</span><span class="o">::</span><span class="n">plotMA</span><span class="p">(</span><span class="n">res</span><span class="p">,</span><span class="w"> </span><span class="n">main</span><span class="o">=</span><span class="s2">"MA Plot"</span><span class="p">,</span><span class="w"> </span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">-2</span><span class="p">,</span><span class="m">2</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<p><img src="DESeq2_course_blog_files/figure-markdown_github/ma_plot-1.png" alt="" /></p>

<p><br /></p>

<h3 id="plotcounts">plotCounts</h3>

<blockquote>
  <p>“It can also be useful to examine the counts of reads for a single gene across the groups. A simple function for making this plot is plotCounts, which normalizes counts by sequencing depth and adds a pseudocount of 1/2 to allow for log scale plotting. The counts are grouped by the variables in intgroup, where more than one variable can be specified.” Love et al., DESeq2 vignette</p>
</blockquote>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">par</span><span class="p">(</span><span class="n">mfrow</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">3</span><span class="p">))</span><span class="w">

</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">){</span><span class="w">
  </span><span class="n">gene</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">res</span><span class="p">)[</span><span class="n">i</span><span class="p">]</span><span class="w">
  </span><span class="n">main</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gene</span><span class="w">
  </span><span class="n">DESeq2</span><span class="o">::</span><span class="n">plotCounts</span><span class="p">(</span><span class="n">data_DESeq</span><span class="p">,</span><span class="w"> </span><span class="n">gene</span><span class="o">=</span><span class="n">gene</span><span class="p">,</span><span class="w"> </span><span class="n">intgroup</span><span class="o">=</span><span class="s2">"treatment"</span><span class="p">,</span><span class="w"> </span><span class="n">main</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">main</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<p><img src="DESeq2_course_blog_files/figure-markdown_github/plot_counts-1.png" alt="" width="700px" /></p>

<hr />

<p><br /></p>

<h2 id="gene-annotations">Gene annotations</h2>

<p>Can be used to add e.g. ENTREZ ID, ENSEMBL ID, etc. to gene name.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">results_anno</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">geneAnnotations</span><span class="p">(</span><span class="n">input</span><span class="o">=</span><span class="n">results</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="o">=</span><span class="n">row.names</span><span class="p">(</span><span class="n">results</span><span class="p">),</span><span class="w"> </span><span class="n">column</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"ENTREZID"</span><span class="p">,</span><span class="w"> </span><span class="s2">"ENSEMBL"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="o">=</span><span class="s2">"SYMBOL"</span><span class="p">,</span><span class="w"> </span><span class="n">organism</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"human"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## 

## 'select()' returned 1:many mapping between keys and columns
## 'select()' returned 1:many mapping between keys and columns
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">head</span><span class="p">(</span><span class="n">results_anno</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##               baseMean log2FoldChange      lfcSE     stat        pvalue
## CATSPER3     2469.0350       4.094351 0.08958149 45.70532  0.000000e+00
## GDA          5578.3392       4.200987 0.13525887 31.05886 8.660849e-212
## CCDC64B       544.6121       5.878983 0.19906885 29.53241 1.104831e-191
## GRB2         1751.2594       2.506512 0.09224234 27.17312 1.350120e-162
## LOC100507387  716.7431       9.537221 0.36844495 25.88506 9.810056e-148
## TMEM102      1346.0139       4.244962 0.16534946 25.67267 2.360945e-145
##                       padj      sig  ENTREZID         ENSEMBL
## CATSPER3      0.000000e+00 FDR&lt;0.05    347732 ENSG00000152705
## GDA          4.254642e-208 FDR&lt;0.05      9615 ENSG00000119125
## CCDC64B      3.618323e-188 FDR&lt;0.05    146439 ENSG00000162069
## GRB2         3.316233e-159 FDR&lt;0.05      2885 ENSG00000177885
## LOC100507387 1.927676e-144 FDR&lt;0.05 100507387 ENSG00000182230
## TMEM102      3.866048e-142 FDR&lt;0.05    284114 ENSG00000181284
</code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="enrichment-analysis-using-clusterpofiler">Enrichment Analysis using clusterPofiler</h2>

<p>See <a href="https://bioconductor.org/packages/release/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html">clusterProfiler</a> instructions for details.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">clusterProfiler</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Loading required package: DOSE
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Loading required package: AnnotationDbi
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">OrgDb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">org.Hs.eg.db</span><span class="w"> </span><span class="c1"># can also be other organisms
</span><span class="w">
</span><span class="n">geneList</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.vector</span><span class="p">(</span><span class="n">results_anno</span><span class="o">$</span><span class="n">log2FoldChange</span><span class="p">)</span><span class="w">
</span><span class="nf">names</span><span class="p">(</span><span class="n">geneList</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">results_anno</span><span class="o">$</span><span class="n">ENTREZID</span><span class="w">
</span><span class="n">gene</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">na.omit</span><span class="p">(</span><span class="n">results_anno</span><span class="o">$</span><span class="n">ENTREZID</span><span class="p">)</span><span class="w">


</span><span class="c1"># Group GO
</span><span class="n">ggo</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">clusterProfiler</span><span class="o">::</span><span class="n">groupGO</span><span class="p">(</span><span class="n">gene</span><span class="w">     </span><span class="o">=</span><span class="w"> </span><span class="n">gene</span><span class="p">,</span><span class="w">
                                </span><span class="n">OrgDb</span><span class="w">    </span><span class="o">=</span><span class="w"> </span><span class="n">OrgDb</span><span class="p">,</span><span class="w">
                                </span><span class="n">ont</span><span class="w">      </span><span class="o">=</span><span class="w"> </span><span class="s2">"BP"</span><span class="p">,</span><span class="w">
                                </span><span class="n">level</span><span class="w">    </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w">
                                </span><span class="n">readable</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">summary</span><span class="p">(</span><span class="n">ggo</span><span class="p">)[,</span><span class="m">-5</span><span class="p">])</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##                    ID                                    Description Count
## GO:0003006 GO:0003006 developmental process involved in reproduction   261
## GO:0019953 GO:0019953                            sexual reproduction   303
## GO:0019954 GO:0019954                           asexual reproduction     0
## GO:0022414 GO:0022414                           reproductive process   547
## GO:0032504 GO:0032504            multicellular organism reproduction   317
## GO:0032505 GO:0032505       reproduction of a single-celled organism     0
##            GeneRatio
## GO:0003006  261/9735
## GO:0019953  303/9735
## GO:0019954    0/9735
## GO:0022414  547/9735
## GO:0032504  317/9735
## GO:0032505    0/9735
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">barplot</span><span class="p">(</span><span class="n">ggo</span><span class="p">,</span><span class="w"> </span><span class="n">drop</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="o">=</span><span class="m">12</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="DESeq2_course_blog_files/figure-markdown_github/clusterProfiler-1.png" alt="" width="700px" /></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># GO over-representation test
</span><span class="n">ego</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">clusterProfiler</span><span class="o">::</span><span class="n">enrichGO</span><span class="p">(</span><span class="n">gene</span><span class="w">          </span><span class="o">=</span><span class="w"> </span><span class="n">gene</span><span class="p">,</span><span class="w">
                                 </span><span class="n">OrgDb</span><span class="w">         </span><span class="o">=</span><span class="w"> </span><span class="n">OrgDb</span><span class="p">,</span><span class="w">
                                 </span><span class="n">ont</span><span class="w">           </span><span class="o">=</span><span class="w"> </span><span class="s2">"BP"</span><span class="p">,</span><span class="w">
                                 </span><span class="n">pAdjustMethod</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"BH"</span><span class="p">,</span><span class="w">
                                 </span><span class="n">pvalueCutoff</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w">
                                 </span><span class="n">qvalueCutoff</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w"> 
                                 </span><span class="n">readable</span><span class="w">      </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">summary</span><span class="p">(</span><span class="n">ego</span><span class="p">)[,</span><span class="m">-8</span><span class="p">])</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##                    ID                                        Description
## GO:0072657 GO:0072657                   protein localization to membrane
## GO:0006893 GO:0006893                 Golgi to plasma membrane transport
## GO:0007051 GO:0007051                               spindle organization
## GO:0044772 GO:0044772                mitotic cell cycle phase transition
## GO:0060999 GO:0060999 positive regulation of dendritic spine development
## GO:0048193 GO:0048193                            Golgi vesicle transport
##            GeneRatio   BgRatio       pvalue   p.adjust     qvalue Count
## GO:0072657  260/7259 487/16655 6.475400e-06 0.02201028 0.02035249   260
## GO:0006893   36/7259  48/16655 9.648441e-06 0.02201028 0.02035249    36
## GO:0007051   81/7259 130/16655 1.246381e-05 0.02201028 0.02035249    81
## GO:0044772  253/7259 477/16655 1.597263e-05 0.02201028 0.02035249   253
## GO:0060999   27/7259  34/16655 2.158731e-05 0.02379785 0.02200542    27
## GO:0048193  179/7259 329/16655 4.371899e-05 0.03540965 0.03274263   179
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">barplot</span><span class="p">(</span><span class="n">ego</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="o">=</span><span class="m">25</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="DESeq2_course_blog_files/figure-markdown_github/clusterProfilerGO-1.png" alt="" width="700px" /></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">clusterProfiler</span><span class="o">::</span><span class="n">dotplot</span><span class="p">(</span><span class="n">ego</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="o">=</span><span class="m">25</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="DESeq2_course_blog_files/figure-markdown_github/clusterProfilerGO-2.png" alt="" width="700px" /></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1">#clusterProfiler::plotGOgraph(ego)
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1">## KEGG over-representation test
</span><span class="n">kk</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">clusterProfiler</span><span class="o">::</span><span class="n">enrichKEGG</span><span class="p">(</span><span class="n">gene</span><span class="w">         </span><span class="o">=</span><span class="w"> </span><span class="n">gene</span><span class="p">,</span><span class="w">
                 </span><span class="n">organism</span><span class="w">     </span><span class="o">=</span><span class="w"> </span><span class="s1">'hsa'</span><span class="p">,</span><span class="w">
                 </span><span class="n">pAdjustMethod</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"BH"</span><span class="p">,</span><span class="w">
                 </span><span class="n">pvalueCutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w">
                 </span><span class="n">qvalueCutoff</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">)</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">summary</span><span class="p">(</span><span class="n">kk</span><span class="p">)[,</span><span class="m">-8</span><span class="p">])</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##                ID                  Description GeneRatio  BgRatio
## hsa04144 hsa04144                  Endocytosis  142/3071 260/7119
## hsa04210 hsa04210                    Apoptosis   80/3071 140/7119
## hsa05169 hsa05169 Epstein-Barr virus infection  111/3071 204/7119
## hsa05134 hsa05134                Legionellosis   36/3071  55/7119
##                pvalue   p.adjust     qvalue Count
## hsa04144 9.858484e-05 0.02858960 0.02345281   142
## hsa04210 5.292257e-04 0.04885909 0.04008042    80
## hsa05169 6.639562e-04 0.04885909 0.04008042   111
## hsa05134 6.739185e-04 0.04885909 0.04008042    36
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">cnetplot</span><span class="p">(</span><span class="n">kk</span><span class="p">,</span><span class="w"> </span><span class="n">categorySize</span><span class="o">=</span><span class="s2">"geneNum"</span><span class="p">,</span><span class="w"> </span><span class="n">foldChange</span><span class="o">=</span><span class="n">geneList</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="DESeq2_course_blog_files/figure-markdown_github/clusterProfilerKEGG-1.png" alt="" width="700px" /></p>

  </div>

  <h1><a href="/rna-seq/microarray/2016/09/28/exprAnalysis">exprAnalysis package</a></h1>
  <p class="author">
    <span class="date">2016-09-28 00:00:00 +0200</span>
  </p>
  <div class="content">
    <p>I created the R package <strong>exprAnalysis</strong> designed to streamline my RNA-seq data analysis pipeline. Below you find the vignette for installation and usage of the package.</p>

<hr />

<p><br /></p>

<p>This package combines functions from various packages used to analyze and visualize expression data from NGS or expression chips.</p>

<ul>
  <li>It supports normalized input as e.g. from Cufflinks or expression chip arrays and raw count data from bam file input.</li>
  <li>It supports mRNA, miRNA, protein or other expression data.</li>
</ul>

<p>So far, it is implemented for human and mouse data only!</p>

<p>It uses functions from the following packages:</p>

<ul>
  <li>AnnotationDbi for annotating gene information</li>
  <li>beadarray for importing Illumina expression chip files from GenomeStudio</li>
  <li>clusterProfiler and DOSE for functional enrichment analysis</li>
  <li>DESeq2 for differential expression analysis of raw count data</li>
  <li>GenomicAlignments, GenomicFeatures, Rsamtools for reading bam files</li>
  <li>pcaGoPromoter for principle component analysis</li>
  <li>limma for differential expression analysis of normalised expression data</li>
  <li>pathview for mapping KEGG pathways</li>
  <li>gplots for heatmaps</li>
  <li>sva for batch correction</li>
  <li>WGCNA for coregulatory network determination</li>
</ul>

<hr />

<p><br /></p>

<h1 id="prepare-session">Prepare session</h1>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">exprAnalysis</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># update all packages
</span><span class="n">source.bioc</span><span class="o">=</span><span class="s2">"https://bioconductor.org/biocLite.R"</span><span class="w">
</span><span class="n">source</span><span class="p">(</span><span class="n">source.bioc</span><span class="p">)</span><span class="w">
</span><span class="n">biocLite</span><span class="p">(</span><span class="s2">"BiocUpgrade"</span><span class="p">)</span><span class="w">
</span><span class="n">biocLite</span><span class="p">(</span><span class="n">ask</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="w"> 

</span><span class="n">update.packages</span><span class="p">()</span><span class="w">

</span><span class="c1"># make sure the workspace is in pristine condition
</span><span class="n">rm</span><span class="p">(</span><span class="n">list</span><span class="o">=</span><span class="n">ls</span><span class="p">(</span><span class="n">all</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">))</span><span class="w">

</span><span class="n">orig_par</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">par</span><span class="p">(</span><span class="n">no.readonly</span><span class="o">=</span><span class="nb">T</span><span class="p">)</span><span class="w">

</span><span class="n">options</span><span class="p">(</span><span class="n">stringsAsFactors</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="the-package-can-be-installed-via-github">The package can be installed via Github:</h2>

<p>Beware that the vignette is rather large and thus takes a minute to compile. You can also just use this page (it is identical to the vignette).</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># install package from github
</span><span class="n">install.packages</span><span class="p">(</span><span class="s2">"devtools"</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">devtools</span><span class="p">)</span><span class="w">

</span><span class="c1"># either the latest stable release that passed TRAVIS CI check
</span><span class="n">devtools</span><span class="o">::</span><span class="n">install_github</span><span class="p">(</span><span class="s2">"ShirinG/exprAnalysis"</span><span class="p">,</span><span class="w"> </span><span class="n">build_vignettes</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">ref</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"stable.version0.1.0"</span><span class="p">)</span><span class="w">

</span><span class="c1"># or the development version
</span><span class="n">devtools</span><span class="o">::</span><span class="n">install_github</span><span class="p">(</span><span class="s2">"ShirinG/exprAnalysis"</span><span class="p">,</span><span class="w"> </span><span class="n">build_vignettes</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">ref</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"master"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>There might be problems with installation of some dependency packages (especially Bioconductor packages and WGCNA and its dependencies from CRAN). In order to install them manually:</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">list.of.packages_bioconductor</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"arrayQualityMetrics"</span><span class="p">,</span><span class="w"> </span><span class="s2">"beadarray"</span><span class="p">,</span><span class="w"> </span><span class="s2">"pcaGoPromoter"</span><span class="p">,</span><span class="w"> </span><span class="s2">"limma"</span><span class="p">,</span><span class="w"> </span><span class="s2">"pathview"</span><span class="p">,</span><span class="w"> </span><span class="s2">"sva"</span><span class="p">,</span><span class="w"> </span><span class="s2">"GO.db"</span><span class="p">,</span><span class="w"> </span><span class="s2">"impute"</span><span class="p">)</span><span class="w">
</span><span class="n">list.of.packages_cran</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"WGCNA"</span><span class="p">,</span><span class="w"> </span><span class="s2">"roxygen2"</span><span class="p">,</span><span class="w"> </span><span class="s2">"testthat"</span><span class="p">,</span><span class="w"> </span><span class="s2">"gplots"</span><span class="p">)</span><span class="w">

</span><span class="n">new.packages_bioconductor</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">list.of.packages_bioconductor</span><span class="p">[</span><span class="o">!</span><span class="p">(</span><span class="n">list.of.packages_bioconductor</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">installed.packages</span><span class="p">()[,</span><span class="s2">"Package"</span><span class="p">])]</span><span class="w">
</span><span class="n">new.packages_cran</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">list.of.packages_cran</span><span class="p">[</span><span class="o">!</span><span class="p">(</span><span class="n">list.of.packages_cran</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">installed.packages</span><span class="p">()[,</span><span class="s2">"Package"</span><span class="p">])]</span><span class="w">

</span><span class="c1"># CRAN
</span><span class="k">if</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">new.packages_cran</span><span class="p">)</span><span class="o">&gt;</span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="n">install.packages</span><span class="p">(</span><span class="n">new.packages_cran</span><span class="p">)</span><span class="w">

</span><span class="c1"># Bioconductor
</span><span class="k">if</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">new.packages_bioconductor</span><span class="p">)</span><span class="o">&gt;</span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
  </span><span class="n">source</span><span class="p">(</span><span class="s2">"https://bioconductor.org/biocLite.R"</span><span class="p">)</span><span class="w">
  </span><span class="n">biocLite</span><span class="p">(</span><span class="n">new.packages_bioconductor</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>
<p>Load the library:</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="s2">"exprAnalysis"</span><span class="p">)</span><span class="w">

</span><span class="c1"># To view the vignette if you built it with the package:
</span><span class="w">
</span><span class="n">vignette</span><span class="p">(</span><span class="s2">"exprAnalysis"</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="o">=</span><span class="s2">"exprAnalysis"</span><span class="p">)</span><span class="w">
</span><span class="n">vignette</span><span class="p">(</span><span class="s2">"CummeRbund"</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="o">=</span><span class="s2">"exprAnalysis"</span><span class="p">)</span><span class="w">

</span><span class="n">browseVignettes</span><span class="p">(</span><span class="s2">"exprAnalysis"</span><span class="p">)</span><span class="w">

</span><span class="n">enableWGCNAThreads</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h1 id="functions">Functions</h1>

<p>See package help (?functionname, ?data) for detailed descriptions of functions and example datasets.</p>

<hr />

<p><br /></p>

<h1 id="input-data">Input data</h1>

<p>Takes an expression data matrix containing numeric values as expression measures (e.g. read count data, FPKM values, Illumina expression data).</p>

<ul>
  <li>Rownames should be gene or isoform identifiers (e.g. gene names)</li>
  <li>Colnames should be sample IDs (sample names)</li>
</ul>

<p>This package contains two example matrices of randomly generated expression data as raw read counts (called <strong>“countmatrix”</strong>) and as normalized read counts (called <strong>“expmatrix”</strong>).</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">data</span><span class="p">(</span><span class="s2">"countmatrix"</span><span class="p">)</span><span class="w">
</span><span class="n">data</span><span class="p">(</span><span class="s2">"expmatrix"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="starting-from-fpkm-data">Starting from FPKM data</h2>

<p>If you have FPKM data (e.g. quantile normalized) from Cufflinks, treat the data such as the <strong>expmatrix</strong> example.</p>

<p><br /></p>

<h3 id="cuffdiff">Cuffdiff</h3>

<p>See CummeRbund vignette included among the vignettes of this package <strong>CummeRbundVignette.html</strong>.</p>

<hr />

<p><br /></p>

<h2 id="starting-from-count-data">Starting from count data</h2>

<p>If you want to do count data analysis, you can either produce a count matrix (e.g. with HTSeq) and proceed to DESeq2 analysis or you can produce the count table directly from the bam files and then proceed to DESeq2.</p>

<p><strong>read_bam_to_countmatrix()</strong> returns a DESeq data set that can directly be used with the DEseq2 pipeline. Or you can manually load the count matrix that was saved by <strong>read_bam_to_countmatrix()</strong> and then go to DESeqDataFrameFromMatrix().</p>

<p>In <strong>read_bam_to_countmatrix()</strong> fragments will be counted only once to each gene, even if they overlap multiple exons of a gene which may themselves be overlapping.</p>

<p><strong>read_bam_to_countmatrix()</strong> uses the packages Rsamtools, GenomicAlignments, GenomicFeatures, Biobase and SummarizedExperiment.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># Locate alignment files
</span><span class="n">dir</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">getwd</span><span class="p">()</span><span class="w">
</span><span class="n">filenames</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">fileslist</span><span class="p">[</span><span class="n">grep</span><span class="p">(</span><span class="s2">".*sorted.bam$"</span><span class="p">,</span><span class="w"> </span><span class="n">list.files</span><span class="p">(</span><span class="n">dir</span><span class="p">))]</span><span class="w">

</span><span class="c1"># Create a sample table
</span><span class="w">
</span><span class="n">samplename</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sub</span><span class="p">(</span><span class="s2">"_accepted_hits.sorted.bam"</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">filenames</span><span class="p">)</span><span class="w">
</span><span class="n">design</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Treatment"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Control"</span><span class="p">)</span><span class="w">
</span><span class="n">sampleid</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sub</span><span class="p">(</span><span class="s2">"Sample"</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">samplename</span><span class="p">)</span><span class="w">

</span><span class="n">sampleTable</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">samplename</span><span class="p">,</span><span class="w"> </span><span class="n">sampleid</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">sampleid</span><span class="p">,</span><span class="w"> </span><span class="n">filenames</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">filenames</span><span class="p">,</span><span class="w"> </span><span class="n">colData</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">design</span><span class="p">,</span><span class="w"> </span><span class="n">stringsAsFactors</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">)</span><span class="w">

</span><span class="c1">#         sampleid                         filenames    colData
#Sample1         1  Sample1_accepted_hits.sorted.bam  Treatment
#Sample2         2  Sample2_accepted_hits.sorted.bam    Control
</span><span class="w">
</span><span class="n">data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read_bam_to_countmatrix</span><span class="p">(</span><span class="n">sampleTable</span><span class="p">,</span><span class="w"> </span><span class="n">gtffile</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Homo_sapiens.GRCh38.83.gtf"</span><span class="p">,</span><span class="w"> </span><span class="n">projectfolder</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">getwd</span><span class="p">(),</span><span class="w"> </span><span class="n">outPrefix</span><span class="o">=</span><span class="s2">"Test"</span><span class="p">)</span><span class="w">

</span><span class="n">countmatrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">SummarizedExperiment</span><span class="o">::</span><span class="n">assay</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<h3 id="continue-count-data-analysis-with-deseq2">Continue count data analysis with DESeq2</h3>

<p>See DESeq2 Vignette for additional details.</p>

<h4 id="if-you-do-not-directly-proceed-from-read_bam_to_countmatrix">If you do not directly proceed from read_bam_to_countmatrix():</h4>

<ul>
  <li>read in saved count matrix</li>
  <li>define experimental design</li>
  <li>convert to DESeq data set</li>
</ul>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">countmatrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w">   </span><span class="n">read.table</span><span class="p">(</span><span class="n">file.path</span><span class="p">(</span><span class="n">projectfolder</span><span class="p">,</span><span class="w"> </span><span class="s2">"Countdata"</span><span class="p">,</span><span class="w"> </span><span class="n">outPrefix</span><span class="p">,</span><span class="w"> </span><span class="n">paste0</span><span class="p">(</span><span class="n">outPrefix</span><span class="p">,</span><span class="w"> </span><span class="s2">"_Summarize_overlaps_count_matrix.txt"</span><span class="p">)),</span><span class="w"> </span><span class="n">header</span><span class="o">=</span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">design</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"(.*)(_[0-9])"</span><span class="p">,</span><span class="w"> </span><span class="s2">"\\1"</span><span class="p">,</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">countmatrix</span><span class="p">))</span><span class="w">
</span><span class="n">ExpDesign</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">row.names</span><span class="o">=</span><span class="n">colnames</span><span class="p">(</span><span class="n">countmatrix</span><span class="p">),</span><span class="w"> </span><span class="n">treatment</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">design</span><span class="p">)</span><span class="w">

</span><span class="n">data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">DESeq2</span><span class="o">::</span><span class="n">DESeqDataSetFromMatrix</span><span class="p">(</span><span class="n">countData</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">countmatrix</span><span class="p">,</span><span class="w"> </span><span class="n">colData</span><span class="o">=</span><span class="n">ExpDesign</span><span class="p">,</span><span class="w"> </span><span class="n">design</span><span class="o">=~</span><span class="n">treatment</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<h3 id="deseq2">DESeq2</h3>

<ul>
  <li>optional, but recommended: remove genes with zero counts over all samples</li>
  <li>run DESeq</li>
  <li>Extracting transformed values</li>
</ul>

<p>Note: the rlog transformation is provided for applications other than differential testing. For differential testing we recommend the DESeq function applied to raw counts, as described later in this workflow, which also takes into account the dependence of the variance of counts on the mean value during the dispersion estimation step.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data</span><span class="p">[</span><span class="n">rowSums</span><span class="p">(</span><span class="n">DESeq2</span><span class="o">::</span><span class="n">counts</span><span class="p">(</span><span class="n">data</span><span class="p">))</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="n">data_DESeq</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">DESeq2</span><span class="o">::</span><span class="n">DESeq</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="w">

</span><span class="n">expmatrix_DESeq</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">DESeq2</span><span class="o">::</span><span class="n">rlog</span><span class="p">(</span><span class="n">data_DESeq</span><span class="p">,</span><span class="w"> </span><span class="n">fitType</span><span class="o">=</span><span class="s2">"local"</span><span class="p">)</span><span class="w">
</span><span class="n">expmatrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">SummarizedExperiment</span><span class="o">::</span><span class="n">assay</span><span class="p">(</span><span class="n">expmatrix_DESeq</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<h4 id="dispersion-plot">Dispersion plot</h4>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">DESeq2</span><span class="o">::</span><span class="n">plotDispEsts</span><span class="p">(</span><span class="n">data_DESeq</span><span class="p">,</span><span class="w"> </span><span class="n">main</span><span class="o">=</span><span class="s2">"Dispersion Estimates"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="starting-from-affymetrix-expression-chips">Starting from Affymetrix expression chips</h2>

<p>If you have CEL files, start with the following code to produce the expression matrix and then treat it like the <strong>expmatrix</strong> example:</p>

<ul>
  <li>Read in the data and create an expression, using RMA for example</li>
</ul>

<p>Currently the rma function implements RMA in the following manner 1. Probe specific correction of the PM probes using a model based on observed intensity being the sum of signal and noise 2. Normalization of corrected PM probes using quantile normalization (Bolstad et al., 2003) 3. Calculation of Expression measure using median polish.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">affy</span><span class="p">)</span><span class="w">

</span><span class="c1"># Choose directory containing CEL files
</span><span class="n">dir</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">getwd</span><span class="p">()</span><span class="w">

</span><span class="c1"># check, that you have the correct directory
</span><span class="n">celfiles</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">file.path</span><span class="p">(</span><span class="n">dir</span><span class="p">,</span><span class="w"> </span><span class="n">list.files</span><span class="p">(</span><span class="n">dir</span><span class="p">,</span><span class="w"> </span><span class="n">recursive</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)[</span><span class="n">grep</span><span class="p">(</span><span class="s2">".*CEL$"</span><span class="p">,</span><span class="w"> </span><span class="n">list.files</span><span class="p">(</span><span class="n">dir</span><span class="p">,</span><span class="w"> </span><span class="n">recursive</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="n">ignore.case</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)])</span><span class="w">

</span><span class="n">affydata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ReadAffy</span><span class="p">(</span><span class="n">filenames</span><span class="o">=</span><span class="n">celfiles</span><span class="p">)</span><span class="w">
</span><span class="n">affy</span><span class="o">::</span><span class="n">MAplot</span><span class="p">(</span><span class="n">affydata</span><span class="p">,</span><span class="n">plot.method</span><span class="o">=</span><span class="s2">"smoothScatter"</span><span class="p">)</span><span class="w">
</span><span class="n">eset</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">affy</span><span class="o">::</span><span class="n">rma</span><span class="p">(</span><span class="n">affydata</span><span class="p">)</span><span class="w">

</span><span class="c1"># If affy doesn' recognize the chip type, try the oligo package:
</span><span class="n">library</span><span class="p">(</span><span class="n">oligo</span><span class="p">)</span><span class="w">
</span><span class="n">rawData</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.celfiles</span><span class="p">(</span><span class="n">celfiles</span><span class="p">)</span><span class="w">
</span><span class="n">oligo</span><span class="o">::</span><span class="n">MAplot</span><span class="p">(</span><span class="n">rawData</span><span class="p">,</span><span class="w"> </span><span class="n">plotFun</span><span class="o">=</span><span class="n">smoothScatter</span><span class="p">)</span><span class="w">
</span><span class="n">eset</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">backgroundCorrect</span><span class="p">(</span><span class="n">rawData</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="o">=</span><span class="s2">"rma"</span><span class="p">)</span><span class="w">

</span><span class="n">quality_control_plots</span><span class="p">(</span><span class="n">eset</span><span class="p">,</span><span class="w"> </span><span class="n">groupColumn</span><span class="o">=</span><span class="nf">c</span><span class="p">())</span><span class="w">

</span><span class="c1"># Optional: Remove bad quality probes/ genes and/ or samples
</span><span class="w">
</span><span class="n">write.exprs</span><span class="p">(</span><span class="n">eset</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="o">=</span><span class="s2">"eset.txt"</span><span class="p">)</span><span class="w">

</span><span class="n">library</span><span class="p">(</span><span class="n">made4</span><span class="p">)</span><span class="w">
</span><span class="n">overview</span><span class="p">(</span><span class="n">eset</span><span class="p">)</span><span class="w">

</span><span class="c1"># RNA degradation plots
</span><span class="n">deg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">AffyRNAdeg</span><span class="p">(</span><span class="n">affydata</span><span class="p">)</span><span class="w">
</span><span class="n">plotAffyRNAdeg</span><span class="p">(</span><span class="n">deg</span><span class="p">)</span><span class="w">


</span><span class="n">expmatrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">exprs</span><span class="p">(</span><span class="n">eset</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="starting-from-illumina-expression-chips">Starting from Illumina expression chips</h2>

<p>If you have Illumina intensity data, load it into GenomeStudio first:</p>

<ul>
  <li>Open New Project -&gt; Gene Expression</li>
  <li>Choose Assay Type -&gt; Next</li>
  <li>Choose Project Repo and name the project -&gt; Next</li>
  <li>On the left side, mark all folders containing data, select “All” and import them into the right window -&gt; Next</li>
  <li>Name Group Set, on the left side, mark all folders containing data, select “All” and import them into the right window (optionally, you can directly create groups here) -&gt; Next</li>
  <li>For now, don’t use normalisation and don’ substract background</li>
  <li>Choose appropriate Content Descriptor (e.g. HumanHT-12_V4_0_R1_15002873_B.bgx) -&gt; Finish</li>
  <li>Export SampleProbeProfile.txt: Choose columns AVG_Signal, Detection Pval, BEAD_STDERR and Avg_NBEADS</li>
  <li>Export ControlProbeProfile.txt: Choose columns AVG_Signal and Detection Pval</li>
  <li>You also need the Samplesheet.csv</li>
</ul>

<p><strong>read_Illumina()</strong> and <strong>normalise_eset()</strong> uses the R package beadarray.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">dataFile</span><span class="w">      </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">file.path</span><span class="p">(</span><span class="s2">"SampleProbeProfile.txt"</span><span class="p">)</span><span class="w">
</span><span class="n">qcFile</span><span class="w">        </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">file.path</span><span class="p">(</span><span class="s2">"ControlProbeProfile.txt"</span><span class="p">)</span><span class="w">
</span><span class="n">sampleSheet</span><span class="w">   </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">file.path</span><span class="p">(</span><span class="s2">"samplesheet.csv"</span><span class="p">)</span><span class="w">

</span><span class="c1"># define Illumina Expression Array: "HumanHT-12 v?", "MouseWG-6 v?", "MouseRef-8 v?"
</span><span class="n">expressionchipType</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"HumanHT-12 v4"</span><span class="w">

</span><span class="n">eset</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read_Illumina</span><span class="p">(</span><span class="n">dataFile</span><span class="p">,</span><span class="w"> </span><span class="n">qcFile</span><span class="p">,</span><span class="w"> </span><span class="n">sampleSheet</span><span class="p">,</span><span class="w"> </span><span class="n">expressionchipType</span><span class="p">,</span><span class="w"> </span><span class="n">ProbeID</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"PROBE_ID"</span><span class="p">,</span><span class="w"> </span><span class="n">skip</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">controlID</span><span class="o">=</span><span class="s2">"ProbeID"</span><span class="p">,</span><span class="w"> </span><span class="n">qc.skip</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">method_norm</span><span class="w"> </span><span class="o">=</span><span class="s2">"none"</span><span class="p">,</span><span class="w"> </span><span class="n">transform</span><span class="o">=</span><span class="s2">"log2"</span><span class="p">)</span><span class="w">

</span><span class="n">quality_control_plots</span><span class="p">(</span><span class="n">eset</span><span class="p">)</span><span class="w">

</span><span class="c1"># Optional: Remove bad quality probes/ genes and/ or samples
</span><span class="w">
</span><span class="n">eset</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">normalise_eset</span><span class="p">(</span><span class="n">eset</span><span class="p">)</span><span class="w">

</span><span class="n">expmatrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">exprs</span><span class="p">(</span><span class="n">eset</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h1 id="batch-correction">Batch correction</h1>

<p><strong>batch_removal()</strong> uses the R package sva.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">pheno</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">sample</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">16</span><span class="p">),</span><span class="w"> </span><span class="n">treatment</span><span class="o">=</span><span class="n">sub</span><span class="p">(</span><span class="s2">"(.*)(_[0-9])"</span><span class="p">,</span><span class="w"> </span><span class="s2">"\\1"</span><span class="p">,</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">)),</span><span class="w"> </span><span class="n">batch</span><span class="o">=</span><span class="n">ifelse</span><span class="p">(</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"Ctrl"</span><span class="p">,</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">))</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="s2">"1"</span><span class="p">,</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"ActLPS"</span><span class="p">,</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">))</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="s2">"1"</span><span class="p">,</span><span class="w"> </span><span class="s2">"2"</span><span class="p">)),</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">))</span><span class="w">

</span><span class="n">expmatrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">batch_removal</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">,</span><span class="w"> </span><span class="n">pheno</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h1 id="exploratory-analysis-of-all-genes">Exploratory analysis of all genes</h1>

<h2 id="variance-vs-mean-gene-expression-across-samples">Variance vs mean gene expression across samples</h2>

<p>Plots variance against mean gene expression across samples and calculates the correlation of a linear regression model.</p>

<p><strong>var_vs_mean()</strong> uses the R package matrixStats.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">var_vs_mean</span><span class="p">(</span><span class="n">countmatrix</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="exprAnalysis_files/figure-markdown_github/var_vs_mean-1.png" alt="" /></p>

<div class="highlighter-rouge"><pre class="highlight"><code>## 
##  Pearson's product-moment correlation
## 
## data:  log2(dispersion[, 1] + 1) and log2(dispersion[, 2] + 1)
## t = 281.2, df = 9998, p-value &lt; 2.2e-16
## alternative hypothesis: true correlation is not equal to 0
## 95 percent confidence interval:
##  0.9399669 0.9443684
## sample estimates:
##       cor 
## 0.9422083
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># var_vs_mean(expmatrix)
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="intersample-variances">Intersample variances</h2>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">corrgram</span><span class="p">)</span><span class="w">

</span><span class="n">Ctrl_cor</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">expmatrix</span><span class="p">[,</span><span class="n">grep</span><span class="p">(</span><span class="s2">"Ctrl"</span><span class="p">,</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">))]</span><span class="w">

</span><span class="n">corrgram</span><span class="o">::</span><span class="n">corrgram</span><span class="p">(</span><span class="n">Ctrl_cor</span><span class="p">,</span><span class="w"> </span><span class="n">order</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">lower.panel</span><span class="o">=</span><span class="n">corrgram</span><span class="o">::</span><span class="n">panel.pie</span><span class="p">,</span><span class="w">
         </span><span class="n">upper.panel</span><span class="o">=</span><span class="n">corrgram</span><span class="o">::</span><span class="n">panel.pts</span><span class="p">,</span><span class="w"> </span><span class="n">text.panel</span><span class="o">=</span><span class="n">corrgram</span><span class="o">::</span><span class="n">panel.txt</span><span class="p">,</span><span class="w">
         </span><span class="n">main</span><span class="o">=</span><span class="s2">"Correlogram of controls"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="principle-component-analysis">Principle Component Analysis</h2>

<p>Uses functions from the R package pcaGoPromoter.</p>

<p>You can only plot the principle components using:</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">groups</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.factor</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="s2">"Ctrl"</span><span class="p">,</span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="s2">"TolLPS"</span><span class="p">,</span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="s2">"TolS100A8"</span><span class="p">,</span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="s2">"ActLPS"</span><span class="p">,</span><span class="m">4</span><span class="p">)))</span><span class="w">

</span><span class="n">pca_plot</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">,</span><span class="w"> </span><span class="n">groups</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="exprAnalysis_files/figure-markdown_github/pca_plot-1.png" alt="" /></p>

<p><br /></p>

<p>Or you can plot the principle components and calculate TF and GO term enrichments of genes (defaults to top 2.5%) with highest and lowest loadings. With this function, the ouput files are directly saved to .pdf and .txt (by default to working directory).</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">pca_plot_enrich</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">,</span><span class="w"> </span><span class="n">groups</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="heatmaps">Heatmaps</h2>

<p><strong>heatmaps()</strong> uses the R package gplots.</p>

<p>Here, of the 30 most highly expressed genes.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">select</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">order</span><span class="p">(</span><span class="n">rowMeans</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">),</span><span class="n">decreasing</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">30</span><span class="p">]</span><span class="w">

</span><span class="n">heatmaps</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">[</span><span class="n">select</span><span class="p">,],</span><span class="w"> </span><span class="n">samplecols</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"#E41A1C"</span><span class="p">,</span><span class="w"> </span><span class="s2">"#377EB8"</span><span class="p">,</span><span class="w"> </span><span class="s2">"#4DAF4A"</span><span class="p">,</span><span class="w"> </span><span class="s2">"#984EA3"</span><span class="p">),</span><span class="w"> </span><span class="n">each</span><span class="o">=</span><span class="m">4</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<p><img src="exprAnalysis_files/figure-markdown_github/heatmaps-1.png" alt="" /></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># Heatmap function from DESeq2, using pheatmap:
</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">pheatmap</span><span class="p">)</span><span class="w">

</span><span class="n">sampleDists</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">dist</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">))</span><span class="w">
</span><span class="n">sampleDistMatrix</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.matrix</span><span class="p">(</span><span class="n">sampleDists</span><span class="p">)</span><span class="w">
</span><span class="n">rownames</span><span class="p">(</span><span class="n">sampleDistMatrix</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="n">expmatrix_DESeq</span><span class="o">$</span><span class="n">treatment</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">sampleDistMatrix</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kc">NULL</span><span class="w">
</span><span class="n">colors</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">grDevices</span><span class="o">::</span><span class="n">colorRampPalette</span><span class="p">(</span><span class="w"> </span><span class="n">rev</span><span class="p">(</span><span class="n">RColorBrewer</span><span class="o">::</span><span class="n">brewer.pal</span><span class="p">(</span><span class="m">9</span><span class="p">,</span><span class="w"> </span><span class="s2">"Blues"</span><span class="p">))</span><span class="w"> </span><span class="p">)(</span><span class="m">255</span><span class="p">)</span><span class="w">
</span><span class="n">pheatmap</span><span class="o">::</span><span class="n">pheatmap</span><span class="p">(</span><span class="n">sampleDistMatrix</span><span class="p">,</span><span class="w">
         </span><span class="n">clustering_distance_rows</span><span class="o">=</span><span class="n">sampleDists</span><span class="p">,</span><span class="w">
         </span><span class="n">clustering_distance_cols</span><span class="o">=</span><span class="n">sampleDists</span><span class="p">,</span><span class="w">
         </span><span class="n">col</span><span class="o">=</span><span class="n">colors</span><span class="p">)</span><span class="w">

</span><span class="n">df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">treatment</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">SummarizedExperiment</span><span class="o">::</span><span class="n">colData</span><span class="p">(</span><span class="n">data_DESeq</span><span class="p">)[,</span><span class="nf">c</span><span class="p">(</span><span class="s2">"treatment"</span><span class="p">)],</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">SummarizedExperiment</span><span class="o">::</span><span class="n">colData</span><span class="p">(</span><span class="n">data_DESeq</span><span class="p">)))</span><span class="w">
</span><span class="n">pheatmap</span><span class="o">::</span><span class="n">pheatmap</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">[</span><span class="n">select</span><span class="p">,],</span><span class="w"> </span><span class="n">cluster_rows</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">show_rownames</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">cluster_cols</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">annotation_col</span><span class="o">=</span><span class="n">df</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># Interactive heatmap
</span><span class="w">
</span><span class="c1">#devtools::install_github('talgalili/heatmaply')
</span><span class="n">library</span><span class="p">(</span><span class="n">heatmaply</span><span class="p">)</span><span class="w">

</span><span class="c1"># sample correlation
</span><span class="n">hm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">heatmapr</span><span class="p">(</span><span class="n">cor</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">[</span><span class="n">select</span><span class="p">,]),</span><span class="w"> </span><span class="n">k_row</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="n">k_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">4</span><span class="p">)</span><span class="w">

</span><span class="c1"># gene correlation
</span><span class="n">hm</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">heatmapr</span><span class="p">(</span><span class="n">cor</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">[</span><span class="n">select</span><span class="p">,])),</span><span class="w"> </span><span class="n">k_row</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">k_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">)</span><span class="w">

</span><span class="n">heatmaply</span><span class="p">(</span><span class="n">hm</span><span class="p">)</span><span class="w">

</span><span class="c1">#Hover over heatmap to see individual values and sample/ gene IDs
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="wgcna">WGCNA</h2>

<p>Check <a href="https://labs.genetics.ucla.edu/horvath/CoexpressionNetwork/">WGCNA page</a> for detailed description of the WGCNA package.</p>

<h3 id="hierarchical-clustering-and-outlier-detection">Hierarchical Clustering and outlier detection</h3>

<p>Uses adjacency matrix function from the R package WGCNA and hierarchical clustering from the R package flashClust.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">datTraits</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">Ctrl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">12</span><span class="p">)),</span><span class="w"> </span><span class="n">TolPS</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">8</span><span class="p">)),</span><span class="w"> </span><span class="n">TolS100A8</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">8</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">)),</span><span class="w"> </span><span class="n">ActLPS</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">12</span><span class="p">),</span><span class="nf">rep</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">)),</span><span class="w"> </span><span class="n">Tol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">8</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">)),</span><span class="w"> </span><span class="n">ExPhenotype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">stats</span><span class="o">::</span><span class="n">rnorm</span><span class="p">(</span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="n">stats</span><span class="o">::</span><span class="n">rnorm</span><span class="p">(</span><span class="m">8</span><span class="p">,</span><span class="w"> </span><span class="m">25</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="n">stats</span><span class="o">::</span><span class="n">rnorm</span><span class="p">(</span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="m">50</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">)),</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">))</span><span class="w">

</span><span class="n">datExpr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">wgcna_sample_dendrogram</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">,</span><span class="w"> </span><span class="n">datTraits</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## 
</code></pre>
</div>

<p><img src="exprAnalysis_files/figure-markdown_github/wgcna_sample_dendrogram-1.png" alt="" /></p>

<div class="highlighter-rouge"><pre class="highlight"><code>##  Flagging genes and samples with too many missing values...
##   ..step 1
## All genes are okay!
## All samples are okay!
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># Optional: Remove outlier samples and repeats: All genes flagged for removal are saved to the object "remove_genes"
#head(remove_genes)
</span></code></pre>
</div>

<p><br /></p>

<h3 id="choosing-a-soft-threshold">Choosing a Soft Threshold</h3>

<p>Ideally, pick a SFT with R^2 &gt; 0.9. In this example, this threshold was not reached, so I pick the highest SFT: 30.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># Choose a set of soft thresholding powers
</span><span class="n">powers</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">30</span><span class="p">)</span><span class="w">

</span><span class="c1"># choose power based on SFT criterion
</span><span class="n">sft</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pickSoftThreshold</span><span class="p">(</span><span class="n">datExpr</span><span class="p">,</span><span class="w"> </span><span class="n">powerVector</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">powers</span><span class="p">,</span><span class="w"> </span><span class="n">verbose</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">5</span><span class="p">,</span><span class="w"> </span><span class="n">networkType</span><span class="w"> </span><span class="o">=</span><span class="s2">"signed"</span><span class="p">,</span><span class="w"> </span><span class="n">corFnc</span><span class="o">=</span><span class="w"> </span><span class="s2">"bicor"</span><span class="p">,</span><span class="w"> </span><span class="n">corOptions</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">maxPOutliers</span><span class="o">=</span><span class="m">0.1</span><span class="p">))</span><span class="w">

</span><span class="c1"># Plot the results:
</span><span class="n">tiff</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="w"> </span><span class="s2">"SFT.tiff"</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">8000</span><span class="w"> </span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">4000</span><span class="p">,</span><span class="w"> </span><span class="n">res</span><span class="o">=</span><span class="m">600</span><span class="p">,</span><span class="w"> </span><span class="n">compression</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"lzw"</span><span class="p">)</span><span class="w">
</span><span class="n">par</span><span class="p">(</span><span class="n">mfrow</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">))</span><span class="w">
</span><span class="c1"># SFT index as a function of different powers
</span><span class="n">plot</span><span class="p">(</span><span class="n">sft</span><span class="o">$</span><span class="n">fitIndices</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="o">-</span><span class="nf">sign</span><span class="p">(</span><span class="n">sft</span><span class="o">$</span><span class="n">fitIndices</span><span class="p">[,</span><span class="m">3</span><span class="p">])</span><span class="o">*</span><span class="n">sft</span><span class="o">$</span><span class="n">fitIndices</span><span class="p">[,</span><span class="m">2</span><span class="p">],</span><span class="w">
     </span><span class="n">xlab</span><span class="o">=</span><span class="s2">"Soft Threshold (power)"</span><span class="p">,</span><span class="n">ylab</span><span class="o">=</span><span class="s2">"SFT, signed R^2"</span><span class="p">,</span><span class="n">type</span><span class="o">=</span><span class="s2">"n"</span><span class="p">,</span><span class="n">main</span><span class="o">=</span><span class="n">paste</span><span class="p">(</span><span class="s2">"Scale independence"</span><span class="p">))</span><span class="w">
</span><span class="n">text</span><span class="p">(</span><span class="n">sft</span><span class="o">$</span><span class="n">fitIndices</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="o">-</span><span class="nf">sign</span><span class="p">(</span><span class="n">sft</span><span class="o">$</span><span class="n">fitIndices</span><span class="p">[,</span><span class="m">3</span><span class="p">])</span><span class="o">*</span><span class="n">sft</span><span class="o">$</span><span class="n">fitIndices</span><span class="p">[,</span><span class="m">2</span><span class="p">],</span><span class="w">
     </span><span class="n">labels</span><span class="o">=</span><span class="n">powers</span><span class="p">,</span><span class="n">col</span><span class="o">=</span><span class="s2">"red"</span><span class="p">)</span><span class="w">
</span><span class="c1"># this line corresponds to using an R^2 cut-off of h
</span><span class="n">abline</span><span class="p">(</span><span class="n">h</span><span class="o">=</span><span class="m">0.90</span><span class="p">,</span><span class="n">col</span><span class="o">=</span><span class="s2">"red"</span><span class="p">)</span><span class="w">
</span><span class="c1"># Mean connectivity as a function of different powers
</span><span class="n">plot</span><span class="p">(</span><span class="n">sft</span><span class="o">$</span><span class="n">fitIndices</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="n">sft</span><span class="o">$</span><span class="n">fitIndices</span><span class="p">[,</span><span class="m">5</span><span class="p">],</span><span class="n">type</span><span class="o">=</span><span class="s2">"n"</span><span class="p">,</span><span class="w">
     </span><span class="n">xlab</span><span class="o">=</span><span class="s2">"Soft Threshold (power)"</span><span class="p">,</span><span class="n">ylab</span><span class="o">=</span><span class="s2">"Mean Connectivity"</span><span class="p">,</span><span class="n">main</span><span class="o">=</span><span class="n">paste</span><span class="p">(</span><span class="s2">"Mean connectivity"</span><span class="p">))</span><span class="w">
</span><span class="n">text</span><span class="p">(</span><span class="n">sft</span><span class="o">$</span><span class="n">fitIndices</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span><span class="n">sft</span><span class="o">$</span><span class="n">fitIndices</span><span class="p">[,</span><span class="m">5</span><span class="p">],</span><span class="n">labels</span><span class="o">=</span><span class="n">powers</span><span class="p">,</span><span class="n">col</span><span class="o">=</span><span class="s2">"red"</span><span class="p">)</span><span class="w">
</span><span class="n">dev.off</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<h3 id="automatic-module-detection-via-dynamic-tree-cutting">Automatic module detection via dynamic tree cutting</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">softpower</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">30</span><span class="w">
</span><span class="n">mergingThresh</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.25</span><span class="w">
</span><span class="n">net</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">blockwiseModules</span><span class="p">(</span><span class="n">datExpr</span><span class="p">,</span><span class="w"> </span><span class="n">checkMissingData</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">corType</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bicor"</span><span class="p">,</span><span class="w"> </span><span class="c1"># or pearson
</span><span class="w">                       </span><span class="n">maxBlockSize</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10000</span><span class="p">,</span><span class="w"> </span><span class="n">networkType</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"signed"</span><span class="p">,</span><span class="w"> </span><span class="n">power</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">softpower</span><span class="p">,</span><span class="w">
                       </span><span class="n">minModuleSize</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">30</span><span class="p">,</span><span class="w"> </span><span class="n">maxPOutliers</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.1</span><span class="p">,</span><span class="w"> </span><span class="n">mergeCutHeight</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">mergingThresh</span><span class="p">,</span><span class="w">
                       </span><span class="n">numericLabels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">saveTOMs</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">randomSeed</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12345</span><span class="p">,</span><span class="w">
                       </span><span class="n">pamRespectsDendro</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">saveTOMFileBase</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"TESTexprsTOM"</span><span class="p">)</span><span class="w">
</span><span class="n">str</span><span class="p">(</span><span class="n">net</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">moduleColors</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">wgcna_plotDendroAndColors</span><span class="p">(</span><span class="n">datExpr</span><span class="p">,</span><span class="w"> </span><span class="n">datTraits</span><span class="p">,</span><span class="w"> </span><span class="n">net</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># Recalculate MEs with color labels
</span><span class="n">MEs0</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">moduleEigengenes</span><span class="p">(</span><span class="n">datExpr</span><span class="p">,</span><span class="w"> </span><span class="n">moduleColors</span><span class="p">,</span><span class="w"> </span><span class="n">softPower</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">softpower</span><span class="p">)</span><span class="o">$</span><span class="n">eigengenes</span><span class="w">
</span><span class="c1"># Reorder given (eigen-)vectors such that similar ones (as measured by correlation) are next to each other:
</span><span class="n">MEs</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">orderMEs</span><span class="p">(</span><span class="n">MEs0</span><span class="p">)</span><span class="w">
</span><span class="n">rownames</span><span class="p">(</span><span class="n">MEs</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">datExpr</span><span class="p">)</span><span class="w">
</span><span class="n">write.table</span><span class="p">(</span><span class="n">MEs</span><span class="p">,</span><span class="w"> </span><span class="s2">"Automatic_ModuleEigengenes_numberLabel.txt"</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="o">=</span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">quote</span><span class="o">=</span><span class="nb">F</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">

</span><span class="c1"># calculate the module membership values (aka. module eigengene based connectivity kME):
</span><span class="w">  </span><span class="n">datKME</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">WGCNA</span><span class="o">::</span><span class="n">signedKME</span><span class="p">(</span><span class="n">datExpr</span><span class="p">,</span><span class="w"> </span><span class="n">MEs</span><span class="p">)</span><span class="w">  </span><span class="c1"># equals geneModuleMembership
</span><span class="w">  </span><span class="n">colnames</span><span class="p">(</span><span class="n">datKME</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sub</span><span class="p">(</span><span class="s2">"kME"</span><span class="p">,</span><span class="w"> </span><span class="s2">"MM."</span><span class="p">,</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">datKME</span><span class="p">))</span><span class="w">

</span><span class="n">wgcna_modulememberships</span><span class="p">(</span><span class="n">datExpr</span><span class="p">,</span><span class="w"> </span><span class="n">MEs</span><span class="p">,</span><span class="w"> </span><span class="n">datKME</span><span class="p">,</span><span class="w"> </span><span class="n">moduleColors</span><span class="p">)</span><span class="w">

</span><span class="n">wgcna_heatmaps</span><span class="p">(</span><span class="n">MEs</span><span class="p">,</span><span class="w"> </span><span class="n">datExpr</span><span class="p">,</span><span class="w"> </span><span class="n">datTraits</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="wgcnaHeatmap.png" alt="Module trait correlation heatmap" /></p>

<p><br /></p>

<h3 id="output-file-for-gene-ontology-analysis">Output file for gene ontology analysis</h3>

<p><strong>geneAnnotations()</strong> uses the R package AnnotationDbi.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># Annotate
</span><span class="n">genes</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">datExpr</span><span class="p">)</span><span class="w">
</span><span class="n">datExpr_anno</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">geneAnnotations</span><span class="p">(</span><span class="n">input</span><span class="o">=</span><span class="n">genes</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="o">=</span><span class="n">genes</span><span class="p">,</span><span class="w"> </span><span class="n">column</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"ENTREZID"</span><span class="p">,</span><span class="w"> </span><span class="s2">"ENSEMBL"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="o">=</span><span class="s2">"SYMBOL"</span><span class="p">,</span><span class="w"> </span><span class="n">organism</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"human"</span><span class="p">)</span><span class="w">

</span><span class="c1"># Correlations of genes and traits
</span><span class="n">traitsCor</span><span class="o">=</span><span class="n">cor</span><span class="p">(</span><span class="n">datExpr</span><span class="p">,</span><span class="n">datTraits</span><span class="p">,</span><span class="n">use</span><span class="o">=</span><span class="s2">"p"</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">traitsCor</span><span class="p">)</span><span class="o">=</span><span class="n">paste</span><span class="p">(</span><span class="s2">"cor"</span><span class="p">,</span><span class="n">colnames</span><span class="p">(</span><span class="n">traitsCor</span><span class="p">),</span><span class="n">sep</span><span class="o">=</span><span class="s2">"."</span><span class="p">)</span><span class="w">

</span><span class="n">dataOutput</span><span class="o">=</span><span class="n">data.frame</span><span class="p">(</span><span class="n">datExpr_anno</span><span class="w"> </span><span class="p">,</span><span class="n">moduleColors</span><span class="p">,</span><span class="w"> </span><span class="n">datKME</span><span class="p">,</span><span class="w"> </span><span class="n">traitsCor</span><span class="p">)</span><span class="w">

</span><span class="n">write.table</span><span class="p">(</span><span class="n">dataOutput</span><span class="p">,</span><span class="s2">"AllGenesResults.txt"</span><span class="p">,</span><span class="n">row.names</span><span class="o">=</span><span class="nb">F</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<h3 id="module-eigengene-expression-plots">Module eigengene expression plots</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">data</span><span class="p">(</span><span class="n">MEs</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">layout</span><span class="p">(</span><span class="n">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">MEs</span><span class="p">)),</span><span class="m">12</span><span class="p">,</span><span class="nf">rep</span><span class="p">(</span><span class="m">13</span><span class="p">,</span><span class="m">3</span><span class="p">)),</span><span class="w"> </span><span class="n">ncol</span><span class="o">=</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">byrow</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="n">heights</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">3</span><span class="p">,</span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="m">0.5</span><span class="p">))</span><span class="w">

</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"ME"</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">MEs</span><span class="p">))){</span><span class="w">
  </span><span class="n">module_eigengene_plot</span><span class="p">(</span><span class="n">groups</span><span class="p">,</span><span class="w"> </span><span class="n">MEs</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span><span class="n">plot.new</span><span class="p">()</span><span class="w">

</span><span class="n">par</span><span class="p">(</span><span class="n">mai</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">))</span><span class="w">
</span><span class="n">plot.new</span><span class="p">()</span><span class="w">
</span><span class="n">legend</span><span class="p">(</span><span class="s2">"bottom"</span><span class="p">,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"mean module eigengenes"</span><span class="p">,</span><span class="w"> </span><span class="s2">"replicate module eigengene values"</span><span class="p">),</span><span class="w"> </span><span class="n">xpd</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">horiz</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">inset</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">0</span><span class="p">),</span><span class="w"> </span><span class="n">bty</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"n"</span><span class="p">,</span><span class="w"> </span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"grey"</span><span class="p">,</span><span class="w"> </span><span class="s2">"black"</span><span class="p">),</span><span class="w"> </span><span class="n">pch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">16</span><span class="p">,</span><span class="w"> </span><span class="n">cex</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="exprAnalysis_files/figure-markdown_github/module_eigengene_plot-1.png" alt="" /></p>

<hr />

<p><br /></p>

<h1 id="differential-expression-analysis">Differential expression analysis</h1>

<h2 id="de-analyis-using-the-r-package-limma">DE analyis using the R package limma</h2>

<p>For normalized expression data (FPKM from Cufflinks or intensities from expression chip arrays).</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">design</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.matrix</span><span class="p">(</span><span class="n">data.frame</span><span class="p">(</span><span class="n">Ctrl</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">12</span><span class="p">)),</span><span class="w"> </span><span class="n">TolLPS</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="w">
              </span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">8</span><span class="p">)),</span><span class="w"> </span><span class="n">TollMRP8</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">8</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">)),</span><span class="w"> </span><span class="n">ActLPS</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="nf">rep</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">12</span><span class="p">),</span><span class="w">
              </span><span class="nf">rep</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">)),</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">)))</span><span class="w">
</span><span class="n">groupcomparisons</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"TolLPS-Ctrl"</span><span class="p">,</span><span class="w"> </span><span class="s2">"TollMRP8-Ctrl"</span><span class="p">,</span><span class="w"> </span><span class="s2">"ActLPS-Ctrl"</span><span class="p">)</span><span class="w">
</span><span class="n">DEgenes_all</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">diff_limma_all</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">,</span><span class="w"> </span><span class="n">design</span><span class="p">,</span><span class="w"> </span><span class="n">groupcomparisons</span><span class="p">)</span><span class="w">

</span><span class="n">comparison</span><span class="o">=</span><span class="s2">"TolLPS-Ctrl"</span><span class="w">
</span><span class="n">Allgenes_limma_pw</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">diff_limma_pw_unfiltered</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">,</span><span class="w"> </span><span class="n">design</span><span class="p">,</span><span class="w"> </span><span class="n">comparison</span><span class="p">)</span><span class="w">

</span><span class="n">DEgenes_pw</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">Allgenes_limma_pw</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="nf">abs</span><span class="p">(</span><span class="n">Allgenes_limma_pw</span><span class="o">$</span><span class="n">logFC</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">log2</span><span class="p">(</span><span class="m">1.5</span><span class="p">)</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">Allgenes_limma_pw</span><span class="o">$</span><span class="n">adj.P.Val</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0.05</span><span class="p">),]</span><span class="w">
</span></code></pre>
</div>

<p>Example output files from diff_limma functions have been saved to the package data and can be accessed via:</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">data</span><span class="p">(</span><span class="s2">"DEgenes_pw"</span><span class="p">)</span><span class="w">
</span><span class="n">data</span><span class="p">(</span><span class="s2">"Allgenes_limma_pw"</span><span class="p">)</span><span class="w">


</span><span class="c1"># Export tables to Latex
</span><span class="n">library</span><span class="p">(</span><span class="n">xtable</span><span class="p">)</span><span class="w">

</span><span class="n">italic</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">){</span><span class="w">
</span><span class="n">paste0</span><span class="p">(</span><span class="s1">'{\\emph{ '</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="s1">'}}'</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">

</span><span class="n">print</span><span class="p">(</span><span class="n">xtable</span><span class="p">(</span><span class="n">DEgenes_pw</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,],</span><span class="w"> 
       </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w">
       </span><span class="n">caption</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"\\tt Long caption"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Short caption"</span><span class="p">),</span><span class="w">
       </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"tab:testtable"</span><span class="p">,</span><span class="w">
       </span><span class="n">auto</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w">
      </span><span class="n">caption.placement</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"top"</span><span class="p">,</span><span class="w">
      </span><span class="n">sanitize.rownames.function</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">italic</span><span class="p">,</span><span class="w">
      </span><span class="n">booktabs</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w">
      </span><span class="n">floating</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> 
      </span><span class="n">latex.environments</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"center"</span><span class="p">,</span><span class="w">
      </span><span class="n">include.rownames</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w">
      </span><span class="n">scalebox</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.8</span><span class="p">,</span><span class="w">
      </span><span class="n">tabular.environment</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"tabularx"</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"\\textwidth"</span><span class="p">)</span><span class="w">

</span><span class="c1"># or to html
</span><span class="n">print</span><span class="p">(</span><span class="n">xtable</span><span class="p">(</span><span class="n">DEgenes_pw</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,],</span><span class="w"> 
       </span><span class="n">digits</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w">
       </span><span class="n">caption</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Long caption"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Short caption"</span><span class="p">),</span><span class="w">
       </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"tab:testtable"</span><span class="p">,</span><span class="w">
       </span><span class="n">auto</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="n">type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"html"</span><span class="p">,</span><span class="w"> </span><span class="n">caption.placement</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"top"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<!-- html table generated in R 3.3.0 by xtable 1.8-2 package -->
<!-- Wed Jun 01 08:44:34 2016 -->
<table border="1">
<caption align="top">
Long caption
</caption>
<tr>
<th>
</th>
<th>
logFC
</th>
<th>
CI.L
</th>
<th>
CI.R
</th>
<th>
AveExpr
</th>
<th>
t
</th>
<th>
P.Value
</th>
<th>
adj.P.Val
</th>
<th>
B
</th>
<th>
ENTREZID
</th>
<th>
ENSEMBL
</th>
</tr>
<tr>
<td>
HCAR2
</td>
<td align="right">
-1.60
</td>
<td align="right">
-1.78
</td>
<td align="right">
-1.41
</td>
<td align="right">
5.17
</td>
<td align="right">
-18.38
</td>
<td align="right">
0.00
</td>
<td align="right">
0.00
</td>
<td align="right">
16.91
</td>
<td>
338442
</td>
<td>
ENSG00000182782
</td>
</tr>
<tr>
<td>
LOC100996342
</td>
<td align="right">
2.54
</td>
<td align="right">
2.18
</td>
<td align="right">
2.90
</td>
<td align="right">
6.89
</td>
<td align="right">
15.03
</td>
<td align="right">
0.00
</td>
<td align="right">
0.00
</td>
<td align="right">
14.33
</td>
<td>
100996342
</td>
<td>
ENSG00000235478
</td>
</tr>
<tr>
<td>
INPP4B
</td>
<td align="right">
-1.54
</td>
<td align="right">
-1.77
</td>
<td align="right">
-1.31
</td>
<td align="right">
4.80
</td>
<td align="right">
-14.19
</td>
<td align="right">
0.00
</td>
<td align="right">
0.00
</td>
<td align="right">
13.58
</td>
<td>
8821
</td>
<td>
ENSG00000109452
</td>
</tr>
<tr>
<td>
LOC100505549
</td>
<td align="right">
-1.26
</td>
<td align="right">
-1.45
</td>
<td align="right">
-1.06
</td>
<td align="right">
3.11
</td>
<td align="right">
-13.91
</td>
<td align="right">
0.00
</td>
<td align="right">
0.00
</td>
<td align="right">
13.32
</td>
<td>
100505549
</td>
<td>
</td>
</tr>
<tr>
<td>
FCGBP
</td>
<td align="right">
-2.06
</td>
<td align="right">
-2.38
</td>
<td align="right">
-1.75
</td>
<td align="right">
3.39
</td>
<td align="right">
-13.87
</td>
<td align="right">
0.00
</td>
<td align="right">
0.00
</td>
<td align="right">
13.28
</td>
<td>
8857
</td>
<td>
ENSG00000275395
</td>
</tr>
<tr>
<td>
ZSCAN31
</td>
<td align="right">
-1.40
</td>
<td align="right">
-1.63
</td>
<td align="right">
-1.18
</td>
<td align="right">
3.44
</td>
<td align="right">
-13.33
</td>
<td align="right">
0.00
</td>
<td align="right">
0.00
</td>
<td align="right">
12.76
</td>
<td>
64288
</td>
<td>
ENSG00000235109
</td>
</tr>
<tr>
<td>
MIR3665
</td>
<td align="right">
-1.32
</td>
<td align="right">
-1.53
</td>
<td align="right">
-1.11
</td>
<td align="right">
2.69
</td>
<td align="right">
-13.27
</td>
<td align="right">
0.00
</td>
<td align="right">
0.00
</td>
<td align="right">
12.70
</td>
<td>
100500861
</td>
<td>
ENSG00000266325
</td>
</tr>
<tr>
<td>
PTH2R
</td>
<td align="right">
-1.62
</td>
<td align="right">
-1.90
</td>
<td align="right">
-1.35
</td>
<td align="right">
3.69
</td>
<td align="right">
-12.56
</td>
<td align="right">
0.00
</td>
<td align="right">
0.00
</td>
<td align="right">
11.98
</td>
<td>
5746
</td>
<td>
ENSG00000144407
</td>
</tr>
<tr>
<td>
GCLC
</td>
<td align="right">
-2.42
</td>
<td align="right">
-2.83
</td>
<td align="right">
-2.00
</td>
<td align="right">
3.13
</td>
<td align="right">
-12.27
</td>
<td align="right">
0.00
</td>
<td align="right">
0.00
</td>
<td align="right">
11.66
</td>
<td>
2729
</td>
<td>
ENSG00000001084
</td>
</tr>
<tr>
<td>
SOX30
</td>
<td align="right">
2.54
</td>
<td align="right">
2.10
</td>
<td align="right">
2.98
</td>
<td align="right">
8.57
</td>
<td align="right">
12.19
</td>
<td align="right">
0.00
</td>
<td align="right">
0.00
</td>
<td align="right">
11.59
</td>
<td>
11063
</td>
<td>
ENSG00000039600
</td>
</tr>
</table>

<hr />

<p><br /></p>

<h2 id="de-analyis-using-deseq2">DE analyis using DESeq2</h2>

<p>For raw read count data.</p>

<h5 id="contrast-de-groups">contrast DE groups:</h5>

<ul>
  <li>lfc = treatment &gt; Ctrl, - lfc = treatment &lt; Ctrl p-value &amp; p.adjust values of NA indicate outliers detected by Cook’s distance NA only for p.adjust means the gene is filtered by automatic independent filtering for having a low mean normalized count</li>
</ul>

<p><br /></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">DESeq2</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">ggrepel</span><span class="p">)</span><span class="w">

</span><span class="c1"># find possible contrasts with
</span><span class="n">DESeq2</span><span class="o">::</span><span class="n">resultsNames</span><span class="p">(</span><span class="n">data_DESeq</span><span class="p">)</span><span class="w">

</span><span class="n">res</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">DESeq2</span><span class="o">::</span><span class="n">results</span><span class="p">(</span><span class="n">data_DESeq</span><span class="p">,</span><span class="w"> </span><span class="n">contrast</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="s2">"treatmentActLPS"</span><span class="p">,</span><span class="w"> </span><span class="s2">"treatmentCtrl"</span><span class="p">),</span><span class="w"> </span><span class="n">cooksCutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.99</span><span class="p">,</span><span class="w"> </span><span class="n">independentFiltering</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span><span class="n">pAdjustMethod</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"BH"</span><span class="p">)</span><span class="w">

</span><span class="c1"># order results table by the smallest adjusted p value:
</span><span class="n">res</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">res</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">res</span><span class="o">$</span><span class="n">padj</span><span class="p">),]</span><span class="w">

</span><span class="n">results</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">dplyr</span><span class="o">::</span><span class="n">mutate</span><span class="p">(</span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">res</span><span class="p">),</span><span class="w"> </span><span class="n">sig</span><span class="o">=</span><span class="n">ifelse</span><span class="p">(</span><span class="n">res</span><span class="o">$</span><span class="n">padj</span><span class="o">&lt;</span><span class="m">0.05</span><span class="p">,</span><span class="w"> </span><span class="s2">"FDR&lt;0.05"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Not Sig"</span><span class="p">)),</span><span class="w"> </span><span class="n">row.names</span><span class="o">=</span><span class="n">rownames</span><span class="p">(</span><span class="n">res</span><span class="p">))</span><span class="w">

</span><span class="n">DEgenes_DESeq</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">results</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="nf">abs</span><span class="p">(</span><span class="n">results</span><span class="o">$</span><span class="n">log2FoldChange</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="n">log2</span><span class="p">(</span><span class="m">1.5</span><span class="p">)</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">results</span><span class="o">$</span><span class="n">padj</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">0.05</span><span class="p">),]</span><span class="w">

</span><span class="n">p</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ggplot2</span><span class="o">::</span><span class="n">ggplot</span><span class="p">(</span><span class="n">results</span><span class="p">,</span><span class="w"> </span><span class="n">ggplot2</span><span class="o">::</span><span class="n">aes</span><span class="p">(</span><span class="n">log2FoldChange</span><span class="p">,</span><span class="w"> </span><span class="o">-</span><span class="n">log10</span><span class="p">(</span><span class="n">pvalue</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">ggplot2</span><span class="o">::</span><span class="n">geom_point</span><span class="p">(</span><span class="n">ggplot2</span><span class="o">::</span><span class="n">aes</span><span class="p">(</span><span class="n">col</span><span class="o">=</span><span class="n">sig</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">ggplot2</span><span class="o">::</span><span class="n">scale_color_manual</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"red"</span><span class="p">,</span><span class="w"> </span><span class="s2">"black"</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">ggplot2</span><span class="o">::</span><span class="n">ggtitle</span><span class="p">(</span><span class="s2">"Volcano Plot of DESeq2 analysis"</span><span class="p">)</span><span class="w">

</span><span class="n">p</span><span class="o">+</span><span class="n">ggrepel</span><span class="o">::</span><span class="n">geom_text_repel</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">results</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="p">],</span><span class="w"> </span><span class="n">ggplot2</span><span class="o">::</span><span class="n">aes</span><span class="p">(</span><span class="n">label</span><span class="o">=</span><span class="n">rownames</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="p">])))</span><span class="w">

</span><span class="c1"># If there aren't too many DE genes:
#p+geom_text_repel(data=dplyr::filter(results, padj&lt;0.05), aes(label=rownames(results[1:10, ])))
</span></code></pre>
</div>

<p><br /></p>

<h4 id="ma-plot">MA-plot</h4>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">DESeq2</span><span class="o">::</span><span class="n">plotMA</span><span class="p">(</span><span class="n">res</span><span class="p">,</span><span class="w"> </span><span class="n">main</span><span class="o">=</span><span class="s2">"MA Plot"</span><span class="p">,</span><span class="w"> </span><span class="n">ylim</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">-2</span><span class="p">,</span><span class="m">2</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<h4 id="plotcounts-which-normalizes-counts-by-sequencing-depth-and-adds-a-pseudocount-of-12-to-allow-for-log-scale-plotting">plotCounts, which normalizes counts by sequencing depth and adds a pseudocount of 1/2 to allow for log scale plotting</h4>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">par</span><span class="p">(</span><span class="n">mfrow</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">3</span><span class="p">))</span><span class="w">

</span><span class="k">for</span><span class="w"> </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">){</span><span class="w">
  </span><span class="n">gene</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">res</span><span class="p">)[</span><span class="n">i</span><span class="p">]</span><span class="w">
  </span><span class="n">main</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gene</span><span class="w">
  </span><span class="n">DESeq2</span><span class="o">::</span><span class="n">plotCounts</span><span class="p">(</span><span class="n">data_DESeq</span><span class="p">,</span><span class="w"> </span><span class="n">gene</span><span class="o">=</span><span class="n">gene</span><span class="p">,</span><span class="w"> </span><span class="n">intgroup</span><span class="o">=</span><span class="s2">"treatment"</span><span class="p">,</span><span class="w"> </span><span class="n">main</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">main</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="gene-annotations">Gene annotations</h2>

<p>Can be used to add e.g. ENTREZ ID, ENSEMBL ID, etc. to gene name.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">DEgenes_pw</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">geneAnnotations</span><span class="p">(</span><span class="n">input</span><span class="o">=</span><span class="n">DEgenes_pw</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="o">=</span><span class="n">row.names</span><span class="p">(</span><span class="n">DEgenes_pw</span><span class="p">),</span><span class="w"> </span><span class="n">column</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"ENTREZID"</span><span class="p">,</span><span class="w"> </span><span class="s2">"ENSEMBL"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="o">=</span><span class="s2">"SYMBOL"</span><span class="p">,</span><span class="w"> </span><span class="n">organism</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"human"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">DEgenes_DESeq</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">geneAnnotations</span><span class="p">(</span><span class="n">input</span><span class="o">=</span><span class="n">DEgenes_DESeq</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="o">=</span><span class="n">row.names</span><span class="p">(</span><span class="n">DEgenes_DESeq</span><span class="p">),</span><span class="w"> </span><span class="n">column</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"ENTREZID"</span><span class="p">,</span><span class="w"> </span><span class="s2">"ENSEMBL"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="o">=</span><span class="s2">"SYMBOL"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="enrichment-analysis-using-clusterpofiler">Enrichment Analysis using clusterPofiler</h2>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">clusterProfiler</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">DOSE</span><span class="p">)</span><span class="w">

</span><span class="n">geneList</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.vector</span><span class="p">(</span><span class="n">Allgenes_limma_pw</span><span class="o">$</span><span class="n">logFC</span><span class="p">)</span><span class="w">
</span><span class="nf">names</span><span class="p">(</span><span class="n">geneList</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">Allgenes_limma_pw</span><span class="p">)</span><span class="w">
</span><span class="n">gene</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">na.omit</span><span class="p">(</span><span class="n">DEgenes_pw</span><span class="o">$</span><span class="n">ENTREZID</span><span class="p">)</span><span class="w">

</span><span class="n">OrgDb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">org.Hs.eg.db</span><span class="o">::</span><span class="n">org.Hs.eg.db</span><span class="w"> </span><span class="c1"># can also be org.Mm.eg.db::org.Mm.eg.db
</span><span class="w">
</span><span class="c1"># Group GO
</span><span class="n">ggo</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">clusterProfiler</span><span class="o">::</span><span class="n">groupGO</span><span class="p">(</span><span class="n">gene</span><span class="w">     </span><span class="o">=</span><span class="w"> </span><span class="n">gene</span><span class="p">,</span><span class="w">
                                </span><span class="n">OrgDb</span><span class="w">    </span><span class="o">=</span><span class="w"> </span><span class="n">OrgDb</span><span class="p">,</span><span class="w">
                                </span><span class="n">ont</span><span class="w">      </span><span class="o">=</span><span class="w"> </span><span class="s2">"BP"</span><span class="p">,</span><span class="w">
                                </span><span class="n">level</span><span class="w">    </span><span class="o">=</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w">
                                </span><span class="n">readable</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">summary</span><span class="p">(</span><span class="n">ggo</span><span class="p">))</span><span class="w">
</span><span class="n">barplot</span><span class="p">(</span><span class="n">ggo</span><span class="p">,</span><span class="w"> </span><span class="n">drop</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="o">=</span><span class="m">12</span><span class="p">)</span><span class="w">

</span><span class="c1"># GO over-representation test
</span><span class="n">ego</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">clusterProfiler</span><span class="o">::</span><span class="n">enrichGO</span><span class="p">(</span><span class="n">gene</span><span class="w">          </span><span class="o">=</span><span class="w"> </span><span class="n">gene</span><span class="p">,</span><span class="w">
                                 </span><span class="n">OrgDb</span><span class="w">         </span><span class="o">=</span><span class="w"> </span><span class="n">OrgDb</span><span class="p">,</span><span class="w">
                                 </span><span class="n">ont</span><span class="w">           </span><span class="o">=</span><span class="w"> </span><span class="s2">"BP"</span><span class="p">,</span><span class="w">
                                 </span><span class="n">pAdjustMethod</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"BH"</span><span class="p">,</span><span class="w">
                                 </span><span class="n">pvalueCutoff</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w">
                                 </span><span class="n">qvalueCutoff</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w"> 
                                 </span><span class="n">readable</span><span class="w">      </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">summary</span><span class="p">(</span><span class="n">ego</span><span class="p">))</span><span class="w">
</span><span class="n">barplot</span><span class="p">(</span><span class="n">ego</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="o">=</span><span class="m">25</span><span class="p">)</span><span class="w">
</span><span class="n">clusterProfiler</span><span class="o">::</span><span class="n">dotplot</span><span class="p">(</span><span class="n">ego</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="o">=</span><span class="m">25</span><span class="p">)</span><span class="w">
</span><span class="c1">#enrichMap(ego)
</span><span class="n">cnetplot</span><span class="p">(</span><span class="n">ego</span><span class="p">,</span><span class="w"> </span><span class="n">categorySize</span><span class="o">=</span><span class="s2">"pvalue"</span><span class="p">,</span><span class="w"> </span><span class="n">foldChange</span><span class="o">=</span><span class="n">geneList</span><span class="p">)</span><span class="w">
</span><span class="n">clusterProfiler</span><span class="o">::</span><span class="n">plotGOgraph</span><span class="p">(</span><span class="n">ego</span><span class="p">)</span><span class="w">

</span><span class="c1"># enrichGO test the whole GO corpus and enriched result may contains very general terms. With dropGO function, user can remove specific GO terms or GO level from results obtained from both enrichGO and compareCluster.
</span><span class="w">

</span><span class="c1">## KEGG over-representation test
</span><span class="n">kk</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">clusterProfiler</span><span class="o">::</span><span class="n">enrichKEGG</span><span class="p">(</span><span class="n">gene</span><span class="w">         </span><span class="o">=</span><span class="w"> </span><span class="n">gene</span><span class="p">,</span><span class="w">
                 </span><span class="n">organism</span><span class="w">     </span><span class="o">=</span><span class="w"> </span><span class="s1">'hsa'</span><span class="p">,</span><span class="w">
                 </span><span class="n">pvalueCutoff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">)</span><span class="w">
</span><span class="n">head</span><span class="p">(</span><span class="n">summary</span><span class="p">(</span><span class="n">kk</span><span class="p">))</span><span class="w">
</span><span class="n">barplot</span><span class="p">(</span><span class="n">kk</span><span class="p">,</span><span class="w"> </span><span class="n">showCategory</span><span class="o">=</span><span class="m">8</span><span class="p">)</span><span class="w">
</span><span class="n">clusterProfiler</span><span class="o">::</span><span class="n">dotplot</span><span class="p">(</span><span class="n">kk</span><span class="p">)</span><span class="w">
</span><span class="n">cnetplot</span><span class="p">(</span><span class="n">kk</span><span class="p">,</span><span class="w"> </span><span class="n">categorySize</span><span class="o">=</span><span class="s2">"geneNum"</span><span class="p">,</span><span class="w"> </span><span class="n">foldChange</span><span class="o">=</span><span class="n">geneList</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="pathview">Pathview</h2>

<p>uses the R package pathview.</p>

<p>For Toll-like-receptor signaling:</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">pathview_func</span><span class="p">(</span><span class="n">DEgenes_pw</span><span class="p">,</span><span class="w"> </span><span class="n">logFCcolumn</span><span class="o">=</span><span class="s2">"logFC"</span><span class="p">,</span><span class="w"> </span><span class="n">pathway.id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"04620"</span><span class="p">,</span><span class="w"> </span><span class="n">out.suffix</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"DE_TolLPS"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="venn-diagrams-and-biological-theme-comparison">Venn diagrams and Biological theme comparison</h2>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">gplots</span><span class="p">)</span><span class="w">

</span><span class="n">venn_list</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">genelist1</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">na.omit</span><span class="p">(</span><span class="n">rownames</span><span class="p">(</span><span class="n">DEgenes_pw</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">50</span><span class="p">]),</span><span class="w"> </span><span class="n">genelist2</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">na.omit</span><span class="p">(</span><span class="n">rownames</span><span class="p">(</span><span class="n">DEgenes_pw</span><span class="p">)[</span><span class="m">30</span><span class="o">:</span><span class="m">80</span><span class="p">]),</span><span class="w"> </span><span class="n">genelist3</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">na.omit</span><span class="p">(</span><span class="n">rownames</span><span class="p">(</span><span class="n">DEgenes_pw</span><span class="p">)[</span><span class="m">48</span><span class="o">:</span><span class="m">100</span><span class="p">]))</span><span class="w">

</span><span class="n">gplots</span><span class="o">::</span><span class="n">venn</span><span class="p">(</span><span class="n">venn_list</span><span class="p">,</span><span class="w"> </span><span class="n">show.plot</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">small</span><span class="o">=</span><span class="m">0.7</span><span class="p">,</span><span class="w"> </span><span class="n">showSetLogicLabel</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">simplify</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="n">mtext</span><span class="p">(</span><span class="s2">"If you want a header, put it here"</span><span class="p">,</span><span class="w"> </span><span class="n">side</span><span class="o">=</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">cex</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.8</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># Biological theme comparison
</span><span class="n">compGO</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">clusterProfiler</span><span class="o">::</span><span class="n">compareCluster</span><span class="p">(</span><span class="n">geneCluster</span><span class="w">   </span><span class="o">=</span><span class="w"> </span><span class="n">venn_list</span><span class="p">,</span><span class="w">
                                          </span><span class="n">fun</span><span class="w">           </span><span class="o">=</span><span class="w"> </span><span class="s2">"enrichGO"</span><span class="p">,</span><span class="w">
                                          </span><span class="n">ont</span><span class="w">           </span><span class="o">=</span><span class="w"> </span><span class="s2">"BP"</span><span class="p">,</span><span class="w">
                                          </span><span class="n">OrgDb</span><span class="w">         </span><span class="o">=</span><span class="w"> </span><span class="n">OrgDb</span><span class="p">,</span><span class="w">
                                          </span><span class="n">qvalueCutoff</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="m">0.001</span><span class="p">,</span><span class="w">
                                          </span><span class="n">pAdjustMethod</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"BH"</span><span class="p">,</span><span class="w">
                                          </span><span class="n">readable</span><span class="w">      </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">


</span><span class="n">compKEGG</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">clusterProfiler</span><span class="o">::</span><span class="n">compareCluster</span><span class="p">(</span><span class="n">geneCluster</span><span class="w">   </span><span class="o">=</span><span class="w"> </span><span class="n">venn_list</span><span class="p">,</span><span class="w">
                                            </span><span class="n">fun</span><span class="w">           </span><span class="o">=</span><span class="w"> </span><span class="s2">"enrichKEGG"</span><span class="p">,</span><span class="w">
                                            </span><span class="n">qvalueCutoff</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w">
                                            </span><span class="n">pAdjustMethod</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"BH"</span><span class="p">)</span><span class="w">
</span><span class="n">compKEGG</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">clusterProfiler</span><span class="o">::</span><span class="n">setReadable</span><span class="p">(</span><span class="n">compKEGG</span><span class="p">,</span><span class="w"> </span><span class="n">OrgDb</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">OrgDb</span><span class="p">)</span><span class="w">


</span><span class="n">compMKEGG</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">clusterProfiler</span><span class="o">::</span><span class="n">compareCluster</span><span class="p">(</span><span class="n">geneCluster</span><span class="w">   </span><span class="o">=</span><span class="w"> </span><span class="n">venn_list</span><span class="p">,</span><span class="w">
                                             </span><span class="n">fun</span><span class="w">           </span><span class="o">=</span><span class="w"> </span><span class="s2">"enrichMKEGG"</span><span class="p">,</span><span class="w">
                                             </span><span class="n">organism</span><span class="o">=</span><span class="s1">'hsa'</span><span class="p">,</span><span class="w"> </span><span class="n">minGSSize</span><span class="o">=</span><span class="m">1</span><span class="p">)</span><span class="w">

</span><span class="n">compDO</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">clusterProfiler</span><span class="o">::</span><span class="n">compareCluster</span><span class="p">(</span><span class="n">geneCluster</span><span class="w">   </span><span class="o">=</span><span class="w"> </span><span class="n">venn_list</span><span class="p">,</span><span class="w">
                                          </span><span class="n">fun</span><span class="w">           </span><span class="o">=</span><span class="w"> </span><span class="s2">"enrichDO"</span><span class="p">,</span><span class="w">
                                          </span><span class="n">qvalueCutoff</span><span class="w">  </span><span class="o">=</span><span class="w"> </span><span class="m">0.05</span><span class="p">,</span><span class="w">
                                          </span><span class="n">pAdjustMethod</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"BH"</span><span class="p">,</span><span class="w">
                                          </span><span class="n">readable</span><span class="w">      </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">

</span><span class="c1"># For visualisation see Enrichment Analysis using clusterProfiler
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h1 id="networks">Networks</h1>

<h2 id="tf-networks">TF networks</h2>

<p>Organism can be human or mouse. For human, the longer published list of TFs is used; for mouse the shorter list provided by Bonn (for which I don’t have any more info on where it comes from).</p>

<p>This function produced two output files:</p>

<ul>
  <li>The .expression matrix of TF expression data.</li>
</ul>

<p>It will have to be opened with <strong>Biolayout3D</strong>: Set Minimum Correlation and Correlation metric (by default 0.7 and Pearson). Then choose a suitable correlation coefficient (Graph Degree Distribution should be close to linear). Save the resulting network as a TGF file.</p>

<p>Open the TGF file in <strong>Cytoscape</strong>: Open network from file. Then got to Advanced Options: untick “Use first column names” and add “ to “Other:”. Now set Column 2 as Source and Column 5 as Target.</p>

<ul>
  <li>Open the node annotation table in .txt format.</li>
</ul>

<p>In cytoscape, open table from file, import data as Node Table Columns. You can then customize the look of your network.</p>

<p>For example, go to Style and</p>

<ul>
  <li>
    <p>go to Border Paint, Column: DE, Mapping type: Discrete and choose a color for 1 and 0</p>
  </li>
  <li>
    <p>go to Fill Color, Column logFC, Mapping Type: Continuous and choose colors.</p>
  </li>
</ul>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">TF_networks</span><span class="p">(</span><span class="n">expmatrix</span><span class="p">,</span><span class="w"> </span><span class="n">nodeAnno</span><span class="o">=</span><span class="n">Allgenes_limma_pw</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="CopyOfnetwork_test.png" alt="TF network Pearson 0.7" /></p>

<hr />

<p><br /></p>

<h2 id="bingo-cytoscape">BiNGO (Cytoscape)</h2>

<ul>
  <li>
    <p>Go to Cytoscape</p>
  </li>
  <li>
    <p>Apps -&gt; <strong>BiNGO</strong></p>
  </li>
  <li>
    <p>The BiNGO Settings panel pops up. Start by filling in a name for your cluster. This name will be used for creating the output file and the visualization of the results in Cytoscape.</p>
  </li>
  <li>
    <p>Check Paste Genes from Text and paste gene names into field below</p>
  </li>
  <li>
    <p>We want to assess overrepresentation of GO categories, and we want to visualize the results in Cytoscape. The corresponding boxes are checked accordingly by default. Then select a statistical test (the Hypergeometric Test is exact and equivalent to an exact Fisher test, the Binomial Test is less accurate but quicker) and a multiple testing correction (we recommend Benjamini &amp; Hochberg’s FDR correction, the Bonferroni correction will be too conservative in most cases), and choose a significance level, e.g. 0.05. Since we only want to visualize those GO categories that are overrepresented after multiple testing correction, and their parents in the GO hierarchy, select the corresponding visualization option. We’re interested in assessing the overrepresentation of functional categories in our cluster with respect to the whole genome, which is why we choose the Complete Annotation as the reference set.</p>
  </li>
  <li>
    <p>Select GO _Biological _Process from the ontology list, and Human from the organism list. We want to consider all evidence codes, so don’t fill in anything in the evidence code box.</p>
  </li>
  <li>
    <p>Finally, select a directory to save the output file in (the file will be named test.bgo if you filled in test as a cluster name), and press Start BiNGO…</p>
  </li>
  <li>
    <p>The program will inform you of its progress while parsing the annotations and calculating the tests, corrections and layout. Finally, a visualization of the overrepresented GO categories is created in Cytoscape. Uncolored nodes are not overrepresented, but they are the parents of overrepresented categories further down. Yellow nodes represent GO categories that are overrepresented at the significance level . For more significant p-values, the node color gets increasingly more orange (see also the Color Legend panel). If you’d like another layout, e.g. hierarchical, select the corresponding option from the Cytoscape Visualization menu. Regardless of the layout you choose, you’ll probably have to tweak the nodes a little in order to avoid overlapping node labels. The list of significantly overrepresented functional categories is shown in the BiNGO output window (more information on the cluster and options you selected, and on which genes did not produce any annotation, is stored in the test.bgo file).</p>
  </li>
</ul>

<hr />

<p><br /></p>

<h1 id="other-functions">Other functions</h1>

<h2 id="mirna-target-identification">miRNA target identification</h2>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># The current version of multiMiR is 1.0.1 and can be downloaded from
</span><span class="n">http</span><span class="o">://</span><span class="n">multimir.ucdenver.edu</span><span class="o">/</span><span class="n">multiMiR_1.0.1.tar.gz.</span><span class="w">

</span><span class="n">install.packages</span><span class="p">(</span><span class="s2">"XML"</span><span class="p">)</span><span class="w">
</span><span class="n">install.packages</span><span class="p">(</span><span class="s2">"RCurl"</span><span class="p">)</span><span class="w">
</span><span class="n">install.packages</span><span class="p">(</span><span class="s2">"/pathname/multiMiR_1.0.1.tar.gz"</span><span class="p">,</span><span class="w"> </span><span class="n">repos</span><span class="o">=</span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"source"</span><span class="p">)</span><span class="w">

</span><span class="n">library</span><span class="p">(</span><span class="n">multiMiR</span><span class="p">)</span><span class="w">

</span><span class="n">miRNA_list</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"hsa-miR-146b-3p"</span><span class="p">,</span><span class="w"> </span><span class="s2">"hsa-miR-155-5p"</span><span class="p">,</span><span class="w"> </span><span class="s2">"hsa-miR-4521"</span><span class="p">))</span><span class="w">

</span><span class="n">miRNA_allTargets</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">get.multimir</span><span class="p">(</span><span class="n">org</span><span class="o">=</span><span class="s2">"hsa"</span><span class="p">,</span><span class="w"> </span><span class="n">mirna</span><span class="o">=</span><span class="n">miRNA_list</span><span class="p">,</span><span class="w"> </span><span class="n">target</span><span class="o">=</span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">table</span><span class="o">=</span><span class="s2">"all"</span><span class="p">,</span><span class="w"> </span><span class="n">summary</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">predicted.cutoff.type</span><span class="o">=</span><span class="s2">"p"</span><span class="p">,</span><span class="w"> </span><span class="n">predicted.cutoff</span><span class="o">=</span><span class="kc">NULL</span><span class="p">)</span><span class="w">
</span><span class="n">str</span><span class="p">(</span><span class="n">miRNA_allTargets</span><span class="p">)</span><span class="w">
    
</span><span class="n">write.table</span><span class="p">(</span><span class="n">miRNA_allTargets</span><span class="o">$</span><span class="n">predicted</span><span class="p">,</span><span class="w"> </span><span class="s2">"miRNAs_allTargets_predictedTargets.txt"</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">,</span><span class="w"> </span><span class="n">col.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span><span class="n">write.table</span><span class="p">(</span><span class="n">miRNA_allTargets</span><span class="o">$</span><span class="n">validated</span><span class="p">,</span><span class="w"> </span><span class="s2">"miRNAs_allTargets_validatedTargets.txt"</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">,</span><span class="w"> </span><span class="n">col.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span><span class="n">write.table</span><span class="p">(</span><span class="n">miRNA_allTargets</span><span class="o">$</span><span class="n">disease.drug</span><span class="p">,</span><span class="w"> </span><span class="s2">"miRNAs_allTargets_diseaseDrugTargets.txt"</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">,</span><span class="w"> </span><span class="n">col.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span><span class="n">write.table</span><span class="p">(</span><span class="n">miRNA_allTargets</span><span class="o">$</span><span class="n">summary</span><span class="p">,</span><span class="w"> </span><span class="s2">"miRNAs_allTargets_summaryTargets.txt"</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">F</span><span class="p">,</span><span class="w"> </span><span class="n">col.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nb">T</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h2 id="hypergeometric-test-for-enrichment">Hypergeometric test for enrichment</h2>

<p>Analogous to Fisher’s Exact Test.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">phyper</span><span class="p">(</span><span class="n">SampleSuccesses</span><span class="w"> </span><span class="m">-1</span><span class="p">,</span><span class="w"> </span><span class="n">PopulationSuccesses</span><span class="p">,</span><span class="w"> </span><span class="n">PopulationSize</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">PopulationSuccesses</span><span class="p">,</span><span class="w"> </span><span class="n">sampleSize</span><span class="p">,</span><span class="w"> </span><span class="n">lower.tail</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<hr />

<p><br /></p>

<h1 id="citations">Citations</h1>

<blockquote>
  <p>Martin Morgan, Herve Pages, Valerie Obenchain and Nathaniel Hayden (2016). Rsamtools: Binary alignment (BAM), FASTA, variant call (BCF), and tabix file import. R package version 1.24.0. <a href="http://bioconductor.org/packages/release/bioc/html/Rsamtools.html">http://bioconductor.org/packages/release/bioc/html/Rsamtools.html</a></p>
</blockquote>

<blockquote>
  <p>Lawrence M, Huber W, Pages H, Aboyoun P, Carlson M, et al. (2013) Software for Computing and Annotating Genomic Ranges. PLoS Comput Biol 9(8): e1003118. &lt;doi:10.1371/journal.pcbi.1003118&gt;</p>
</blockquote>

<blockquote>
  <p>Martin Morgan, Valerie Obenchain, Michel Lang and Ryan Thompson (2016). BiocParallel: Bioconductor facilities for parallel evaluation. R package version 1.6.2. <a href="https://github.com/Bioconductor/BiocParallel">https://github.com/Bioconductor/BiocParallel</a></p>
</blockquote>

<blockquote>
  <p>Martin Morgan, Valerie Obenchain, Jim Hester and Herve Pages (2016). SummarizedExperiment: SummarizedExperiment container. R package version 1.2.2.</p>
</blockquote>

<blockquote>
  <p>Michael I Love, Wolfgang Huber and Simon Anders (2014): Moderated estimation of fold change and dispersion for RNA-Seq data with DESeq2. Genome Biology</p>
</blockquote>

<blockquote>
  <p>Gautier, L., Cope, L., Bolstad, B. M., and Irizarry, R. A. 2004. affy—analysis of Affymetrix GeneChip data at the probe level. Bioinformatics 20, 3 (Feb. 2004), 307-315.</p>
</blockquote>

<blockquote>
  <p>Carvalho B. S., and Irizarry, R. A. 2010. A Framework for Oligonucleotide Microarray Preprocessing. Bioinformatics.</p>
</blockquote>

<blockquote>
  <p>Kauffmann, A., Gentleman, R.,, Huber, W. (2009) arrayQualityMetrics–a bioconductor package for quality assessment of microarray data. Bioinformatics, 25(3):415-6.</p>
</blockquote>

<blockquote>
  <p>Dunning, M.J., Smith, M.L., Ritchie, M.E., Tavare, S. beadarray: R classes and methods for Illumina bead-based data, Bioinformatics. 2007, 23(16):2183-4.</p>
</blockquote>

<blockquote>
  <p>Jeffrey T. Leek, W. Evan Johnson, Hilary S. Parker, Elana J. Fertig, Andrew E. Jaffe and John D. Storey (2016). sva: Surrogate Variable Analysis. R package version 3.20.0.</p>
</blockquote>

<blockquote>
  <p>Henrik Bengtsson (2016). matrixStats: Functions that Apply to Rows and Columns of Matrices (and to Vectors). R package version 0.50.2. <a href="https://CRAN.R-project.org/package=matrixStats">https://CRAN.R-project.org/package=matrixStats</a></p>
</blockquote>

<blockquote>
  <p>Kevin Wright (2015). corrgram: Plot a Correlogram. R package version 1.8. <a href="https://CRAN.R-project.org/package=corrgram">https://CRAN.R-project.org/package=corrgram</a></p>
</blockquote>

<blockquote>
  <p>Hansen M, Gerds TA, Nielsen OH, Seidelin JB, Troelsen JT, et al. (2012) pcaGoPromoter - An R Package for Biological and Regulatory Interpretation of Principal Components in Genome-Wide Gene Expression Data. PLoS ONE 7(2): e32394. &lt;doi:10.1371/journal.pone.0032394&gt;</p>
</blockquote>

<blockquote>
  <p>Raivo Kolde (2015). pheatmap: Pretty Heatmaps. R package version 1.0.8. <a href="https://CRAN.R-project.org/package=pheatmap">https://CRAN.R-project.org/package=pheatmap</a></p>
</blockquote>

<blockquote>
  <p>Gregory R. Warnes, Ben Bolker, Lodewijk Bonebakker, Robert Gentleman, Wolfgang Huber Andy Liaw, Thomas Lumley, Martin Maechler, Arni Magnusson, Steffen Moeller, Marc Schwartz and Bill Venables (2016). gplots: Various R Programming Tools for Plotting Data. R package version 3.0.1. <a href="https://CRAN.R-project.org/package=gplots">https://CRAN.R-project.org/package=gplots</a></p>
</blockquote>

<blockquote>
  <p>Tal Galili (2016). heatmaply: Interactive Heat Maps Using ‘plotly’. R package version 0.3.2. <a href="https://CRAN.R-project.org/package=heatmaply">https://CRAN.R-project.org/package=heatmaply</a></p>
</blockquote>

<blockquote>
  <p>Langfelder P and Horvath S, WGCNA: an R package for weighted correlation network analysis. BMC Bioinformatics 2008, 9:559 &lt;doi:10.1186/1471-2105-9-559&gt;</p>
</blockquote>

<blockquote>
  <p>Peter Langfelder, Steve Horvath (2012). Fast R Functions for Robust Correlations and Hierarchical Clustering. Journal of Statistical Software, 46(11), 1-17. URL <a href="http://www.jstatsoft.org/v46/i11/">http://www.jstatsoft.org/v46/i11/</a>.</p>
</blockquote>

<blockquote>
  <p>Herve Pages, Marc Carlson, Seth Falcon and Nianhua Li (2016). AnnotationDbi: Annotation Database Interface. R package version 1.34.3.</p>
</blockquote>

<blockquote>
  <p>Ritchie, M.E., Phipson, B., Wu, D., Hu, Y., Law, C.W., Shi, W., and Smyth, G.K. (2015). limma powers differential expression analyses for RNA-sequencing and microarray studies. Nucleic Acids Research 43(7), e47.</p>
</blockquote>

<blockquote>
  <p>David B. Dahl (2016). xtable: Export Tables to LaTeX or HTML. R package version 1.8-2. <a href="https://CRAN.R-project.org/package=xtable">https://CRAN.R-project.org/package=xtable</a></p>
</blockquote>

<blockquote>
  <p>H. Wickham. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2009.</p>
</blockquote>

<blockquote>
  <p>Kamil Slowikowski (2016). ggrepel: Repulsive Text and Label Geoms for ‘ggplot2’. R package version 0.5. <a href="https://CRAN.R-project.org/package=ggrepel">https://CRAN.R-project.org/package=ggrepel</a></p>
</blockquote>

<blockquote>
  <p>Guangchuang Yu, Li-Gen Wang, Yanyan Han and Qing-Yu He. clusterProfiler: an R package for comparing biological themes among gene clusters. OMICS: A Journal of Integrative Biology 2012, 16(5):284-287</p>
</blockquote>

<blockquote>
  <p>Guangchuang Yu, Li-Gen Wang, Guang-Rong Yan, Qing-Yu He. DOSE: an R/Bioconductor package for Disease Ontology Semantic and Enrichment analysis. Bioinformatics 2015 31(4):608-609</p>
</blockquote>

<blockquote>
  <p>Marc Carlson (2016). org.Hs.eg.db: Genome wide annotation for Human. R package version 3.3.0.</p>
</blockquote>

<blockquote>
  <p>Marc Carlson (2016). org.Mm.eg.db: Genome wide annotation for Mouse. R package version 3.3.0.</p>
</blockquote>

<blockquote>
  <p>Luo, W. and Brouwer C., Pathview: an R/Bioconductor package for pathway-based data integration and visualization. Bioinformatics, 2013, 29(14): 1830-1831, doi: 10.1093/bioinformatics/btt285</p>
</blockquote>

<blockquote>
  <p>Yuanbin Ru and Katerina Kechris (2014). multiMiR: Integration of multiple microRNA-target databases with their disease and drug associations. R package version 1.0.1.</p>
</blockquote>

<h1 id="session-info">Session Info</h1>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sessionInfo</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## R version 3.3.1 (2016-06-21)
## Platform: x86_64-w64-mingw32/x64 (64-bit)
## Running under: Windows 7 x64 (build 7601) Service Pack 1
## 
## locale:
## [1] LC_COLLATE=English_United States.1252 
## [2] LC_CTYPE=English_United States.1252   
## [3] LC_MONETARY=English_United States.1252
## [4] LC_NUMERIC=C                          
## [5] LC_TIME=English_United States.1252    
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] exprAnalysis_0.1.0
## 
## loaded via a namespace (and not attached):
##  [1] fastcluster_1.1.20    gtools_3.5.0          splines_3.3.1        
##  [4] lattice_0.20-34       colorspace_1.2-6      htmltools_0.3.5      
##  [7] stats4_3.3.1          yaml_2.1.13           chron_2.3-47         
## [10] survival_2.39-5       pcaGoPromoter_1.16.0  foreign_0.8-66       
## [13] DBI_0.5               BiocGenerics_0.18.0   RColorBrewer_1.1-2   
## [16] matrixStats_0.50.2    foreach_1.4.3         plyr_1.8.4           
## [19] stringr_1.1.0         zlibbioc_1.18.0       Biostrings_2.40.2    
## [22] munsell_0.4.3         gtable_0.2.0          caTools_1.17.1       
## [25] codetools_0.2-14      evaluate_0.9          latticeExtra_0.6-28  
## [28] Biobase_2.32.0        knitr_1.14            IRanges_2.6.1        
## [31] doParallel_1.0.10     parallel_3.3.1        AnnotationDbi_1.34.4 
## [34] preprocessCore_1.34.0 Rcpp_0.12.7           acepack_1.3-3.3      
## [37] KernSmooth_2.23-15    flashClust_1.01-2     scales_0.4.0         
## [40] formatR_1.4           gdata_2.17.0          org.Hs.eg.db_3.3.0   
## [43] S4Vectors_0.10.3      Hmisc_3.17-4          WGCNA_1.51           
## [46] XVector_0.12.1        gplots_3.0.1          gridExtra_2.2.1      
## [49] impute_1.46.0         ellipse_0.3-8         ggplot2_2.1.0        
## [52] digest_0.6.10         stringi_1.1.1         grid_3.3.1           
## [55] tools_3.3.1           bitops_1.0-6          magrittr_1.5         
## [58] RSQLite_1.0.0         dynamicTreeCut_1.63-1 Formula_1.2-1        
## [61] cluster_2.0.4         GO.db_3.3.0           Matrix_1.2-7.1       
## [64] data.table_1.9.6      rmarkdown_1.0         iterators_1.0.8      
## [67] rpart_4.1-10          nnet_7.3-12
</code></pre>
</div>

  </div>


<!-- Pagination links -->
<div class="pagination">
  
    <a href="/blog/page4/" class="previous">Previous</a>
  
  <span class="page_number ">Page: 5 of 5</span>
  
    <span class="next ">Next</span>
  
</div>


<div class="pagination">
  
    <a href="/blog/page4/">&laquo; Prev</a>
  

  
    
      <a href="/blog/page4/">1</a>
    
  
    
      <a href="/blog/page2/">2</a>
    
  
    
      <a href="/blog/page3/">3</a>
    
  
    
      <a href="/blog/page4/">4</a>
    
  
    
      <em>5</em>
    
  

  
    <span>Next &raquo;</span>
  
</div>

      </div>

    </div>
	  
	   



	  
  </body>

<div class="footer">
    <div id="footer">
      <div class="container">
        <p>&copy; 2017 Shirin Glander
		<a href="mailto:shirin.glander@gmail.com"><img src="http://localhost:4000/assets/images/200px-Email_Shiny_Icon.png" /></a>
		<a href="http://stackoverflow.com/users/6623620/shirin-glander"><img src="http://localhost:4000/assets/images/so-logo.png" /></a>
		<a href="https://github.com/ShirinG"><img src="http://localhost:4000/assets/images/GitHub_Logo.png" /></a>
		<a href="http://www.xing.com/profile/Shirin_Glander"><img src="http://localhost:4000/assets/images/xing.png" /></a>
		<a href="http://de.linkedin.com/in/shirin-glander-01120881"><img src="http://localhost:4000/assets/images/Logo-2C-101px-R.png" /></a>
          with help from <a href="http://jekyllbootstrap.com" target="_blank" title="The Definitive Jekyll Blogging Framework">Jekyll Bootstrap</a>
          and <a href="http://getbootstrap.com" target="_blank">Bootstrap</a>
        </p>
      </div>
    </div>
	
    <!-- Latest compiled and minified JavaScript, requires jQuery 1.x (2.x not supported in IE8) -->
    <!-- Placed at the end of the document so the pages load faster -->
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.10.2/jquery.min.js"></script>
    <script src="/assets/themes/bootstrap-3/bootstrap/js/bootstrap.min.js"></script>
</div>
</html>

