---
layout: post
title: "How to use plink2 to do human genetics research"
author: Shicheng Guo
date: 1928-02-28
categories: bioinformatics
tags: plink2

---

* how do I prune second-degree-related samples?  —rel-cutoff is obsolete and see [—king-cutoff](https://www.cog-genomics.org/plink/2.0/distance#make_king)

```
plink2 —bfile ... —king-cutoff 0.088 --maf 0.01 —make-bed --out myplink
```


* How to make dbSNP153 vcf format to plink format:
```
## change X to 23
bcftools annotate --rename-chrs chr_name_conv.txt dbSNP153.hg19.chrX.vcf -Oz -o dbSNP153.hg19.chr23.vcf

## db152plink.sh
wget https://raw.githubusercontent.com/Shicheng-Guo/Gscutility/master/db152plink.pl
perl db152plink.pl >  db152.vcf
plink --vcf db152.vcf --make-bed --out db152

## db153plink.sh
wget https://raw.githubusercontent.com/Shicheng-Guo/Gscutility/master/db152plink.pl
perl db152plink.pl >  db152.vcf
plink --vcf db152.vcf --make-bed --out db152

mkdir temp
for i in `ls dbSNP153.hg19.chr*.vcf`
do
echo $i
echo \#PBS -N $i > $i.job
echo \#PBS -l nodes=1:ppn=1 >> $i.job
echo \#PBS -M Guo.shicheng\@marshfieldresearch.org >> $i.job
echo \#PBS -m abe  >> $i.job
echo \#PBS -o $(pwd)/temp/ >>$i.job
echo \#PBS -e $(pwd)/temp/ >>$i.job
echo cd $(pwd) >> $i.job
echo perl addfakegenotype.pl $i\> $i.vcf >>$i.job
qsub  $i.job
done
```
* How to let bcftools output rs ID
```
bcftools query -f '%ID\n' gnomad.genomes.r2.1.sites.chr16.rec.hsa.gff3.sort.rmdup.biallelic.vcf.bgz
```
* vcf2bed

```
for i in `ls gnomad.genomes.r2.1.sites.chr*.rec.hsa.gff3.sort.rmdup.biallelic.vcf.bgz`
do 
bcftools query -f '%CHROM\t%POS\t%ID\n' $i | awk '{print $1,$2-1,$2,$3}' OFS="\t"
done
```
* GenomeAsia100K 
```
# GenomeAsia100K data are available at the links below. Users can download GA100K data in compressed Variant Call Format (VCF) file.
mkdir ~/hpc/db/GenomeAsia100K
cd ~/hpc/db/GenomeAsia100K
for i in {1..22}
do
wget --no-check-certificate https://browser.genomeasia100k.org/service/web/download_files/$i.substitutions.annot.cont_withmaf.vcf.gz &
done
```
* merge plink files from different chrosomes.
```
# Round 1: run merging and try to find dupliates SNPs and then remove them 
rm mergelist.txt
for i in {1..22}
do
echo chr$i >> mergelist.txt
done
plink --merge-list mergelist.txt --make-bed --out 1000plink
# Round 2: remove duplicates SNPs which stored in `1000plink-merge.missnp`
mkdir temp
for i in {1..22}
do
echo $i
echo \#PBS -N $i > $i.job
echo \#PBS -l nodes=1:ppn=1 >> $i.job
echo \#PBS -M Guo.shicheng\@marshfieldresearch.org >> $i.job
echo \#PBS -m abe  >> $i.job
echo \#PBS -o $(pwd)/temp/ >>$i.job
echo \#PBS -e $(pwd)/temp/ >>$i.job
echo \#PBS -m abe  >> $i.job
echo cd $(pwd) >> $i.job
echo plink --bfile chr$i --alleleACGT --snps-only just-acgt --exclude 1000plink-merge.missnp --make-bed --out chr$i.uni >> $i.job
qsub  $i.job
done
# Round 3: Try to merge again
rm mergelist.txt
for i in {1..22}
do
echo chr$i >> mergelist.txt
done
plink --merge-list mergelist.txt --make-bed --out 1000plink
```
