---
layout: post
title: "How to use plink2 to do human genetics research"
author: Shicheng Guo
date: 1928-02-28
categories: bioinformatics
tags: plink2

---

* how do I prune second-degree-related samples?  —rel-cutoff is obsolete and see [—king-cutoff](https://www.cog-genomics.org/plink/2.0/distance#make_king)

```
plink2 —bfile ... —king-cutoff 0.088 --maf 0.01 —make-bed --out myplink
```


* How to make dbSNP153 vcf format to plink format:
```
## change X to 23
bcftools annotate --rename-chrs chr_name_conv.txt dbSNP153.hg19.chrX.vcf -Oz -o dbSNP153.hg19.chr23.vcf

## db152plink.sh
wget https://raw.githubusercontent.com/Shicheng-Guo/Gscutility/master/db152plink.pl
perl db152plink.pl >  db152.vcf
plink --vcf db152.vcf --make-bed --out db152

## db153plink.sh
wget https://raw.githubusercontent.com/Shicheng-Guo/Gscutility/master/db152plink.pl
perl db152plink.pl >  db152.vcf
plink --vcf db152.vcf --make-bed --out db152

mkdir temp
for i in `ls dbSNP153.hg19.chr*.vcf`
do
echo $i
echo \#PBS -N $i > $i.job
echo \#PBS -l nodes=1:ppn=1 >> $i.job
echo \#PBS -M Guo.shicheng\@marshfieldresearch.org >> $i.job
echo \#PBS -m abe  >> $i.job
echo \#PBS -o $(pwd)/temp/ >>$i.job
echo \#PBS -e $(pwd)/temp/ >>$i.job
echo cd $(pwd) >> $i.job
echo perl addfakegenotype.pl $i\> $i.vcf >>$i.job
qsub  $i.job
done
```
