---
layout: post
title: "How to generate dbSNP153(hg19) from dbSNP153(hg38,GRCH38)"
author: Shicheng Guo
date: 2019-07-09
categories: bioinformatics
tags: dbSNP153 GRCH37 hg19 GRCH38 hg38
image: images/dbSNP153_Shicheng_Guo_2019_07_09.JPG

---

Recently (07/09/2019), dbSNP have been updated [dbSNP153](https://ftp.ncbi.nih.gov/snp/redesign/latest_release/VCF/GCF_000001405.38.gz) from dbSNP152. However, NCBI only provided dbSNP153 in hg38 ([GRCH38](http://mirrors.vbi.vt.edu/mirrors/ftp.ncbi.nih.gov/genomes/refseq/vertebrate_mammalian/Homo_sapiens/all_assembly_versions/GCF_000001405.38_GRCh38.p12/GCF_000001405.38_GRCh38.p12_assembly_report.txt)) version without any source for hg19 version. Here, I prepared a approach to generate dbSNP153 in hg19..

#### Method to generate classic dbSNP153(hg19,GRCH37) from dbSNP153(hg19,GCF)

```
wget https://ftp.ncbi.nih.gov/snp/redesign/latest_release/VCF/GCF_000001405.25.gz
wget https://ftp.ncbi.nih.gov/snp/redesign/latest_release/VCF/GCF_000001405.25.gz.tbi
wget https://raw.githubusercontent.com/Shicheng-Guo/AnnotationDatabase/master/GCF_000001405.25_GRCh37.p13_assembly_report.txt

awk -v RS="(\r)?\n" 'BEGIN { FS="\t" } !/^#/ { if ($10 != "na") print $7,$10; else print $7,$5 }' GCF_000001405.25_GRCh37.p13_assembly_report.txt > dbSNP-to-UCSC-GRCh37.p13.map
perl -p -i -e '{s/chr//}' dbSNP-to-UCSC-GRCh37.p13.map

bcftools annotate --rename-chrs dbSNP-to-UCSC-GRCh37.p13.map GCF_000001405.25.gz -Oz -o dbSNP153.hg19.vcf.gz
```
#### Method to generate dbSNP153(hg19,GRCH37) from dbSNP153(hg38,GRCH38)

Bash, Perl, Python and (GATK or CrossMap), awk and wget are required in this approach. Crossmap is Python based method which I don't recommend to use since the version problem may waste tons of your time. I prefer to use GATK4 LiftoverVcf.

```
wget https://ftp.ncbi.nih.gov/snp/redesign/latest_release/VCF/GCF_000001405.38.gz
wget https://ftp.ncbi.nih.gov/snp/redesign/latest_release/VCF/GCF_000001405.38.gz.tbi
wget https://raw.githubusercontent.com/Shicheng-Guo/AnnotationDatabase/master/GCF_000001405.38_GRCh38.p12_assembly_report.txt
awk -v RS="(\r)?\n" 'BEGIN { FS="\t" } !/^#/ { if ($10 != "na") print $7,$10; else print $7,$5 }' GCF_000001405.38_GRCh38.p12_assembly_report.txt > dbSNP-to-UCSC-GRCh38.p12.map
bcftools annotate --rename-chrs dbSNP-to-UCSC-GRCh38.p12.map GCF_000001405.38.gz | gawk '/^#/ && !/^##contig=/ { print } !/^#/ { if( $1!="na" ) print }' | bgzip -c > GCF_000001405.38.dbSNP153.GRCh38p12b.GATK.vcf.gz
gatk LiftoverVcf -I GCF_000001405.38.dbSNP152.GRCh38p12b.GATK.vcf.gz -O dbSNP153.hg19.vcf -C hg38ToHg19.over.chain.gz --REJECT rejected.vcf -R ~/hpc/db/hg19/hg19.fa
```
liftOver with `python CrossMap.py`
```
python CrossMap.py vcf hg38Tohg19.over.chain.gz GCF_000001405.38.dbSNP153.GRCh38p12b.GATK.vcf.gz ~/hpc/db/hg19/hg19.fa  GCF_000001405.38.dbSNP153.hg19.gz
perl -p -i -e '{s/chr//}' dbSNP-to-UCSC-GRCh38.p12.map
```
02/16/2020
```
wget https://ftp.ncbi.nih.gov/snp/redesign/latest_release/VCF/GCF_000001405.25.gz
wget https://ftp.ncbi.nih.gov/snp/redesign/latest_release/VCF/GCF_000001405.25.gz.tbi
wget https://raw.githubusercontent.com/Shicheng-Guo/AnnotationDatabase/master/GCF_000001405.25_GRCh37.p13_assembly_report.txt
awk -v RS="(\r)?\n" 'BEGIN { FS="\t" } !/^#/ { if ($10 != "na") print $7,$10; else print $7,$5 }' GCF_000001405.25_GRCh37.p13_assembly_report.txt > dbSNP-to-UCSC-GRCh37.p13.map
perl -p -i -e '{s/chr//}' dbSNP-to-UCSC-GRCh37.p13.map
bcftools annotate --rename-chrs dbSNP-to-UCSC-GRCh37.p13.map GCF_000001405.25.gz | gawk '/^#/ && !/^##contig=/ { print } !/^#/ { if( $1!="na" ) print }' | bgzip -c > dbSNP153.hg19.vcf.gz
```
03/11/2020 (again, I moved to UW-Madison, department of Medical Genetics, I need to rebuild the dbSNP153 in deepThought server)

#### How to share dbSNP153(hg19,GRCH37) and how to download dbSNP153(hg19,GRCH37)
In order to share dbSNP153(hg19,GRCH37), I uploaded dbSNP153(hg19,GRCH37) to . You can download dbSNP153(hg19,GRCH37) with the follow link. Good luck for your analysis. Finally, Thanks to Dr. Raony Guimar√£es for the help on the ideas of CrossMap to liftover vcf files.. 











