---
layout: post
title: "How to Install HTseq in Linux and STAR for RNA-seq"
author: Shicheng Guo
date: 1941-01-08
categories: research
tags: HTseq STAR RNAseq FPKM TPM GTF HG19 Human Expression Transcript

---

The current post want to use STAR and HTseq together to estimation gene expression for RNA-seq: 

How to install lastest version of HTseq (Htseq-count)? HTseq is python based approach (>2.7 or 3.4). 

* Requirement:

```
pip install deeptoolsintervals
pip install matplotlib
pip install numpydoc
pip install plotly
pip install py2bit
pip install pyBigWig
pip install scipy
```

* and then 

```
pip install HTSeq
```

* Here I suppose you use STAR to mapping RNA-seq fastq to human genome (hg19) 

* before run `htseq-count`, you'd better to download human gtf files like this way: 
```
wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/GRCh37_mapping/gencode.v32lift37.annotation.gtf.gz
gunzip gencode.v32lift37.annotation.gtf.gz
```
* and now run `htseq` to the bam files created by `STAR` to generate FPKM matrix
```
htseq-count -m intersection-nonempty -t exon -i gene_id -f bam STAR.output.bam gencode.v32lift37.annotation.gtf -o output
```

* if you do not want to use HTseq, you can also try `cufflinks`

```
cufflinks --library-type fr-firststrand --outFilterIntronMotifs RemoveNoncanonical
```

* How to infer the RNA-seq library type by [salmon](https://salmon.readthedocs.io/en/latest/salmon.html) if you don't have such information? 

* before run salmon, you need to download fasta file for the transcripts like the following example:
```
wget http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/GRCh37_mapping/gencode.v32lift37.transcripts.fa.gz
gunzip gencode.v32lift37.transcripts.fa.gz
```

* then now you can run selmon to infer the RNA-seq library type:

```
salmon quant -t gencode.v32lift37.transcripts.fa -l A -a UW040LPS_ATTCCT_L002Aligned.sortedByCoord.out.bam -o salmon_quant
```



