---
layout: post
title: "Genome-wide cell-free DNA fragmentation in patients with cancer"
author: Shicheng Guo
date: 2019-08-19
categories: flexdashboard
tags: cfDNA fragmentation cancer
image: images/Nature_cfDNA_fragmentation_nature_letter_2019.jpg

---
Today, I will give a talk about "Genome-wide cell-free DNA fragmentation in patients with cancer" in MCRI postdoc club meeting. The paper was published in [Nature](https://www.nature.com/articles/s41586-019-1272-6) recently. See the [slice](https://github.com/Shicheng-Guo/Shicheng-Guo.Github.io/blob/master/slice/Postdoc_Club_2019_08_19_cfDNA_Fragement.pptx) here or [download here](https://raw.githubusercontent.com/Shicheng-Guo/Shicheng-Guo.Github.io/master/slice/Postdoc_Club_2019_08_19_cfDNA_Fragement.pptx) for more details..

### cfDNA fragmentation in human cancers
Cell-free DNA in the blood provides a non-invasive diagnostic avenue for patients with cancer1. However, characteristics of the origins and molecular features of cell-free DNA are poorly understood. Here we developed an approach to evaluate fragmentation patterns of cell-free DNA across the genome, and found that profiles of healthy individuals reflected nucleosomal patterns of white blood cells, whereas patients with cancer had altered fragmentation profiles. We used this method to analyse the fragmentation profiles of 236 patients with breast, colorectal, lung, ovarian, pancreatic, gastric or bile duct cancer and 245 healthy individuals. A machine learning model that incorporated genome-wide fragmentation features had sensitivities of detection ranging from 57% to more than 99% among the seven cancer types at 98% specificity, with an overall area under the curve value of 0.94. Fragmentation profiles could be used to identify the tissue of origin of the cancers to a limited number of sites in 75% of cases. Combining our approach with mutation-based cell-free DNA analyses detected 91% of patients with cancer. The results of these analyses highlight important properties of cell-free DNA and provide a proof-of-principle approach for the screening, early detection and monitoring of human cancer.

### Cell-Free DNA General Background
* Pan et al ["Brain Tumor Mutations Detected in Cerebral Spinal Fluid"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/cfDNA_background/Pan%20et%20al.pdf)
* Schilling and Rehli ["Global, comparative analysis of tissue-specific
promoter CpG methylation"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/cfDNA_background/Schilling%20and%20Rehli.pdf)
* Jahr et al ["DNA Fragments in the Blood Plasma of Cancer
Patients: Quantitations and Evidence for Their Origin from Apoptotic
and Necrotic Cells"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/cfDNA_background/Jahr%20et%20al%20.pdf)
* Van der Vaart and Pretorius ["Circulating DNAIts Origin and Fluctuation"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/cfDNA_background/Van_der_Vaart_and_Pretorious.pdf)
* Stroun et al ["The Origin and Mechanism of Circulating DNA"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/cfDNA_background/Stroun%20et%20al.pdf)
* Gravina et al ["The dark side of circulating nucleic acids" ](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/cfDNA_background/Gravina%20et%20al.pdf)
* Sun et al ["Plasma DNA tissue mapping by genome-wide methylation
sequencing for noninvasive prenatal, cancer, and
transplantation assessments"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/cfDNA_background/Sun_et_al.pdf)
* Wyatt et al ["Concordance of Circulating
Tumor DNA and MatchedMetastatic Tissue
Biopsy in Prostate Cancer"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/cfDNA_background/Wyatt_et_al.pdf)
* Ma et al ["Cell-Free DNA Provides a Good Representationof the Tumor Genome Despite Its Biased Fragmentation Patterns"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/cfDNA_background/Ma_et_al.pdf)
#### Cell-Free DNA and Methylation
* Lehman-Werman et al  ["Identification of tissue-specific cell
death using methylation patterns of circulating DNA"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/tissue_specific_cfDNA/Lehman-Werman_et_al.pdf)
* Synder et al ["Cell-free DNA Comprises an In Vivo Nucleosome
Footprint that Informs Its Tissues-Of-Origin"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/tissue_specific_cfDNA/Synder_et_al.pdf)
* Jensen et al ["Whole genome bisulfite sequencing of
cell-free DNA and its cellular contributors
uncovers placenta hypomethylated domains"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/tissue_specific_cfDNA/Jensen_et_al.pdf)
* Guo et al ["Identification of methylation haplotype
blocks aids in deconvolution of heterogeneous
tissue samples and tumor
tissue-of-origin mapping from plasma DNA"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/tissue_specific_cfDNA/Guo_et_al.pdf)
* Seoighe et al (response to Guo et al)["DNA methylation haplotypes as cancer markers"](https://www.nature.com/articles/s41588-018-0185-x)
* Diep and Zhang (response to Seoighe et al) ["Reply to â€˜DNA methylation haplotypes as cancer markers'"](https://www.nature.com/articles/s41588-018-0186-9)

* Kessler et al ["CpG methylation differences betweenneurons and glia are highlyconserved from mouse to human"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/tissue_specific_cfDNA/Kessler_et_al.pdf)
* Luck et al ["A Stochastic Model for the Formation of SpatialMethylation Patterns"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/cfDNA_background/Luck_preprint.pdf)
* Doherty and Couldrey ["Exploring genome wide bisulfite sequencing for DNA methylation analysis in livestock: a technical assessment"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/cfDNA_background/Doherty_and_Couldrey.pdf)
* Intro methylation review: ["DNA methylation: a form ofepigenetic control of gene expression"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/cfDNA_background/intro_methylation_review.pdf)
* Koch et al ["Analysis of DNA methylation in cancer: location revisited"](https://www.nature.com/articles/s41571-018-0004-4)
* Smith and Meissner ["DNA methylation: roles in mammalian development"](https://www.nature.com/articles/nrg3354)

#### Non-CpG Methylation
* Xu and Corces ["Nascent DNA methylome mapping reveals inheritance of hemimethylation at CTCF/cohesin sites"](http://science.sciencemag.org/content/359/6380/1166.full)

#### Deconvolution algorithms
* Houseman et al ["Reference-freedeconvolution of DNA methylationdata and mediation by cellcomposition effects" ](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/cfDNA_tissue_of_origin_deconvolution/Houseman_et_al.pdf)
* Rahmani et al ["Sparse PCA Corrects forCell-Type Heterogeneity in Epigenome-WideAssociation Studies"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/cfDNA_tissue_of_origin_deconvolution/Rahmani_et_al.pdf)
* Gond and Szustakowski ["DeconRNASeq: a statistical framework fordeconvolution of heterogeneoustissue samples based on mRNA-Seq data"](https://github.com/christacaggiano/cell-free-dna-readinglist/blob/master/cfDNA_tissue_of_origin_deconvolution/Gong%20and%20Szustakowski.pdf)

#### WGBS Algorithms
* Ziller et al ["Coverage recommendations for methylation analysis by whole genome bisulfite sequencing"](https://www.nature.com/articles/nmeth.3152)
* Merkel et al ["gemBS - high throughput processing for DNA methylation data from Bisulfite Sequencing
"](https://academic-oup-com.ucsf.idm.oclc.org/bioinformatics/advance-article/doi/10.1093/bioinformatics/bty690/5077236)
* Lee et al ["An integrative approach for efficient
analysis of whole genome bisulfite sequencing data"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/cfDNA_tissue_of_origin_deconvolution/Lee_et_al.pdf)
* Kunde-Ramamoorty et al ["Comparison and quantitative
verification of mapping algorithms
for whole-genome bisulfite sequencing"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/cfDNA_tissue_of_origin_deconvolution/Kunde-Ramamoorthy_et_al.pdf)
* Xi and Li ["BSMAP: whole genome bisulfite
sequence MAPping program"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/cfDNA_tissue_of_origin_deconvolution/Xi%20and%20Li.pdf)
* Guo et al []"BS-Seeker2: a versatile aligning pipeline for bisulfite sequencing data"](https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-14-774)
* Krueger and Andrews ["Bismark: a
flexible aligner and methylation
caller for Bisulfite-Seq applications"](https://github.com/christacaggiano/cell-free-dna-reading-list/blob/master/cfDNA_tissue_of_origin_deconvolution/Krueger%20and%20Andrews.pdf)
* Feng et al ["Disease prediction by cell-free DNA methylation"](https://academic-oup-com.ucsf.idm.oclc.org/bib/advance-article/doi/10.1093/bib/bby029/4973009)

> see the [slice](https://github.com/Shicheng-Guo/Shicheng-Guo.Github.io/blob/master/slice/Postdoc_Club_2019_08_19_cfDNA_Fragement.pptx) for the presentation. Also you can [download here](https://raw.githubusercontent.com/Shicheng-Guo/Shicheng-Guo.Github.io/master/slice/Postdoc_Club_2019_08_19_cfDNA_Fragement.pptx) for more details..
