

<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">

    <title>Exploring the human genome (Part 2) - Transcripts</title>
    
    <meta name="author" content="Shirin Glander">

    <!-- Enable responsive viewport -->
    <meta name="viewport" content="width=device-width, initial-scale=1.0">

    <!-- Bootstrap styles -->
    <link href="/assets/themes/bootstrap-3/bootstrap/css/bootstrap.min.css" rel="stylesheet">
    <!-- Optional theme -->
    <link href="/assets/themes/bootstrap-3/bootstrap/css/bootstrap-theme.min.css" rel="stylesheet">
    <!-- Sticky Footer -->
    <link href="/assets/themes/bootstrap-3/bootstrap/css/bs-sticky-footer.css" rel="stylesheet">

    <!-- Custom styles -->
    <link href="/assets/themes/bootstrap-3/css/style.css?body=1" rel="stylesheet" type="text/css" media="all">

    <!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
      <script src="https://oss.maxcdn.com/libs/respond.js/1.3.0/respond.min.js"></script>
    <![endif]-->
	  
    <!-- Fav and touch icons -->
    <!-- Update these with your own images      -->
      <link rel="shortcut icon" type="image/x-icon" href="http://localhost:4000/assets/images/iconified/favicon.ico">
      <!-- <link rel="shortcut icon" type="image/png" href="http://localhost:4000/assets/images/iconified/apple-touch-icon.png"> -->
      <link rel="apple-touch-icon" type="image/png" href="http://localhost:4000/assets/images/iconified/apple-touch-icon.png">
      <link rel="apple-touch-icon" type="image/png" sizes="76x76" href="http://localhost:4000/images/iconified/apple-touch-icon-76x76.png">
      <link rel="apple-touch-icon" type="image/png" sizes="114x114" href="http://localhost:4000/images/iconified/apple-touch-icon-114x114.png">

    <!-- atom & rss feed -->
    <link href="/atom.xml" type="application/atom+xml" rel="alternate" title="Sitewide ATOM Feed">
    <link href="/rss.xml" type="application/rss+xml" rel="alternate" title="Sitewide RSS Feed">
  </head>
  
 <div class="header">
      <div id="wrap">
      <nav class="navbar navbar-default" role="navigation">
        <!-- Brand and toggle get grouped for better mobile display -->
        <div class="navbar-header">
          <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#jb-navbar-collapse">
            <span class="sr-only">Toggle navigation</span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
          </button>
          <!-- <a class="navbar-brand" href="/">Shirin's playgRound</a> -->
		  <a class="navbar-brand" href="/"><img src="https://raw.githubusercontent.com/ShirinG/ShirinG.github.io/master/assets/images/logo.png" alt="logo" />Shirin's playgRound</a>
        </div>

        <!-- Collect the nav links, forms, and other content for toggling -->
        <div class="collapse navbar-collapse" id="jb-navbar-collapse">
          <ul class="nav navbar-nav">
            
            
            


  
    
      
    
  
    
      
    
  
    
      
      	
      	<li><a href="/about">About me</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/archive">Archive</a></li>
      	
      
    
  
    
      
    
  
    
      
      	
      	<li><a href="/categories">Categories</a></li>
      	
      
    
  
    
      
      	
      	<li><a href="/feeds">Feeds</a></li>
      	
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
      	
      	<li><a href="/tags">Tags</a></li>
      	
      
    
  
    
  
    
  
    
  
    
      
    
  
    
      
    
  
    
      
    
  
    
      
    
  



          </ul>
          <form class="navbar-form navbar-right" role="search">
            <div class="form-group">
              <input type="text" class="form-control" placeholder="Search">
            </div>
            <button type="submit" class="btn btn-default">Submit</button>
          </form>
        </div><!-- /.navbar-collapse -->
      </nav>

    </div>
</div>

  <body>
    <div id="wrap">
      <div class="container">
        

<div class="page-header">
  <h1>Exploring the human genome (Part 2) - Transcripts </h1>
</div>

<!-- Paste the 3 next lines where you want the sharing button(s) to appear -->
    <div class="post-sharing">
     

  
  		<div id="fb-root"></div>

<ul class="post-share ulno mob">

<!-- Twitter -->
<li class="tw"><a href="https://twitter.com/share" class="twitter-share-button" data-text="Exploring the human genome (Part 2) - Transcripts" data-via="" data-related="" data-count="" data-size="">Tweet</a>

<!-- Google+ -->
<li class="gp"><div class="g-plusone" data-size="medium" data-annotation="bubble" data-width=""></div>

<!-- Facebook -->
<li class="fb"><div class="fb-like" data-send="false" data-layout="button_count" data-width="90" data-show-faces="false" data-font=""></div>

<!-- Reddit -->
<li><script type="text/javascript" src="http://www.reddit.com/buttonlite.js?i=4"></script>
</ul>

<script>
  
(function(doc, script) {
 	
    // Async Social Buttons
    var js, 
        fjs = doc.getElementsByTagName(script)[0],
        add = function(url, id) {
            if (doc.getElementById(id)) {return;}
            js = doc.createElement(script);
            js.src = url;
            id && (js.id = id);
            fjs.parentNode.insertBefore(js, fjs);
        };

    // Twitter SDK
    add('//platform.twitter.com/widgets.js', 'twitter-wjs');
    
    // Google+ button
    add('https://apis.google.com/js/plusone.js');
    
    // Facebook SDK
    add('https://connect.facebook.net/en_GB/all.js#xfbml=1&appId=1424112504267565', 'facebook-jssdk');
    
}(document, 'script'));

</script>
  



    </div>

<div class="row post-full">
  <div class="col-xs-12">
    <div class="date">
      <span>01 November 2016</span>
    </div>
    <div class="content">
      <h1 id="how-many-transcripts-and-proteins-do-genes-have">How many transcripts and proteins do genes have?</h1>

<p>In <a href="https://shiring.github.io/genome/2016/10/23/AnnotationDbi">Exploring the human genome (Part 1) - Gene Annotations</a> I examined Ensembl, Entrez and HGNC gene annotations with <a href="http://www.bioconductor.org/packages/release/bioc/html/AnnotationDbi.html">AnnotationDbi</a> via three R packages: <a href="http://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html">org.Hs.eg.db</a>, <a href="https://bioconductor.org/packages/release/data/annotation/html/EnsDb.Hsapiens.v79.html">EnsDb.Hsapiens.v79</a> and <a href="http://bioconductor.org/packages/release/data/annotation/html/TxDb.Hsapiens.UCSC.hg38.knownGene.html">TxDb.Hsapiens.UCSC.hg38.knownGene</a>.</p>

<p>Now, I want to know how many transcripts there are for genes in these databases.</p>

<p><br /></p>

<h3 id="what-is-a-transcript">What is a transcript?</h3>

<p>While a gene is defined as a unit of DNA information which encodes for the production of a protein, it is really more a concept than an actual physical unit. Human genes conists of <a href="https://en.wikipedia.org/wiki/Exon">exons</a> and <a href="https://en.wikipedia.org/wiki/Intron">introns</a>, which can often be transcribed in different combinations - a process called <a href="https://en.wikipedia.org/wiki/Alternative_splicing">alternative splicing</a>.</p>

<blockquote>
  <p>“While the concept of a gene has been helpful in defining the relationship of a portion of a genome to a phenotype, this traditional term may not be as useful as it once was. Currently, “gene” has come to refer principally to a genomic region producing a polyadenylated mRNA that encodes a protein. However, the recent emergence of a large collection of unannotated transcripts with apparently little protein coding capacity, collectively called transcripts of unknown function (TUFs), has begun to blur the physical boundaries and genomic organization of genic regions with noncoding transcripts often overlapping protein-coding genes on the same (sense) and opposite strand (antisense). Moreover, they are often located in intergenic regions, making the genic portions of the human genome an interleaved network of both annotated polyadenylated and nonpolyadenylated transcripts, including splice variants with novel 5′ ends extending hundreds of kilobase. This complex transcriptional organization and other recently observed features of genomes argue for the reconsideration of the term “gene” and suggests that transcripts may be used to define the operational unit of a genome.” <a href="http://genome.cshlp.org/content/17/6/682.full">Thomas Gingeras, Genome Res. 2007</a></p>
</blockquote>

<p><br /></p>

<h2 id="orghsegdb">org.Hs.eg.db</h2>

<p>With org.Hs.eg.db I am using Entrez and Ensembl IDs to obtain Ensembl transcript IDs, which show splice variants of a gene, including protein-coding and non-coding transcripts.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">AnnotationDbi</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">)</span><span class="w">

</span><span class="n">ENTREZID_org</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">)</span><span class="w">
</span><span class="n">ENSEMBL_org</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENSEMBL"</span><span class="p">)</span><span class="w">

</span><span class="c1"># Summarize number of transcripts per gene Entrez ID
</span><span class="n">org_trans_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">AnnotationDbi</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ENTREZID_org</span><span class="p">,</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"ENSEMBLTRANS"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">)</span><span class="w">
</span><span class="n">org_transcript_num_table_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">org_trans_entrez</span><span class="o">$</span><span class="n">ENTREZID</span><span class="p">))</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">org_transcript_num_table_entrez</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Entrez"</span><span class="p">,</span><span class="w"> </span><span class="s2">"orgDb"</span><span class="p">)</span><span class="w">

</span><span class="c1"># Summarize number of transcripts per gene Ensembl ID
</span><span class="n">org_trans_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">AnnotationDbi</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">org.Hs.eg.db</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ENSEMBL_org</span><span class="p">,</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"ENSEMBLTRANS"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENSEMBL"</span><span class="p">)</span><span class="w">
</span><span class="n">org_transcript_num_table_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">org_trans_ensembl</span><span class="o">$</span><span class="n">ENSEMBL</span><span class="p">))</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">org_transcript_num_table_ensembl</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Ensembl"</span><span class="p">,</span><span class="w"> </span><span class="s2">"orgDb"</span><span class="p">)</span><span class="w">

</span><span class="c1"># how many NAs are in each column?
</span><span class="n">sapply</span><span class="p">(</span><span class="n">org_trans_entrez</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##     ENTREZID ENSEMBLTRANS 
##            0        52081
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sapply</span><span class="p">(</span><span class="n">org_trans_ensembl</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##      ENSEMBL ENSEMBLTRANS 
##            0        17678
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">head</span><span class="p">(</span><span class="n">org_trans_entrez</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   ENTREZID    ENSEMBLTRANS
## 1        1            &lt;NA&gt;
## 2        2            &lt;NA&gt;
## 3        3 ENST00000543404
## 4        3 ENST00000566278
## 5        3 ENST00000545343
## 6        3 ENST00000544183
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">head</span><span class="p">(</span><span class="n">org_trans_ensembl</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##           ENSEMBL    ENSEMBLTRANS
## 1 ENSG00000121410            &lt;NA&gt;
## 2 ENSG00000175899            &lt;NA&gt;
## 3 ENSG00000256069 ENST00000543404
## 4 ENSG00000256069 ENST00000566278
## 5 ENSG00000256069 ENST00000545343
## 6 ENSG00000256069 ENST00000544183
</code></pre>
</div>

<p>Strangely, some genes, like A1BG (Entrez ID 1) are listed with one gene ID but no Ensembl transcript ID. This is weird, especially since I happen to know for this particular gene that it has several transcripts. Also, because each gene must have at least one transcript, all NAs are counted as 1 transcript in the summary table.</p>

<p>Let’s check other databases…</p>

<p><br /></p>

<h2 id="txdbhsapiensucschg38knowngene">TxDb.Hsapiens.UCSC.hg38.knownGene</h2>

<p>TxDb.Hsapiens.UCSC.hg38.knownGene only has Entrez IDs to identify genes and UCSC transcript ID to identify transcripts.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="p">)</span><span class="w">

</span><span class="n">ENTREZID_TxDb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">
</span><span class="n">TxDb_trans</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">AnnotationDbi</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ENTREZID_TxDb</span><span class="p">,</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"TXID"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">

</span><span class="c1"># Summarize number of transcripts per gene Entrez ID
</span><span class="n">TxDb_transcript_num_table_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">TxDb_trans</span><span class="o">$</span><span class="n">GENEID</span><span class="p">))</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">TxDb_transcript_num_table_entrez</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Entrez"</span><span class="p">,</span><span class="w"> </span><span class="s2">"TxDb"</span><span class="p">)</span><span class="w">

</span><span class="c1"># how many NAs are in each column?
</span><span class="n">sapply</span><span class="p">(</span><span class="n">TxDb_trans</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GENEID   TXID 
##      0      0
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">head</span><span class="p">(</span><span class="n">TxDb_trans</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   GENEID   TXID
## 1      1 166436
## 2      1 166437
## 3      1 166438
## 4      1 166439
## 5      1 166440
## 6      1 166441
</code></pre>
</div>

<p>Here, we don’t have NAs in the data and in contrast to org.Hs.eg.db we find at least six transcripts for A1BG.</p>

<p><br /></p>

<h2 id="ensdbhsapiensv79">EnsDb.Hsapiens.v79</h2>

<p>Like org.Hs.eg.db EnsDb.Hsapiens.v79 has Entrez and Ensembl IDs in reference to Ensembl transcript IDs.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">)</span><span class="w">

</span><span class="n">ENSEMBL_EnsDb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">
</span><span class="n">ENTREZ_EnsDb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">)</span><span class="w">

</span><span class="n">EnsDb_trans_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ENSEMBL_EnsDb</span><span class="p">,</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"TXID"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">
</span><span class="n">EnsDb_trans_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ENTREZ_EnsDb</span><span class="p">,</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"TXID"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">)</span><span class="w">

</span><span class="c1"># somehow there are empty fields in the Entrez ID column, replacing them with NA
</span><span class="n">EnsDb_trans_entrez</span><span class="p">[</span><span class="n">EnsDb_trans_entrez</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">""</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kc">NA</span><span class="w">

</span><span class="c1"># how many NAs are in each column?
</span><span class="n">sapply</span><span class="p">(</span><span class="n">EnsDb_trans_entrez</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## ENTREZID     TXID 
##    47828        0
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># and removing NA rows
</span><span class="n">EnsDb_trans_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">EnsDb_trans_entrez</span><span class="p">[</span><span class="o">!</span><span class="nf">is.na</span><span class="p">(</span><span class="n">EnsDb_trans_entrez</span><span class="o">$</span><span class="n">ENTREZID</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># order by Entrez ID to compare with other databases
</span><span class="n">EnsDb_trans_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">EnsDb_trans_entrez</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">EnsDb_trans_entrez</span><span class="o">$</span><span class="n">ENTREZID</span><span class="p">),]</span><span class="w">

</span><span class="n">head</span><span class="p">(</span><span class="n">EnsDb_trans_ensembl</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##            GENEID            TXID
## 1 ENSG00000000003 ENST00000373020
## 2 ENSG00000000003 ENST00000496771
## 3 ENSG00000000003 ENST00000494424
## 4 ENSG00000000003 ENST00000612152
## 5 ENSG00000000003 ENST00000614008
## 6 ENSG00000000005 ENST00000373031
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">head</span><span class="p">(</span><span class="n">EnsDb_trans_entrez</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##        ENTREZID            TXID
## 95915         1 ENST00000596924
## 95916         1 ENST00000263100
## 95917         1 ENST00000595014
## 95918         1 ENST00000598345
## 95919         1 ENST00000600966
## 135994       10 ENST00000286479
</code></pre>
</div>

<p>Here, we find five transcripts for A1BG.</p>

<p><br /></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># Summarize number of transcripts per gene Entrez ID
</span><span class="n">EnsDb_transcript_num_table_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">EnsDb_trans_entrez</span><span class="o">$</span><span class="n">ENTREZID</span><span class="p">))</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">EnsDb_transcript_num_table_entrez</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Entrez"</span><span class="p">,</span><span class="w"> </span><span class="s2">"EnsDb"</span><span class="p">)</span><span class="w">

</span><span class="c1"># Summarize number of transcripts per gene Ensembl ID
</span><span class="n">EnsDb_transcript_num_table_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">EnsDb_trans_ensembl</span><span class="o">$</span><span class="n">GENEID</span><span class="p">))</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">EnsDb_transcript_num_table_ensembl</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Ensembl"</span><span class="p">,</span><span class="w"> </span><span class="s2">"EnsDb"</span><span class="p">)</span><span class="w">

</span><span class="c1"># In the Entrez column, there are some with multiple entries
# divide entries with multiple gene names into one row per gene/ entry
</span><span class="n">head</span><span class="p">(</span><span class="n">EnsDb_transcript_num_table_entrez</span><span class="p">[</span><span class="n">grep</span><span class="p">(</span><span class="s2">";"</span><span class="p">,</span><span class="w"> </span><span class="n">EnsDb_transcript_num_table_entrez</span><span class="o">$</span><span class="n">Entrez</span><span class="p">),</span><span class="w"> </span><span class="p">])</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##                                     Entrez EnsDb
## 13                     100033415;100033421     1
## 15                     100033417;100033419     2
## 18                     100033421;100033415     1
## 40                     100033446;100033449     1
## 42 100033448;100033803;100033817;100033810     2
## 43                     100033449;100033446     1
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">splitstackshape</span><span class="p">)</span><span class="w">
</span><span class="n">out</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">cSplit</span><span class="p">(</span><span class="n">EnsDb_transcript_num_table_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">splitCols</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Entrez"</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">";"</span><span class="p">,</span><span class="w"> </span><span class="n">direction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"long"</span><span class="p">),</span><span class="w"> </span><span class="n">stringsAsFactors</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span><span class="n">out</span><span class="o">$</span><span class="n">Entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">out</span><span class="o">$</span><span class="n">Entrez</span><span class="p">)</span><span class="w">

</span><span class="c1"># remove duplicates and take the mean
</span><span class="n">library</span><span class="p">(</span><span class="n">plyr</span><span class="p">)</span><span class="w">
</span><span class="n">EnsDb_transcript_num_table_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ddply</span><span class="p">(</span><span class="n">out</span><span class="p">,</span><span class="w"> </span><span class="s2">"Entrez"</span><span class="p">,</span><span class="w"> </span><span class="n">numcolwise</span><span class="p">(</span><span class="n">mean</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<h2 id="comparison-of-transcript-numbers-per-database-package">Comparison of transcript numbers per database package</h2>

<p>The majority of genes have only one transcript. The number of genes with more transcripts decreases with number of transcripts; this can be seen in the plots below.</p>

<h3 id="entrez-ids">Entrez IDs</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># merging datasets by Entrez ID
</span><span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span><span class="w">
</span><span class="n">transcript_num_table_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">full_join</span><span class="p">(</span><span class="n">org_transcript_num_table_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">TxDb_transcript_num_table_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Entrez"</span><span class="p">)</span><span class="w">
</span><span class="n">transcript_num_table_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">full_join</span><span class="p">(</span><span class="n">transcript_num_table_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">EnsDb_transcript_num_table_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Entrez"</span><span class="p">)</span><span class="w">

</span><span class="c1"># gather for plotting
</span><span class="n">library</span><span class="p">(</span><span class="n">tidyr</span><span class="p">)</span><span class="w">
</span><span class="n">transcript_num_table_gather_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">transcript_num_table_entrez</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">DB</span><span class="p">,</span><span class="w"> </span><span class="n">count</span><span class="p">,</span><span class="w"> </span><span class="n">orgDb</span><span class="o">:</span><span class="n">EnsDb</span><span class="p">)</span><span class="w">

</span><span class="c1"># How many counts are NA?
</span><span class="n">sapply</span><span class="p">(</span><span class="n">transcript_num_table_gather_entrez</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Entrez     DB  count 
##      0      0  71066
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># removing rows with NA counts
</span><span class="n">transcript_num_table_gather_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">transcript_num_table_gather_entrez</span><span class="p">[</span><span class="o">!</span><span class="nf">is.na</span><span class="p">(</span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">count</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="c1"># because there are only a handful of genes with many transcripts, they can't be plotted together with genes with few transcripts
# separating them in high and low
</span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">count</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">25</span><span class="p">,</span><span class="w"> </span><span class="s2">"high"</span><span class="p">,</span><span class="w"> </span><span class="s2">"low"</span><span class="p">)</span><span class="w">

</span><span class="c1"># setting factor levels
</span><span class="n">f</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"low"</span><span class="p">,</span><span class="w"> </span><span class="s2">"high"</span><span class="p">)</span><span class="w">
</span><span class="n">transcript_num_table_gather_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">within</span><span class="p">(</span><span class="n">transcript_num_table_gather_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">f</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># setting my custom theme of choice
</span><span class="n">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span><span class="w">

</span><span class="n">my_theme</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">base_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">,</span><span class="w"> </span><span class="n">base_family</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"sans"</span><span class="p">){</span><span class="w">
  </span><span class="n">theme_grey</span><span class="p">(</span><span class="n">base_size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">base_size</span><span class="p">,</span><span class="w"> </span><span class="n">base_family</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">base_family</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">theme</span><span class="p">(</span><span class="w">
    </span><span class="n">axis.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">),</span><span class="w">
    </span><span class="n">axis.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">14</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.grid.major</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_line</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.grid.minor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_blank</span><span class="p">(),</span><span class="w">
    </span><span class="n">panel.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"aliceblue"</span><span class="p">),</span><span class="w">
    </span><span class="n">strip.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"lightgrey"</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">),</span><span class="w">
    </span><span class="n">strip.text</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_text</span><span class="p">(</span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bold"</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">12</span><span class="p">,</span><span class="w"> </span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"navy"</span><span class="p">),</span><span class="w">
    </span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"bottom"</span><span class="p">,</span><span class="w">
    </span><span class="n">panel.margin</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">unit</span><span class="p">(</span><span class="m">.05</span><span class="p">,</span><span class="w"> </span><span class="s2">"lines"</span><span class="p">),</span><span class="w">
    </span><span class="n">panel.border</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">element_rect</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"grey"</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">)</span><span class="w">
  </span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">transcript_num_table_gather_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">count</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_histogram</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"deepskyblue4"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Histogram of number of transcripts per gene (Entrez ID)"</span><span class="p">,</span><span class="w"> 
       </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of transcripts per gene"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Count"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_wrap</span><span class="p">(</span><span class="n">DB</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">scales</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"free"</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">

</span><span class="n">ann_text_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">20</span><span class="p">,</span><span class="w"> </span><span class="m">300</span><span class="p">,</span><span class="w"> </span><span class="m">20</span><span class="p">,</span><span class="w"> </span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="m">20</span><span class="p">,</span><span class="w"> </span><span class="m">100</span><span class="p">),</span><span class="w">
                       </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">4000</span><span class="p">,</span><span class="w"> </span><span class="m">350</span><span class="p">,</span><span class="w"> </span><span class="m">40000</span><span class="p">,</span><span class="w"> </span><span class="m">15</span><span class="p">,</span><span class="w"> </span><span class="m">5000</span><span class="p">,</span><span class="w"> </span><span class="m">280</span><span class="p">),</span><span class="w">
                       </span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"low"</span><span class="p">,</span><span class="w"> </span><span class="s2">"high"</span><span class="p">),</span><span class="w"> </span><span class="m">3</span><span class="p">),</span><span class="w">
                       </span><span class="n">DB</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"EnsDb"</span><span class="p">,</span><span class="w"> </span><span class="s2">"orgDb"</span><span class="p">,</span><span class="w"> </span><span class="s2">"TxDb"</span><span class="p">),</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w">
                       </span><span class="n">labs</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"low"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> 
                                                     </span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"EnsDb"</span><span class="p">))),</span><span class="w"> 
                                </span><span class="n">paste</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"high"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> 
                                                     </span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"EnsDb"</span><span class="p">))),</span><span class="w">
                                </span><span class="n">paste</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"low"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> 
                                                     </span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"orgDb"</span><span class="p">))),</span><span class="w"> 
                                </span><span class="n">paste</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"high"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> 
                                                     </span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"orgDb"</span><span class="p">))),</span><span class="w">
                                </span><span class="n">paste</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"low"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> 
                                                     </span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"TxDb"</span><span class="p">))),</span><span class="w"> 
                                </span><span class="n">paste</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"high"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> 
                                                     </span><span class="n">transcript_num_table_gather_entrez</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"TxDb"</span><span class="p">)))))</span><span class="w">

</span><span class="n">p</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">geom_text</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ann_text_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">labs</span><span class="p">,</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">8</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="AnnotationDbi_part2_blog_files/figure-markdown_github/transcript_histogram-1.png" style="display: block; margin: auto;" /></p>

<p>This time, orgDb has the highest number of Entrez ID gene entries with corresponding transcript information, the majority of which have fewer than 25 transcripts (59984 genes); only 152 genes have more than 25 transcripts. EnsDb and TxDB have comparable numbers of gene entries, also for genes with few (23971 and 24669 genes) and many transcripts (566 and 552 genes).</p>

<p><br /></p>

<h3 id="ensembl-ids">Ensembl IDs</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># merging datasets by Ensembl ID
</span><span class="n">transcript_num_table_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">full_join</span><span class="p">(</span><span class="n">org_transcript_num_table_ensembl</span><span class="p">,</span><span class="w"> </span><span class="n">EnsDb_transcript_num_table_ensembl</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Ensembl"</span><span class="p">)</span><span class="w">

</span><span class="c1"># gather for plotting
</span><span class="n">transcript_num_table_gather_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">transcript_num_table_ensembl</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">DB</span><span class="p">,</span><span class="w"> </span><span class="n">count</span><span class="p">,</span><span class="w"> </span><span class="n">orgDb</span><span class="o">:</span><span class="n">EnsDb</span><span class="p">)</span><span class="w">

</span><span class="c1"># How many counts are NA?
</span><span class="n">sapply</span><span class="p">(</span><span class="n">transcript_num_table_gather_ensembl</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Ensembl      DB   count 
##       0       0   38659
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># removing rows with NA counts
</span><span class="n">transcript_num_table_gather_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">transcript_num_table_gather_ensembl</span><span class="p">[</span><span class="o">!</span><span class="nf">is.na</span><span class="p">(</span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">count</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="c1"># because there are only a handful of genes with many transcripts, they can't be plotted together with genes with few transcripts
# separating them in high and low
</span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ifelse</span><span class="p">(</span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">count</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">25</span><span class="p">,</span><span class="w"> </span><span class="s2">"high"</span><span class="p">,</span><span class="w"> </span><span class="s2">"low"</span><span class="p">)</span><span class="w">

</span><span class="c1"># setting factor levels
</span><span class="n">f</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"low"</span><span class="p">,</span><span class="w"> </span><span class="s2">"high"</span><span class="p">)</span><span class="w">
</span><span class="n">transcript_num_table_gather_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">within</span><span class="p">(</span><span class="n">transcript_num_table_gather_ensembl</span><span class="p">,</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">levels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">f</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">transcript_num_table_gather_ensembl</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="nf">as.numeric</span><span class="p">(</span><span class="n">count</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_histogram</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"deepskyblue4"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Histogram of number of transcripts per gene (Ensembl ID)"</span><span class="p">,</span><span class="w"> 
       </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of transcripts per gene"</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Count"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">facet_wrap</span><span class="p">(</span><span class="n">DB</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">group</span><span class="p">,</span><span class="w"> </span><span class="n">scales</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"free"</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">

</span><span class="n">ann_text_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="m">75</span><span class="p">,</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="m">100</span><span class="p">),</span><span class="w">
                              </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">30000</span><span class="p">,</span><span class="w"> </span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="m">15000</span><span class="p">,</span><span class="w"> </span><span class="m">100</span><span class="p">),</span><span class="w">
                              </span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"low"</span><span class="p">,</span><span class="w"> </span><span class="s2">"high"</span><span class="p">),</span><span class="w"> </span><span class="m">4</span><span class="p">),</span><span class="w">
                              </span><span class="n">DB</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">rep</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"EnsDb"</span><span class="p">,</span><span class="w"> </span><span class="s2">"orgDb"</span><span class="p">),</span><span class="w"> </span><span class="n">each</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w">
                              </span><span class="n">labs</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"low"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> 
                                                            </span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"EnsDb"</span><span class="p">))),</span><span class="w"> 
                                       </span><span class="n">paste</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"high"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> 
                                                            </span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"EnsDb"</span><span class="p">))),</span><span class="w">
                                       </span><span class="n">paste</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"low"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> 
                                                            </span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"orgDb"</span><span class="p">))),</span><span class="w"> 
                                       </span><span class="n">paste</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">which</span><span class="p">(</span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">group</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"high"</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> 
                                                            </span><span class="n">transcript_num_table_gather_ensembl</span><span class="o">$</span><span class="n">DB</span><span class="w"> </span><span class="o">==</span><span class="w">  </span><span class="s2">"orgDb"</span><span class="p">)))))</span><span class="w">

</span><span class="n">p</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">geom_text</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ann_text_ensembl</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="p">,</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">labs</span><span class="p">,</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">),</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">8</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="AnnotationDbi_part2_blog_files/figure-markdown_github/transcript_histogram2-1.png" style="display: block; margin: auto;" /></p>

<p>For Ensembl gene IDs EnsDb has the higher number of entries with transcript information; most of these genes (65374) have fewer than 25 transcripts. While orgDb has fewer Ensembl-transcript entries, proportionally more have a high number of transcripts (1035 vs 400), while 26980 genes have low transcript numbers.</p>

<p><br /></p>

<h2 id="comparison-of-transcript-numbers-per-gene">Comparison of transcript numbers per gene</h2>

<p>Now that we know what the distribution of transcript numbers per gene is in different database packages, I want to directly compare the numbers of transcripts that are identified for each gene.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># for comparison replacing NAs with 0
</span><span class="n">transcript_num_table_ensembl_NAtozero</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">transcript_num_table_ensembl</span><span class="w">
</span><span class="n">transcript_num_table_ensembl_NAtozero</span><span class="p">[</span><span class="nf">is.na</span><span class="p">(</span><span class="n">transcript_num_table_ensembl_NAtozero</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span><span class="w">

</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">transcript_num_table_entrez</span><span class="w">
</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">[</span><span class="nf">is.na</span><span class="p">(</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">ggrepel</span><span class="p">)</span><span class="w">

</span><span class="n">p</span><span class="m">1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">transcript_num_table_ensembl_NAtozero</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">orgDb</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_abline</span><span class="p">(</span><span class="n">linetype</span><span class="o">=</span><span class="s2">"dashed"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"deepskyblue4"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_smooth</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">            </span><span class="c1"># Add a loess smoothed fit curve with confidence region
</span><span class="w">  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of transcripts per Ensembl ID"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_text_repel</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">transcript_num_table_ensembl_NAtozero</span><span class="p">,</span><span class="w"> </span><span class="n">orgDb</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">140</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">EnsDb</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">150</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Ensembl</span><span class="p">))</span><span class="w">

</span><span class="n">p</span><span class="m">2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">orgDb</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">TxDb</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_abline</span><span class="p">(</span><span class="n">linetype</span><span class="o">=</span><span class="s2">"dashed"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"deepskyblue4"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_smooth</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of transcripts per Entrez ID"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_text_repel</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">,</span><span class="w"> </span><span class="n">orgDb</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">100</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">TxDb</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">200</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Entrez</span><span class="p">))</span><span class="w">

</span><span class="n">p</span><span class="m">3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">orgDb</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_abline</span><span class="p">(</span><span class="n">linetype</span><span class="o">=</span><span class="s2">"dashed"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"deepskyblue4"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_smooth</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of transcripts per Entrez ID"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_text_repel</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">,</span><span class="w"> </span><span class="n">orgDb</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">100</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">EnsDb</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">900</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Entrez</span><span class="p">))</span><span class="w">

</span><span class="n">p</span><span class="m">4</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">TxDb</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_abline</span><span class="p">(</span><span class="n">linetype</span><span class="o">=</span><span class="s2">"dashed"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"deepskyblue4"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_smooth</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of transcripts per Entrez ID"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_text_repel</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">,</span><span class="w"> </span><span class="n">TxDb</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">150</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">EnsDb</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">900</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Entrez</span><span class="p">))</span><span class="w">

</span><span class="n">library</span><span class="p">(</span><span class="n">gridExtra</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">grid</span><span class="p">)</span><span class="w">

</span><span class="n">grid.arrange</span><span class="p">(</span><span class="n">p</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">2</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="AnnotationDbi_part2_blog_files/figure-markdown_github/unnamed-chunk-2-1.png" style="display: block; margin: auto;" /></p>

<p>The plots above show scatterplots with fitted loess curves with confidence regions of the numbers of transcripts for each gene with either Entrez or Ensembl entries from different database packages; dashed lines show 1:1 ratio. For most extreme genes, the Ensembl/ Entrez ID is given in the plots.</p>

<p>While some genes do have the same number of transcripts in different database packages, the majority do not. Entrez IDs seem to give a better correlation than Ensembl IDs, except for a few extreme outliers.</p>

<p><br /></p>

<h2 id="is-there-a-correlation-between-number-of-transcripts-and-number-of-exons-of-a-gene">Is there a correlation between number of transcripts and number of exons of a gene?</h2>

<p>Because the number of exons a gene has exponentially increases the number of potential transcripts produced via alternative splicing, it would make sense if the number of exons correlates with the number of transcripts of a gene. However, not all combinations of exons will produce a functioning protein, so there won’t be as many transcripts as theoretically possible.</p>

<p>Exon IDs are only available in TxDb.Hsapiens.UCSC.hg38.knownGene and EnsDb.Hsapiens.v79.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">TxDb_exons</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">AnnotationDbi</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ENTREZID_TxDb</span><span class="p">,</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"EXONID"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">

</span><span class="n">EnsDb_exons_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ENSEMBL_EnsDb</span><span class="p">,</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"EXONID"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">
</span><span class="n">EnsDb_exons_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ENTREZ_EnsDb</span><span class="p">,</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"EXONID"</span><span class="p">),</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">)</span><span class="w">

</span><span class="c1"># summarize number of exons per gene
</span><span class="n">TxDb_exons_num_table_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">TxDb_exons</span><span class="o">$</span><span class="n">GENEID</span><span class="p">))</span><span class="w">

</span><span class="n">EnsDb_exons_num_table_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">EnsDb_exons_ensembl</span><span class="o">$</span><span class="n">GENEID</span><span class="p">))</span><span class="w">
</span><span class="n">EnsDb_exons_num_table_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">EnsDb_exons_entrez</span><span class="o">$</span><span class="n">ENTREZID</span><span class="p">))</span><span class="w">

</span><span class="c1"># there are over 98000 entries without corresponding Entrez ID, removing those
</span><span class="n">EnsDb_exons_num_table_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">EnsDb_exons_num_table_entrez</span><span class="p">[</span><span class="m">-1</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">EnsDb_trans_vs_exons_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">full_join</span><span class="p">(</span><span class="n">EnsDb_transcript_num_table_ensembl</span><span class="p">,</span><span class="w"> </span><span class="n">EnsDb_exons_num_table_ensembl</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Ensembl"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">))</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">EnsDb_trans_vs_exons_ensembl</span><span class="p">)[</span><span class="m">2</span><span class="o">:</span><span class="m">3</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"EnsDb_transcripts"</span><span class="p">,</span><span class="w"> </span><span class="s2">"EnsDb_exons"</span><span class="p">)</span><span class="w">

</span><span class="n">EnsDb_trans_vs_exons_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">full_join</span><span class="p">(</span><span class="n">EnsDb_transcript_num_table_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">EnsDb_exons_num_table_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Entrez"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">))</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">EnsDb_trans_vs_exons_entrez</span><span class="p">)[</span><span class="m">2</span><span class="o">:</span><span class="m">3</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"EnsDb_transcripts"</span><span class="p">,</span><span class="w"> </span><span class="s2">"EnsDb_exons"</span><span class="p">)</span><span class="w">

</span><span class="n">TxDb_trans_vs_exons_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">full_join</span><span class="p">(</span><span class="n">TxDb_transcript_num_table_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">TxDb_exons_num_table_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"Entrez"</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">))</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">TxDb_trans_vs_exons_entrez</span><span class="p">)[</span><span class="m">2</span><span class="o">:</span><span class="m">3</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"TxDb_transcripts"</span><span class="p">,</span><span class="w"> </span><span class="s2">"TxDb_exons"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">p</span><span class="m">1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">EnsDb_trans_vs_exons_ensembl</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb_transcripts</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb_exons</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_abline</span><span class="p">(</span><span class="n">linetype</span><span class="o">=</span><span class="s2">"dashed"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"deepskyblue4"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_smooth</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of transcripts vs number of exons\nper Ensembl ID in EnsDb"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_text_repel</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">EnsDb_trans_vs_exons_ensembl</span><span class="p">,</span><span class="w"> </span><span class="n">EnsDb_transcripts</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">150</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">EnsDb_exons</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">300</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Ensembl</span><span class="p">))</span><span class="w">

</span><span class="n">p</span><span class="m">2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">EnsDb_trans_vs_exons_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb_transcripts</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb_exons</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_abline</span><span class="p">(</span><span class="n">linetype</span><span class="o">=</span><span class="s2">"dashed"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"deepskyblue4"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_smooth</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of transcripts vs number of exons\nper Entrez ID in EnsDb"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_text_repel</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">EnsDb_trans_vs_exons_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">EnsDb_transcripts</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">300</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">EnsDb_exons</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">1000</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Entrez</span><span class="p">))</span><span class="w">

</span><span class="n">p</span><span class="m">3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">TxDb_trans_vs_exons_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">TxDb_transcripts</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">TxDb_exons</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_abline</span><span class="p">(</span><span class="n">linetype</span><span class="o">=</span><span class="s2">"dashed"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_point</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"deepskyblue4"</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.3</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_smooth</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"black"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Number of transcripts vs number of exons\nper Entrez ID in TxDb"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">geom_text_repel</span><span class="p">(</span><span class="n">data</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">subset</span><span class="p">(</span><span class="n">TxDb_trans_vs_exons_entrez</span><span class="p">,</span><span class="w"> </span><span class="n">TxDb_transcripts</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">200</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">TxDb_exons</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">400</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Entrez</span><span class="p">))</span><span class="w">

</span><span class="n">grid.arrange</span><span class="p">(</span><span class="n">p</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="AnnotationDbi_part2_blog_files/figure-markdown_github/unnamed-chunk-5-1.png" style="display: block; margin: auto;" /></p>

<p>Indeed, there is a reasonably good correlation between number of transcripts and number of exons ins genes. The scatterplots above show transcript vs exon number per gene in the three database packages with fitted loess curves with confidence regions. As expected, the increase is not exponential as theoretically possible, but stronger than 1:1. For Entrez IDs, in both EnsDb and TxDbm genes with up to about 100 transcripts show a steeper correlation than genes with more than 100 transcripts. For Ensembl IDs the cutoff is around 40 transcripts.</p>

<p><br /></p>

<h3 id="what-genes-have-many-exons-but-few-transcripts">What genes have many exons but few transcripts?</h3>

<p>Interestingly, there are a couple of genes with many exons but only few transcripts. To find out more about these genes, I will compare their proportion of gene biotypes.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># gene biotypes are only available in EnsDb
</span><span class="n">EnsDb_trans_vs_exons_ensembl</span><span class="o">$</span><span class="n">ratio</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">EnsDb_trans_vs_exons_ensembl</span><span class="o">$</span><span class="n">EnsDb_exons</span><span class="o">/</span><span class="n">EnsDb_trans_vs_exons_ensembl</span><span class="o">$</span><span class="n">EnsDb_transcripts</span><span class="w">
</span><span class="n">EnsDb_trans_vs_exons_ensembl_genes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">EnsDb_trans_vs_exons_ensembl</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">EnsDb_trans_vs_exons_ensembl</span><span class="o">$</span><span class="n">ratio</span><span class="p">,</span><span class="w"> </span><span class="n">decreasing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="s2">"Ensembl"</span><span class="p">][</span><span class="m">1</span><span class="o">:</span><span class="m">100</span><span class="p">])</span><span class="w">

</span><span class="n">EnsDb_trans_vs_exons_entrez</span><span class="o">$</span><span class="n">ratio</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">EnsDb_trans_vs_exons_entrez</span><span class="o">$</span><span class="n">EnsDb_exons</span><span class="o">/</span><span class="n">EnsDb_trans_vs_exons_entrez</span><span class="o">$</span><span class="n">EnsDb_transcripts</span><span class="w">
</span><span class="n">EnsDb_trans_vs_exons_entrez_genes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">EnsDb_trans_vs_exons_entrez</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">EnsDb_trans_vs_exons_entrez</span><span class="o">$</span><span class="n">ratio</span><span class="p">,</span><span class="w"> </span><span class="n">decreasing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="s2">"Entrez"</span><span class="p">][</span><span class="m">1</span><span class="o">:</span><span class="m">100</span><span class="p">])</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># get all gene biotypes for all gene IDs
</span><span class="n">biotypes_exons_trans_EnsDb_ensembl_all</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb_trans_vs_exons_ensembl_genes</span><span class="p">,</span><span class="w"> 
                                          </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEBIOTYPE"</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">

</span><span class="n">biotypes_exons_trans_EnsDb_entrez_all</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">EnsDb_trans_vs_exons_entrez_genes</span><span class="p">,</span><span class="w"> 
                                          </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEBIOTYPE"</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># get all gene biotypes for all gene IDs
</span><span class="n">biotypes_EnsDb_ensembl_all</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">),</span><span class="w"> 
                                          </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEBIOTYPE"</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">

</span><span class="n">biotypes_EnsDb_entrez_all</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">keys</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">),</span><span class="w"> 
                                          </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEBIOTYPE"</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># how many NAs are in each column?
</span><span class="n">sapply</span><span class="p">(</span><span class="n">biotypes_exons_trans_EnsDb_ensembl_all</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##      GENEID GENEBIOTYPE 
##           0           0
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sapply</span><span class="p">(</span><span class="n">biotypes_exons_trans_EnsDb_entrez_all</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##    ENTREZID GENEBIOTYPE 
##           0           0
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># summary table of number of genebiotypes
</span><span class="n">biotypes_exons_trans_EnsDb_ensembl_all_num_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">biotypes_exons_trans_EnsDb_ensembl_all</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">))</span><span class="w">
</span><span class="n">biotypes_exons_trans_EnsDb_entrez_all_num_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">biotypes_exons_trans_EnsDb_entrez_all</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># subset to keep only those biotypes that are among the top genes
</span><span class="n">biotypes_EnsDb_ensembl_all_subs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">biotypes_EnsDb_ensembl_all</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">biotypes_EnsDb_ensembl_all</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="w"> </span><span class="o">%in%</span><span class="w"> 
                                                                      </span><span class="n">unique</span><span class="p">(</span><span class="n">biotypes_exons_trans_EnsDb_ensembl_all</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">)),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="n">biotypes_EnsDb_entrez_all_subs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">biotypes_EnsDb_entrez_all</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">biotypes_EnsDb_entrez_all</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="w"> </span><span class="o">%in%</span><span class="w"> 
                                                                      </span><span class="n">unique</span><span class="p">(</span><span class="n">biotypes_exons_trans_EnsDb_entrez_all</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">)),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="c1"># summary table of biotypes background (all genes)
</span><span class="n">biotypes_EnsDb_ensembl_all_subs_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">biotypes_EnsDb_ensembl_all_subs</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">))</span><span class="w">
</span><span class="n">biotypes_EnsDb_entrez_all_subs_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">biotypes_EnsDb_entrez_all_subs</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">))</span><span class="w">

</span><span class="c1"># add background number of biotypes to number among top genes
</span><span class="n">biotypes_EnsDb_ensembl_top_num_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">biotypes_exons_trans_EnsDb_ensembl_all_num_table</span><span class="p">,</span><span class="w"> </span><span class="n">biotypes_EnsDb_ensembl_all_subs_table</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">,</span><span class="w"> </span><span class="n">all.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">biotypes_EnsDb_entrez_top_num_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">biotypes_exons_trans_EnsDb_entrez_all_num_table</span><span class="p">,</span><span class="w"> </span><span class="n">biotypes_EnsDb_entrez_all_subs_table</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">,</span><span class="w"> </span><span class="n">all.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># merge
</span><span class="n">biotypes_top_num_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">biotypes_EnsDb_ensembl_top_num_table</span><span class="p">,</span><span class="w"> </span><span class="n">biotypes_EnsDb_entrez_top_num_table</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">,</span><span class="w"> </span><span class="n">all</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">biotypes_top_num_table</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"GENEBIOTYPE"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Ensembl_top"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Ensembl_bg"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Entrez_top"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Entrez_bg"</span><span class="p">)</span><span class="w">

</span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Ensembl_top</span><span class="o">/</span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Ensembl_bg</span><span class="w">
</span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Entrez_top</span><span class="o">/</span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Entrez_bg</span><span class="w">

</span><span class="c1"># gather for plotting
</span><span class="n">biotypes_top_num_table_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">biotypes_top_num_table</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">6</span><span class="p">,</span><span class="w"> </span><span class="m">7</span><span class="p">)]</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">GENEBIOTYPE</span><span class="p">,</span><span class="w"> </span><span class="n">count</span><span class="p">,</span><span class="w"> </span><span class="n">Ensembl</span><span class="o">:</span><span class="n">Entrez</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">biotypes_top_num_table_gather</span><span class="p">)[</span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"Gene_ID"</span><span class="w">

</span><span class="c1"># replace NA with 0
</span><span class="n">biotypes_top_num_table_gather</span><span class="p">[</span><span class="nf">is.na</span><span class="p">(</span><span class="n">biotypes_top_num_table_gather</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">biotypes_top_num_table_gather</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">GENEBIOTYPE</span><span class="p">),</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">count</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Gene_ID</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"identity"</span><span class="p">,</span><span class="w"> </span><span class="n">position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"dodge"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"right"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">coord_flip</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Gene biotypes of genes with many exons but few transcripts"</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Ratio of biotype count vs background number"</span><span class="p">,</span><span class="w">
       </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Gene ID"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">scale_x_discrete</span><span class="p">(</span><span class="n">breaks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">biotypes_top_num_table_gather</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">,</span><span class="w">
                      </span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"_"</span><span class="p">,</span><span class="w"> </span><span class="s2">" "</span><span class="p">,</span><span class="w"> </span><span class="n">biotypes_top_num_table_gather</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_brewer</span><span class="p">(</span><span class="n">palette</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Set1"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="AnnotationDbi_part2_blog_files/figure-markdown_github/unnamed-chunk-12-1.png" style="display: block; margin: auto;" /></p>

<p>Compared to the background distribution of biotypes, the proportionally most strongly represented group are LRG genes. As defined by <a href="http://www.ensembl.org/common/Help/Glossary?db=core">Ensembl’s glossary</a>:</p>

<blockquote>
  <p>“An LRG (Locus Reference Genomic) is a fixed sequence, independent of the genome, specifically created for the diagnostic community to record DNA sequence variation on a fixed framework. Sequence variants in LSDBs (Locus Specific Databases) are reported using LRG sequences.”</p>
</blockquote>

<p>Other than this, there are only protein coding, processed and pseudogenes among genes with many exons and few transcripts. So, nothing exceptional about these genes’s biotypes.</p>

<p><br /></p>

<h2 id="what-types-of-genes-have-the-most-transcripts">What types of genes have the most transcripts?</h2>

<p>Is there a biological function associated with having many transcripts? In order to find out more about these genes, I am looking at</p>

<ol>
  <li>their biotypes and</li>
  <li>for over-represented gene ontology (GO) categories.</li>
</ol>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># mean number of transcripts over all database packages
</span><span class="n">mean_num_transcripts_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">mean_num_transcripts</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rowMeans</span><span class="p">(</span><span class="n">transcript_num_table_ensembl_NAtozero</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]),</span><span class="w">
               </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">transcript_num_table_ensembl_NAtozero</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="p">])</span><span class="w">

</span><span class="n">mean_num_transcripts_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">mean_num_transcripts</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rowMeans</span><span class="p">(</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">]),</span><span class="w">
               </span><span class="n">row.names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">[,</span><span class="w"> </span><span class="m">1</span><span class="p">])</span><span class="w">

</span><span class="c1"># standard error of mean
</span><span class="n">mean_num_transcripts_ensembl</span><span class="o">$</span><span class="n">sem</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">transcript_num_table_ensembl_NAtozero</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">sd</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">/</span><span class="nf">sqrt</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span><span class="n">mean_num_transcripts_entrez</span><span class="o">$</span><span class="n">sem</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">transcript_num_table_entrez_NAtozero</span><span class="p">[,</span><span class="w"> </span><span class="m">-1</span><span class="p">],</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">sd</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">/</span><span class="nf">sqrt</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">

</span><span class="c1"># ranking the mean
</span><span class="n">mean_num_transcripts_ensembl</span><span class="o">$</span><span class="n">rank</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rank</span><span class="p">(</span><span class="n">mean_num_transcripts_ensembl</span><span class="o">$</span><span class="n">mean_num_transcripts</span><span class="p">)</span><span class="w">
</span><span class="n">mean_num_transcripts_ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mean_num_transcripts_ensembl</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">mean_num_transcripts_ensembl</span><span class="o">$</span><span class="n">rank</span><span class="p">,</span><span class="w"> </span><span class="n">decreasing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="n">mean_num_transcripts_entrez</span><span class="o">$</span><span class="n">rank</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rank</span><span class="p">(</span><span class="n">mean_num_transcripts_entrez</span><span class="o">$</span><span class="n">mean_num_transcripts</span><span class="p">)</span><span class="w">
</span><span class="n">mean_num_transcripts_entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mean_num_transcripts_entrez</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">mean_num_transcripts_entrez</span><span class="o">$</span><span class="n">rank</span><span class="p">,</span><span class="w"> </span><span class="n">decreasing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<p><br /></p>

<p>I am picking the top 1000 genes with highest mean number of transcripts.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># picking top genes
</span><span class="n">mean_num_transcripts_ensembl_top</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mean_num_transcripts_ensembl</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">1000</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="n">mean_num_transcripts_entrez_top</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">mean_num_transcripts_entrez</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">1000</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="c1"># distribution of mean number of transcripts
</span><span class="n">p</span><span class="m">1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">mean_num_transcripts_ensembl_top</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">mean_num_transcripts</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">geom_density</span><span class="p">(</span><span class="n">fill</span><span class="o">=</span><span class="s2">"navy"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Mean number of transcripts\nper Ensembl ID for the top genes"</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"mean number of transcripts pr gene"</span><span class="p">)</span><span class="w">
</span><span class="n">p</span><span class="m">2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ggplot</span><span class="p">(</span><span class="n">mean_num_transcripts_entrez_top</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">mean_num_transcripts</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">geom_density</span><span class="p">(</span><span class="n">fill</span><span class="o">=</span><span class="s2">"navy"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Mean number of transcripts\nper Entrez ID for the top genes"</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"mean number of transcripts per gene"</span><span class="p">)</span><span class="w">
</span><span class="n">grid.arrange</span><span class="p">(</span><span class="n">p</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">p</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="AnnotationDbi_part2_blog_files/figure-markdown_github/unnamed-chunk-15-1.png" style="display: block; margin: auto;" /></p>

<p>The density distribution of the mean number of transcripts per gene (mean over database packages) for the top genes with highest transcript numbers shows the stark difference between Ensembl and Entrez IDs.</p>

<p><br /></p>

<h3 id="gene-biotypes-of-top-genes-with-highest-mean-number-of-transcripts">Gene biotypes of top genes with highest mean number of transcripts</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">biotypes_EnsDb_ensembl_top</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">mean_num_transcripts_ensembl_top</span><span class="p">),</span><span class="w"> 
                                          </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEBIOTYPE"</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEID"</span><span class="p">)</span><span class="w">

</span><span class="n">biotypes_EnsDb_entrez_top</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">ensembldb</span><span class="o">::</span><span class="n">select</span><span class="p">(</span><span class="n">EnsDb.Hsapiens.v79</span><span class="p">,</span><span class="w"> </span><span class="n">keys</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">rownames</span><span class="p">(</span><span class="n">mean_num_transcripts_entrez_top</span><span class="p">),</span><span class="w"> 
                                          </span><span class="n">columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"GENEBIOTYPE"</span><span class="p">,</span><span class="w"> </span><span class="n">keytype</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"ENTREZID"</span><span class="p">)</span><span class="w">

</span><span class="c1"># how many NAs are in each column?
</span><span class="n">sapply</span><span class="p">(</span><span class="n">biotypes_EnsDb_ensembl_top</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##      GENEID GENEBIOTYPE 
##           0           0
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sapply</span><span class="p">(</span><span class="n">biotypes_EnsDb_entrez_top</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##    ENTREZID GENEBIOTYPE 
##           0           0
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># summary table of number of genebiotypes
</span><span class="n">biotypes_EnsDb_ensembl_top_num_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">biotypes_EnsDb_ensembl_top</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">))</span><span class="w">
</span><span class="n">biotypes_EnsDb_entrez_top_num_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">biotypes_EnsDb_entrez_top</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># subset to keep only those biotypes that are among the top genes
</span><span class="n">biotypes_EnsDb_ensembl_all_subs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">biotypes_EnsDb_ensembl_all</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">biotypes_EnsDb_ensembl_all</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="w"> </span><span class="o">%in%</span><span class="w"> 
                                                                      </span><span class="n">unique</span><span class="p">(</span><span class="n">biotypes_EnsDb_ensembl_top_num_table</span><span class="o">$</span><span class="n">Var1</span><span class="p">)),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="n">biotypes_EnsDb_entrez_all_subs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">biotypes_EnsDb_entrez_all</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">biotypes_EnsDb_entrez_all</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="w"> </span><span class="o">%in%</span><span class="w"> 
                                                                      </span><span class="n">unique</span><span class="p">(</span><span class="n">biotypes_EnsDb_entrez_top_num_table</span><span class="o">$</span><span class="n">Var1</span><span class="p">)),</span><span class="w"> </span><span class="p">]</span><span class="w">

</span><span class="c1"># summary table of biotypes background (all genes)
</span><span class="n">biotypes_EnsDb_ensembl_all_subs_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">biotypes_EnsDb_ensembl_all_subs</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">))</span><span class="w">
</span><span class="n">biotypes_EnsDb_entrez_all_subs_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">table</span><span class="p">(</span><span class="n">biotypes_EnsDb_entrez_all_subs</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">))</span><span class="w">

</span><span class="c1"># add background number of biotypes to number among top genes
</span><span class="n">biotypes_EnsDb_ensembl_top_num_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">biotypes_EnsDb_ensembl_top_num_table</span><span class="p">,</span><span class="w"> </span><span class="n">biotypes_EnsDb_ensembl_all_subs_table</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">,</span><span class="w"> </span><span class="n">all.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">biotypes_EnsDb_entrez_top_num_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">biotypes_EnsDb_entrez_top_num_table</span><span class="p">,</span><span class="w"> </span><span class="n">biotypes_EnsDb_entrez_all_subs_table</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">,</span><span class="w"> </span><span class="n">all.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="c1"># merge
</span><span class="n">biotypes_top_num_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">merge</span><span class="p">(</span><span class="n">biotypes_EnsDb_ensembl_top_num_table</span><span class="p">,</span><span class="w"> </span><span class="n">biotypes_EnsDb_entrez_top_num_table</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Var1"</span><span class="p">,</span><span class="w"> </span><span class="n">all</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">biotypes_top_num_table</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"GENEBIOTYPE"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Ensembl_top"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Ensembl_bg"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Entrez_top"</span><span class="p">,</span><span class="w"> </span><span class="s2">"Entrez_bg"</span><span class="p">)</span><span class="w">

</span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Ensembl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Ensembl_top</span><span class="o">/</span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Ensembl_bg</span><span class="w">
</span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Entrez</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Entrez_top</span><span class="o">/</span><span class="n">biotypes_top_num_table</span><span class="o">$</span><span class="n">Entrez_bg</span><span class="w">

</span><span class="c1"># gather for plotting
</span><span class="n">biotypes_top_num_table_gather</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">biotypes_top_num_table</span><span class="p">[,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">6</span><span class="p">,</span><span class="w"> </span><span class="m">7</span><span class="p">)]</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
  </span><span class="n">gather</span><span class="p">(</span><span class="n">GENEBIOTYPE</span><span class="p">,</span><span class="w"> </span><span class="n">count</span><span class="p">,</span><span class="w"> </span><span class="n">Ensembl</span><span class="o">:</span><span class="n">Entrez</span><span class="p">)</span><span class="w">
</span><span class="n">colnames</span><span class="p">(</span><span class="n">biotypes_top_num_table_gather</span><span class="p">)[</span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"Gene_ID"</span><span class="w">

</span><span class="c1"># replace NA with 0
</span><span class="n">biotypes_top_num_table_gather</span><span class="p">[</span><span class="nf">is.na</span><span class="p">(</span><span class="n">biotypes_top_num_table_gather</span><span class="p">)]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">0</span><span class="w">
</span></code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">ggplot</span><span class="p">(</span><span class="n">biotypes_top_num_table_gather</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">factor</span><span class="p">(</span><span class="n">GENEBIOTYPE</span><span class="p">),</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">count</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Gene_ID</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"identity"</span><span class="p">,</span><span class="w"> </span><span class="n">position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"dodge"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">my_theme</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"right"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">coord_flip</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Gene biotypes of top genes\nwith highest transcript numbers"</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Ratio of biotype count vs background number"</span><span class="p">,</span><span class="w">
       </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Gene ID"</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> 
  </span><span class="n">scale_x_discrete</span><span class="p">(</span><span class="n">breaks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">biotypes_top_num_table_gather</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">,</span><span class="w">
                      </span><span class="n">labels</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">gsub</span><span class="p">(</span><span class="s2">"_"</span><span class="p">,</span><span class="w"> </span><span class="s2">" "</span><span class="p">,</span><span class="w"> </span><span class="n">biotypes_top_num_table_gather</span><span class="o">$</span><span class="n">GENEBIOTYPE</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w">
  </span><span class="n">scale_fill_brewer</span><span class="p">(</span><span class="n">palette</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Set1"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="AnnotationDbi_part2_blog_files/figure-markdown_github/unnamed-chunk-19-1.png" style="display: block; margin: auto;" /></p>

<p>Again, compared to background, we find a high proportion of LRG, protein coding, processed and pseudogenes. But here, we also see a few other biotypes represented among genes with most transcripts. Especially interesting is that we find <a href="http://portals.broadinstitute.org/genome_bio/human_lincrnas/">lincRNAs</a> and <a href="https://en.wikipedia.org/wiki/MicroRNA">miRNAs</a> among them, not something I would have expected…</p>

<p><br /></p>

<h3 id="enrichment-analysis-of-top-genes-with-highest-mean-number-of-transcripts">Enrichment Analysis of top genes with highest mean number of transcripts</h3>

<p>Functional profiles for the top 100 genes with most transcripts were created with <a href="https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html">clusterProfiler</a>’s <strong>compareCluster</strong> function.</p>

<p><br /></p>

<h4 id="grouping-go-terms">Grouping GO terms</h4>

<p>With <strong>groupGO</strong>, we can see the distribution of GO categories among input genes.</p>

<p><img src="AnnotationDbi_part2_blog_files/figure-markdown_github/unnamed-chunk-22-1.png" style="display: block; margin: auto;" /></p>

<p><br /></p>

<h4 id="go-category-enrichment">GO category enrichment</h4>

<p><a href="http://bioconductor.org/packages/release/bioc/vignettes/DOSE/inst/doc/enrichmentAnalysis.html">Over-representation</a> of GO categories is calculated with <strong>enrichGO</strong>.</p>

<p><img src="AnnotationDbi_part2_blog_files/figure-markdown_github/unnamed-chunk-24-1.png" style="display: block; margin: auto;" /></p>

<p>Especially interesting to me is that the GO enrichment analysis of genes with many transcripts finds these genes to be overproportionally involved in immune system functions like <a href="http://www.cell.com/cell/abstract/S0092-8674(11)00300-X?_returnURL=http%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS009286741100300X%3Fshowall%3Dtrue">MCH class II</a> and <a href="https://www.ncbi.nlm.nih.gov/books/NBK27160/">antigen binding</a>. These functions are known for their high variability, which is necessary for efficient immune defenses against a wide variety of intruders (pathogens).</p>

<p>We also see an enrichment of genes involved in <a href="https://www.ncbi.nlm.nih.gov/books/NBK26937/">cell-cell-adhesion</a>, specifically <a href="https://en.wikipedia.org/wiki/Cadherin">cadherin binding</a>. These have also been known for their high variability via alternative splicing.</p>

<hr />

<p><br /></p>

<div class="highlighter-rouge"><pre class="highlight"><code>## R version 3.3.1 (2016-06-21)
## Platform: x86_64-apple-darwin13.4.0 (64-bit)
## Running under: OS X 10.12 (Sierra)
## 
## locale:
## [1] de_DE.UTF-8/de_DE.UTF-8/de_DE.UTF-8/C/de_DE.UTF-8/de_DE.UTF-8
## 
## attached base packages:
##  [1] grid      parallel  stats4    stats     graphics  grDevices utils    
##  [8] datasets  methods   base     
## 
## other attached packages:
##  [1] clusterProfiler_3.2.0                  
##  [2] DOSE_3.0.0                             
##  [3] gridExtra_2.2.1                        
##  [4] ggrepel_0.5                            
##  [5] ggplot2_2.1.0                          
##  [6] tidyr_0.6.0                            
##  [7] dplyr_0.5.0                            
##  [8] plyr_1.8.4                             
##  [9] splitstackshape_1.4.2                  
## [10] data.table_1.9.6                       
## [11] EnsDb.Hsapiens.v79_1.1.0               
## [12] ensembldb_1.6.0                        
## [13] TxDb.Hsapiens.UCSC.hg38.knownGene_3.4.0
## [14] GenomicFeatures_1.26.0                 
## [15] GenomicRanges_1.26.0                   
## [16] GenomeInfoDb_1.10.0                    
## [17] org.Hs.eg.db_3.4.0                     
## [18] AnnotationDbi_1.36.0                   
## [19] IRanges_2.8.0                          
## [20] S4Vectors_0.12.0                       
## [21] Biobase_2.34.0                         
## [22] BiocGenerics_0.20.0                    
## 
## loaded via a namespace (and not attached):
##  [1] httr_1.2.1                    AnnotationHub_2.6.0          
##  [3] splines_3.3.1                 shiny_0.14.1                 
##  [5] assertthat_0.1                DO.db_2.9                    
##  [7] interactiveDisplayBase_1.12.0 Rsamtools_1.26.0             
##  [9] yaml_2.1.13                   RSQLite_1.0.0                
## [11] lattice_0.20-34               chron_2.3-47                 
## [13] digest_0.6.10                 RColorBrewer_1.1-2           
## [15] XVector_0.14.0                qvalue_2.6.0                 
## [17] colorspace_1.2-7              htmltools_0.3.5              
## [19] httpuv_1.3.3                  Matrix_1.2-7.1               
## [21] XML_3.98-1.4                  biomaRt_2.30.0               
## [23] zlibbioc_1.20.0               xtable_1.8-2                 
## [25] GO.db_3.4.0                   scales_0.4.0                 
## [27] BiocParallel_1.8.0            tibble_1.2                   
## [29] mgcv_1.8-15                   SummarizedExperiment_1.4.0   
## [31] lazyeval_0.2.0                magrittr_1.5                 
## [33] mime_0.5                      evaluate_0.10                
## [35] nlme_3.1-128                  BiocInstaller_1.24.0         
## [37] tools_3.3.1                   formatR_1.4                  
## [39] stringr_1.1.0                 munsell_0.4.3                
## [41] Biostrings_2.42.0             RCurl_1.95-4.8               
## [43] igraph_1.0.1                  bitops_1.0-6                 
## [45] labeling_0.3                  rmarkdown_1.1                
## [47] gtable_0.2.0                  DBI_0.5-1                    
## [49] reshape2_1.4.1                R6_2.2.0                     
## [51] GenomicAlignments_1.10.0      knitr_1.14                   
## [53] rtracklayer_1.34.0            fastmatch_1.0-4              
## [55] fgsea_1.0.0                   stringi_1.1.2                
## [57] GOSemSim_2.0.0                Rcpp_0.12.7
</code></pre>
</div>

    </div>

  
    <ul class="tag_box inline">
      <li><i class="glyphicon glyphicon-open"></i></li>
      
      


  
     
    	<li><a href="/categories.html#genome-ref">
    		genome <span>6</span>
    	</a></li>
    
  


    </ul>
    

  
    <ul class="tag_box inline">
      <li><i class="glyphicon glyphicon-tags"></i></li>
      
      


  
     
    	<li><a href="/tags.html#AnnotationDbi-ref">AnnotationDbi <span>2</span></a></li>
     
    	<li><a href="/tags.html#ggplot2-ref">ggplot2 <span>32</span></a></li>
     
    	<li><a href="/tags.html#genes-ref">genes <span>2</span></a></li>
     
    	<li><a href="/tags.html#genomics-ref">genomics <span>2</span></a></li>
     
    	<li><a href="/tags.html#genetics-ref">genetics <span>2</span></a></li>
    
  



    </ul>
    
  
    <hr>
    <ul class="pagination">
    
      <li class="prev"><a href="/genome/2016/10/23/AnnotationDbi" title="Exploring the human genome (Part 1) - Gene Annotations">&laquo; Previous</a></li>
    
      <li><a href="/archive.html">Archive</a></li>
    
      <li class="next"><a href="/text_analysis/2016/11/06/sysk_earwax" title="Is 'Yeah' Josh and Chuck's favorite word?">Next &raquo;</a></li>
    
    </ul>
    <hr>
    


  <div id="disqus_thread"></div>
<script type="text/javascript">
    
    var disqus_developer = 1;
    var disqus_shortname = 'shirinsplayground'; // required: replace example with your forum shortname
    
    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = 'https://' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="http://disqus.com" class="dsq-brlink">blog comments powered by <span class="logo-disqus">Disqus</span></a>




  </div>
</div>


      </div>

    </div>
	  
	   



	  
  </body>

<div class="footer">
    <div id="footer">
      <div class="container">
        <p>&copy; 2017 Shirin Glander
		<a href="mailto:shirin.glander@gmail.com"><img src="http://localhost:4000/assets/images/200px-Email_Shiny_Icon.png" /></a>
		<a href="http://stackoverflow.com/users/6623620/shirin-glander"><img src="http://localhost:4000/assets/images/so-logo.png" /></a>
		<a href="https://github.com/ShirinG"><img src="http://localhost:4000/assets/images/GitHub_Logo.png" /></a>
		<a href="http://www.xing.com/profile/Shirin_Glander"><img src="http://localhost:4000/assets/images/xing.png" /></a>
		<a href="http://de.linkedin.com/in/shirin-glander-01120881"><img src="http://localhost:4000/assets/images/Logo-2C-101px-R.png" /></a>
          with help from <a href="http://jekyllbootstrap.com" target="_blank" title="The Definitive Jekyll Blogging Framework">Jekyll Bootstrap</a>
          and <a href="http://getbootstrap.com" target="_blank">Bootstrap</a>
        </p>
      </div>
    </div>
	
    <!-- Latest compiled and minified JavaScript, requires jQuery 1.x (2.x not supported in IE8) -->
    <!-- Placed at the end of the document so the pages load faster -->
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.10.2/jquery.min.js"></script>
    <script src="/assets/themes/bootstrap-3/bootstrap/js/bootstrap.min.js"></script>
</div>
</html>

